{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# APPROXIMATE MATCHING"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import string\n",
    "\n",
    "def z_array(s):\n",
    "    \"\"\" Use Z algorithm (Gusfield theorem 1.4.1) to preprocess s \"\"\"\n",
    "    assert len(s) > 1\n",
    "    z = [len(s)] + [0] * (len(s)-1)\n",
    "    # Initial comparison of s[1:] with prefix\n",
    "    for i in range(1, len(s)):\n",
    "        if s[i] == s[i-1]:\n",
    "            z[1] += 1\n",
    "        else:\n",
    "            break\n",
    "    r, l = 0, 0\n",
    "    if z[1] > 0:\n",
    "        r, l = z[1], 1\n",
    "    for k in range(2, len(s)):\n",
    "        assert z[k] == 0\n",
    "        if k > r:\n",
    "            # Case 1\n",
    "            for i in range(k, len(s)):\n",
    "                if s[i] == s[i-k]:\n",
    "                    z[k] += 1\n",
    "                else:\n",
    "                    break\n",
    "            r, l = k + z[k] - 1, k\n",
    "        else:\n",
    "            # Case 2\n",
    "            # Calculate length of beta\n",
    "            nbeta = r - k + 1\n",
    "            zkp = z[k - l]\n",
    "            if nbeta > zkp:\n",
    "                # Case 2a: Zkp wins\n",
    "                z[k] = zkp\n",
    "            else:\n",
    "                # Case 2b: Compare characters just past r\n",
    "                nmatch = 0\n",
    "                for i in range(r+1, len(s)):\n",
    "                    if s[i] == s[i - k]:\n",
    "                        nmatch += 1\n",
    "                    else:\n",
    "                        break\n",
    "                l, r = k, r + nmatch\n",
    "                z[k] = r - k + 1\n",
    "    return z\n",
    "\n",
    "\n",
    "def n_array(s):\n",
    "    \"\"\" Compile the N array (Gusfield theorem 2.2.2) from the Z array \"\"\"\n",
    "    return z_array(s[::-1])[::-1]\n",
    "\n",
    "\n",
    "def big_l_prime_array(p, n):\n",
    "    \"\"\" Compile L' array (Gusfield theorem 2.2.2) using p and N array.\n",
    "        L'[i] = largest index j less than n such that N[j] = |P[i:]| \"\"\"\n",
    "    lp = [0] * len(p)\n",
    "    for j in range(len(p)-1):\n",
    "        i = len(p) - n[j]\n",
    "        if i < len(p):\n",
    "            lp[i] = j + 1\n",
    "    return lp\n",
    "\n",
    "\n",
    "def big_l_array(p, lp):\n",
    "    \"\"\" Compile L array (Gusfield theorem 2.2.2) using p and L' array.\n",
    "        L[i] = largest index j less than n such that N[j] >= |P[i:]| \"\"\"\n",
    "    l = [0] * len(p)\n",
    "    l[1] = lp[1]\n",
    "    for i in range(2, len(p)):\n",
    "        l[i] = max(l[i-1], lp[i])\n",
    "    return l\n",
    "\n",
    "\n",
    "def small_l_prime_array(n):\n",
    "    \"\"\" Compile lp' array (Gusfield theorem 2.2.4) using N array. \"\"\"\n",
    "    small_lp = [0] * len(n)\n",
    "    for i in range(len(n)):\n",
    "        if n[i] == i+1:  # prefix matching a suffix\n",
    "            small_lp[len(n)-i-1] = i+1\n",
    "    for i in range(len(n)-2, -1, -1):  # \"smear\" them out to the left\n",
    "        if small_lp[i] == 0:\n",
    "            small_lp[i] = small_lp[i+1]\n",
    "    return small_lp\n",
    "\n",
    "\n",
    "def good_suffix_table(p):\n",
    "    \"\"\" Return tables needed to apply good suffix rule. \"\"\"\n",
    "    n = n_array(p)\n",
    "    lp = big_l_prime_array(p, n)\n",
    "    return lp, big_l_array(p, lp), small_l_prime_array(n)\n",
    "\n",
    "\n",
    "def good_suffix_mismatch(i, big_l_prime, small_l_prime):\n",
    "    \"\"\" Given a mismatch at offset i, and given L/L' and l' arrays,\n",
    "        return amount to shift as determined by good suffix rule. \"\"\"\n",
    "    length = len(big_l_prime)\n",
    "    assert i < length\n",
    "    if i == length - 1:\n",
    "        return 0\n",
    "    i += 1  # i points to leftmost matching position of P\n",
    "    if big_l_prime[i] > 0:\n",
    "        return length - big_l_prime[i]\n",
    "    return length - small_l_prime[i]\n",
    "\n",
    "\n",
    "def good_suffix_match(small_l_prime):\n",
    "    \"\"\" Given a full match of P to T, return amount to shift as\n",
    "        determined by good suffix rule. \"\"\"\n",
    "    return len(small_l_prime) - small_l_prime[1]\n",
    "\n",
    "\n",
    "def dense_bad_char_tab(p, amap):\n",
    "    \"\"\" Given pattern string and list with ordered alphabet characters, create\n",
    "        and return a dense bad character table.  Table is indexed by offset\n",
    "        then by character. \"\"\"\n",
    "    tab = []\n",
    "    nxt = [0] * len(amap)\n",
    "    for i in range(0, len(p)):\n",
    "        c = p[i]\n",
    "        assert c in amap\n",
    "        tab.append(nxt[:])\n",
    "        nxt[amap[c]] = i+1\n",
    "    return tab\n",
    "\n",
    "\n",
    "class BoyerMoore(object):\n",
    "    \"\"\" Encapsulates pattern and associated Boyer-Moore preprocessing. \"\"\"\n",
    "    \n",
    "    def __init__(self, p, alphabet='ACGT'):\n",
    "        self.p = p\n",
    "        self.alphabet = alphabet\n",
    "        # Create map from alphabet characters to integers\n",
    "        self.amap = {}\n",
    "        for i in range(len(self.alphabet)):\n",
    "            self.amap[self.alphabet[i]] = i\n",
    "        # Make bad character rule table\n",
    "        self.bad_char = dense_bad_char_tab(p, self.amap)\n",
    "        # Create good suffix rule table\n",
    "        _, self.big_l, self.small_l_prime = good_suffix_table(p)\n",
    "    \n",
    "    def bad_character_rule(self, i, c):\n",
    "        \"\"\" Return # skips given by bad character rule at offset i \"\"\"\n",
    "        assert c in self.amap\n",
    "        ci = self.amap[c]\n",
    "        assert i > (self.bad_char[i][ci]-1)\n",
    "        return i - (self.bad_char[i][ci]-1)\n",
    "    \n",
    "    def good_suffix_rule(self, i):\n",
    "        \"\"\" Given a mismatch at offset i, return amount to shift\n",
    "            as determined by (weak) good suffix rule. \"\"\"\n",
    "        length = len(self.big_l)\n",
    "        assert i < length\n",
    "        if i == length - 1:\n",
    "            return 0\n",
    "        i += 1  # i points to leftmost matching position of P\n",
    "        if self.big_l[i] > 0:\n",
    "            return length - self.big_l[i]\n",
    "        return length - self.small_l_prime[i]\n",
    "    \n",
    "    def match_skip(self):\n",
    "        \"\"\" Return amount to shift in case where P matches T \"\"\"\n",
    "        return len(self.small_l_prime) - self.small_l_prime[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def boyer_moore(p, p_bm, t):\n",
    "    \"\"\" Do Boyer-Moore matching \"\"\"\n",
    "    i = 0\n",
    "    occurrences = []\n",
    "    while i < len(t) - len(p) + 1:\n",
    "        shift = 1\n",
    "        mismatched = False\n",
    "        for j in range(len(p)-1, -1, -1):\n",
    "            if p[j] != t[i+j]:\n",
    "                skip_bc = p_bm.bad_character_rule(j, t[i+j])\n",
    "                skip_gs = p_bm.good_suffix_rule(j)\n",
    "                shift = max(shift, skip_bc, skip_gs)\n",
    "                mismatched = True\n",
    "                break\n",
    "        if not mismatched:\n",
    "            occurrences.append(i)\n",
    "            skip_gs = p_bm.match_skip()\n",
    "            shift = max(shift, skip_gs)\n",
    "        i += shift\n",
    "    return occurrences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def approximate_match(p, t, n):\n",
    "    segment_length = int(round(len(p) / (n+1)))\n",
    "    all_matches = set()\n",
    "    for i in range(n+1):\n",
    "        start = i*segment_length\n",
    "        end = min((i+1)*segment_length, len(p))\n",
    "        p_bm = BoyerMoore(p[start:end], alphabet='ACGT')\n",
    "        matches = boyer_moore(p[start:end], p_bm, t)\n",
    "        # Extend matching segments to see if whole p matches\n",
    "        for m in matches:\n",
    "            if m < start or m-start+len(p) > len(t):\n",
    "                continue\n",
    "            mismatches = 0\n",
    "            for j in range(0, start):\n",
    "                if not p[j] == t[m-start+j]:\n",
    "                    mismatches += 1\n",
    "                    if mismatches > n:\n",
    "                        break\n",
    "            for j in range(end, len(p)):\n",
    "                if not p[j] == t[m-start+j]:\n",
    "                    mismatches += 1\n",
    "                    if mismatches > n:\n",
    "                        break\n",
    "            if mismatches <= n:\n",
    "                all_matches.add(m - start)\n",
    "    return list(all_matches)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0, 5]\n"
     ]
    }
   ],
   "source": [
    "p = 'AACTTG'\n",
    "t = 'CACTTAATTTG'\n",
    "print(approximate_match(p, t, 2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def editDistRecursive(x, y):\n",
    "    # This implementation is very slow\n",
    "    if len(x) == 0:\n",
    "        return len(y)\n",
    "    elif len(y) == 0:\n",
    "        return len(x)\n",
    "    else:\n",
    "        distHor = editDistRecursive(x[:-1], y) + 1\n",
    "        distVer = editDistRecursive(x, y[:-1]) + 1\n",
    "        if x[-1] == y[-1]:\n",
    "            distDiag = editDistRecursive(x[:-1], y[:-1])\n",
    "        else:\n",
    "            distDiag = editDistRecursive(x[:-1], y[:-1]) + 1\n",
    "        return min(distHor, distVer, distDiag)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def editDistance(x, y):\n",
    "    # Create distance matrix\n",
    "    D = []\n",
    "    for i in range(len(x)+1):\n",
    "        D.append([0]*(len(y)+1))\n",
    "    # Initialize first row and column of matrix\n",
    "    for i in range(len(x)+1):\n",
    "        D[i][0] = i\n",
    "    for i in range(len(y)+1):\n",
    "        D[0][i] = i\n",
    "    # Fill in the rest of the matrix\n",
    "    for i in range(1, len(x)+1):\n",
    "        for j in range(1, len(y)+1):\n",
    "            distHor = D[i][j-1] + 1\n",
    "            distVer = D[i-1][j] + 1\n",
    "            if x[i-1] == y[j-1]:\n",
    "                distDiag = D[i-1][j-1]\n",
    "            else:\n",
    "                distDiag = D[i-1][j-1] + 1\n",
    "            D[i][j] = min(distHor, distVer, distDiag)\n",
    "    # Edit distance is the value in the bottom right corner of the matrix\n",
    "    return D[-1][-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wall time: 8.01 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "x = 'shake spea'\n",
    "y = 'Shakespear'\n",
    "editDistRecursive(x, y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wall time: 0 ns\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "x = 'shake spea'\n",
    "y = 'Shakespear'\n",
    "editDistance(x, y)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PATTERN-MATCHING:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Test dataset results - Occurences and leftmost offset: \n",
      "4 258 \n",
      "\n",
      "Q1: The 'AGGT' or 'ACCT' occurs 2 times \n",
      "\n",
      "Q2: The 'TTAA' occurs 1 times \n",
      "\n",
      "Q3: The offset of the leftmost occurrence of GGCGCGC is 71 \n",
      "\n",
      "Q4: The offset of the leftmost occurrence of GTTTGC is 320 \n",
      "\n",
      "Q5: The 'TTCAAGCC' occurs 1 times with mismatch less than 2 \n",
      "\n",
      "Q6: The offset of the leftmost occurrence of AGGAGGTT with mismatch less than 2 is 4 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "class findPattern ():\n",
    "    \"\"\"\n",
    "    This class finds the occurance and position of a given pattern in a given \n",
    "    genomic sequence in a file.\n",
    "    \"\"\"\n",
    "    def __init__(self, pattern, filename):\n",
    "        # initiate parameters\n",
    "        self.pattern = pattern\n",
    "        self.filename = filename\n",
    "        \n",
    "    def readGenome (self):\n",
    "        \"\"\"\n",
    "        read genomic DNA sequence to a string\n",
    "        \"\"\"\n",
    "        genome = \"\"\n",
    "        with open (self.filename, \"r\") as f:\n",
    "            for line in f:\n",
    "                # skip header\n",
    "                if not line[0] == \">\":\n",
    "                    genome += line.rstrip()\n",
    "            f.close()\n",
    "        return genome\n",
    "        \n",
    "    \n",
    "    def reverseComplement (self):\n",
    "        \"\"\"\n",
    "        generate reverse complement sequence for a given dna sequence\n",
    "        \"\"\"\n",
    "        complement = {\"A\": \"T\", \"C\": \"G\", \"T\": \"A\", \"G\": \"C\"}\n",
    "        revComPattern = \"\" # reversed compliment pattern\n",
    "        for nt in self.pattern:\n",
    "            revComPattern = complement[nt] + revComPattern\n",
    "        \n",
    "        return revComPattern\n",
    "        \n",
    "    def match(self, string1, string2, numOfMismatch):\n",
    "        \"\"\"\n",
    "        return True or False for matching results of two strings under the offset\n",
    "        of numOfMismatch\n",
    "        \"\"\"\n",
    "        counter = 0\n",
    "        if len(string1) != len(string2):\n",
    "            return False\n",
    "        # loop over string to compare character\n",
    "        for i in range(len(string1)):\n",
    "            if string1[i] != string2[i]:\n",
    "                counter += 1\n",
    "        if counter > numOfMismatch:\n",
    "            return False\n",
    "        return True\n",
    "        \n",
    "    def patternIdentifier (self, numOfMismatch):\n",
    "        \"\"\"\n",
    "        find positions of a given pattern and the reversed complement pattern \n",
    "        in a given genome\n",
    "        \"\"\"\n",
    "        patternLength = len(self.pattern)\n",
    "        genome = self.readGenome()\n",
    "        revComPattern = self.reverseComplement()\n",
    "        occurances = []\n",
    "        \n",
    "        for i in range (patternLength): # loop over pattern index\n",
    "            # loop over genome patterns\n",
    "            for j in range (i, len(genome), patternLength): \n",
    "                genomeMotif = genome[j: j+patternLength]\n",
    "                # compare genomic motif and patterns\n",
    "                if numOfMismatch == 0:\n",
    "                    if (self.match(genomeMotif, self.pattern, 0) or \\\n",
    "                        self.match(genomeMotif, revComPattern, 0))\\\n",
    "                       and j not in occurances: # avoid duplicate records\n",
    "                        occurances.append(j)\n",
    "                else:\n",
    "                    if self.match(genomeMotif, self.pattern, numOfMismatch)\\\n",
    "                       and j not in occurances: # avoid duplicate records:\n",
    "                        occurances.append(j)\n",
    "        return occurances\n",
    "    \n",
    "class checkQuality ():\n",
    "    \"\"\"\n",
    "    The checkQuality class exams quality of sequencing for each cycle\n",
    "    \"\"\"\n",
    "    def __init__ (self, filename):\n",
    "        self.filename = filename\n",
    "    \n",
    "    def readFastq(self):\n",
    "        \"\"\"\n",
    "        read dna sequence and quality base from a fastq sequencing file to lists\n",
    "        \"\"\"\n",
    "        with open (self.filename, \"r\") as f:\n",
    "             sequences = []\n",
    "             qualities = []\n",
    "             while True:\n",
    "                 f.readline() # skip name line\n",
    "                 seq = f.readline().rstrip() # read sequence line\n",
    "                 f.readline() # skip strand line\n",
    "                 qual = f.readline().rstrip() # read quality line\n",
    "                 if len(seq) == 0: #finish read\n",
    "                     break\n",
    "                 # add seqence and quality information to list\n",
    "                 sequences.append(seq)\n",
    "                 qualities.append(qual)\n",
    "             f.close()\n",
    "        return sequences, qualities\n",
    "        \n",
    "    def phre33ToQ (self, qualString):\n",
    "        \"\"\"\n",
    "        transform quality string to quality base score\n",
    "        \"\"\"\n",
    "        qScore = []\n",
    "        for qual in qualString:\n",
    "            qScore.append(ord(qual) - 33)\n",
    "        return qScore\n",
    "    def findPoorQuality(self):\n",
    "        \"\"\"\n",
    "        find the index of poorest q score in each sequencing\n",
    "        \"\"\"\n",
    "        _, qualities = self.readFastq()\n",
    "        lowestQScoreIndex = []\n",
    "        for qualString in qualities:\n",
    "            qScore = self.phre33ToQ(qualString)\n",
    "            lowestQScoreIndex.append(qScore.index(min(qScore)))\n",
    "        return lowestQScoreIndex\n",
    "        \n",
    "    def countPoorQuality(self):    \n",
    "        \"\"\"\n",
    "        count number of poorest q score in each cycle\n",
    "        \"\"\"\n",
    "        import collections\n",
    "        return collections.Counter(self.findPoorQuality())\n",
    "        \n",
    "    def plotHist(self):\n",
    "        \"\"\"\n",
    "         show the distribution of poorest q score\n",
    "        \"\"\"\n",
    "        import matplotlib.pyplot as plt\n",
    "        data = self.countPoorQuality()\n",
    "        plt.bar(data.keys(), data.values())\n",
    "        plt.show()\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    #Test\n",
    "    filename = \"CERVICALDNA.fasta\"\n",
    "    pattern = \"ATTA\"\n",
    "    patterns = findPattern(pattern,filename)\n",
    "    print(\"Test dataset results - Occurences and leftmost offset: \")\n",
    "    print(len(patterns.patternIdentifier(0)), min(patterns.patternIdentifier(0)), \"\\n\")\n",
    "    \n",
    "    #Questions 1-6\n",
    "    filename1 = \"CERVICALDNA.fasta\"\n",
    "    #Q1: How many times does AGGT or its reverse complement (ACCT) occur in the \n",
    "    #Cervical Cancer genome? E.g. if AGGT occurs 10 times and ACCT occurs 12 times, \n",
    "    #you should report 22.\n",
    "    pattern = \"AGGT\"\n",
    "    patterns = findPattern(pattern,filename1)\n",
    "    print(\"Q1: The 'AGGT' or 'ACCT' occurs %d times \\n\" \\\n",
    "           %len(patterns.patternIdentifier(0)))\n",
    "    \n",
    "    #Q2: How many times does TTAA or its reverse complement occur in the Cervical Cancer genome? \n",
    "    #Hint: TTAA and its reverse complement are equal, so remember \n",
    "    #not to double count.    \n",
    "    pattern = \"TTAA\"\n",
    "    patterns = findPattern(pattern,filename1)\n",
    "    print(\"Q2: The 'TTAA' occurs %d times \\n\" \\\n",
    "           %len(patterns.patternIdentifier(0)))\n",
    "    \n",
    "    #Q3: What is the offset of the leftmost occurrence of GGCGCGC or its reverse\n",
    "    #complement in the Cervical Cancer genome? E.g. if the leftmost occurrence of \n",
    "    #GGCGCGC is at offset 40 (0-based) and the leftmost occurrence of its reverse \n",
    "    #complement GGCGCGC is at offset 29, then report 29.\n",
    "    pattern = \"GGCGCGC\"\n",
    "    patterns = findPattern(pattern,filename1)\n",
    "    print(\"Q3: The offset of the leftmost occurrence of GGCGCGC is %d \\n\" \\\n",
    "           %min(patterns.patternIdentifier(0)))\n",
    "    \n",
    "    #Q4: What is the offset of the leftmost occurrence of GTTTGC or its reverse \n",
    "    #complement in the Cervical Cancer genome?\n",
    "    pattern = \"GTTTGC\"\n",
    "    patterns = findPattern(pattern,filename1)\n",
    "    print(\"Q4: The offset of the leftmost occurrence of GTTTGC is %d \\n\" \\\n",
    "           %min(patterns.patternIdentifier(0)))\n",
    "           \n",
    "    #Q5: How many times does TTCAAGCC occur in the Cervical Cancer genome when \n",
    "    #allowing up to 2 mismatches?    \n",
    "    pattern = \"TTCAAGCC\"\n",
    "    patterns = findPattern(pattern,filename1)\n",
    "    print(\"Q5: The 'TTCAAGCC' occurs %d times with mismatch less than 2 \\n\" \\\n",
    "           %len(patterns.patternIdentifier(2)))\n",
    "       \n",
    "    #Q6: What is the offset of the leftmost occurrence of AGGAGGTT in the \n",
    "    #Cervical Cancer genome when allowing up to 2 mismatches?   \n",
    "    pattern = \"AGGAGGTT\"\n",
    "    patterns = findPattern(pattern,filename1)\n",
    "    print(\"Q6: The offset of the leftmost occurrence of AGGAGGTT with mismatch less than 2 is %d \\n\" \\\n",
    "          %min(patterns.patternIdentifier(2)))      \n",
    "      \n",
    "    #Q7: Report which sequencing cycle has the problem.    \n",
    "    filename2 = \"CERVICALDNA.fasta\"\n",
    "    qualities = checkQuality(filename2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Q1: the edit distance of the best match between pattern and the genome is 3\n",
      "\n",
      "Q2: the edit distance of the best match between pattern and the genome is 2\n",
      "\n",
      "Running time for naive overlap mapping: 4 sec\n",
      "\n",
      "Running time for phrase reads: 0 sec\n",
      "\n",
      "Running time for optimized algorithm: 0 sec\n",
      "\n",
      "Q3: the total edges are 3\n",
      "\n",
      "Q4: the total nodes are 2\n"
     ]
    }
   ],
   "source": [
    "from itertools import permutations\n",
    "\n",
    "class findPatternV3 ():\n",
    "    \"\"\"\n",
    "    This class finds the edit distance of a given pattern in a given genomic \n",
    "    sequence and constructs overlap graphs.\n",
    "    \"\"\"\n",
    "    def __init__(self, filename, pattern = False):\n",
    "        # initiate parameters\n",
    "        self.pattern = pattern\n",
    "        self.filename = filename\n",
    "        \n",
    "    def readGenome (self):\n",
    "        \"\"\"\n",
    "        read genomic DNA sequence to a string\n",
    "        \"\"\"\n",
    "        genome = \"\"\n",
    "        with open (self.filename, \"r\") as f:\n",
    "            for line in f:\n",
    "                # skip header\n",
    "                if not line[0] == \">\":\n",
    "                    genome += line.rstrip()\n",
    "            f.close()\n",
    "        return genome\n",
    "        \n",
    "    def readFastq(self):\n",
    "        \"\"\"\n",
    "        read dna sequence and quality base from a fastq sequencing file to lists\n",
    "        \"\"\"\n",
    "        with open (self.filename, \"r\") as f:\n",
    "             sequences = []\n",
    "             qualities = []\n",
    "             while True:\n",
    "                 f.readline() # skip name line\n",
    "                 seq = f.readline().rstrip() # read sequence line\n",
    "                 f.readline() # skip strand line\n",
    "                 qual = f.readline().rstrip() # read quality line\n",
    "                 if len(seq) == 0: #finish read\n",
    "                     break\n",
    "                 # add seqence and quality information to list\n",
    "                 sequences.append(seq)\n",
    "                 qualities.append(qual)\n",
    "             f.close()\n",
    "        return sequences, qualities\n",
    "        \n",
    "    def editDistance(self):\n",
    "        \"\"\"\n",
    "        Implement dynamic algorithm to calculate edit distance between a given\n",
    "        pattern and a given genome\n",
    "        \"\"\"\n",
    "        pattern = self.pattern\n",
    "        genome = self.readGenome()\n",
    "        pattern_length = len(pattern) + 1\n",
    "        genome_length = len(genome) + 1\n",
    "        #generate matrix\n",
    "        matrix = [[0]*genome_length for i in range(pattern_length)]\n",
    "        # initiate the first column\n",
    "        for i in range(pattern_length):\n",
    "            matrix[i][0] = i\n",
    "        for i in range(1, pattern_length):\n",
    "            for j in range(1, genome_length):\n",
    "                dist_hor = matrix[i][j-1] + 1\n",
    "                dist_vel = matrix[i-1][j] + 1\n",
    "                dist_diag = matrix[i-1][j-1] + 1 if pattern[i-1] != genome[j-1]\\\n",
    "                            else matrix[i-1][j-1]\n",
    "                matrix[i][j] = min(dist_hor, dist_vel, dist_diag)\n",
    "        return min(matrix[-1])                \n",
    "            \n",
    "    def phraseReads(self, k_mer):\n",
    "        \"\"\"\n",
    "        construct the prefix and suffix of a read to a dictornary with read as\n",
    "        key and pre-,suffix as values\n",
    "        \"\"\"\n",
    "        reads, _ = self.readFastq()\n",
    "        reads_dict = {}\n",
    "        for read in reads:\n",
    "            for i in range(len(read) - k_mer + 1):\n",
    "                substring = read[i:i+k_mer]\n",
    "                if substring not in reads_dict:\n",
    "                    reads_dict[substring] = set([read])\n",
    "                else:\n",
    "                    reads_dict[substring].add(read)\n",
    "        return reads_dict\n",
    "    def overlap(self, read1, read2, k_mer):\n",
    "        \"\"\"\n",
    "        find overlaped leftmost offset\n",
    "        \"\"\"\n",
    "        start = 0\n",
    "        while True:\n",
    "            start = read1.find(read2[:k_mer], start)\n",
    "            if start == -1:\n",
    "                return 0 # without overlap\n",
    "            if read2.startswith(read1[start:]):\n",
    "                return len(read1) - start\n",
    "            start += 1\n",
    "            \n",
    "    def overlapGraph(self, k_mer):\n",
    "        \"\"\"\n",
    "        construct graph with key as a read (node) and values as all other \n",
    "        reads overlapped with the previous read (node)\n",
    "        \"\"\"\n",
    "        reads_dict = self.phraseReads(k_mer)\n",
    "        reads, _= self.readFastq()\n",
    "        graph = {}\n",
    "        for read1 in reads:\n",
    "            k_mer_string = read1[len(read1) - k_mer:]\n",
    "            if k_mer_string in reads_dict:\n",
    "                edges = set([])\n",
    "                reads_set = reads_dict[k_mer_string]\n",
    "                for read2 in reads_set:\n",
    "                    if read1 != read2: #skip self comparison\n",
    "                        offset = self.overlap(read1, read2, k_mer)           \n",
    "                        if offset > 0: # skip non-overlapped pairs\n",
    "                            edges.add(read2) #add overlapped reads to be values\n",
    "                            graph[read1] = edges           \n",
    "        return graph\n",
    "                        \n",
    "    def naive_overlap_map(self, k_mer):\n",
    "        \"\"\"\n",
    "        construct graph with key as a pair of reads with overlap and values as\n",
    "        the leftmost offset of the overlap\n",
    "        \"\"\"\n",
    "        graph = {}\n",
    "        reads, _ = self.readFastq()\n",
    "        for read1, read2 in permutations(reads, 2):\n",
    "            #skip non-overlapped reads\n",
    "            if read1[len(read1) - k_mer:] in read2:\n",
    "                offset = self.overlap(read1, read2, k_mer)\n",
    "                    # check if reads[i] overlapped with reads[j]\n",
    "                if offset != 0:\n",
    "                    graph[(read1,read2)] = offset  \n",
    "        return graph\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    \n",
    "    #Q1: What is the edit distance of the best match between pattern \n",
    "    #ATTTATCCGCCGATTCC and the excerpt of human chromosome 1? \n",
    "    #(Don't consider reverse complements.)\n",
    "    pattern = \"ATTTATCCGCTTCC\"\n",
    "    filename = \"CERVICALDNA.fasta\"\n",
    "    patterns = findPatternV3 (filename, pattern)\n",
    "    edit_dist = patterns.editDistance()\n",
    "    print(\"Q1: the edit distance of the best match between pattern and the genome is %d\\n\"\\\n",
    "           %edit_dist)\n",
    "           \n",
    "    #Q2: What is the edit distance of the best match between pattern \n",
    "    #GCCGATTCCGGGCG and the excerpt of human chromosome 1? \n",
    "    #(Don't consider reverse complements.)\n",
    "    pattern = \"GCCGATTCGGCG\"\n",
    "    filename = \"CERVICALDNA.fasta\"\n",
    "    patterns = findPatternV3 (filename, pattern)\n",
    "    edit_dist = patterns.editDistance()\n",
    "    print(\"Q2: the edit distance of the best match between pattern and the genome is %d\\n\"\\\n",
    "           %edit_dist)\n",
    "    \n",
    "    #Q3: Picture the overlap graph corresponding to the overlaps just calculated. \n",
    "    # How many edges are in the graph? In other words, how many distinct pairs \n",
    "    # of reads overlap?\n",
    "    #Q4: Picture the overlap graph corresponding to the overlaps computed for \n",
    "    #the previous question. How many nodes in this graph have at least one \n",
    "    #outgoing edge? (In other words, how many reads have a suffix involved in \n",
    "    #an overlap?)\n",
    "    import time\n",
    "    t1 = time.time()\n",
    "    filename = \"ConcatDNA.fastq\"\n",
    "    patterns = findPatternV3 (filename)\n",
    "    k_mer = 30\n",
    "    graph = patterns.naive_overlap_map(k_mer)\n",
    "    t2 = time.time()\n",
    "    print(\"Running time for naive overlap mapping: %d sec\\n\"%(t2 - t1))\n",
    "    \n",
    "    reads = patterns.phraseReads(k_mer)\n",
    "    t3 = time.time()\n",
    "    print(\"Running time for phrase reads: %d sec\\n\"%(t3 - t2))\n",
    "    graph = patterns.overlapGraph(k_mer)\n",
    "    t4 = time.time()\n",
    "    print(\"Running time for optimized algorithm: %d sec\\n\"%(t4 - t2))\n",
    "    numberOfNodes = len(graph)\n",
    "    numberOfEdges = sum([len(edges) for edges in graph.values()])\n",
    "    print(\"Q3: the total edges are %d\\n\"%numberOfEdges)\n",
    "    print(\"Q4: the total nodes are %d\"%numberOfNodes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
