{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BIOPYTHON-BASICS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('AGTACACTGGT')"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "my_seq = Seq(\"AGTACACTGGT\")\n",
    "my_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Alphabet()"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_seq.alphabet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('AGTACACTGGT', IUPACUnambiguousDNA())"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio.Alphabet import IUPAC\n",
    "my_seq = Seq(\"AGTACACTGGT\", IUPAC.unambiguous_dna)\n",
    "my_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "IUPACUnambiguousDNA()"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_seq.alphabet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 A\n",
      "1 G\n",
      "2 T\n",
      "3 A\n",
      "4 C\n",
      "5 A\n",
      "6 C\n",
      "7 T\n",
      "8 G\n",
      "9 G\n",
      "10 T\n"
     ]
    }
   ],
   "source": [
    "for index, letter in enumerate(my_seq):\n",
    "   print(\"%i %s\" % (index, letter))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "11"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(my_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_seq.count(\"A\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_seq.count(\"GT\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Seq(\"AAAA\").count(\"AA\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('TACACT', IUPACUnambiguousDNA())"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_seq[2:8]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('TACACTAGTACACTGGT', IUPACUnambiguousDNA())"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p_seq = Seq(\"EVRNAK\", IUPAC.protein)\n",
    "d_seq = Seq('TACACT', IUPAC.unambiguous_dna)\n",
    "d_seq + my_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "Incompatible alphabets IUPACProtein() and IUPACUnambiguousDNA()",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-13-ed678bbfe19e>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0mp_seq\u001b[0m \u001b[1;33m+\u001b[0m \u001b[0mmy_seq\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;32mC:\\ProgramData\\Anaconda3\\lib\\site-packages\\Bio\\Seq.py\u001b[0m in \u001b[0;36m__add__\u001b[1;34m(self, other)\u001b[0m\n\u001b[0;32m    309\u001b[0m                 raise TypeError(\n\u001b[0;32m    310\u001b[0m                     \"Incompatible alphabets {0!r} and {1!r}\".format(\n\u001b[1;32m--> 311\u001b[1;33m                         self.alphabet, other.alphabet))\n\u001b[0m\u001b[0;32m    312\u001b[0m             \u001b[1;31m# They should be the same sequence type (or one of them is generic)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    313\u001b[0m             \u001b[0ma\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mAlphabet\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_consensus_alphabet\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0malphabet\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mother\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0malphabet\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mTypeError\u001b[0m: Incompatible alphabets IUPACProtein() and IUPACUnambiguousDNA()"
     ]
    }
   ],
   "source": [
    "p_seq + my_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('TCATGTGACCA', IUPACUnambiguousDNA())"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_seq.complement()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('ACCAGTGTACT', IUPACUnambiguousDNA())"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_seq.reverse_complement()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Proteins do not have complements!",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-16-2828b73a4b37>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[0mp_seq\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mSeq\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"EVRNAK\"\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mIUPAC\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mprotein\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 2\u001b[1;33m \u001b[0mp_seq\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mreverse_complement\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;32mC:\\ProgramData\\Anaconda3\\lib\\site-packages\\Bio\\Seq.py\u001b[0m in \u001b[0;36mreverse_complement\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    946\u001b[0m         \"\"\"\n\u001b[0;32m    947\u001b[0m         \u001b[1;31m# Use -1 stride/step to reverse the complement\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 948\u001b[1;33m         \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcomplement\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m-\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    949\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    950\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0mtranscribe\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mC:\\ProgramData\\Anaconda3\\lib\\site-packages\\Bio\\Seq.py\u001b[0m in \u001b[0;36mcomplement\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    896\u001b[0m         \u001b[0mbase\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mAlphabet\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_get_base_alphabet\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0malphabet\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    897\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mbase\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mAlphabet\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mProteinAlphabet\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 898\u001b[1;33m             \u001b[1;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"Proteins do not have complements!\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    899\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mbase\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mAlphabet\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mDNAAlphabet\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    900\u001b[0m             \u001b[0mttable\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0m_dna_complement_table\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mValueError\u001b[0m: Proteins do not have complements!"
     ]
    }
   ],
   "source": [
    "p_seq = Seq(\"EVRNAK\", IUPAC.protein)\n",
    "p_seq.reverse_complement()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SeqRecord(seq=Seq('AGTACACTGGT', IUPACUnambiguousDNA()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio.SeqRecord import SeqRecord\n",
    "simple_seq_r = SeqRecord(my_seq)\n",
    "simple_seq_r"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NZ_ADRZ01000932.1\n",
      "Seq('GAGTTTTATCGCTTCCATGACGCAGAAGTTAACACTTTCGGATATTTCTGATGA...GCA', SingleLetterAlphabet())\n",
      "5386\n",
      "0\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "for seq_record in SeqIO.parse(\"DNA.fa\",\"fasta\"):\n",
    "    print(seq_record.id)\n",
    "    print(repr(seq_record.seq))\n",
    "    print(len(seq_record))\n",
    "    print(len(seq_record.features))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['NZ_ADRZ01000932.1']"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "identifiers = [seq_record.id for seq_record in SeqIO.parse(\"DNA.fa\",\"fasta\")]\n",
    "identifiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[SeqRecord(seq=Seq('GAGTTTTATCGCTTCCATGACGCAGAAGTTAACACTTTCGGATATTTCTGATGA...GCA', SingleLetterAlphabet()), id='NZ_ADRZ01000932.1', name='NZ_ADRZ01000932.1', description='NZ_ADRZ01000932.1 Yersinia pestis biovar Antiqua str. E1979001 Contig_E1979001_19275, whole genome shotgun sequence', dbxrefs=[])]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "records = list(SeqIO.parse(\"DNA.fa\",\"fasta\"))\n",
    "records[:50]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['NZ_ADRZ01000932.1'])"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "records_dict = SeqIO.to_dict(SeqIO.parse(\"DNA.fa\",\"fasta\"))\n",
    "records_dict.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SeqRecord(seq=Seq('GAGTTTTATCGCTTCCATGACGCAGAAGTTAACACTTTCGGATATTTCTGATGA...GCA', SingleLetterAlphabet()), id='NZ_ADRZ01000932.1', name='NZ_ADRZ01000932.1', description='NZ_ADRZ01000932.1 Yersinia pestis biovar Antiqua str. E1979001 Contig_E1979001_19275, whole genome shotgun sequence', dbxrefs=[])"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "records_dict['NZ_ADRZ01000932.1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['NZ_ADRZ01000932.1']"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "records_dict = SeqIO.index(\"DNA.fa\",\"fasta\")\n",
    "list(records_dict.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SeqRecord(seq=Seq('GAGTTTTATCGCTTCCATGACGCAGAAGTTAACACTTTCGGATATTTCTGATGA...GCA', SingleLetterAlphabet()), id='NZ_ADRZ01000932.1', name='NZ_ADRZ01000932.1', description='NZ_ADRZ01000932.1 Yersinia pestis biovar Antiqua str. E1979001 Contig_E1979001_19275, whole genome shotgun sequence', dbxrefs=[])"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "records_dict['NZ_ADRZ01000932.1']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PARSING FASTA-FILES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['>NP_001273979.1 stabilizer of axonemal microtubules 1 isoform c [Homo '\n",
      " 'sapiens]\\n',\n",
      " '\\n',\n",
      " 'ATGAAAACCAAATGCATTTGCGAACTGTGCAGCTGCGGCCGCCATCATTGCCCGCATCTG\\n',\n",
      " 'CCGACCAAAATTTATGATAAAACCGAAAAACCGTGCCTGCTGAGCGAATATACCGAAAAC\\n',\n",
      " 'TATCCGTTTTATCATAGCTATCTGCCGCGCGAAAGCTTTAAACCGCGCCGCGAATATCAG\\n',\n",
      " 'AAAGGCCCGATTCCGATGGAAGGCCTGACCACCAGCAGC\\n',\n",
      " '\\n',\n",
      " '\\n',\n",
      " '\\n',\n",
      " '>NP_001017978.1 kita-kyushu lung cancer antigen 1 [Homo sapiens]\\n',\n",
      " '\\n',\n",
      " 'ATGAACTTTTATCTGCTGCTGGCGAGCAGCATTCTGTGCGCGCTGATTGTGTTTTGGAAA\\n',\n",
      " 'TATCGCCGCTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCACCGCGCTGGCGCTG\\n',\n",
      " 'GTGCGCCCGAGCAGCAGCGGCCTGATTAACAGCAACACCGATAACAACCTGGCGGTGTAT\\n',\n",
      " 'GATCTGAGCCGCGATATTCTGAACAACTTTCCGCATAGCATTGCGCGCCAGAAACGCATT\\n',\n",
      " 'CTGGTGAACCTGAGCATGGTGGAAAACAAACTGGTGGAACTGGAACATACCCTGCTGAGC\\n',\n",
      " 'AAAGGCTTTCGCGGCGCGAGCCCGCATCGCAAAAGCACC\\n',\n",
      " '\\n',\n",
      " '\\n',\n",
      " '>sp|Q5H943.1|KKLC1_HUMAN RecName: Full=Kita-kyushu lung cancer antigen 1; '\n",
      " 'Short=KK-LC-1; AltName: Full=Cancer/testis antigen 83\\n',\n",
      " '\\n',\n",
      " 'ATGAACTTTTATCTGCTGCTGGCGAGCAGCATTCTGTGCGCGCTGATTGTGTTTTGGAAA\\n',\n",
      " 'TATCGCCGCTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCACCGCGCTGGCGCTG\\n',\n",
      " 'GTGCGCCCGAGCAGCAGCGGCCTGATTAACAGCAACACCGATAACAACCTGGCGGTGTAT\\n',\n",
      " 'GATCTGAGCCGCGATATTCTGAACAACTTTCCGCATAGCATTGCGCGCCAGAAACGCATT\\n',\n",
      " 'CTGGTGAACCTGAGCATGGTGGAAAACAAACTGGTGGAACTGGAACATACCCTGCTGAGC\\n',\n",
      " 'AAAGGCTTTCGCGGCGCGAGCCCGCATCGCAAAAGCACC\\n',\n",
      " '\\n',\n",
      " '\\n',\n",
      " '>AAC35497.1 lung cancer candidate FUS1 [Homo sapiens]\\n',\n",
      " '\\n',\n",
      " 'ATGGGCGCGAGCGGCAGCAAAGCGCGCGGCCTGTGGCCGTTTGCGAGCGCGGCGGGCGGC\\n',\n",
      " 'GGCGGCAGCGAAGCGGCGGGCGCGGAACAGGCGCTGGTGCGCCCGCGCGGCCGCGCGGTG\\n',\n",
      " 'CCGCCGTTTGTGTTTACCCGCCGCGGCAGCATGTTTTATGATGAAGATGGCGATCTGGCG\\n',\n",
      " 'CATGAATTTTATGAAGAAACCATTGTGACCAAAAACGGCCAGAAACGCGCGAAACTGCGC\\n',\n",
      " 'CGCGTGCATAAAAACCTGATTCCGCAGGGCATTGTGAAACTGGATCATCCGCGCATTCAT\\n',\n",
      " 'GTGGATTTTCCGGTGATTCTGTATGAAGTG\\n',\n",
      " '\\n',\n",
      " '\\n',\n",
      " '>XP_002699503.1 kita-kyushu lung cancer antigen 1 [Bos taurus]\\n',\n",
      " '\\n',\n",
      " 'ATGAGCTTTCTGTTTCTGCTGGTGAGCAGCATTCTGTTTGCGTTTATGTTTGTGTTTTGG\\n',\n",
      " 'AAAACCTGGTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCACCAGCCTGGCGCTG\\n',\n",
      " 'GTGCGCAGCAGCAGCAGCACCCAGAGCACCAAAAGCAACATTGATAAAAGCCTGCTGGTG\\n',\n",
      " 'AACAGCCTGAGCCGCGATATTCTGATTAACTTTCCGCATAGCATTGCGATGCAGCGCCGC\\n',\n",
      " 'ATTCTGGTGAACCTGCGCATTGTGGATTATAAACTGGCGGAACTGGAACAGTTTCTGGTG\\n',\n",
      " 'ACCAAAGGCCTGAACGGCGCGCTGGTGAACCGCAAAAGCACCGATAAACCGACCGAAGTG\\n',\n",
      " '\\n',\n",
      " '\\n',\n",
      " '>sp|Q4R717.1|KKLC1_MACFA RecName: Full=Kita-kyushu lung cancer antigen 1 '\n",
      " 'homolog\\n',\n",
      " '\\n',\n",
      " 'ATGAACGTGTATCTGCTGCTGGCGAGCGGCATTCTGTGCGCGCTGATGACCGTGTTTTGG\\n',\n",
      " 'AAATATCGCCGCTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCACCGCGCTGGCG\\n',\n",
      " 'CTGGTGCGCCCGAGCAGCACCGGCCTGATTAACAGCAACACCGATAACAACCTGAGCGTG\\n',\n",
      " 'TATGATCTGAGCCGCGATATTCTGAACAACTTTCCGCATAGCATTGCGATGCAGAAACGC\\n',\n",
      " 'ATTCTGGTGAACCTGACCACCGTGGAAAACAAACTGGTGGAACTGGAACATATTCTGGTG\\n',\n",
      " 'AGCAAAGGCTTTCGCAGCGCGAGCGCGCATCGCAAAAGCACC\\n',\n",
      " '\\n',\n",
      " '\\n',\n",
      " '\\n',\n",
      " '>AAN12271.1 lung cancer oncogene 5 [Homo sapiens]\\n',\n",
      " '\\n',\n",
      " 'ATGCTGCTGACCCGCTTTGAACAGAAAAACGGCCATCTGGCGCAGGTGGAAGTGGATGAA\\n',\n",
      " 'GTGTTTGGCTTTATGTGCCATATTACCACCGAAGTGCCGCCGCATGATGCGATGCCGGGC\\n',\n",
      " 'GGCATTGTGCTGCTGGTGAAATTTCTGCTGGATATGAGCCGCAACGTGCTGCTGTATGTG\\n',\n",
      " 'GTGTTTCTGCAGCGCCTGAGCAGCGCGCTGCATCGCGTGCTGCTGCATCTGTTTCGCCAT\\n',\n",
      " 'GTGCGCATTTTTGATCATGGCCTGAGCGTGGCGCATGGCTATCGCGGCGAAGGCGGCTGG\\n',\n",
      " 'CCGACCGCGACCGTGAGCTGGGGCCGCTGC\\n',\n",
      " '\\n',\n",
      " '\\n',\n",
      " '>NP_001270680.1 kita-kyushu lung cancer antigen 1 homolog [Macaca '\n",
      " 'fascicularis]\\n',\n",
      " '\\n',\n",
      " 'ATGAACGTGTATCTGCTGCTGGCGAGCGGCATTCTGTGCGCGCTGATGACCGTGTTTTGG\\n',\n",
      " 'AAATATCGCCGCTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCACCGCGCTGGCG\\n',\n",
      " 'CTGGTGCGCCCGAGCAGCACCGGCCTGATTAACAGCAACACCGATAACAACCTGAGCGTG\\n',\n",
      " 'TATGATCTGAGCCGCGATATTCTGAACAACTTTCCGCATAGCATTGCGATGCAGAAACGC\\n',\n",
      " 'ATTCTGGTGAACCTGACCACCGTGGAAAACAAACTGGTGGAACTGGAACATATTCTGGTG\\n',\n",
      " 'AGCAAAGGCTTTCGCAGCGCGAGCGCGCATCGCAAAAGCACC\\n',\n",
      " '\\n',\n",
      " '\\n',\n",
      " '>Sp|P63244.3|Rack1_Human Recname: Full=Receptor Of Activated Protein C '\n",
      " 'Kinase 1; Altname: Full=Cell Proliferation-Inducing Gene 21 Protein; '\n",
      " 'Altname: Full=Guanine Nucleotide-Binding Protein Subunit Beta-2-Like 1; '\n",
      " 'Altname: Full=Guanine Nucleotide-Binding Protein Subunit Beta-Like Protein '\n",
      " '12.3; Altname: Full=Human Lung Cancer Oncogene 7 Protein; Short=Hlc-7; '\n",
      " 'Altname: Full=Receptor For Activated C Kinase; Altname: Full=Small Ribosomal '\n",
      " 'Subunit Protein Rack1; Contains: Recname: Full=Receptor Of Activated Protein '\n",
      " 'C Kinase 1, N-Terminally Processed; Altname: Full=Guanine Nucleotide-Binding '\n",
      " 'Protein Subunit Beta-2-Like 1, N-Terminally Processed\\n',\n",
      " '\\n',\n",
      " 'ATGACCGAACAGATGACCCTGCGCGGCACCCTGAAAGGCCATAACGGCTGGGTGACCCAG\\n',\n",
      " 'ATTGCGACCACCCCGCAGTTTCCGGATATGATTCTGAGCGCGAGCCGCGATAAAACCATT\\n',\n",
      " 'ATTATGTGGAAACTGACCCGCGATGAAACCAACTATGGCATTCCGCAGCGCGCGCTGCGC\\n',\n",
      " 'GGCCATAGCCATTTTGTGAGCGATGTGGTGATTAGCAGCGATGGCCAGTTTGCGCTGAGC\\n',\n",
      " 'GGCAGCTGGGATGGCACCCTGCGCCTGTGGGATCTGACCACCGGCACCACCACCCGCCGC\\n',\n",
      " 'TTTGTGGGCCATACCAAAGATGTGCTGAGCGTGGCGTTTAGCAGCGATAACCGCCAGATT\\n',\n",
      " 'GTGAGCGGCAGCCGCGATAAAACCATTAAACTGTGGAACACCCTGGGCGTGTGCAAATAT\\n',\n",
      " 'ACCGTGCAGGATGAAAGCCATAGCGAATGGGTGAGCTGCGTGCGCTTTAGCCCGAACAGC\\n',\n",
      " 'AGCAACCCGATTATTGTGAGCTGCGGCTGGGATAAACTGGTGAAAGTGTGGAACCTGGCG\\n',\n",
      " 'AACTGCAAACTGAAAACCAACCATATTGGCCATACCGGCTATCTGAACACCGTGACCGTG\\n',\n",
      " 'AGCCCGGATGGCAGCCTGTGCGCGAGCGGCGGCAAAGATGGCCAGGCGATGCTGTGGGAT\\n',\n",
      " 'CTGAACGAAGGCAAACATCTGTATACCCTGGATGGCGGCGATATTATTAACGCGCTGTGC\\n',\n",
      " 'TTTAGCCCGAACCGCTATTGGCTGTGCGCGGCGACCGGCCCGAGCATTAAAATTTGGGAT\\n',\n",
      " 'CTGGAAGGCAAAATTATTGTGGATGAACTGAAACAGGAAGTGATTAGCACCAGCAGCAAA\\n',\n",
      " 'GAACCGCCGCAGTGCACCAGCCTGGCGTGGAGCGCGGATGGCCAGACCCTGTTTGCGGGC\\n',\n",
      " 'TATACCGATAACCTGGTGCGCGTGTGGCAGGTGACCATTGGCACCCGC\\n',\n",
      " '\\n',\n",
      " '>XP_027389897.1 kita-kyushu lung cancer antigen 1 [Bos indicus x Bos '\n",
      " 'taurus]\\n',\n",
      " '\\n',\n",
      " 'ATGAGCTTTCTGTTTCTGCTGGTGAGCAGCATTCTGTTTGCGTTTATGTTTGTGTTTTGG\\n',\n",
      " 'AAAACCTGGTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCACCAGCCTGGCGCTG\\n',\n",
      " 'GTGCGCAGCAGCAGCAGCACCCAGAGCACCAAAAGCAACATTGATAAAAGCCTGCTGGTG\\n',\n",
      " 'AACAGCCTGAGCCGCGATATTCTGATTAACTTTCCGCATAGCATTGCGATGCAGCGCCGC\\n',\n",
      " 'ATTCTGGTGAACCTGCGCATTGTGGATTATAAACTGGCGGAACTGGAACAGTTTCTGGTG\\n',\n",
      " 'ACCAAAGGCCTGAACGGCGCGCTGGTGAACCGCAAAAGCACCGATAAACCGACCGAAGTG\\n',\n",
      " '\\n',\n",
      " '\\n',\n",
      " '>XP_007647427.1 kita-kyushu lung cancer antigen 1 [Cricetulus griseus]\\n',\n",
      " '\\n',\n",
      " 'ATGAACATTCTGTTTATTCTGGTGGGCAGCATTATTTTTGCGAGCATTTTTAGCATTTTT\\n',\n",
      " 'AAAGTGGTGTTTCAGATTCCGGATCATGAACGCCCGAGCAACCCGACCAGCCCGACCCTG\\n',\n",
      " 'CTGCGCGCGAACAGCTTTTGGAGCTATCGCAACACCGGCCTGAGCTTTCCGGGCAACAGC\\n',\n",
      " 'TATAACCAGGATATGATGAACAACTTTCCGCGCACCCTGGCGCTGCATAAACGCATTCTG\\n',\n",
      " 'GTGAACCTGAACATTATGGAATGCCAGCTGACCAAACTGGAACAGTTTCTGAGCAGCAAA\\n',\n",
      " 'AACCCGCAGGGCCATACCAAACATCGCCGCCGCGCGACCGTGCGCATTCCGATGGAAAGC\\n',\n",
      " 'GATAGCGGCAGCAACCAG\\n',\n",
      " '\\n',\n",
      " '\\n',\n",
      " '>XP_014918346.1 kita-kyushu lung cancer antigen 1 [Acinonyx jubatus]\\n',\n",
      " '\\n',\n",
      " 'ATGAGCATGCTGCTGCTGCTGCTGAGCTTTATTGTGTTTACCCTGCTGCTGGTGGTGTGG\\n',\n",
      " 'AAAAGCCGCTTTCAGAGCAGCGTGGGCGAAATGAGCAGCAACAGCACCAGCCTGGCGCTG\\n',\n",
      " 'GTGCGCGCGACCAGCAGCACCGGCAGCACCAAAAGCAACACCGATAAAAACCTGAGCGTG\\n',\n",
      " 'GATAGCATTAGCCGCGGCATTCTGATTAACAGCCCGCATACCATTGCGATGCAGAAACGC\\n',\n",
      " 'ATTCTGATTAACCTGCGCATTGTGGAATATAAACTGGCGGAACTGGAATATTTTCTGATT\\n',\n",
      " 'ATTAAAGGCCTGAACGGCACCCTGGCGAACTGCAAAAGCACCGATCGCCTGCGCGAATGC\\n',\n",
      " 'AACGAAAGCGAAGGCAACCAT\\n',\n",
      " '\\n',\n",
      " '>XP_0026997503.1 kita-kyushu lung cancer antigen 1 [Bos taurus]\\n',\n",
      " '\\n',\n",
      " 'ATGAGCTTTCTGTTTCTGCTGGTGAGCAGCATTCTGTTTGCGTTTATGTTTGTGTTTTGG\\n',\n",
      " 'AAAACCTGGTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCACCAGCCTGGCGCTG\\n',\n",
      " 'GTGCGCAGCAGCAGCAGCACCCAGAGCACCAAAAGCAACATTGATAAAAGCCTGCTGGTG\\n',\n",
      " 'AACAGCCTGAGCCGCGATATTCTGATTAACTTTCCGCATAGCATTGCGATGCAGCGCCGC\\n',\n",
      " 'ATTCTGGTGAACCTGCGCATTGTGGATTATAAACTGGCGGAACTGGAACAGTTTCTGGTG\\n',\n",
      " 'ACCAAAGGCCTGAACGGCGCGCTGGTGAACCGCAAAAGCACCGATAAACCGACCGAAGTG\\n',\n",
      " '\\n',\n",
      " '>XP_002699603.1 kita-kyushu lung cancer antigen 1 [Bos taurus]\\n',\n",
      " '\\n',\n",
      " 'ATGAGCTTTCTGTTTCTGCTGGTGAGCAGCATTCTGTTTGCGTTTATGTTTGTGTTTTGG\\n',\n",
      " 'AAAACCTGGTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCACCAGCAGCCATTAT\\n',\n",
      " 'CGCGCGTGCGTGGTGGTGGTGGGCCATAAACTGAGCCTGAGCAAATTTAACTTTAACTTT\\n',\n",
      " 'TGGCATCAGCGCGTGAAA\\n',\n",
      " '\\n',\n",
      " '>XP_004699503.1 kita-kyushu lung cancer antigen 1 [Bos taurus]\\n',\n",
      " '\\n',\n",
      " 'ATGAGCTTTCTGTTTCTGCTGGTGAGCAGCATTCTGTTTGCGTTTATGTTTGTGTTTTGG\\n',\n",
      " 'AAAACCTGGTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCGATAAACCGACCGAA\\n',\n",
      " 'GTGAACGGCGCGCTGGTGAACCGCAAAAGC\\n']\n"
     ]
    }
   ],
   "source": [
    "\"\"\"Display Identifiers and Nucleotides along with Trailing Newline-Characters\"\"\"\n",
    "#The pprint module provides a capability to “pretty-print” arbitrary Python data structures in a form \n",
    "#which can be used as input to the interpreter.\n",
    "from pprint import pprint\n",
    "\n",
    "fasta_file = \"DNA.fasta\"\n",
    "\n",
    "def parse_fasta(fname):\n",
    "    contents = []\n",
    "    with open(fname, \"r\") as fh:\n",
    "        for line in fh:\n",
    "            contents.append(line)\n",
    "    return contents\n",
    "\n",
    "fasta = parse_fasta(fasta_file)\n",
    "pprint(fasta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['>NP_001273979.1 stabilizer of axonemal microtubules 1 isoform c [Homo '\n",
      " 'sapiens]\\n',\n",
      " '>NP_001017978.1 kita-kyushu lung cancer antigen 1 [Homo sapiens]\\n',\n",
      " '>sp|Q5H943.1|KKLC1_HUMAN RecName: Full=Kita-kyushu lung cancer antigen 1; '\n",
      " 'Short=KK-LC-1; AltName: Full=Cancer/testis antigen 83\\n',\n",
      " '>AAC35497.1 lung cancer candidate FUS1 [Homo sapiens]\\n',\n",
      " '>XP_002699503.1 kita-kyushu lung cancer antigen 1 [Bos taurus]\\n',\n",
      " '>sp|Q4R717.1|KKLC1_MACFA RecName: Full=Kita-kyushu lung cancer antigen 1 '\n",
      " 'homolog\\n',\n",
      " '>AAN12271.1 lung cancer oncogene 5 [Homo sapiens]\\n',\n",
      " '>NP_001270680.1 kita-kyushu lung cancer antigen 1 homolog [Macaca '\n",
      " 'fascicularis]\\n',\n",
      " '>Sp|P63244.3|Rack1_Human Recname: Full=Receptor Of Activated Protein C '\n",
      " 'Kinase 1; Altname: Full=Cell Proliferation-Inducing Gene 21 Protein; '\n",
      " 'Altname: Full=Guanine Nucleotide-Binding Protein Subunit Beta-2-Like 1; '\n",
      " 'Altname: Full=Guanine Nucleotide-Binding Protein Subunit Beta-Like Protein '\n",
      " '12.3; Altname: Full=Human Lung Cancer Oncogene 7 Protein; Short=Hlc-7; '\n",
      " 'Altname: Full=Receptor For Activated C Kinase; Altname: Full=Small Ribosomal '\n",
      " 'Subunit Protein Rack1; Contains: Recname: Full=Receptor Of Activated Protein '\n",
      " 'C Kinase 1, N-Terminally Processed; Altname: Full=Guanine Nucleotide-Binding '\n",
      " 'Protein Subunit Beta-2-Like 1, N-Terminally Processed\\n',\n",
      " '>XP_027389897.1 kita-kyushu lung cancer antigen 1 [Bos indicus x Bos '\n",
      " 'taurus]\\n',\n",
      " '>XP_007647427.1 kita-kyushu lung cancer antigen 1 [Cricetulus griseus]\\n',\n",
      " '>XP_014918346.1 kita-kyushu lung cancer antigen 1 [Acinonyx jubatus]\\n',\n",
      " '>XP_0026997503.1 kita-kyushu lung cancer antigen 1 [Bos taurus]\\n',\n",
      " '>XP_002699603.1 kita-kyushu lung cancer antigen 1 [Bos taurus]\\n',\n",
      " '>XP_004699503.1 kita-kyushu lung cancer antigen 1 [Bos taurus]\\n']\n"
     ]
    }
   ],
   "source": [
    "\"\"\"Display Only Identifiers with > \"\"\"\n",
    "#The pprint module provides a capability to “pretty-print” arbitrary Python data structures in a form \n",
    "#which can be used as input to the interpreter.\n",
    "from pprint import pprint\n",
    "\n",
    "fasta_file = \"DNA.fasta\"\n",
    "\n",
    "def parse_fasta(fname):\n",
    "    identifiers = []\n",
    "    with open(fname, \"r\") as fh:\n",
    "        for line in fh:\n",
    "            if line.startswith(\">\"):\n",
    "                identifiers.append(line)\n",
    "    return identifiers\n",
    "\n",
    "fasta = parse_fasta(fasta_file)\n",
    "pprint(fasta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('>NP_001273979.1 stabilizer of axonemal microtubules 1 isoform c [Homo '\n",
      " 'sapiens]',\n",
      " ['',\n",
      "  'ATGAAAACCAAATGCATTTGCGAACTGTGCAGCTGCGGCCGCCATCATTGCCCGCATCTG',\n",
      "  'CCGACCAAAATTTATGATAAAACCGAAAAACCGTGCCTGCTGAGCGAATATACCGAAAAC',\n",
      "  'TATCCGTTTTATCATAGCTATCTGCCGCGCGAAAGCTTTAAACCGCGCCGCGAATATCAG',\n",
      "  'AAAGGCCCGATTCCGATGGAAGGCCTGACCACCAGCAGC',\n",
      "  '',\n",
      "  '',\n",
      "  ''])\n",
      "('>NP_001017978.1 kita-kyushu lung cancer antigen 1 [Homo sapiens]',\n",
      " ['',\n",
      "  'ATGAACTTTTATCTGCTGCTGGCGAGCAGCATTCTGTGCGCGCTGATTGTGTTTTGGAAA',\n",
      "  'TATCGCCGCTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCACCGCGCTGGCGCTG',\n",
      "  'GTGCGCCCGAGCAGCAGCGGCCTGATTAACAGCAACACCGATAACAACCTGGCGGTGTAT',\n",
      "  'GATCTGAGCCGCGATATTCTGAACAACTTTCCGCATAGCATTGCGCGCCAGAAACGCATT',\n",
      "  'CTGGTGAACCTGAGCATGGTGGAAAACAAACTGGTGGAACTGGAACATACCCTGCTGAGC',\n",
      "  'AAAGGCTTTCGCGGCGCGAGCCCGCATCGCAAAAGCACC',\n",
      "  '',\n",
      "  ''])\n",
      "('>sp|Q5H943.1|KKLC1_HUMAN RecName: Full=Kita-kyushu lung cancer antigen 1; '\n",
      " 'Short=KK-LC-1; AltName: Full=Cancer/testis antigen 83',\n",
      " ['',\n",
      "  'ATGAACTTTTATCTGCTGCTGGCGAGCAGCATTCTGTGCGCGCTGATTGTGTTTTGGAAA',\n",
      "  'TATCGCCGCTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCACCGCGCTGGCGCTG',\n",
      "  'GTGCGCCCGAGCAGCAGCGGCCTGATTAACAGCAACACCGATAACAACCTGGCGGTGTAT',\n",
      "  'GATCTGAGCCGCGATATTCTGAACAACTTTCCGCATAGCATTGCGCGCCAGAAACGCATT',\n",
      "  'CTGGTGAACCTGAGCATGGTGGAAAACAAACTGGTGGAACTGGAACATACCCTGCTGAGC',\n",
      "  'AAAGGCTTTCGCGGCGCGAGCCCGCATCGCAAAAGCACC',\n",
      "  '',\n",
      "  ''])\n",
      "('>AAC35497.1 lung cancer candidate FUS1 [Homo sapiens]',\n",
      " ['',\n",
      "  'ATGGGCGCGAGCGGCAGCAAAGCGCGCGGCCTGTGGCCGTTTGCGAGCGCGGCGGGCGGC',\n",
      "  'GGCGGCAGCGAAGCGGCGGGCGCGGAACAGGCGCTGGTGCGCCCGCGCGGCCGCGCGGTG',\n",
      "  'CCGCCGTTTGTGTTTACCCGCCGCGGCAGCATGTTTTATGATGAAGATGGCGATCTGGCG',\n",
      "  'CATGAATTTTATGAAGAAACCATTGTGACCAAAAACGGCCAGAAACGCGCGAAACTGCGC',\n",
      "  'CGCGTGCATAAAAACCTGATTCCGCAGGGCATTGTGAAACTGGATCATCCGCGCATTCAT',\n",
      "  'GTGGATTTTCCGGTGATTCTGTATGAAGTG',\n",
      "  '',\n",
      "  ''])\n",
      "('>XP_002699503.1 kita-kyushu lung cancer antigen 1 [Bos taurus]',\n",
      " ['',\n",
      "  'ATGAGCTTTCTGTTTCTGCTGGTGAGCAGCATTCTGTTTGCGTTTATGTTTGTGTTTTGG',\n",
      "  'AAAACCTGGTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCACCAGCCTGGCGCTG',\n",
      "  'GTGCGCAGCAGCAGCAGCACCCAGAGCACCAAAAGCAACATTGATAAAAGCCTGCTGGTG',\n",
      "  'AACAGCCTGAGCCGCGATATTCTGATTAACTTTCCGCATAGCATTGCGATGCAGCGCCGC',\n",
      "  'ATTCTGGTGAACCTGCGCATTGTGGATTATAAACTGGCGGAACTGGAACAGTTTCTGGTG',\n",
      "  'ACCAAAGGCCTGAACGGCGCGCTGGTGAACCGCAAAAGCACCGATAAACCGACCGAAGTG',\n",
      "  '',\n",
      "  ''])\n",
      "('>sp|Q4R717.1|KKLC1_MACFA RecName: Full=Kita-kyushu lung cancer antigen 1 '\n",
      " 'homolog',\n",
      " ['',\n",
      "  'ATGAACGTGTATCTGCTGCTGGCGAGCGGCATTCTGTGCGCGCTGATGACCGTGTTTTGG',\n",
      "  'AAATATCGCCGCTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCACCGCGCTGGCG',\n",
      "  'CTGGTGCGCCCGAGCAGCACCGGCCTGATTAACAGCAACACCGATAACAACCTGAGCGTG',\n",
      "  'TATGATCTGAGCCGCGATATTCTGAACAACTTTCCGCATAGCATTGCGATGCAGAAACGC',\n",
      "  'ATTCTGGTGAACCTGACCACCGTGGAAAACAAACTGGTGGAACTGGAACATATTCTGGTG',\n",
      "  'AGCAAAGGCTTTCGCAGCGCGAGCGCGCATCGCAAAAGCACC',\n",
      "  '',\n",
      "  '',\n",
      "  ''])\n",
      "('>AAN12271.1 lung cancer oncogene 5 [Homo sapiens]',\n",
      " ['',\n",
      "  'ATGCTGCTGACCCGCTTTGAACAGAAAAACGGCCATCTGGCGCAGGTGGAAGTGGATGAA',\n",
      "  'GTGTTTGGCTTTATGTGCCATATTACCACCGAAGTGCCGCCGCATGATGCGATGCCGGGC',\n",
      "  'GGCATTGTGCTGCTGGTGAAATTTCTGCTGGATATGAGCCGCAACGTGCTGCTGTATGTG',\n",
      "  'GTGTTTCTGCAGCGCCTGAGCAGCGCGCTGCATCGCGTGCTGCTGCATCTGTTTCGCCAT',\n",
      "  'GTGCGCATTTTTGATCATGGCCTGAGCGTGGCGCATGGCTATCGCGGCGAAGGCGGCTGG',\n",
      "  'CCGACCGCGACCGTGAGCTGGGGCCGCTGC',\n",
      "  '',\n",
      "  ''])\n",
      "('>NP_001270680.1 kita-kyushu lung cancer antigen 1 homolog [Macaca '\n",
      " 'fascicularis]',\n",
      " ['',\n",
      "  'ATGAACGTGTATCTGCTGCTGGCGAGCGGCATTCTGTGCGCGCTGATGACCGTGTTTTGG',\n",
      "  'AAATATCGCCGCTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCACCGCGCTGGCG',\n",
      "  'CTGGTGCGCCCGAGCAGCACCGGCCTGATTAACAGCAACACCGATAACAACCTGAGCGTG',\n",
      "  'TATGATCTGAGCCGCGATATTCTGAACAACTTTCCGCATAGCATTGCGATGCAGAAACGC',\n",
      "  'ATTCTGGTGAACCTGACCACCGTGGAAAACAAACTGGTGGAACTGGAACATATTCTGGTG',\n",
      "  'AGCAAAGGCTTTCGCAGCGCGAGCGCGCATCGCAAAAGCACC',\n",
      "  '',\n",
      "  ''])\n",
      "('>Sp|P63244.3|Rack1_Human Recname: Full=Receptor Of Activated Protein C '\n",
      " 'Kinase 1; Altname: Full=Cell Proliferation-Inducing Gene 21 Protein; '\n",
      " 'Altname: Full=Guanine Nucleotide-Binding Protein Subunit Beta-2-Like 1; '\n",
      " 'Altname: Full=Guanine Nucleotide-Binding Protein Subunit Beta-Like Protein '\n",
      " '12.3; Altname: Full=Human Lung Cancer Oncogene 7 Protein; Short=Hlc-7; '\n",
      " 'Altname: Full=Receptor For Activated C Kinase; Altname: Full=Small Ribosomal '\n",
      " 'Subunit Protein Rack1; Contains: Recname: Full=Receptor Of Activated Protein '\n",
      " 'C Kinase 1, N-Terminally Processed; Altname: Full=Guanine Nucleotide-Binding '\n",
      " 'Protein Subunit Beta-2-Like 1, N-Terminally Processed',\n",
      " ['',\n",
      "  'ATGACCGAACAGATGACCCTGCGCGGCACCCTGAAAGGCCATAACGGCTGGGTGACCCAG',\n",
      "  'ATTGCGACCACCCCGCAGTTTCCGGATATGATTCTGAGCGCGAGCCGCGATAAAACCATT',\n",
      "  'ATTATGTGGAAACTGACCCGCGATGAAACCAACTATGGCATTCCGCAGCGCGCGCTGCGC',\n",
      "  'GGCCATAGCCATTTTGTGAGCGATGTGGTGATTAGCAGCGATGGCCAGTTTGCGCTGAGC',\n",
      "  'GGCAGCTGGGATGGCACCCTGCGCCTGTGGGATCTGACCACCGGCACCACCACCCGCCGC',\n",
      "  'TTTGTGGGCCATACCAAAGATGTGCTGAGCGTGGCGTTTAGCAGCGATAACCGCCAGATT',\n",
      "  'GTGAGCGGCAGCCGCGATAAAACCATTAAACTGTGGAACACCCTGGGCGTGTGCAAATAT',\n",
      "  'ACCGTGCAGGATGAAAGCCATAGCGAATGGGTGAGCTGCGTGCGCTTTAGCCCGAACAGC',\n",
      "  'AGCAACCCGATTATTGTGAGCTGCGGCTGGGATAAACTGGTGAAAGTGTGGAACCTGGCG',\n",
      "  'AACTGCAAACTGAAAACCAACCATATTGGCCATACCGGCTATCTGAACACCGTGACCGTG',\n",
      "  'AGCCCGGATGGCAGCCTGTGCGCGAGCGGCGGCAAAGATGGCCAGGCGATGCTGTGGGAT',\n",
      "  'CTGAACGAAGGCAAACATCTGTATACCCTGGATGGCGGCGATATTATTAACGCGCTGTGC',\n",
      "  'TTTAGCCCGAACCGCTATTGGCTGTGCGCGGCGACCGGCCCGAGCATTAAAATTTGGGAT',\n",
      "  'CTGGAAGGCAAAATTATTGTGGATGAACTGAAACAGGAAGTGATTAGCACCAGCAGCAAA',\n",
      "  'GAACCGCCGCAGTGCACCAGCCTGGCGTGGAGCGCGGATGGCCAGACCCTGTTTGCGGGC',\n",
      "  'TATACCGATAACCTGGTGCGCGTGTGGCAGGTGACCATTGGCACCCGC',\n",
      "  ''])\n",
      "('>XP_027389897.1 kita-kyushu lung cancer antigen 1 [Bos indicus x Bos taurus]',\n",
      " ['',\n",
      "  'ATGAGCTTTCTGTTTCTGCTGGTGAGCAGCATTCTGTTTGCGTTTATGTTTGTGTTTTGG',\n",
      "  'AAAACCTGGTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCACCAGCCTGGCGCTG',\n",
      "  'GTGCGCAGCAGCAGCAGCACCCAGAGCACCAAAAGCAACATTGATAAAAGCCTGCTGGTG',\n",
      "  'AACAGCCTGAGCCGCGATATTCTGATTAACTTTCCGCATAGCATTGCGATGCAGCGCCGC',\n",
      "  'ATTCTGGTGAACCTGCGCATTGTGGATTATAAACTGGCGGAACTGGAACAGTTTCTGGTG',\n",
      "  'ACCAAAGGCCTGAACGGCGCGCTGGTGAACCGCAAAAGCACCGATAAACCGACCGAAGTG',\n",
      "  '',\n",
      "  ''])\n",
      "('>XP_007647427.1 kita-kyushu lung cancer antigen 1 [Cricetulus griseus]',\n",
      " ['',\n",
      "  'ATGAACATTCTGTTTATTCTGGTGGGCAGCATTATTTTTGCGAGCATTTTTAGCATTTTT',\n",
      "  'AAAGTGGTGTTTCAGATTCCGGATCATGAACGCCCGAGCAACCCGACCAGCCCGACCCTG',\n",
      "  'CTGCGCGCGAACAGCTTTTGGAGCTATCGCAACACCGGCCTGAGCTTTCCGGGCAACAGC',\n",
      "  'TATAACCAGGATATGATGAACAACTTTCCGCGCACCCTGGCGCTGCATAAACGCATTCTG',\n",
      "  'GTGAACCTGAACATTATGGAATGCCAGCTGACCAAACTGGAACAGTTTCTGAGCAGCAAA',\n",
      "  'AACCCGCAGGGCCATACCAAACATCGCCGCCGCGCGACCGTGCGCATTCCGATGGAAAGC',\n",
      "  'GATAGCGGCAGCAACCAG',\n",
      "  '',\n",
      "  ''])\n",
      "('>XP_014918346.1 kita-kyushu lung cancer antigen 1 [Acinonyx jubatus]',\n",
      " ['',\n",
      "  'ATGAGCATGCTGCTGCTGCTGCTGAGCTTTATTGTGTTTACCCTGCTGCTGGTGGTGTGG',\n",
      "  'AAAAGCCGCTTTCAGAGCAGCGTGGGCGAAATGAGCAGCAACAGCACCAGCCTGGCGCTG',\n",
      "  'GTGCGCGCGACCAGCAGCACCGGCAGCACCAAAAGCAACACCGATAAAAACCTGAGCGTG',\n",
      "  'GATAGCATTAGCCGCGGCATTCTGATTAACAGCCCGCATACCATTGCGATGCAGAAACGC',\n",
      "  'ATTCTGATTAACCTGCGCATTGTGGAATATAAACTGGCGGAACTGGAATATTTTCTGATT',\n",
      "  'ATTAAAGGCCTGAACGGCACCCTGGCGAACTGCAAAAGCACCGATCGCCTGCGCGAATGC',\n",
      "  'AACGAAAGCGAAGGCAACCAT',\n",
      "  ''])\n",
      "('>XP_0026997503.1 kita-kyushu lung cancer antigen 1 [Bos taurus]',\n",
      " ['',\n",
      "  'ATGAGCTTTCTGTTTCTGCTGGTGAGCAGCATTCTGTTTGCGTTTATGTTTGTGTTTTGG',\n",
      "  'AAAACCTGGTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCACCAGCCTGGCGCTG',\n",
      "  'GTGCGCAGCAGCAGCAGCACCCAGAGCACCAAAAGCAACATTGATAAAAGCCTGCTGGTG',\n",
      "  'AACAGCCTGAGCCGCGATATTCTGATTAACTTTCCGCATAGCATTGCGATGCAGCGCCGC',\n",
      "  'ATTCTGGTGAACCTGCGCATTGTGGATTATAAACTGGCGGAACTGGAACAGTTTCTGGTG',\n",
      "  'ACCAAAGGCCTGAACGGCGCGCTGGTGAACCGCAAAAGCACCGATAAACCGACCGAAGTG',\n",
      "  ''])\n",
      "('>XP_002699603.1 kita-kyushu lung cancer antigen 1 [Bos taurus]',\n",
      " ['',\n",
      "  'ATGAGCTTTCTGTTTCTGCTGGTGAGCAGCATTCTGTTTGCGTTTATGTTTGTGTTTTGG',\n",
      "  'AAAACCTGGTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCACCAGCAGCCATTAT',\n",
      "  'CGCGCGTGCGTGGTGGTGGTGGGCCATAAACTGAGCCTGAGCAAATTTAACTTTAACTTT',\n",
      "  'TGGCATCAGCGCGTGAAA',\n",
      "  ''])\n"
     ]
    }
   ],
   "source": [
    "\"\"\"Display Identifiers and Nucleotides as Individual Sequence without Trailing Newline Characters\"\"\"\n",
    "#The pprint module provides a capability to “pretty-print” arbitrary Python data structures in a form \n",
    "#which can be used as input to the interpreter.\n",
    "from pprint import pprint\n",
    "\n",
    "fasta_file = \"DNA.fasta\"\n",
    "\n",
    "def parse_fasta(fname):\n",
    "    with open(fname, \"r\") as fh:\n",
    "\n",
    "        # Create variables for storing the identifiers and the sequence.\n",
    "        identifier = None\n",
    "        sequence = []\n",
    "\n",
    "        for line in fh:\n",
    "            line = line.strip()  # Remove trailing newline characters.\n",
    "            if line.startswith(\">\"):\n",
    "                if identifier is None:\n",
    "                    # This only happens when the first line of the\n",
    "                    # FASTA file is parsed.\n",
    "                    identifier = line\n",
    "                else:\n",
    "                    # This happens every time a new FASTA record is\n",
    "                    # encountered.\n",
    "\n",
    "                    # Start by yielding the entry that has been built up.\n",
    "                    yield identifier, sequence\n",
    "\n",
    "                    # Then reinitialise the identifier and sequence\n",
    "                    # variables to build up a new record.\n",
    "                    identifier = line\n",
    "                    sequence = []\n",
    "            else:\n",
    "                # This happens every time a sequence line is encountered.\n",
    "                sequence.append(line)\n",
    "\n",
    "for entry in parse_fasta(fasta_file):\n",
    "    pprint(entry)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('>NP_001273979.1 stabilizer of axonemal microtubules 1 isoform c [Homo sapiens]', ['A', 'T', 'G', 'A', 'A', 'A', 'A', 'C', 'C', 'A', 'A', 'A', 'T', 'G', 'C', 'A', 'T', 'T', 'T', 'G', 'C', 'G', 'A', 'A', 'C', 'T', 'G', 'T', 'G', 'C', 'A', 'G', 'C', 'T', 'G', 'C', 'G', 'G', 'C', 'C', 'G', 'C', 'C', 'A', 'T', 'C', 'A', 'T', 'T', 'G', 'C', 'C', 'C', 'G', 'C', 'A', 'T', 'C', 'T', 'G', 'C', 'C', 'G', 'A', 'C', 'C', 'A', 'A', 'A', 'A', 'T', 'T', 'T', 'A', 'T', 'G', 'A', 'T', 'A', 'A', 'A', 'A', 'C', 'C', 'G', 'A', 'A', 'A', 'A', 'A', 'C', 'C', 'G', 'T', 'G', 'C', 'C', 'T', 'G', 'C', 'T', 'G', 'A', 'G', 'C', 'G', 'A', 'A', 'T', 'A', 'T', 'A', 'C', 'C', 'G', 'A', 'A', 'A', 'A', 'C', 'T', 'A', 'T', 'C', 'C', 'G', 'T', 'T', 'T', 'T', 'A', 'T', 'C', 'A', 'T', 'A', 'G', 'C', 'T', 'A', 'T', 'C', 'T', 'G', 'C', 'C', 'G', 'C', 'G', 'C', 'G', 'A', 'A', 'A', 'G', 'C', 'T', 'T', 'T', 'A', 'A', 'A', 'C', 'C', 'G', 'C', 'G', 'C', 'C', 'G', 'C', 'G', 'A', 'A', 'T', 'A', 'T', 'C', 'A', 'G', 'A', 'A', 'A', 'G', 'G', 'C', 'C', 'C', 'G', 'A', 'T', 'T', 'C', 'C', 'G', 'A', 'T', 'G', 'G', 'A', 'A', 'G', 'G', 'C', 'C', 'T', 'G', 'A', 'C', 'C', 'A', 'C', 'C', 'A', 'G', 'C', 'A', 'G', 'C'])\n",
      "('>NP_001017978.1 kita-kyushu lung cancer antigen 1 [Homo sapiens]', ['A', 'T', 'G', 'A', 'A', 'C', 'T', 'T', 'T', 'T', 'A', 'T', 'C', 'T', 'G', 'C', 'T', 'G', 'C', 'T', 'G', 'G', 'C', 'G', 'A', 'G', 'C', 'A', 'G', 'C', 'A', 'T', 'T', 'C', 'T', 'G', 'T', 'G', 'C', 'G', 'C', 'G', 'C', 'T', 'G', 'A', 'T', 'T', 'G', 'T', 'G', 'T', 'T', 'T', 'T', 'G', 'G', 'A', 'A', 'A', 'T', 'A', 'T', 'C', 'G', 'C', 'C', 'G', 'C', 'T', 'T', 'T', 'C', 'A', 'G', 'C', 'G', 'C', 'A', 'A', 'C', 'A', 'C', 'C', 'G', 'G', 'C', 'G', 'A', 'A', 'A', 'T', 'G', 'A', 'G', 'C', 'A', 'G', 'C', 'A', 'A', 'C', 'A', 'G', 'C', 'A', 'C', 'C', 'G', 'C', 'G', 'C', 'T', 'G', 'G', 'C', 'G', 'C', 'T', 'G', 'G', 'T', 'G', 'C', 'G', 'C', 'C', 'C', 'G', 'A', 'G', 'C', 'A', 'G', 'C', 'A', 'G', 'C', 'G', 'G', 'C', 'C', 'T', 'G', 'A', 'T', 'T', 'A', 'A', 'C', 'A', 'G', 'C', 'A', 'A', 'C', 'A', 'C', 'C', 'G', 'A', 'T', 'A', 'A', 'C', 'A', 'A', 'C', 'C', 'T', 'G', 'G', 'C', 'G', 'G', 'T', 'G', 'T', 'A', 'T', 'G', 'A', 'T', 'C', 'T', 'G', 'A', 'G', 'C', 'C', 'G', 'C', 'G', 'A', 'T', 'A', 'T', 'T', 'C', 'T', 'G', 'A', 'A', 'C', 'A', 'A', 'C', 'T', 'T', 'T', 'C', 'C', 'G', 'C', 'A', 'T', 'A', 'G', 'C', 'A', 'T', 'T', 'G', 'C', 'G', 'C', 'G', 'C', 'C', 'A', 'G', 'A', 'A', 'A', 'C', 'G', 'C', 'A', 'T', 'T', 'C', 'T', 'G', 'G', 'T', 'G', 'A', 'A', 'C', 'C', 'T', 'G', 'A', 'G', 'C', 'A', 'T', 'G', 'G', 'T', 'G', 'G', 'A', 'A', 'A', 'A', 'C', 'A', 'A', 'A', 'C', 'T', 'G', 'G', 'T', 'G', 'G', 'A', 'A', 'C', 'T', 'G', 'G', 'A', 'A', 'C', 'A', 'T', 'A', 'C', 'C', 'C', 'T', 'G', 'C', 'T', 'G', 'A', 'G', 'C', 'A', 'A', 'A', 'G', 'G', 'C', 'T', 'T', 'T', 'C', 'G', 'C', 'G', 'G', 'C', 'G', 'C', 'G', 'A', 'G', 'C', 'C', 'C', 'G', 'C', 'A', 'T', 'C', 'G', 'C', 'A', 'A', 'A', 'A', 'G', 'C', 'A', 'C', 'C'])\n",
      "('>sp|Q5H943.1|KKLC1_HUMAN RecName: Full=Kita-kyushu lung cancer antigen 1; Short=KK-LC-1; AltName: Full=Cancer/testis antigen 83', ['A', 'T', 'G', 'A', 'A', 'C', 'T', 'T', 'T', 'T', 'A', 'T', 'C', 'T', 'G', 'C', 'T', 'G', 'C', 'T', 'G', 'G', 'C', 'G', 'A', 'G', 'C', 'A', 'G', 'C', 'A', 'T', 'T', 'C', 'T', 'G', 'T', 'G', 'C', 'G', 'C', 'G', 'C', 'T', 'G', 'A', 'T', 'T', 'G', 'T', 'G', 'T', 'T', 'T', 'T', 'G', 'G', 'A', 'A', 'A', 'T', 'A', 'T', 'C', 'G', 'C', 'C', 'G', 'C', 'T', 'T', 'T', 'C', 'A', 'G', 'C', 'G', 'C', 'A', 'A', 'C', 'A', 'C', 'C', 'G', 'G', 'C', 'G', 'A', 'A', 'A', 'T', 'G', 'A', 'G', 'C', 'A', 'G', 'C', 'A', 'A', 'C', 'A', 'G', 'C', 'A', 'C', 'C', 'G', 'C', 'G', 'C', 'T', 'G', 'G', 'C', 'G', 'C', 'T', 'G', 'G', 'T', 'G', 'C', 'G', 'C', 'C', 'C', 'G', 'A', 'G', 'C', 'A', 'G', 'C', 'A', 'G', 'C', 'G', 'G', 'C', 'C', 'T', 'G', 'A', 'T', 'T', 'A', 'A', 'C', 'A', 'G', 'C', 'A', 'A', 'C', 'A', 'C', 'C', 'G', 'A', 'T', 'A', 'A', 'C', 'A', 'A', 'C', 'C', 'T', 'G', 'G', 'C', 'G', 'G', 'T', 'G', 'T', 'A', 'T', 'G', 'A', 'T', 'C', 'T', 'G', 'A', 'G', 'C', 'C', 'G', 'C', 'G', 'A', 'T', 'A', 'T', 'T', 'C', 'T', 'G', 'A', 'A', 'C', 'A', 'A', 'C', 'T', 'T', 'T', 'C', 'C', 'G', 'C', 'A', 'T', 'A', 'G', 'C', 'A', 'T', 'T', 'G', 'C', 'G', 'C', 'G', 'C', 'C', 'A', 'G', 'A', 'A', 'A', 'C', 'G', 'C', 'A', 'T', 'T', 'C', 'T', 'G', 'G', 'T', 'G', 'A', 'A', 'C', 'C', 'T', 'G', 'A', 'G', 'C', 'A', 'T', 'G', 'G', 'T', 'G', 'G', 'A', 'A', 'A', 'A', 'C', 'A', 'A', 'A', 'C', 'T', 'G', 'G', 'T', 'G', 'G', 'A', 'A', 'C', 'T', 'G', 'G', 'A', 'A', 'C', 'A', 'T', 'A', 'C', 'C', 'C', 'T', 'G', 'C', 'T', 'G', 'A', 'G', 'C', 'A', 'A', 'A', 'G', 'G', 'C', 'T', 'T', 'T', 'C', 'G', 'C', 'G', 'G', 'C', 'G', 'C', 'G', 'A', 'G', 'C', 'C', 'C', 'G', 'C', 'A', 'T', 'C', 'G', 'C', 'A', 'A', 'A', 'A', 'G', 'C', 'A', 'C', 'C'])\n",
      "('>AAC35497.1 lung cancer candidate FUS1 [Homo sapiens]', ['A', 'T', 'G', 'G', 'G', 'C', 'G', 'C', 'G', 'A', 'G', 'C', 'G', 'G', 'C', 'A', 'G', 'C', 'A', 'A', 'A', 'G', 'C', 'G', 'C', 'G', 'C', 'G', 'G', 'C', 'C', 'T', 'G', 'T', 'G', 'G', 'C', 'C', 'G', 'T', 'T', 'T', 'G', 'C', 'G', 'A', 'G', 'C', 'G', 'C', 'G', 'G', 'C', 'G', 'G', 'G', 'C', 'G', 'G', 'C', 'G', 'G', 'C', 'G', 'G', 'C', 'A', 'G', 'C', 'G', 'A', 'A', 'G', 'C', 'G', 'G', 'C', 'G', 'G', 'G', 'C', 'G', 'C', 'G', 'G', 'A', 'A', 'C', 'A', 'G', 'G', 'C', 'G', 'C', 'T', 'G', 'G', 'T', 'G', 'C', 'G', 'C', 'C', 'C', 'G', 'C', 'G', 'C', 'G', 'G', 'C', 'C', 'G', 'C', 'G', 'C', 'G', 'G', 'T', 'G', 'C', 'C', 'G', 'C', 'C', 'G', 'T', 'T', 'T', 'G', 'T', 'G', 'T', 'T', 'T', 'A', 'C', 'C', 'C', 'G', 'C', 'C', 'G', 'C', 'G', 'G', 'C', 'A', 'G', 'C', 'A', 'T', 'G', 'T', 'T', 'T', 'T', 'A', 'T', 'G', 'A', 'T', 'G', 'A', 'A', 'G', 'A', 'T', 'G', 'G', 'C', 'G', 'A', 'T', 'C', 'T', 'G', 'G', 'C', 'G', 'C', 'A', 'T', 'G', 'A', 'A', 'T', 'T', 'T', 'T', 'A', 'T', 'G', 'A', 'A', 'G', 'A', 'A', 'A', 'C', 'C', 'A', 'T', 'T', 'G', 'T', 'G', 'A', 'C', 'C', 'A', 'A', 'A', 'A', 'A', 'C', 'G', 'G', 'C', 'C', 'A', 'G', 'A', 'A', 'A', 'C', 'G', 'C', 'G', 'C', 'G', 'A', 'A', 'A', 'C', 'T', 'G', 'C', 'G', 'C', 'C', 'G', 'C', 'G', 'T', 'G', 'C', 'A', 'T', 'A', 'A', 'A', 'A', 'A', 'C', 'C', 'T', 'G', 'A', 'T', 'T', 'C', 'C', 'G', 'C', 'A', 'G', 'G', 'G', 'C', 'A', 'T', 'T', 'G', 'T', 'G', 'A', 'A', 'A', 'C', 'T', 'G', 'G', 'A', 'T', 'C', 'A', 'T', 'C', 'C', 'G', 'C', 'G', 'C', 'A', 'T', 'T', 'C', 'A', 'T', 'G', 'T', 'G', 'G', 'A', 'T', 'T', 'T', 'T', 'C', 'C', 'G', 'G', 'T', 'G', 'A', 'T', 'T', 'C', 'T', 'G', 'T', 'A', 'T', 'G', 'A', 'A', 'G', 'T', 'G'])\n",
      "('>XP_002699503.1 kita-kyushu lung cancer antigen 1 [Bos taurus]', ['A', 'T', 'G', 'A', 'G', 'C', 'T', 'T', 'T', 'C', 'T', 'G', 'T', 'T', 'T', 'C', 'T', 'G', 'C', 'T', 'G', 'G', 'T', 'G', 'A', 'G', 'C', 'A', 'G', 'C', 'A', 'T', 'T', 'C', 'T', 'G', 'T', 'T', 'T', 'G', 'C', 'G', 'T', 'T', 'T', 'A', 'T', 'G', 'T', 'T', 'T', 'G', 'T', 'G', 'T', 'T', 'T', 'T', 'G', 'G', 'A', 'A', 'A', 'A', 'C', 'C', 'T', 'G', 'G', 'T', 'T', 'T', 'C', 'A', 'G', 'C', 'G', 'C', 'A', 'A', 'C', 'A', 'C', 'C', 'G', 'G', 'C', 'G', 'A', 'A', 'A', 'T', 'G', 'A', 'G', 'C', 'A', 'G', 'C', 'A', 'A', 'C', 'A', 'G', 'C', 'A', 'C', 'C', 'A', 'G', 'C', 'C', 'T', 'G', 'G', 'C', 'G', 'C', 'T', 'G', 'G', 'T', 'G', 'C', 'G', 'C', 'A', 'G', 'C', 'A', 'G', 'C', 'A', 'G', 'C', 'A', 'G', 'C', 'A', 'C', 'C', 'C', 'A', 'G', 'A', 'G', 'C', 'A', 'C', 'C', 'A', 'A', 'A', 'A', 'G', 'C', 'A', 'A', 'C', 'A', 'T', 'T', 'G', 'A', 'T', 'A', 'A', 'A', 'A', 'G', 'C', 'C', 'T', 'G', 'C', 'T', 'G', 'G', 'T', 'G', 'A', 'A', 'C', 'A', 'G', 'C', 'C', 'T', 'G', 'A', 'G', 'C', 'C', 'G', 'C', 'G', 'A', 'T', 'A', 'T', 'T', 'C', 'T', 'G', 'A', 'T', 'T', 'A', 'A', 'C', 'T', 'T', 'T', 'C', 'C', 'G', 'C', 'A', 'T', 'A', 'G', 'C', 'A', 'T', 'T', 'G', 'C', 'G', 'A', 'T', 'G', 'C', 'A', 'G', 'C', 'G', 'C', 'C', 'G', 'C', 'A', 'T', 'T', 'C', 'T', 'G', 'G', 'T', 'G', 'A', 'A', 'C', 'C', 'T', 'G', 'C', 'G', 'C', 'A', 'T', 'T', 'G', 'T', 'G', 'G', 'A', 'T', 'T', 'A', 'T', 'A', 'A', 'A', 'C', 'T', 'G', 'G', 'C', 'G', 'G', 'A', 'A', 'C', 'T', 'G', 'G', 'A', 'A', 'C', 'A', 'G', 'T', 'T', 'T', 'C', 'T', 'G', 'G', 'T', 'G', 'A', 'C', 'C', 'A', 'A', 'A', 'G', 'G', 'C', 'C', 'T', 'G', 'A', 'A', 'C', 'G', 'G', 'C', 'G', 'C', 'G', 'C', 'T', 'G', 'G', 'T', 'G', 'A', 'A', 'C', 'C', 'G', 'C', 'A', 'A', 'A', 'A', 'G', 'C', 'A', 'C', 'C', 'G', 'A', 'T', 'A', 'A', 'A', 'C', 'C', 'G', 'A', 'C', 'C', 'G', 'A', 'A', 'G', 'T', 'G'])\n",
      "('>sp|Q4R717.1|KKLC1_MACFA RecName: Full=Kita-kyushu lung cancer antigen 1 homolog', ['A', 'T', 'G', 'A', 'A', 'C', 'G', 'T', 'G', 'T', 'A', 'T', 'C', 'T', 'G', 'C', 'T', 'G', 'C', 'T', 'G', 'G', 'C', 'G', 'A', 'G', 'C', 'G', 'G', 'C', 'A', 'T', 'T', 'C', 'T', 'G', 'T', 'G', 'C', 'G', 'C', 'G', 'C', 'T', 'G', 'A', 'T', 'G', 'A', 'C', 'C', 'G', 'T', 'G', 'T', 'T', 'T', 'T', 'G', 'G', 'A', 'A', 'A', 'T', 'A', 'T', 'C', 'G', 'C', 'C', 'G', 'C', 'T', 'T', 'T', 'C', 'A', 'G', 'C', 'G', 'C', 'A', 'A', 'C', 'A', 'C', 'C', 'G', 'G', 'C', 'G', 'A', 'A', 'A', 'T', 'G', 'A', 'G', 'C', 'A', 'G', 'C', 'A', 'A', 'C', 'A', 'G', 'C', 'A', 'C', 'C', 'G', 'C', 'G', 'C', 'T', 'G', 'G', 'C', 'G', 'C', 'T', 'G', 'G', 'T', 'G', 'C', 'G', 'C', 'C', 'C', 'G', 'A', 'G', 'C', 'A', 'G', 'C', 'A', 'C', 'C', 'G', 'G', 'C', 'C', 'T', 'G', 'A', 'T', 'T', 'A', 'A', 'C', 'A', 'G', 'C', 'A', 'A', 'C', 'A', 'C', 'C', 'G', 'A', 'T', 'A', 'A', 'C', 'A', 'A', 'C', 'C', 'T', 'G', 'A', 'G', 'C', 'G', 'T', 'G', 'T', 'A', 'T', 'G', 'A', 'T', 'C', 'T', 'G', 'A', 'G', 'C', 'C', 'G', 'C', 'G', 'A', 'T', 'A', 'T', 'T', 'C', 'T', 'G', 'A', 'A', 'C', 'A', 'A', 'C', 'T', 'T', 'T', 'C', 'C', 'G', 'C', 'A', 'T', 'A', 'G', 'C', 'A', 'T', 'T', 'G', 'C', 'G', 'A', 'T', 'G', 'C', 'A', 'G', 'A', 'A', 'A', 'C', 'G', 'C', 'A', 'T', 'T', 'C', 'T', 'G', 'G', 'T', 'G', 'A', 'A', 'C', 'C', 'T', 'G', 'A', 'C', 'C', 'A', 'C', 'C', 'G', 'T', 'G', 'G', 'A', 'A', 'A', 'A', 'C', 'A', 'A', 'A', 'C', 'T', 'G', 'G', 'T', 'G', 'G', 'A', 'A', 'C', 'T', 'G', 'G', 'A', 'A', 'C', 'A', 'T', 'A', 'T', 'T', 'C', 'T', 'G', 'G', 'T', 'G', 'A', 'G', 'C', 'A', 'A', 'A', 'G', 'G', 'C', 'T', 'T', 'T', 'C', 'G', 'C', 'A', 'G', 'C', 'G', 'C', 'G', 'A', 'G', 'C', 'G', 'C', 'G', 'C', 'A', 'T', 'C', 'G', 'C', 'A', 'A', 'A', 'A', 'G', 'C', 'A', 'C', 'C'])\n",
      "('>AAN12271.1 lung cancer oncogene 5 [Homo sapiens]', ['A', 'T', 'G', 'C', 'T', 'G', 'C', 'T', 'G', 'A', 'C', 'C', 'C', 'G', 'C', 'T', 'T', 'T', 'G', 'A', 'A', 'C', 'A', 'G', 'A', 'A', 'A', 'A', 'A', 'C', 'G', 'G', 'C', 'C', 'A', 'T', 'C', 'T', 'G', 'G', 'C', 'G', 'C', 'A', 'G', 'G', 'T', 'G', 'G', 'A', 'A', 'G', 'T', 'G', 'G', 'A', 'T', 'G', 'A', 'A', 'G', 'T', 'G', 'T', 'T', 'T', 'G', 'G', 'C', 'T', 'T', 'T', 'A', 'T', 'G', 'T', 'G', 'C', 'C', 'A', 'T', 'A', 'T', 'T', 'A', 'C', 'C', 'A', 'C', 'C', 'G', 'A', 'A', 'G', 'T', 'G', 'C', 'C', 'G', 'C', 'C', 'G', 'C', 'A', 'T', 'G', 'A', 'T', 'G', 'C', 'G', 'A', 'T', 'G', 'C', 'C', 'G', 'G', 'G', 'C', 'G', 'G', 'C', 'A', 'T', 'T', 'G', 'T', 'G', 'C', 'T', 'G', 'C', 'T', 'G', 'G', 'T', 'G', 'A', 'A', 'A', 'T', 'T', 'T', 'C', 'T', 'G', 'C', 'T', 'G', 'G', 'A', 'T', 'A', 'T', 'G', 'A', 'G', 'C', 'C', 'G', 'C', 'A', 'A', 'C', 'G', 'T', 'G', 'C', 'T', 'G', 'C', 'T', 'G', 'T', 'A', 'T', 'G', 'T', 'G', 'G', 'T', 'G', 'T', 'T', 'T', 'C', 'T', 'G', 'C', 'A', 'G', 'C', 'G', 'C', 'C', 'T', 'G', 'A', 'G', 'C', 'A', 'G', 'C', 'G', 'C', 'G', 'C', 'T', 'G', 'C', 'A', 'T', 'C', 'G', 'C', 'G', 'T', 'G', 'C', 'T', 'G', 'C', 'T', 'G', 'C', 'A', 'T', 'C', 'T', 'G', 'T', 'T', 'T', 'C', 'G', 'C', 'C', 'A', 'T', 'G', 'T', 'G', 'C', 'G', 'C', 'A', 'T', 'T', 'T', 'T', 'T', 'G', 'A', 'T', 'C', 'A', 'T', 'G', 'G', 'C', 'C', 'T', 'G', 'A', 'G', 'C', 'G', 'T', 'G', 'G', 'C', 'G', 'C', 'A', 'T', 'G', 'G', 'C', 'T', 'A', 'T', 'C', 'G', 'C', 'G', 'G', 'C', 'G', 'A', 'A', 'G', 'G', 'C', 'G', 'G', 'C', 'T', 'G', 'G', 'C', 'C', 'G', 'A', 'C', 'C', 'G', 'C', 'G', 'A', 'C', 'C', 'G', 'T', 'G', 'A', 'G', 'C', 'T', 'G', 'G', 'G', 'G', 'C', 'C', 'G', 'C', 'T', 'G', 'C'])\n",
      "('>NP_001270680.1 kita-kyushu lung cancer antigen 1 homolog [Macaca fascicularis]', ['A', 'T', 'G', 'A', 'A', 'C', 'G', 'T', 'G', 'T', 'A', 'T', 'C', 'T', 'G', 'C', 'T', 'G', 'C', 'T', 'G', 'G', 'C', 'G', 'A', 'G', 'C', 'G', 'G', 'C', 'A', 'T', 'T', 'C', 'T', 'G', 'T', 'G', 'C', 'G', 'C', 'G', 'C', 'T', 'G', 'A', 'T', 'G', 'A', 'C', 'C', 'G', 'T', 'G', 'T', 'T', 'T', 'T', 'G', 'G', 'A', 'A', 'A', 'T', 'A', 'T', 'C', 'G', 'C', 'C', 'G', 'C', 'T', 'T', 'T', 'C', 'A', 'G', 'C', 'G', 'C', 'A', 'A', 'C', 'A', 'C', 'C', 'G', 'G', 'C', 'G', 'A', 'A', 'A', 'T', 'G', 'A', 'G', 'C', 'A', 'G', 'C', 'A', 'A', 'C', 'A', 'G', 'C', 'A', 'C', 'C', 'G', 'C', 'G', 'C', 'T', 'G', 'G', 'C', 'G', 'C', 'T', 'G', 'G', 'T', 'G', 'C', 'G', 'C', 'C', 'C', 'G', 'A', 'G', 'C', 'A', 'G', 'C', 'A', 'C', 'C', 'G', 'G', 'C', 'C', 'T', 'G', 'A', 'T', 'T', 'A', 'A', 'C', 'A', 'G', 'C', 'A', 'A', 'C', 'A', 'C', 'C', 'G', 'A', 'T', 'A', 'A', 'C', 'A', 'A', 'C', 'C', 'T', 'G', 'A', 'G', 'C', 'G', 'T', 'G', 'T', 'A', 'T', 'G', 'A', 'T', 'C', 'T', 'G', 'A', 'G', 'C', 'C', 'G', 'C', 'G', 'A', 'T', 'A', 'T', 'T', 'C', 'T', 'G', 'A', 'A', 'C', 'A', 'A', 'C', 'T', 'T', 'T', 'C', 'C', 'G', 'C', 'A', 'T', 'A', 'G', 'C', 'A', 'T', 'T', 'G', 'C', 'G', 'A', 'T', 'G', 'C', 'A', 'G', 'A', 'A', 'A', 'C', 'G', 'C', 'A', 'T', 'T', 'C', 'T', 'G', 'G', 'T', 'G', 'A', 'A', 'C', 'C', 'T', 'G', 'A', 'C', 'C', 'A', 'C', 'C', 'G', 'T', 'G', 'G', 'A', 'A', 'A', 'A', 'C', 'A', 'A', 'A', 'C', 'T', 'G', 'G', 'T', 'G', 'G', 'A', 'A', 'C', 'T', 'G', 'G', 'A', 'A', 'C', 'A', 'T', 'A', 'T', 'T', 'C', 'T', 'G', 'G', 'T', 'G', 'A', 'G', 'C', 'A', 'A', 'A', 'G', 'G', 'C', 'T', 'T', 'T', 'C', 'G', 'C', 'A', 'G', 'C', 'G', 'C', 'G', 'A', 'G', 'C', 'G', 'C', 'G', 'C', 'A', 'T', 'C', 'G', 'C', 'A', 'A', 'A', 'A', 'G', 'C', 'A', 'C', 'C'])\n",
      "('>Sp|P63244.3|Rack1_Human Recname: Full=Receptor Of Activated Protein C Kinase 1; Altname: Full=Cell Proliferation-Inducing Gene 21 Protein; Altname: Full=Guanine Nucleotide-Binding Protein Subunit Beta-2-Like 1; Altname: Full=Guanine Nucleotide-Binding Protein Subunit Beta-Like Protein 12.3; Altname: Full=Human Lung Cancer Oncogene 7 Protein; Short=Hlc-7; Altname: Full=Receptor For Activated C Kinase; Altname: Full=Small Ribosomal Subunit Protein Rack1; Contains: Recname: Full=Receptor Of Activated Protein C Kinase 1, N-Terminally Processed; Altname: Full=Guanine Nucleotide-Binding Protein Subunit Beta-2-Like 1, N-Terminally Processed', ['A', 'T', 'G', 'A', 'C', 'C', 'G', 'A', 'A', 'C', 'A', 'G', 'A', 'T', 'G', 'A', 'C', 'C', 'C', 'T', 'G', 'C', 'G', 'C', 'G', 'G', 'C', 'A', 'C', 'C', 'C', 'T', 'G', 'A', 'A', 'A', 'G', 'G', 'C', 'C', 'A', 'T', 'A', 'A', 'C', 'G', 'G', 'C', 'T', 'G', 'G', 'G', 'T', 'G', 'A', 'C', 'C', 'C', 'A', 'G', 'A', 'T', 'T', 'G', 'C', 'G', 'A', 'C', 'C', 'A', 'C', 'C', 'C', 'C', 'G', 'C', 'A', 'G', 'T', 'T', 'T', 'C', 'C', 'G', 'G', 'A', 'T', 'A', 'T', 'G', 'A', 'T', 'T', 'C', 'T', 'G', 'A', 'G', 'C', 'G', 'C', 'G', 'A', 'G', 'C', 'C', 'G', 'C', 'G', 'A', 'T', 'A', 'A', 'A', 'A', 'C', 'C', 'A', 'T', 'T', 'A', 'T', 'T', 'A', 'T', 'G', 'T', 'G', 'G', 'A', 'A', 'A', 'C', 'T', 'G', 'A', 'C', 'C', 'C', 'G', 'C', 'G', 'A', 'T', 'G', 'A', 'A', 'A', 'C', 'C', 'A', 'A', 'C', 'T', 'A', 'T', 'G', 'G', 'C', 'A', 'T', 'T', 'C', 'C', 'G', 'C', 'A', 'G', 'C', 'G', 'C', 'G', 'C', 'G', 'C', 'T', 'G', 'C', 'G', 'C', 'G', 'G', 'C', 'C', 'A', 'T', 'A', 'G', 'C', 'C', 'A', 'T', 'T', 'T', 'T', 'G', 'T', 'G', 'A', 'G', 'C', 'G', 'A', 'T', 'G', 'T', 'G', 'G', 'T', 'G', 'A', 'T', 'T', 'A', 'G', 'C', 'A', 'G', 'C', 'G', 'A', 'T', 'G', 'G', 'C', 'C', 'A', 'G', 'T', 'T', 'T', 'G', 'C', 'G', 'C', 'T', 'G', 'A', 'G', 'C', 'G', 'G', 'C', 'A', 'G', 'C', 'T', 'G', 'G', 'G', 'A', 'T', 'G', 'G', 'C', 'A', 'C', 'C', 'C', 'T', 'G', 'C', 'G', 'C', 'C', 'T', 'G', 'T', 'G', 'G', 'G', 'A', 'T', 'C', 'T', 'G', 'A', 'C', 'C', 'A', 'C', 'C', 'G', 'G', 'C', 'A', 'C', 'C', 'A', 'C', 'C', 'A', 'C', 'C', 'C', 'G', 'C', 'C', 'G', 'C', 'T', 'T', 'T', 'G', 'T', 'G', 'G', 'G', 'C', 'C', 'A', 'T', 'A', 'C', 'C', 'A', 'A', 'A', 'G', 'A', 'T', 'G', 'T', 'G', 'C', 'T', 'G', 'A', 'G', 'C', 'G', 'T', 'G', 'G', 'C', 'G', 'T', 'T', 'T', 'A', 'G', 'C', 'A', 'G', 'C', 'G', 'A', 'T', 'A', 'A', 'C', 'C', 'G', 'C', 'C', 'A', 'G', 'A', 'T', 'T', 'G', 'T', 'G', 'A', 'G', 'C', 'G', 'G', 'C', 'A', 'G', 'C', 'C', 'G', 'C', 'G', 'A', 'T', 'A', 'A', 'A', 'A', 'C', 'C', 'A', 'T', 'T', 'A', 'A', 'A', 'C', 'T', 'G', 'T', 'G', 'G', 'A', 'A', 'C', 'A', 'C', 'C', 'C', 'T', 'G', 'G', 'G', 'C', 'G', 'T', 'G', 'T', 'G', 'C', 'A', 'A', 'A', 'T', 'A', 'T', 'A', 'C', 'C', 'G', 'T', 'G', 'C', 'A', 'G', 'G', 'A', 'T', 'G', 'A', 'A', 'A', 'G', 'C', 'C', 'A', 'T', 'A', 'G', 'C', 'G', 'A', 'A', 'T', 'G', 'G', 'G', 'T', 'G', 'A', 'G', 'C', 'T', 'G', 'C', 'G', 'T', 'G', 'C', 'G', 'C', 'T', 'T', 'T', 'A', 'G', 'C', 'C', 'C', 'G', 'A', 'A', 'C', 'A', 'G', 'C', 'A', 'G', 'C', 'A', 'A', 'C', 'C', 'C', 'G', 'A', 'T', 'T', 'A', 'T', 'T', 'G', 'T', 'G', 'A', 'G', 'C', 'T', 'G', 'C', 'G', 'G', 'C', 'T', 'G', 'G', 'G', 'A', 'T', 'A', 'A', 'A', 'C', 'T', 'G', 'G', 'T', 'G', 'A', 'A', 'A', 'G', 'T', 'G', 'T', 'G', 'G', 'A', 'A', 'C', 'C', 'T', 'G', 'G', 'C', 'G', 'A', 'A', 'C', 'T', 'G', 'C', 'A', 'A', 'A', 'C', 'T', 'G', 'A', 'A', 'A', 'A', 'C', 'C', 'A', 'A', 'C', 'C', 'A', 'T', 'A', 'T', 'T', 'G', 'G', 'C', 'C', 'A', 'T', 'A', 'C', 'C', 'G', 'G', 'C', 'T', 'A', 'T', 'C', 'T', 'G', 'A', 'A', 'C', 'A', 'C', 'C', 'G', 'T', 'G', 'A', 'C', 'C', 'G', 'T', 'G', 'A', 'G', 'C', 'C', 'C', 'G', 'G', 'A', 'T', 'G', 'G', 'C', 'A', 'G', 'C', 'C', 'T', 'G', 'T', 'G', 'C', 'G', 'C', 'G', 'A', 'G', 'C', 'G', 'G', 'C', 'G', 'G', 'C', 'A', 'A', 'A', 'G', 'A', 'T', 'G', 'G', 'C', 'C', 'A', 'G', 'G', 'C', 'G', 'A', 'T', 'G', 'C', 'T', 'G', 'T', 'G', 'G', 'G', 'A', 'T', 'C', 'T', 'G', 'A', 'A', 'C', 'G', 'A', 'A', 'G', 'G', 'C', 'A', 'A', 'A', 'C', 'A', 'T', 'C', 'T', 'G', 'T', 'A', 'T', 'A', 'C', 'C', 'C', 'T', 'G', 'G', 'A', 'T', 'G', 'G', 'C', 'G', 'G', 'C', 'G', 'A', 'T', 'A', 'T', 'T', 'A', 'T', 'T', 'A', 'A', 'C', 'G', 'C', 'G', 'C', 'T', 'G', 'T', 'G', 'C', 'T', 'T', 'T', 'A', 'G', 'C', 'C', 'C', 'G', 'A', 'A', 'C', 'C', 'G', 'C', 'T', 'A', 'T', 'T', 'G', 'G', 'C', 'T', 'G', 'T', 'G', 'C', 'G', 'C', 'G', 'G', 'C', 'G', 'A', 'C', 'C', 'G', 'G', 'C', 'C', 'C', 'G', 'A', 'G', 'C', 'A', 'T', 'T', 'A', 'A', 'A', 'A', 'T', 'T', 'T', 'G', 'G', 'G', 'A', 'T', 'C', 'T', 'G', 'G', 'A', 'A', 'G', 'G', 'C', 'A', 'A', 'A', 'A', 'T', 'T', 'A', 'T', 'T', 'G', 'T', 'G', 'G', 'A', 'T', 'G', 'A', 'A', 'C', 'T', 'G', 'A', 'A', 'A', 'C', 'A', 'G', 'G', 'A', 'A', 'G', 'T', 'G', 'A', 'T', 'T', 'A', 'G', 'C', 'A', 'C', 'C', 'A', 'G', 'C', 'A', 'G', 'C', 'A', 'A', 'A', 'G', 'A', 'A', 'C', 'C', 'G', 'C', 'C', 'G', 'C', 'A', 'G', 'T', 'G', 'C', 'A', 'C', 'C', 'A', 'G', 'C', 'C', 'T', 'G', 'G', 'C', 'G', 'T', 'G', 'G', 'A', 'G', 'C', 'G', 'C', 'G', 'G', 'A', 'T', 'G', 'G', 'C', 'C', 'A', 'G', 'A', 'C', 'C', 'C', 'T', 'G', 'T', 'T', 'T', 'G', 'C', 'G', 'G', 'G', 'C', 'T', 'A', 'T', 'A', 'C', 'C', 'G', 'A', 'T', 'A', 'A', 'C', 'C', 'T', 'G', 'G', 'T', 'G', 'C', 'G', 'C', 'G', 'T', 'G', 'T', 'G', 'G', 'C', 'A', 'G', 'G', 'T', 'G', 'A', 'C', 'C', 'A', 'T', 'T', 'G', 'G', 'C', 'A', 'C', 'C', 'C', 'G', 'C'])\n",
      "('>XP_027389897.1 kita-kyushu lung cancer antigen 1 [Bos indicus x Bos taurus]', ['A', 'T', 'G', 'A', 'G', 'C', 'T', 'T', 'T', 'C', 'T', 'G', 'T', 'T', 'T', 'C', 'T', 'G', 'C', 'T', 'G', 'G', 'T', 'G', 'A', 'G', 'C', 'A', 'G', 'C', 'A', 'T', 'T', 'C', 'T', 'G', 'T', 'T', 'T', 'G', 'C', 'G', 'T', 'T', 'T', 'A', 'T', 'G', 'T', 'T', 'T', 'G', 'T', 'G', 'T', 'T', 'T', 'T', 'G', 'G', 'A', 'A', 'A', 'A', 'C', 'C', 'T', 'G', 'G', 'T', 'T', 'T', 'C', 'A', 'G', 'C', 'G', 'C', 'A', 'A', 'C', 'A', 'C', 'C', 'G', 'G', 'C', 'G', 'A', 'A', 'A', 'T', 'G', 'A', 'G', 'C', 'A', 'G', 'C', 'A', 'A', 'C', 'A', 'G', 'C', 'A', 'C', 'C', 'A', 'G', 'C', 'C', 'T', 'G', 'G', 'C', 'G', 'C', 'T', 'G', 'G', 'T', 'G', 'C', 'G', 'C', 'A', 'G', 'C', 'A', 'G', 'C', 'A', 'G', 'C', 'A', 'G', 'C', 'A', 'C', 'C', 'C', 'A', 'G', 'A', 'G', 'C', 'A', 'C', 'C', 'A', 'A', 'A', 'A', 'G', 'C', 'A', 'A', 'C', 'A', 'T', 'T', 'G', 'A', 'T', 'A', 'A', 'A', 'A', 'G', 'C', 'C', 'T', 'G', 'C', 'T', 'G', 'G', 'T', 'G', 'A', 'A', 'C', 'A', 'G', 'C', 'C', 'T', 'G', 'A', 'G', 'C', 'C', 'G', 'C', 'G', 'A', 'T', 'A', 'T', 'T', 'C', 'T', 'G', 'A', 'T', 'T', 'A', 'A', 'C', 'T', 'T', 'T', 'C', 'C', 'G', 'C', 'A', 'T', 'A', 'G', 'C', 'A', 'T', 'T', 'G', 'C', 'G', 'A', 'T', 'G', 'C', 'A', 'G', 'C', 'G', 'C', 'C', 'G', 'C', 'A', 'T', 'T', 'C', 'T', 'G', 'G', 'T', 'G', 'A', 'A', 'C', 'C', 'T', 'G', 'C', 'G', 'C', 'A', 'T', 'T', 'G', 'T', 'G', 'G', 'A', 'T', 'T', 'A', 'T', 'A', 'A', 'A', 'C', 'T', 'G', 'G', 'C', 'G', 'G', 'A', 'A', 'C', 'T', 'G', 'G', 'A', 'A', 'C', 'A', 'G', 'T', 'T', 'T', 'C', 'T', 'G', 'G', 'T', 'G', 'A', 'C', 'C', 'A', 'A', 'A', 'G', 'G', 'C', 'C', 'T', 'G', 'A', 'A', 'C', 'G', 'G', 'C', 'G', 'C', 'G', 'C', 'T', 'G', 'G', 'T', 'G', 'A', 'A', 'C', 'C', 'G', 'C', 'A', 'A', 'A', 'A', 'G', 'C', 'A', 'C', 'C', 'G', 'A', 'T', 'A', 'A', 'A', 'C', 'C', 'G', 'A', 'C', 'C', 'G', 'A', 'A', 'G', 'T', 'G'])\n",
      "('>XP_007647427.1 kita-kyushu lung cancer antigen 1 [Cricetulus griseus]', ['A', 'T', 'G', 'A', 'A', 'C', 'A', 'T', 'T', 'C', 'T', 'G', 'T', 'T', 'T', 'A', 'T', 'T', 'C', 'T', 'G', 'G', 'T', 'G', 'G', 'G', 'C', 'A', 'G', 'C', 'A', 'T', 'T', 'A', 'T', 'T', 'T', 'T', 'T', 'G', 'C', 'G', 'A', 'G', 'C', 'A', 'T', 'T', 'T', 'T', 'T', 'A', 'G', 'C', 'A', 'T', 'T', 'T', 'T', 'T', 'A', 'A', 'A', 'G', 'T', 'G', 'G', 'T', 'G', 'T', 'T', 'T', 'C', 'A', 'G', 'A', 'T', 'T', 'C', 'C', 'G', 'G', 'A', 'T', 'C', 'A', 'T', 'G', 'A', 'A', 'C', 'G', 'C', 'C', 'C', 'G', 'A', 'G', 'C', 'A', 'A', 'C', 'C', 'C', 'G', 'A', 'C', 'C', 'A', 'G', 'C', 'C', 'C', 'G', 'A', 'C', 'C', 'C', 'T', 'G', 'C', 'T', 'G', 'C', 'G', 'C', 'G', 'C', 'G', 'A', 'A', 'C', 'A', 'G', 'C', 'T', 'T', 'T', 'T', 'G', 'G', 'A', 'G', 'C', 'T', 'A', 'T', 'C', 'G', 'C', 'A', 'A', 'C', 'A', 'C', 'C', 'G', 'G', 'C', 'C', 'T', 'G', 'A', 'G', 'C', 'T', 'T', 'T', 'C', 'C', 'G', 'G', 'G', 'C', 'A', 'A', 'C', 'A', 'G', 'C', 'T', 'A', 'T', 'A', 'A', 'C', 'C', 'A', 'G', 'G', 'A', 'T', 'A', 'T', 'G', 'A', 'T', 'G', 'A', 'A', 'C', 'A', 'A', 'C', 'T', 'T', 'T', 'C', 'C', 'G', 'C', 'G', 'C', 'A', 'C', 'C', 'C', 'T', 'G', 'G', 'C', 'G', 'C', 'T', 'G', 'C', 'A', 'T', 'A', 'A', 'A', 'C', 'G', 'C', 'A', 'T', 'T', 'C', 'T', 'G', 'G', 'T', 'G', 'A', 'A', 'C', 'C', 'T', 'G', 'A', 'A', 'C', 'A', 'T', 'T', 'A', 'T', 'G', 'G', 'A', 'A', 'T', 'G', 'C', 'C', 'A', 'G', 'C', 'T', 'G', 'A', 'C', 'C', 'A', 'A', 'A', 'C', 'T', 'G', 'G', 'A', 'A', 'C', 'A', 'G', 'T', 'T', 'T', 'C', 'T', 'G', 'A', 'G', 'C', 'A', 'G', 'C', 'A', 'A', 'A', 'A', 'A', 'C', 'C', 'C', 'G', 'C', 'A', 'G', 'G', 'G', 'C', 'C', 'A', 'T', 'A', 'C', 'C', 'A', 'A', 'A', 'C', 'A', 'T', 'C', 'G', 'C', 'C', 'G', 'C', 'C', 'G', 'C', 'G', 'C', 'G', 'A', 'C', 'C', 'G', 'T', 'G', 'C', 'G', 'C', 'A', 'T', 'T', 'C', 'C', 'G', 'A', 'T', 'G', 'G', 'A', 'A', 'A', 'G', 'C', 'G', 'A', 'T', 'A', 'G', 'C', 'G', 'G', 'C', 'A', 'G', 'C', 'A', 'A', 'C', 'C', 'A', 'G'])\n",
      "('>XP_014918346.1 kita-kyushu lung cancer antigen 1 [Acinonyx jubatus]', ['A', 'T', 'G', 'A', 'G', 'C', 'A', 'T', 'G', 'C', 'T', 'G', 'C', 'T', 'G', 'C', 'T', 'G', 'C', 'T', 'G', 'C', 'T', 'G', 'A', 'G', 'C', 'T', 'T', 'T', 'A', 'T', 'T', 'G', 'T', 'G', 'T', 'T', 'T', 'A', 'C', 'C', 'C', 'T', 'G', 'C', 'T', 'G', 'C', 'T', 'G', 'G', 'T', 'G', 'G', 'T', 'G', 'T', 'G', 'G', 'A', 'A', 'A', 'A', 'G', 'C', 'C', 'G', 'C', 'T', 'T', 'T', 'C', 'A', 'G', 'A', 'G', 'C', 'A', 'G', 'C', 'G', 'T', 'G', 'G', 'G', 'C', 'G', 'A', 'A', 'A', 'T', 'G', 'A', 'G', 'C', 'A', 'G', 'C', 'A', 'A', 'C', 'A', 'G', 'C', 'A', 'C', 'C', 'A', 'G', 'C', 'C', 'T', 'G', 'G', 'C', 'G', 'C', 'T', 'G', 'G', 'T', 'G', 'C', 'G', 'C', 'G', 'C', 'G', 'A', 'C', 'C', 'A', 'G', 'C', 'A', 'G', 'C', 'A', 'C', 'C', 'G', 'G', 'C', 'A', 'G', 'C', 'A', 'C', 'C', 'A', 'A', 'A', 'A', 'G', 'C', 'A', 'A', 'C', 'A', 'C', 'C', 'G', 'A', 'T', 'A', 'A', 'A', 'A', 'A', 'C', 'C', 'T', 'G', 'A', 'G', 'C', 'G', 'T', 'G', 'G', 'A', 'T', 'A', 'G', 'C', 'A', 'T', 'T', 'A', 'G', 'C', 'C', 'G', 'C', 'G', 'G', 'C', 'A', 'T', 'T', 'C', 'T', 'G', 'A', 'T', 'T', 'A', 'A', 'C', 'A', 'G', 'C', 'C', 'C', 'G', 'C', 'A', 'T', 'A', 'C', 'C', 'A', 'T', 'T', 'G', 'C', 'G', 'A', 'T', 'G', 'C', 'A', 'G', 'A', 'A', 'A', 'C', 'G', 'C', 'A', 'T', 'T', 'C', 'T', 'G', 'A', 'T', 'T', 'A', 'A', 'C', 'C', 'T', 'G', 'C', 'G', 'C', 'A', 'T', 'T', 'G', 'T', 'G', 'G', 'A', 'A', 'T', 'A', 'T', 'A', 'A', 'A', 'C', 'T', 'G', 'G', 'C', 'G', 'G', 'A', 'A', 'C', 'T', 'G', 'G', 'A', 'A', 'T', 'A', 'T', 'T', 'T', 'T', 'C', 'T', 'G', 'A', 'T', 'T', 'A', 'T', 'T', 'A', 'A', 'A', 'G', 'G', 'C', 'C', 'T', 'G', 'A', 'A', 'C', 'G', 'G', 'C', 'A', 'C', 'C', 'C', 'T', 'G', 'G', 'C', 'G', 'A', 'A', 'C', 'T', 'G', 'C', 'A', 'A', 'A', 'A', 'G', 'C', 'A', 'C', 'C', 'G', 'A', 'T', 'C', 'G', 'C', 'C', 'T', 'G', 'C', 'G', 'C', 'G', 'A', 'A', 'T', 'G', 'C', 'A', 'A', 'C', 'G', 'A', 'A', 'A', 'G', 'C', 'G', 'A', 'A', 'G', 'G', 'C', 'A', 'A', 'C', 'C', 'A', 'T'])\n",
      "('>XP_0026997503.1 kita-kyushu lung cancer antigen 1 [Bos taurus]', ['A', 'T', 'G', 'A', 'G', 'C', 'T', 'T', 'T', 'C', 'T', 'G', 'T', 'T', 'T', 'C', 'T', 'G', 'C', 'T', 'G', 'G', 'T', 'G', 'A', 'G', 'C', 'A', 'G', 'C', 'A', 'T', 'T', 'C', 'T', 'G', 'T', 'T', 'T', 'G', 'C', 'G', 'T', 'T', 'T', 'A', 'T', 'G', 'T', 'T', 'T', 'G', 'T', 'G', 'T', 'T', 'T', 'T', 'G', 'G', 'A', 'A', 'A', 'A', 'C', 'C', 'T', 'G', 'G', 'T', 'T', 'T', 'C', 'A', 'G', 'C', 'G', 'C', 'A', 'A', 'C', 'A', 'C', 'C', 'G', 'G', 'C', 'G', 'A', 'A', 'A', 'T', 'G', 'A', 'G', 'C', 'A', 'G', 'C', 'A', 'A', 'C', 'A', 'G', 'C', 'A', 'C', 'C', 'A', 'G', 'C', 'C', 'T', 'G', 'G', 'C', 'G', 'C', 'T', 'G', 'G', 'T', 'G', 'C', 'G', 'C', 'A', 'G', 'C', 'A', 'G', 'C', 'A', 'G', 'C', 'A', 'G', 'C', 'A', 'C', 'C', 'C', 'A', 'G', 'A', 'G', 'C', 'A', 'C', 'C', 'A', 'A', 'A', 'A', 'G', 'C', 'A', 'A', 'C', 'A', 'T', 'T', 'G', 'A', 'T', 'A', 'A', 'A', 'A', 'G', 'C', 'C', 'T', 'G', 'C', 'T', 'G', 'G', 'T', 'G', 'A', 'A', 'C', 'A', 'G', 'C', 'C', 'T', 'G', 'A', 'G', 'C', 'C', 'G', 'C', 'G', 'A', 'T', 'A', 'T', 'T', 'C', 'T', 'G', 'A', 'T', 'T', 'A', 'A', 'C', 'T', 'T', 'T', 'C', 'C', 'G', 'C', 'A', 'T', 'A', 'G', 'C', 'A', 'T', 'T', 'G', 'C', 'G', 'A', 'T', 'G', 'C', 'A', 'G', 'C', 'G', 'C', 'C', 'G', 'C', 'A', 'T', 'T', 'C', 'T', 'G', 'G', 'T', 'G', 'A', 'A', 'C', 'C', 'T', 'G', 'C', 'G', 'C', 'A', 'T', 'T', 'G', 'T', 'G', 'G', 'A', 'T', 'T', 'A', 'T', 'A', 'A', 'A', 'C', 'T', 'G', 'G', 'C', 'G', 'G', 'A', 'A', 'C', 'T', 'G', 'G', 'A', 'A', 'C', 'A', 'G', 'T', 'T', 'T', 'C', 'T', 'G', 'G', 'T', 'G', 'A', 'C', 'C', 'A', 'A', 'A', 'G', 'G', 'C', 'C', 'T', 'G', 'A', 'A', 'C', 'G', 'G', 'C', 'G', 'C', 'G', 'C', 'T', 'G', 'G', 'T', 'G', 'A', 'A', 'C', 'C', 'G', 'C', 'A', 'A', 'A', 'A', 'G', 'C', 'A', 'C', 'C', 'G', 'A', 'T', 'A', 'A', 'A', 'C', 'C', 'G', 'A', 'C', 'C', 'G', 'A', 'A', 'G', 'T', 'G'])\n",
      "('>XP_002699603.1 kita-kyushu lung cancer antigen 1 [Bos taurus]', ['A', 'T', 'G', 'A', 'G', 'C', 'T', 'T', 'T', 'C', 'T', 'G', 'T', 'T', 'T', 'C', 'T', 'G', 'C', 'T', 'G', 'G', 'T', 'G', 'A', 'G', 'C', 'A', 'G', 'C', 'A', 'T', 'T', 'C', 'T', 'G', 'T', 'T', 'T', 'G', 'C', 'G', 'T', 'T', 'T', 'A', 'T', 'G', 'T', 'T', 'T', 'G', 'T', 'G', 'T', 'T', 'T', 'T', 'G', 'G', 'A', 'A', 'A', 'A', 'C', 'C', 'T', 'G', 'G', 'T', 'T', 'T', 'C', 'A', 'G', 'C', 'G', 'C', 'A', 'A', 'C', 'A', 'C', 'C', 'G', 'G', 'C', 'G', 'A', 'A', 'A', 'T', 'G', 'A', 'G', 'C', 'A', 'G', 'C', 'A', 'A', 'C', 'A', 'G', 'C', 'A', 'C', 'C', 'A', 'G', 'C', 'A', 'G', 'C', 'C', 'A', 'T', 'T', 'A', 'T', 'C', 'G', 'C', 'G', 'C', 'G', 'T', 'G', 'C', 'G', 'T', 'G', 'G', 'T', 'G', 'G', 'T', 'G', 'G', 'T', 'G', 'G', 'G', 'C', 'C', 'A', 'T', 'A', 'A', 'A', 'C', 'T', 'G', 'A', 'G', 'C', 'C', 'T', 'G', 'A', 'G', 'C', 'A', 'A', 'A', 'T', 'T', 'T', 'A', 'A', 'C', 'T', 'T', 'T', 'A', 'A', 'C', 'T', 'T', 'T', 'T', 'G', 'G', 'C', 'A', 'T', 'C', 'A', 'G', 'C', 'G', 'C', 'G', 'T', 'G', 'A', 'A', 'A'])\n"
     ]
    }
   ],
   "source": [
    "\"\"\"Display Identifiers and Nucleotides as Separate Sequences with ,\"\"\"\n",
    "#The pprint module provides a capability to “pretty-print” arbitrary Python data structures in a form \n",
    "#which can be used as input to the interpreter.\n",
    "fasta_file = \"DNA.fasta\"\n",
    "\n",
    "def parse_fasta(fname):\n",
    "    with open(fname, \"r\") as fh:\n",
    "\n",
    "        # Create variables for storing the identifiers and the sequence.\n",
    "        identifier = None\n",
    "        sequence = []\n",
    "\n",
    "        for line in fh:\n",
    "            line = line.strip()  # Remove trailing newline characters.\n",
    "            if line.startswith(\">\"):\n",
    "                if identifier is None:\n",
    "                    # This only happens when the first line of the\n",
    "                    # FASTA file is parsed.\n",
    "                    identifier = line\n",
    "                else:\n",
    "                    # This happens every time a new FASTA record is\n",
    "                    # encountered.\n",
    "\n",
    "                    # Start by yielding the entry that has been built up.\n",
    "                    yield identifier, sequence\n",
    "\n",
    "                    # Then reinitialise the identifier and sequence\n",
    "                    # variables to build up a new record.\n",
    "                    identifier = line\n",
    "                    sequence = []\n",
    "            else:\n",
    "                # This happens every time a sequence line is encountered.\n",
    "                seq_on_line = list(line)\n",
    "                sequence.extend(seq_on_line)\n",
    "\n",
    "for entry in parse_fasta(fasta_file):\n",
    "    print(entry)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MKTKCICELCSCGRHHCPHLPTKIYDKTEKPCLLSEYTENYPFYHSYLPRESFKPRREYQKGPIPMEGLTTSSMNFYLLLASSILCALIVFWKYRRFQRNTGEMSSNSTALALVRPSSSGLINSNTDNNLAVYDLSRDILNNFPHSIARQKRILVNLSMVENKLVELEHTLLSKGFRGASPHRKSTMNFYLLLASSILCALIVFWKYRRFQRNTGEMSSNSTALALVRPSSSGLINSNTDNNLAVYDLSRDILNNFPHSIARQKRILVNLSMVENKLVELEHTLLSKGFRGASPHRKSTMGASGSKARGLWPFASAAGGGGSEAAGAEQALVRPRGRAVPPFVFTRRGSMFYDEDGDLAHEFYEETIVTKNGQKRAKLRRVHKNLIPQGIVKLDHPRIHVDFPVILYEVMSFLFLLVSSILFAFMFVFWKTWFQRNTGEMSSNSTSLALVRSSSSTQSTKSNIDKSLLVNSLSRDILINFPHSIAMQRRILVNLRIVDYKLAELEQFLVTKGLNGALVNRKSTDKPTEVMNVYLLLASGILCALMTVFWKYRRFQRNTGEMSSNSTALALVRPSSTGLINSNTDNNLSVYDLSRDILNNFPHSIAMQKRILVNLTTVENKLVELEHILVSKGFRSASAHRKSTMLLTRFEQKNGHLAQVEVDEVFGFMCHITTEVPPHDAMPGGIVLLVKFLLDMSRNVLLYVVFLQRLSSALHRVLLHLFRHVRIFDHGLSVAHGYRGEGGWPTATVSWGRCMNVYLLLASGILCALMTVFWKYRRFQRNTGEMSSNSTALALVRPSSTGLINSNTDNNLSVYDLSRDILNNFPHSIAMQKRILVNLTTVENKLVELEHILVSKGFRSASAHRKSTMTEQMTLRGTLKGHNGWVTQIATTPQFPDMILSASRDKTIIMWKLTRDETNYGIPQRALRGHSHFVSDVVISSDGQFALSGSWDGTLRLWDLTTGTTTRRFVGHTKDVLSVAFSSDNRQIVSGSRDKTIKLWNTLGVCKYTVQDESHSEWVSCVRFSPNSSNPIIVSCGWDKLVKVWNLANCKLKTNHIGHTGYLNTVTVSPDGSLCASGGKDGQAMLWDLNEGKHLYTLDGGDIINALCFSPNRYWLCAATGPSIKIWDLEGKIIVDELKQEVISTSSKAEPPQCTSLAWSADGQTLFAGYTDNLVRVWQVTIGTRMSFLFLLVSSILFAFMFVFWKTWFQRNTGEMSSNSTSLALVRSSSSTQSTKSNIDKSLLVNSLSRDILINFPHSIAMQRRILVNLRIVDYKLAELEQFLVTKGLNGALVNRKSTDKPTEVMNILFILVGSIIFASIFSIFKVVFQIPDHERPSNPTSPTLLRANSFWSYRNTGLSFPGNSYNQDMMNNFPRTLALHKRILVNLNIMECQLTKLEQFLSSKNPQGHTKHRRRATVRIPMESDSGSNQMSMLLLLLSFIVFTLLLVVWKSRFQSSVGEMSSNSTSLALVRATSSTGSTKSNTDKNLSVDSISRGILINSPHTIAMQKRILINLRIVEYKLAELEYFLIIKGLNGTLANCKSTDRLRECNESEGNHMSFLFLLVSSILFAFMFVFWKTWFQRNTGEMSSNSTSLALVRSSSSTQSTKSNIDKSLLVNSLSRDILINFPHSIAMQRRILVNLRIVDYKLAELEQFLVTKGLNGALVNRKSTDKPTEVMSFLFLLVSSILFAFMFVFWKTWFQRNTGEMSSNSTSSHYRACVBVBVBVGHJKLSLSKFNFNFWHQRVKMSFLFLLVSSILFAFMFVFWKTWFQRNTGEMSSNSDKPTEVNGALVNRKS'"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"Display all Protein Sequences as Whole\"\"\"\n",
    "def readGenome(filename):\n",
    "    genome = ''\n",
    "    with open(filename, 'r') as f:\n",
    "        for line in f:\n",
    "            # ignore header line with genome information\n",
    "            if not line[0] == '>':\n",
    "                genome += line.rstrip()\n",
    "    return genome\n",
    "genome = readGenome('dataset.fa')\n",
    "genome[0:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ATGAAAACCAAATGCATTTGCGAACTGTGCAGCTGCGGCCGCCATCATTGCCCGCATCTGCCGACCAAAATTTATGATAAAACCGAAAAACCGTGCCTGCTGAGCGAATATACCGAAAACTATCCGTTTTATCATAGCTATCTGCCGCGCGAAAGCTTTAAACCGCGCCGCGAATATCAGAAAGGCCCGATTCCGATGGAAGGCCTGACCACCAGCAGCATGAACTTTTATCTGCTGCTGGCGAGCAGCATTCTGTGCGCGCTGATTGTGTTTTGGAAATATCGCCGCTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCACCGCGCTGGCGCTGGTGCGCCCGAGCAGCAGCGGCCTGATTAACAGCAACACCGATAACAACCTGGCGGTGTATGATCTGAGCCGCGATATTCTGAACAACTTTCCGCATAGCATTGCGCGCCAGAAACGCATTCTGGTGAACCTGAGCATGGTGGAAAACAAACTGGTGGAACTGGAACATACCCTGCTGAGCAAAGGCTTTCGCGGCGCGAGCCCGCATCGCAAAAGCACCATGAACTTTTATCTGCTGCTGGCGAGCAGCATTCTGTGCGCGCTGATTGTGTTTTGGAAATATCGCCGCTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCACCGCGCTGGCGCTGGTGCGCCCGAGCAGCAGCGGCCTGATTAACAGCAACACCGATAACAACCTGGCGGTGTATGATCTGAGCCGCGATATTCTGAACAACTTTCCGCATAGCATTGCGCGCCAGAAACGCATTCTGGTGAACCTGAGCATGGTGGAAAACAAACTGGTGGAACTGGAACATACCCTGCTGAGCAAAGGCTTTCGCGGCGCGAGCCCGCATCGCAAAAGCACCATGGGCGCGAGCGGCAGCAAAGCGCGCGGCCTGTGGCCGTTTGCGAGCGCGGCGGGCGGCGGCGGCAGCGAAGCGGCGGGCGCGGAACAGGCGCTGGTGCGCCCGCGCGGCCGCGCGGTGCCGCCGTTTGTGTTTACCCGCCGCGGCAGCATGTTTTATGATGAAGATGGCGATCTGGCGCATGAATTTTATGAAGAAACCATTGTGACCAAAAACGGCCAGAAACGCGCGAAACTGCGCCGCGTGCATAAAAACCTGATTCCGCAGGGCATTGTGAAACTGGATCATCCGCGCATTCATGTGGATTTTCCGGTGATTCTGTATGAAGTGATGAGCTTTCTGTTTCTGCTGGTGAGCAGCATTCTGTTTGCGTTTATGTTTGTGTTTTGGAAAACCTGGTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCACCAGCCTGGCGCTGGTGCGCAGCAGCAGCAGCACCCAGAGCACCAAAAGCAACATTGATAAAAGCCTGCTGGTGAACAGCCTGAGCCGCGATATTCTGATTAACTTTCCGCATAGCATTGCGATGCAGCGCCGCATTCTGGTGAACCTGCGCATTGTGGATTATAAACTGGCGGAACTGGAACAGTTTCTGGTGACCAAAGGCCTGAACGGCGCGCTGGTGAACCGCAAAAGCACCGATAAACCGACCGAAGTGATGAACGTGTATCTGCTGCTGGCGAGCGGCATTCTGTGCGCGCTGATGACCGTGTTTTGGAAATATCGCCGCTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCACCGCGCTGGCGCTGGTGCGCCCGAGCAGCACCGGCCTGATTAACAGCAACACCGATAACAACCTGAGCGTGTATGATCTGAGCCGCGATATTCTGAACAACTTTCCGCATAGCATTGCGATGCAGAAACGCATTCTGGTGAACCTGACCACCGTGGAAAACAAACTGGTGGAACTGGAACATATTCTGGTGAGCAAAGGCTTTCGCAGCGCGAGCGCGCATCGCAAAAGCACCATGCTGCTGACCCGCTTTGAACAGAAAAACGGCCATCTGGCGCAGGTGGAAGTGGATGAAGTGTTTGGCTTTATGTGCCATATTACCACCGAAGTGCCGCCGCATGATGCGATGCCGGGCGGCATTGTGCTGCTGGTGAAATTTCTGCTGGATATGAGCCGCAACGTGCTGCTGTATGTGGTGTTTCTGCAGCGCCTGAGCAGCGCGCTGCATCGCGTGCTGCTGCATCTGTTTCGCCATGTGCGCATTTTTGATCATGGCCTGAGCGTGGCGCATGGCTATCGCGGCGAAGGCGGCTGGCCGACCGCGACCGTGAGCTGGGGCCGCTGCATGAACGTGTATCTGCTGCTGGCGAGCGGCATTCTGTGCGCGCTGATGACCGTGTTTTGGAAATATCGCCGCTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCACCGCGCTGGCGCTGGTGCGCCCGAGCAGCACCGGCCTGATTAACAGCAACACCGATAACAACCTGAGCGTGTATGATCTGAGCCGCGATATTCTGAACAACTTTCCGCATAGCATTGCGATGCAGAAACGCATTCTGGTGAACCTGACCACCGTGGAAAACAAACTGGTGGAACTGGAACATATTCTGGTGAGCAAAGGCTTTCGCAGCGCGAGCGCGCATCGCAAAAGCACCATGACCGAACAGATGACCCTGCGCGGCACCCTGAAAGGCCATAACGGCTGGGTGACCCAGATTGCGACCACCCCGCAGTTTCCGGATATGATTCTGAGCGCGAGCCGCGATAAAACCATTATTATGTGGAAACTGACCCGCGATGAAACCAACTATGGCATTCCGCAGCGCGCGCTGCGCGGCCATAGCCATTTTGTGAGCGATGTGGTGATTAGCAGCGATGGCCAGTTTGCGCTGAGCGGCAGCTGGGATGGCACCCTGCGCCTGTGGGATCTGACCACCGGCACCACCACCCGCCGCTTTGTGGGCCATACCAAAGATGTGCTGAGCGTGGCGTTTAGCAGCGATAACCGCCAGATTGTGAGCGGCAGCCGCGATAAAACCATTAAACTGTGGAACACCCTGGGCGTGTGCAAATATACCGTGCAGGATGAAAGCCATAGCGAATGGGTGAGCTGCGTGCGCTTTAGCCCGAACAGCAGCAACCCGATTATTGTGAGCTGCGGCTGGGATAAACTGGTGAAAGTGTGGAACCTGGCGAACTGCAAACTGAAAACCAACCATATTGGCCATACCGGCTATCTGAACACCGTGACCGTGAGCCCGGATGGCAGCCTGTGCGCGAGCGGCGGCAAAGATGGCCAGGCGATGCTGTGGGATCTGAACGAAGGCAAACATCTGTATACCCTGGATGGCGGCGATATTATTAACGCGCTGTGCTTTAGCCCGAACCGCTATTGGCTGTGCGCGGCGACCGGCCCGAGCATTAAAATTTGGGATCTGGAAGGCAAAATTATTGTGGATGAACTGAAACAGGAAGTGATTAGCACCAGCAGCAAAGAACCGCCGCAGTGCACCAGCCTGGCGTGGAGCGCGGATGGCCAGACCCTGTTTGCGGGCTATACCGATAACCTGGTGCGCGTGTGGCAGGTGACCATTGGCACCCGCATGAGCTTTCTGTTTCTGCTGGTGAGCAGCATTCTGTTTGCGTTTATGTTTGTGTTTTGGAAAACCTGGTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCACCAGCCTGGCGCTGGTGCGCAGCAGCAGCAGCACCCAGAGCACCAAAAGCAACATTGATAAAAGCCTGCTGGTGAACAGCCTGAGCCGCGATATTCTGATTAACTTTCCGCATAGCATTGCGATGCAGCGCCGCATTCTGGTGAACCTGCGCATTGTGGATTATAAACTGGCGGAACTGGAACAGTTTCTGGTGACCAAAGGCCTGAACGGCGCGCTGGTGAACCGCAAAAGCACCGATAAACCGACCGAAGTGATGAACATTCTGTTTATTCTGGTGGGCAGCATTATTTTTGCGAGCATTTTTAGCATTTTTAAAGTGGTGTTTCAGATTCCGGATCATGAACGCCCGAGCAACCCGACCAGCCCGACCCTGCTGCGCGCGAACAGCTTTTGGAGCTATCGCAACACCGGCCTGAGCTTTCCGGGCAACAGCTATAACCAGGATATGATGAACAACTTTCCGCGCACCCTGGCGCTGCATAAACGCATTCTGGTGAACCTGAACATTATGGAATGCCAGCTGACCAAACTGGAACAGTTTCTGAGCAGCAAAAACCCGCAGGGCCATACCAAACATCGCCGCCGCGCGACCGTGCGCATTCCGATGGAAAGCGATAGCGGCAGCAACCAGATGAGCATGCTGCTGCTGCTGCTGAGCTTTATTGTGTTTACCCTGCTGCTGGTGGTGTGGAAAAGCCGCTTTCAGAGCAGCGTGGGCGAAATGAGCAGCAACAGCACCAGCCTGGCGCTGGTGCGCGCGACCAGCAGCACCGGCAGCACCAAAAGCAACACCGATAAAAACCTGAGCGTGGATAGCATTAGCCGCGGCATTCTGATTAACAGCCCGCATACCATTGCGATGCAGAAACGCATTCTGATTAACCTGCGCATTGTGGAATATAAACTGGCGGAACTGGAATATTTTCTGATTATTAAAGGCCTGAACGGCACCCTGGCGAACTGCAAAAGCACCGATCGCCTGCGCGAATGCAACGAAAGCGAAGGCAACCATATGAGCTTTCTGTTTCTGCTGGTGAGCAGCATTCTGTTTGCGTTTATGTTTGTGTTTTGGAAAACCTGGTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCACCAGCCTGGCGCTGGTGCGCAGCAGCAGCAGCACCCAGAGCACCAAAAGCAACATTGATAAAAGCCTGCTGGTGAACAGCCTGAGCCGCGATATTCTGATTAACTTTCCGCATAGCATTGCGATGCAGCGCCGCATTCTGGTGAACCTGCGCATTGTGGATTATAAACTGGCGGAACTGGAACAGTTTCTGGTGACCAAAGGCCTGAACGGCGCGCTGGTGAACCGCAAAAGCACCGATAAACCGACCGAAGTGATGAGCTTTCTGTTTCTGCTGGTGAGCAGCATTCTGTTTGCGTTTATGTTTGTGTTTTGGAAAACCTGGTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCACCAGCAGCCATTATCGCGCGTGCGTGGTGGTGGTGGGCCATAAACTGAGCCTGAGCAAATTTAACTTTAACTTTTGGCATCAGCGCGTGAAAATGAGCTTTCTGTTTCTGCTGGTGAGCAGCATTCTGTTTGCGTTTATGTTTGTGTTTTGGAAAACCTGGTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCGATAAACCGACCGAAGTGAACGGCGCGCTGGTGAACCGCAAAAGC'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"Display all DNA Sequences as Whole\"\"\"\n",
    "def readGenome(filename):\n",
    "    genome = ''\n",
    "    with open(filename, 'r') as f:\n",
    "        for line in f:\n",
    "            # ignore header line with genome information\n",
    "            if not line[0] == '>':\n",
    "                genome += line.rstrip()\n",
    "    return genome\n",
    "genome = readGenome('DNA.fasta')\n",
    "genome[0:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ATGAARACNAARTGYATHTGYGARYTNTGYWSNTGYGGNMGNCAYCAYTGYCCNCAYYTNCCNACNAARATHTAYGAYAARACNGARAARCCNTGYYTNYTNWSNGARTAYACNGARAAYTAYCCNTTYTAYCAYWSNTAYYTNCCNMGNGARWSNTTYAARCCNMGNMGNGARTAYCARAARGGNCCNATHCCNATGGARGGNYTNACNACNWSNWSNATGAAYTTYTAYYTNYTNYTNGCNWSNWSNATHYTNTGYGCNYTNATHGTNTTYTGGAARTAYMGNMGNTTYCARMGNAAYACNGGNGARATGWSNWSNAAYWSNACNGCNYTNGCNYTNGTNMGNCCNWSNWSNWSNGGNYTNATHAAYWSNAAYACNGAYAAYAAYYTNGCNGTNTAYGAYYTNWSNMGNGAYATHYTNAAYAAYTTYCCNCAYWSNATHGCNMGNCARAARMGNATHYTNGTNAAYYTNWSNATGGTNGARAAYAARYTNGTNGARYTNGARCAYACNYTNYTNWSNAARGGNTTYMGNGGNGCNWSNCCNCAYMGNAARWSNACNATGAAYTTYTAYYTNYTNYTNGCNWSNWSNATHYTNTGYGCNYTNATHGTNTTYTGGAARTAYMGNMGNTTYCARMGNAAYACNGGNGARATGWSNWSNAAYWSNACNGCNYTNGCNYTNGTNMGNCCNWSNWSNWSNGGNYTNATHAAYWSNAAYACNGAYAAYAAYYTNGCNGTNTAYGAYYTNWSNMGNGAYATHYTNAAYAAYTTYCCNCAYWSNATHGCNMGNCARAARMGNATHYTNGTNAAYYTNWSNATGGTNGARAAYAARYTNGTNGARYTNGARCAYACNYTNYTNWSNAARGGNTTYMGNGGNGCNWSNCCNCAYMGNAARWSNACNATGGGNGCNWSNGGNWSNAARGCNMGNGGNYTNTGGCCNTTYGCNWSNGCNGCNGGNGGNGGNGGNWSNGARGCNGCNGGNGCNGARCARGCNYTNGTNMGNCCNMGNGGNMGNGCNGTNCCNCCNTTYGTNTTYACNMGNMGNGGNWSNATGTTYTAYGAYGARGAYGGNGAYYTNGCNCAYGARTTYTAYGARGARACNATHGTNACNAARAAYGGNCARAARMGNGCNAARYTNMGNMGNGTNCAYAARAAYYTNATHCCNCARGGNATHGTNAARYTNGAYCAYCCNMGNATHCAYGTNGAYTTYCCNGTNATHYTNTAYGARGTNATGWSNTTYYTNTTYYTNYTNGTNWSNWSNATHYTNTTYGCNTTYATGTTYGTNTTYTGGAARACNTGGTTYCARMGNAAYACNGGNGARATGWSNWSNAAYWSNACNWSNYTNGCNYTNGTNMGNWSNWSNWSNWSNACNCARWSNACNAARWSNAAYATHGAYAARWSNYTNYTNGTNAAYWSNYTNWSNMGNGAYATHYTNATHAAYTTYCCNCAYWSNATHGCNATGCARMGNMGNATHYTNGTNAAYYTNMGNATHGTNGAYTAYAARYTNGCNGARYTNGARCARTTYYTNGTNACNAARGGNYTNAAYGGNGCNYTNGTNAAYMGNAARWSNACNGAYAARCCNACNGARGTNATGAAYGTNTAYYTNYTNYTNGCNWSNGGNATHYTNTGYGCNYTNATGACNGTNTTYTGGAARTAYMGNMGNTTYCARMGNAAYACNGGNGARATGWSNWSNAAYWSNACNGCNYTNGCNYTNGTNMGNCCNWSNWSNACNGGNYTNATHAAYWSNAAYACNGAYAAYAAYYTNWSNGTNTAYGAYYTNWSNMGNGAYATHYTNAAYAAYTTYCCNCAYWSNATHGCNATGCARAARMGNATHYTNGTNAAYYTNACNACNGTNGARAAYAARYTNGTNGARYTNGARCAYATHYTNGTNWSNAARGGNTTYMGNWSNGCNWSNGCNCAYMGNAARWSNACNATGYTNYTNACNMGNTTYGARCARAARAAYGGNCAYYTNGCNCARGTNGARGTNGAYGARGTNTTYGGNTTYATGTGYCAYATHACNACNGARGTNCCNCCNCAYGAYGCNATGCCNGGNGGNATHGTNYTNYTNGTNAARTTYYTNYTNGAYATGWSNMGNAAYGTNYTNYTNTAYGTNGTNTTYYTNCARMGNYTNWSNWSNGCNYTNCAYMGNGTNYTNYTNCAYYTNTTYMGNCAYGTNMGNATHTTYGAYCAYGGNYTNWSNGTNGCNCAYGGNTAYMGNGGNGARGGNGGNTGGCCNACNGCNACNGTNWSNTGGGGNMGNTGYATGAAYGTNTAYYTNYTNYTNGCNWSNGGNATHYTNTGYGCNYTNATGACNGTNTTYTGGAARTAYMGNMGNTTYCARMGNAAYACNGGNGARATGWSNWSNAAYWSNACNGCNYTNGCNYTNGTNMGNCCNWSNWSNACNGGNYTNATHAAYWSNAAYACNGAYAAYAAYYTNWSNGTNTAYGAYYTNWSNMGNGAYATHYTNAAYAAYTTYCCNCAYWSNATHGCNATGCARAARMGNATHYTNGTNAAYYTNACNACNGTNGARAAYAARYTNGTNGARYTNGARCAYATHYTNGTNWSNAARGGNTTYMGNWSNGCNWSNGCNCAYMGNAARWSNACNATGACNGARCARATGACNYTNMGNGGNACNYTNAARGGNCAYAAYGGNTGGGTNACNCARATHGCNACNACNCCNCARTTYCCNGAYATGATHYTNWSNGCNWSNMGNGAYAARACNATHATHATGTGGAARYTNACNMGNGAYGARACNAAYTAYGGNATHCCNCARMGNGCNYTNMGNGGNCAYWSNCAYTTYGTNWSNGAYGTNGTNATHWSNWSNGAYGGNCARTTYGCNYTNWSNGGNWSNTGGGAYGGNACNYTNMGNYTNTGGGAYYTNACNACNGGNACNACNACNMGNMGNTTYGTNGGNCAYACNAARGAYGTNYTNWSNGTNGCNTTYWSNWSNGAYAAYMGNCARATHGTNWSNGGNWSNMGNGAYAARACNATHAARYTNTGGAAYACNYTNGGNGTNTGYAARTAYACNGTNCARGAYGARWSNCAYWSNGARTGGGTNWSNTGYGTNMGNTTYWSNCCNAAYWSNWSNAAYCCNATHATHGTNWSNTGYGGNTGGGAYAARYTNGTNAARGTNTGGAAYYTNGCNAAYTGYAARYTNAARACNAAYCAYATHGGNCAYACNGGNTAYYTNAAYACNGTNACNGTNWSNCCNGAYGGNWSNYTNTGYGCNWSNGGNGGNAARGAYGGNCARGCNATGYTNTGGGAYYTNAAYGARGGNAARCAYYTNTAYACNYTNGAYGGNGGNGAYATHATHAAYGCNYTNTGYTTYWSNCCNAAYMGNTAYTGGYTNTGYGCNGCNACNGGNCCNWSNATHAARATHTGGGAYYTNGARGGNAARATHATHGTNGAYGARYTNAARCARGARGTNATHWSNACNWSNWSNAARGARCCNCCNCARTGYACNWSNYTNGCNTGGWSNGCNGAYGGNCARACNYTNTTYGCNGGNTAYACNGAYAAYYTNGTNMGNGTNTGGCARGTNACNATHGGNACNMGNATGWSNTTYYTNTTYYTNYTNGTNWSNWSNATHYTNTTYGCNTTYATGTTYGTNTTYTGGAARACNTGGTTYCARMGNAAYACNGGNGARATGWSNWSNAAYWSNACNWSNYTNGCNYTNGTNMGNWSNWSNWSNWSNACNCARWSNACNAARWSNAAYATHGAYAARWSNYTNYTNGTNAAYWSNYTNWSNMGNGAYATHYTNATHAAYTTYCCNCAYWSNATHGCNATGCARMGNMGNATHYTNGTNAAYYTNMGNATHGTNGAYTAYAARYTNGCNGARYTNGARCARTTYYTNGTNACNAARGGNYTNAAYGGNGCNYTNGTNAAYMGNAARWSNACNGAYAARCCNACNGARGTNATGAAYATHYTNTTYATHYTNGTNGGNWSNATHATHTTYGCNWSNATHTTYWSNATHTTYAARGTNGTNTTYCARATHCCNGAYCAYGARMGNCCNWSNAAYCCNACNWSNCCNACNYTNYTNMGNGCNAAYWSNTTYTGGWSNTAYMGNAAYACNGGNYTNWSNTTYCCNGGNAAYWSNTAYAAYCARGAYATGATGAAYAAYTTYCCNMGNACNYTNGCNYTNCAYAARMGNATHYTNGTNAAYYTNAAYATHATGGARTGYCARYTNACNAARYTNGARCARTTYYTNWSNWSNAARAAYCCNCARGGNCAYACNAARCAYMGNMGNMGNGCNACNGTNMGNATHCCNATGGARWSNGAYWSNGGNWSNAAYCARATGWSNATGYTNYTNYTNYTNYTNWSNTTYATHGTNTTYACNYTNYTNYTNGTNGTNTGGAARWSNMGNTTYCARWSNWSNGTNGGNGARATGWSNWSNAAYWSNACNWSNYTNGCNYTNGTNMGNGCNACNWSNWSNACNGGNWSNACNAARWSNAAYACNGAYAARAAYYTNWSNGTNGAYWSNATHWSNMGNGGNATHYTNATHAAYWSNCCNCAYACNATHGCNATGCARAARMGNATHYTNATHAAYYTNMGNATHGTNGARTAYAARYTNGCNGARYTNGARTAYTTYYTNATHATHAARGGNYTNAAYGGNACNYTNGCNAAYTGYAARWSNACNGAYMGNYTNMGNGARTGYAAYGARWSNGARGGNAAYCAYATGWSNTTYYTNTTYYTNYTNGTNWSNWSNATHYTNTTYGCNTTYATGTTYGTNTTYTGGAARACNTGGTTYCARMGNAAYACNGGNGARATGWSNWSNAAYWSNACNWSNYTNGCNYTNGTNMGNWSNWSNWSNWSNACNCARWSNACNAARWSNAAYATHGAYAARWSNYTNYTNGTNAAYWSNYTNWSNMGNGAYATHYTNATHAAYTTYCCNCAYWSNATHGCNATGCARMGNMGNATHYTNGTNAAYYTNMGNATHGTNGAYTAYAARYTNGCNGARYTNGARCARTTYYTNGTNACNAARGGNYTNAAYGGNGCNYTNGTNAAYMGNAARWSNACNGAYAARCCNACNGARGTNATGWSNTTYYTNTTYYTNYTNGTNWSNWSNATHYTNTTYGCNTTYATGTTYGTNTTYTGGAARACNTGGTTYCARMGNAAYACNGGNGARATGWSNWSNAAYWSNACNWSNWSNCAYTAYMGNGCNTGYGTNGTNGTNGTNGGNCAYAARYTNWSNYTNWSNAARTTYAAYTTYAAYTTYTGGCAYCARMGNGTNAARATGWSNTTYYTNTTYYTNYTNGTNWSNWSNATHYTNTTYGCNTTYATGTTYGTNTTYTGGAARACNTGGTTYCARMGNAAYACNGGNGARATGWSNWSNAAYWSNGAYAARCCNACNGARGTNAAYGGNGCNYTNGTNAAYMGNAARWSN'"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"Display all Cancer affected-DNA Sequences as Whole\"\"\"\n",
    "def readGenome(filename):\n",
    "    genome = ''\n",
    "    with open(filename, 'r') as f:\n",
    "        for line in f:\n",
    "            # ignore header line with genome information\n",
    "            if not line[0] == '>':\n",
    "                genome += line.rstrip()\n",
    "    return genome\n",
    "genome = readGenome('Mutated.fasta')\n",
    "genome[0:]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import FASTA Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NP_001273979.1\n",
      "Seq('ATGAAAACCAAATGCATTTGCGAACTGTGCAGCTGCGGCCGCCATCATTGCCCG...AGC', SingleLetterAlphabet())\n",
      "219\n",
      "NP_001017978.1\n",
      "Seq('ATGAACTTTTATCTGCTGCTGGCGAGCAGCATTCTGTGCGCGCTGATTGTGTTT...ACC', SingleLetterAlphabet())\n",
      "339\n",
      "sp|Q5H943.1|KKLC1_HUMAN\n",
      "Seq('ATGAACTTTTATCTGCTGCTGGCGAGCAGCATTCTGTGCGCGCTGATTGTGTTT...ACC', SingleLetterAlphabet())\n",
      "339\n",
      "AAC35497.1\n",
      "Seq('ATGGGCGCGAGCGGCAGCAAAGCGCGCGGCCTGTGGCCGTTTGCGAGCGCGGCG...GTG', SingleLetterAlphabet())\n",
      "330\n",
      "XP_002699503.1\n",
      "Seq('ATGAGCTTTCTGTTTCTGCTGGTGAGCAGCATTCTGTTTGCGTTTATGTTTGTG...GTG', SingleLetterAlphabet())\n",
      "360\n",
      "sp|Q4R717.1|KKLC1_MACFA\n",
      "Seq('ATGAACGTGTATCTGCTGCTGGCGAGCGGCATTCTGTGCGCGCTGATGACCGTG...ACC', SingleLetterAlphabet())\n",
      "342\n",
      "AAN12271.1\n",
      "Seq('ATGCTGCTGACCCGCTTTGAACAGAAAAACGGCCATCTGGCGCAGGTGGAAGTG...TGC', SingleLetterAlphabet())\n",
      "330\n",
      "NP_001270680.1\n",
      "Seq('ATGAACGTGTATCTGCTGCTGGCGAGCGGCATTCTGTGCGCGCTGATGACCGTG...ACC', SingleLetterAlphabet())\n",
      "342\n",
      "Sp|P63244.3|Rack1_Human\n",
      "Seq('ATGACCGAACAGATGACCCTGCGCGGCACCCTGAAAGGCCATAACGGCTGGGTG...CGC', SingleLetterAlphabet())\n",
      "948\n",
      "XP_027389897.1\n",
      "Seq('ATGAGCTTTCTGTTTCTGCTGGTGAGCAGCATTCTGTTTGCGTTTATGTTTGTG...GTG', SingleLetterAlphabet())\n",
      "360\n",
      "XP_007647427.1\n",
      "Seq('ATGAACATTCTGTTTATTCTGGTGGGCAGCATTATTTTTGCGAGCATTTTTAGC...CAG', SingleLetterAlphabet())\n",
      "378\n",
      "XP_014918346.1\n",
      "Seq('ATGAGCATGCTGCTGCTGCTGCTGAGCTTTATTGTGTTTACCCTGCTGCTGGTG...CAT', SingleLetterAlphabet())\n",
      "381\n",
      "XP_0026997503.1\n",
      "Seq('ATGAGCTTTCTGTTTCTGCTGGTGAGCAGCATTCTGTTTGCGTTTATGTTTGTG...GTG', SingleLetterAlphabet())\n",
      "360\n",
      "XP_002699603.1\n",
      "Seq('ATGAGCTTTCTGTTTCTGCTGGTGAGCAGCATTCTGTTTGCGTTTATGTTTGTG...AAA', SingleLetterAlphabet())\n",
      "198\n",
      "XP_004699503.1\n",
      "Seq('ATGAGCTTTCTGTTTCTGCTGGTGAGCAGCATTCTGTTTGCGTTTATGTTTGTG...AGC', SingleLetterAlphabet())\n",
      "150\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "for seq_record in SeqIO.parse(open('DNA.fasta'),'fasta'):\n",
    "    print(seq_record.id)\n",
    "    print(repr(seq_record.seq))\n",
    "    print(len(seq_record))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NP_001273979.1\n",
      "NP_001273979.1 stabilizer of axonemal microtubules 1 isoform c [Homo sapiens]\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "record_iterator = SeqIO.parse(\"DNA.fasta\", \"fasta\")\n",
    "first_record = next(record_iterator)\n",
    "print(first_record.id)\n",
    "print(first_record.description)\n",
    "#second_record = next(record_iterator)\n",
    "#print(second_record.id)\n",
    "#print(second_record.description)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 15 records\n",
      "The last record\n",
      "XP_004699503.1\n",
      "Seq('ATGAGCTTTCTGTTTCTGCTGGTGAGCAGCATTCTGTTTGCGTTTATGTTTGTG...AGC', SingleLetterAlphabet())\n",
      "150\n",
      "The first record\n",
      "NP_001273979.1\n",
      "Seq('ATGAAAACCAAATGCATTTGCGAACTGTGCAGCTGCGGCCGCCATCATTGCCCG...AGC', SingleLetterAlphabet())\n",
      "219\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "records = list(SeqIO.parse(\"DNA.fasta\", \"fasta\"))\n",
    "print(\"Found %i records\" % len(records))\n",
    "print(\"The last record\")\n",
    "last_record = records[-1] #using Python's list tricks\n",
    "print(last_record.id)\n",
    "print(repr(last_record.seq))\n",
    "print(len(last_record))\n",
    "print(\"The first record\")\n",
    "first_record = records[0] #remember, Python counts from zero\n",
    "print(first_record.id)\n",
    "print(repr(first_record.seq))\n",
    "print(len(first_record))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: NP_001273979.1\n",
      "Name: NP_001273979.1\n",
      "Description: NP_001273979.1 stabilizer of axonemal microtubules 1 isoform c [Homo sapiens]\n",
      "Number of features: 0\n",
      "Seq('ATGAAAACCAAATGCATTTGCGAACTGTGCAGCTGCGGCCGCCATCATTGCCCG...AGC', SingleLetterAlphabet())\n",
      "{}\n",
      "dict_keys([])\n",
      "dict_values([])\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "record_iterator = SeqIO.parse(\"DNA.fasta\", \"fasta\")\n",
    "first_record = next(record_iterator)\n",
    "print(first_record)\n",
    "print(first_record.annotations)\n",
    "print(first_record.annotations.keys())\n",
    "print(first_record.annotations.values())\n",
    "#print(first_record.annotations[\"source\"])\n",
    "#print(first_record.annotations[\"organism\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['stabilizer']\n",
      "['stabilizer', 'kita-kyushu']\n",
      "['stabilizer', 'kita-kyushu', 'RecName:']\n",
      "['stabilizer', 'kita-kyushu', 'RecName:', 'lung']\n",
      "['stabilizer', 'kita-kyushu', 'RecName:', 'lung', 'kita-kyushu']\n",
      "['stabilizer', 'kita-kyushu', 'RecName:', 'lung', 'kita-kyushu', 'RecName:']\n",
      "['stabilizer', 'kita-kyushu', 'RecName:', 'lung', 'kita-kyushu', 'RecName:', 'lung']\n",
      "['stabilizer', 'kita-kyushu', 'RecName:', 'lung', 'kita-kyushu', 'RecName:', 'lung', 'kita-kyushu']\n",
      "['stabilizer', 'kita-kyushu', 'RecName:', 'lung', 'kita-kyushu', 'RecName:', 'lung', 'kita-kyushu', 'Recname:']\n",
      "['stabilizer', 'kita-kyushu', 'RecName:', 'lung', 'kita-kyushu', 'RecName:', 'lung', 'kita-kyushu', 'Recname:', 'kita-kyushu']\n",
      "['stabilizer', 'kita-kyushu', 'RecName:', 'lung', 'kita-kyushu', 'RecName:', 'lung', 'kita-kyushu', 'Recname:', 'kita-kyushu', 'kita-kyushu']\n",
      "['stabilizer', 'kita-kyushu', 'RecName:', 'lung', 'kita-kyushu', 'RecName:', 'lung', 'kita-kyushu', 'Recname:', 'kita-kyushu', 'kita-kyushu', 'kita-kyushu']\n",
      "['stabilizer', 'kita-kyushu', 'RecName:', 'lung', 'kita-kyushu', 'RecName:', 'lung', 'kita-kyushu', 'Recname:', 'kita-kyushu', 'kita-kyushu', 'kita-kyushu', 'kita-kyushu']\n",
      "['stabilizer', 'kita-kyushu', 'RecName:', 'lung', 'kita-kyushu', 'RecName:', 'lung', 'kita-kyushu', 'Recname:', 'kita-kyushu', 'kita-kyushu', 'kita-kyushu', 'kita-kyushu', 'kita-kyushu']\n",
      "['stabilizer', 'kita-kyushu', 'RecName:', 'lung', 'kita-kyushu', 'RecName:', 'lung', 'kita-kyushu', 'Recname:', 'kita-kyushu', 'kita-kyushu', 'kita-kyushu', 'kita-kyushu', 'kita-kyushu', 'kita-kyushu']\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "all_species = []\n",
    "for seq_record in SeqIO.parse(\"DNA.fasta\", \"fasta\"):\n",
    "    all_species.append(seq_record.description.split()[1])\n",
    "    print(all_species)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import SeqFeatures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[>5:(8^9)]\n",
      "[5:9]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "5"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio import SeqFeature\n",
    "start_pos = SeqFeature.AfterPosition(5)\n",
    "end_pos = SeqFeature.BetweenPosition(9, left=8, right=9)\n",
    "my_location = SeqFeature.FeatureLocation(start_pos, end_pos)\n",
    "print(my_location)\n",
    "my_location.start\n",
    "my_location.end\n",
    "int(my_location.start)\n",
    "int(my_location.end)\n",
    "my_location.nofuzzy_start\n",
    "my_location.nofuzzy_end\n",
    "exact_location = SeqFeature.FeatureLocation(5, 9)\n",
    "print(exact_location)\n",
    "exact_location.start\n",
    "int(exact_location.start)\n",
    "exact_location.nofuzzy_start"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Adding New Sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('EVRNAKACGT')"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio.Alphabet import IUPAC\n",
    "from Bio.Seq import Seq\n",
    "protein_seq = Seq(\"EVRNAK\", IUPAC.protein)\n",
    "dna_seq = Seq(\"ACGT\", IUPAC.unambiguous_dna)\n",
    "#protein_seq + dna_seq\n",
    "from Bio.Alphabet import generic_alphabet\n",
    "protein_seq.alphabet = generic_alphabet\n",
    "dna_seq.alphabet = generic_alphabet\n",
    "protein_seq + dna_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('GATCGATGCACGT', NucleotideAlphabet())"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import generic_nucleotide\n",
    "from Bio.Alphabet import IUPAC\n",
    "nuc_seq = Seq(\"GATCGATGC\", generic_nucleotide)\n",
    "dna_seq = Seq(\"ACGT\", IUPAC.unambiguous_dna)\n",
    "nuc_seq\n",
    "dna_seq\n",
    "nuc_seq + dna_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('ACGTAACCGGTT', DNAAlphabet())"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import generic_dna\n",
    "list_of_seqs = [Seq(\"ACGT\", generic_dna), Seq(\"AACC\", generic_dna), Seq(\"GGTT\", generic_dna)]\n",
    "concatenated = Seq(\"\", generic_dna)\n",
    "for s in list_of_seqs:\n",
    "    concatenated += s\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import generic_dna\n",
    "list_of_seqs = [Seq(\"ACGT\", generic_dna), Seq(\"AACC\", generic_dna), Seq(\"GGTT\", generic_dna)]\n",
    "sum(list_of_seqs, Seq(\"\", generic_dna))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Finding Complements"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import generic_dna\n",
    "dna_seq = Seq(\"acgtACGT\", generic_dna)\n",
    "dna_seq\n",
    "dna_seq.upper()\n",
    "dna_seq.lower()\n",
    "\"GTAC\" in dna_seq\n",
    "\"GTAC\" in dna_seq.upper()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG', IUPACUnambiguousDNA())"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import IUPAC\n",
    "dna_seq = Seq(\"ACGT\", IUPAC.unambiguous_dna)\n",
    "dna_seq\n",
    "dna_seq.lower()\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import IUPAC\n",
    "my_seq = Seq(\"GATCGATGGGCCTATATAGGATCGAAAATCGC\", IUPAC.unambiguous_dna)\n",
    "my_seq\n",
    "my_seq.complement()\n",
    "my_seq.reverse_complement()\n",
    "my_seq[::-1]\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import IUPAC\n",
    "protein_seq = Seq(\"EVRNAK\", IUPAC.protein)\n",
    "#protein_seq.complement()\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import IUPAC\n",
    "coding_dna = Seq(\"ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\", IUPAC.unambiguous_dna)\n",
    "coding_dna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG', IUPACUnambiguousDNA())"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "messenger_rna = coding_dna.transcribe()\n",
    "messenger_rna\n",
    "#template_dna.reverse_complement().transcribe()\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import IUPAC\n",
    "messenger_rna = Seq(\"AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG\", IUPAC.unambiguous_rna)\n",
    "messenger_rna\n",
    "messenger_rna.back_transcribe()\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import IUPAC\n",
    "messenger_rna = Seq(\"AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG\", IUPAC.unambiguous_rna)\n",
    "messenger_rna\n",
    "messenger_rna.translate()\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import IUPAC\n",
    "coding_dna = Seq(\"ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\", IUPAC.unambiguous_dna)\n",
    "coding_dna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "O23729\n",
      "CHS3_BROFI\n",
      "RecName: Full=Chalcone synthase 3; EC=2.3.1.74; AltName: Full=Naringenin-chalcone synthase 3;\n",
      "Seq('MAPAMEEIRQAQRAEGPAAVLAIGTSTPPNALYQADYPDYYFRITKSEHLTELK...GAE', ProteinAlphabet())\n",
      "Length 394\n",
      "['Acyltransferase', 'Flavonoid biosynthesis', 'Transferase']\n"
     ]
    }
   ],
   "source": [
    "from Bio import ExPASy\n",
    "from Bio import SeqIO\n",
    "with ExPASy.get_sprot_raw(\"O23729\") as handle:\n",
    "    seq_record = SeqIO.read(handle, \"swiss\")\n",
    "print(seq_record.id)\n",
    "print(seq_record.name)\n",
    "print(seq_record.description)\n",
    "print(repr(seq_record.seq))\n",
    "print(\"Length %i\" % len(seq_record))\n",
    "print(seq_record.annotations[\"keywords\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[SeqRecord(seq=Seq('ATGAAAACCAAATGCATTTGCGAACTGTGCAGCTGCGGCCGCCATCATTGCCCG...AGC', SingleLetterAlphabet()), id='NP_001273979.1', name='NP_001273979.1', description='NP_001273979.1 stabilizer of axonemal microtubules 1 isoform c [Homo sapiens]', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('ATGAACTTTTATCTGCTGCTGGCGAGCAGCATTCTGTGCGCGCTGATTGTGTTT...ACC', SingleLetterAlphabet()), id='NP_001017978.1', name='NP_001017978.1', description='NP_001017978.1 kita-kyushu lung cancer antigen 1 [Homo sapiens]', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('ATGAACTTTTATCTGCTGCTGGCGAGCAGCATTCTGTGCGCGCTGATTGTGTTT...ACC', SingleLetterAlphabet()), id='sp|Q5H943.1|KKLC1_HUMAN', name='sp|Q5H943.1|KKLC1_HUMAN', description='sp|Q5H943.1|KKLC1_HUMAN RecName: Full=Kita-kyushu lung cancer antigen 1; Short=KK-LC-1; AltName: Full=Cancer/testis antigen 83', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('ATGGGCGCGAGCGGCAGCAAAGCGCGCGGCCTGTGGCCGTTTGCGAGCGCGGCG...GTG', SingleLetterAlphabet()), id='AAC35497.1', name='AAC35497.1', description='AAC35497.1 lung cancer candidate FUS1 [Homo sapiens]', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('ATGAGCTTTCTGTTTCTGCTGGTGAGCAGCATTCTGTTTGCGTTTATGTTTGTG...GTG', SingleLetterAlphabet()), id='XP_002699503.1', name='XP_002699503.1', description='XP_002699503.1 kita-kyushu lung cancer antigen 1 [Bos taurus]', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('ATGAACGTGTATCTGCTGCTGGCGAGCGGCATTCTGTGCGCGCTGATGACCGTG...ACC', SingleLetterAlphabet()), id='sp|Q4R717.1|KKLC1_MACFA', name='sp|Q4R717.1|KKLC1_MACFA', description='sp|Q4R717.1|KKLC1_MACFA RecName: Full=Kita-kyushu lung cancer antigen 1 homolog', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('ATGCTGCTGACCCGCTTTGAACAGAAAAACGGCCATCTGGCGCAGGTGGAAGTG...TGC', SingleLetterAlphabet()), id='AAN12271.1', name='AAN12271.1', description='AAN12271.1 lung cancer oncogene 5 [Homo sapiens]', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('ATGAACGTGTATCTGCTGCTGGCGAGCGGCATTCTGTGCGCGCTGATGACCGTG...ACC', SingleLetterAlphabet()), id='NP_001270680.1', name='NP_001270680.1', description='NP_001270680.1 kita-kyushu lung cancer antigen 1 homolog [Macaca fascicularis]', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('ATGACCGAACAGATGACCCTGCGCGGCACCCTGAAAGGCCATAACGGCTGGGTG...CGC', SingleLetterAlphabet()), id='Sp|P63244.3|Rack1_Human', name='Sp|P63244.3|Rack1_Human', description='Sp|P63244.3|Rack1_Human Recname: Full=Receptor Of Activated Protein C Kinase 1; Altname: Full=Cell Proliferation-Inducing Gene 21 Protein; Altname: Full=Guanine Nucleotide-Binding Protein Subunit Beta-2-Like 1; Altname: Full=Guanine Nucleotide-Binding Protein Subunit Beta-Like Protein 12.3; Altname: Full=Human Lung Cancer Oncogene 7 Protein; Short=Hlc-7; Altname: Full=Receptor For Activated C Kinase; Altname: Full=Small Ribosomal Subunit Protein Rack1; Contains: Recname: Full=Receptor Of Activated Protein C Kinase 1, N-Terminally Processed; Altname: Full=Guanine Nucleotide-Binding Protein Subunit Beta-2-Like 1, N-Terminally Processed', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('ATGAGCTTTCTGTTTCTGCTGGTGAGCAGCATTCTGTTTGCGTTTATGTTTGTG...GTG', SingleLetterAlphabet()), id='XP_027389897.1', name='XP_027389897.1', description='XP_027389897.1 kita-kyushu lung cancer antigen 1 [Bos indicus x Bos taurus]', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('ATGAACATTCTGTTTATTCTGGTGGGCAGCATTATTTTTGCGAGCATTTTTAGC...CAG', SingleLetterAlphabet()), id='XP_007647427.1', name='XP_007647427.1', description='XP_007647427.1 kita-kyushu lung cancer antigen 1 [Cricetulus griseus]', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('ATGAGCATGCTGCTGCTGCTGCTGAGCTTTATTGTGTTTACCCTGCTGCTGGTG...CAT', SingleLetterAlphabet()), id='XP_014918346.1', name='XP_014918346.1', description='XP_014918346.1 kita-kyushu lung cancer antigen 1 [Acinonyx jubatus]', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('ATGAGCTTTCTGTTTCTGCTGGTGAGCAGCATTCTGTTTGCGTTTATGTTTGTG...GTG', SingleLetterAlphabet()), id='XP_0026997503.1', name='XP_0026997503.1', description='XP_0026997503.1 kita-kyushu lung cancer antigen 1 [Bos taurus]', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('ATGAGCTTTCTGTTTCTGCTGGTGAGCAGCATTCTGTTTGCGTTTATGTTTGTG...AAA', SingleLetterAlphabet()), id='XP_002699603.1', name='XP_002699603.1', description='XP_002699603.1 kita-kyushu lung cancer antigen 1 [Bos taurus]', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('ATGAGCTTTCTGTTTCTGCTGGTGAGCAGCATTCTGTTTGCGTTTATGTTTGTG...AGC', SingleLetterAlphabet()), id='XP_004699503.1', name='XP_004699503.1', description='XP_004699503.1 kita-kyushu lung cancer antigen 1 [Bos taurus]', dbxrefs=[])]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "data_fasta = SeqIO.to_dict(SeqIO.parse(\"DNA.fasta\", \"fasta\"))\n",
    "len(data_fasta)\n",
    "list(data_fasta.keys())\n",
    "list(data_fasta.values())\n",
    "#print(data_fasta.description)\n",
    "#print(repr(data_fasta.seq))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Adding New Records to FASTA file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Alphabet import generic_protein\n",
    "rec1 = SeqRecord(Seq(\"MMYQQGCFAGGTVLRLAKDLAENNRGARVLVVCSEITAVTFRGPSETHLDSMVGQALFGD\" \\\n",
    "+\"GAGAVIVGSDPDLSVERPLYELVWTGATLLPDSEGAIDGHLREVGLTFHLLKDVPGLISK\" \\\n",
    "+\"NIEKSLKEAFTPLGISDWNSTFWIAHPGGPAILDQVEAKLGLKEEKMRATREVLSEYGNM\" \\\n",
    "+\"SSAC\", generic_protein),\n",
    "id=\"gi|14150838|gb|AAK54648.1|AF376133_1\",\n",
    "description=\"chalcone synthase [Cucumis sativus]\")\n",
    "\n",
    "rec2 = SeqRecord(Seq(\"YPDYYFRITNREHKAELKEKFQRMCDKSMIKKRYMYLTEEILKENPSMCEYMAPSLDARQ\" \\\n",
    "+\"DMVVVEIPKLGKEAAVKAIKEWGQ\", generic_protein),\n",
    "id=\"gi|13919613|gb|AAK33142.1|\",\n",
    "description=\"chalcone synthase [Fragaria vesca subsp. bracteata]\")\n",
    "\n",
    "rec3 = SeqRecord(Seq(\"MVTVEEFRRAQCAEGPATVMAIGTATPSNCVDQSTYPDYYFRITNSEHKVELKEKFKRMC\" \\\n",
    "+\"EKSMIKKRYMHLTEEILKENPNICAYMAPSLDARQDIVVVEVPKLGKEAAQKAIKEWGQP\" \\\n",
    "+\"KSKITHLVFCTTSGVDMPGCDYQLTKLLGLRPSVKRFMMYQQGCFAGGTVLRMAKDLAEN\" \\\n",
    "+\"NKGARVLVVCSEITAVTFRGPNDTHLDSLVGQALFGDGAAAVIIGSDPIPEVERPLFELV\" \\\n",
    "+\"SAAQTLLPDSEGAIDGHLREVGLTFHLLKDVPGLISKNIEKSLVEAFQPLGISDWNSLFW\" \\\n",
    "+\"IAHPGGPAILDQVELKLGLKQEKLKATRKVLSNYGNMSSACVLFILDEMRKASAKEGLGT\" \\\n",
    "+\"TGEGLEWGVLFGFGPGLTVETVVLHSVAT\", generic_protein),\n",
    "id=\"gi|13925890|gb|AAK49457.1|\",\n",
    "description=\"chalcone synthase [Nicotiana tabacum]\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_records = [rec1, rec2, rec3]\n",
    "from Bio import SeqIO\n",
    "SeqIO.write(my_records, \"ADDED.fa\", \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Converted 15 records\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "records = SeqIO.parse(\"DNA.fasta\", \"fasta\")\n",
    "count = SeqIO.write(records, \"my_example.fasta\", \"fasta\")\n",
    "print(\"Converted %i records\" % count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function convert in module Bio.SeqIO:\n",
      "\n",
      "convert(in_file, in_format, out_file, out_format, alphabet=None)\n",
      "    Convert between two sequence file formats, return number of records.\n",
      "    \n",
      "    Arguments:\n",
      "     - in_file - an input handle or filename\n",
      "     - in_format - input file format, lower case string\n",
      "     - out_file - an output handle or filename\n",
      "     - out_format - output file format, lower case string\n",
      "     - alphabet - optional alphabet to assume\n",
      "    \n",
      "    **NOTE** - If you provide an output filename, it will be opened which will\n",
      "    overwrite any existing file without warning. This may happen if even\n",
      "    the conversion is aborted (e.g. an invalid out_format name is given).\n",
      "    \n",
      "    For example, going from a filename to a handle:\n",
      "    \n",
      "    >>> from Bio import SeqIO\n",
      "    >>> try:\n",
      "    ...     from StringIO import StringIO # Python 2\n",
      "    ... except ImportError:\n",
      "    ...     from io import StringIO # Python 3\n",
      "    ...\n",
      "    >>> handle = StringIO(\"\")\n",
      "    >>> SeqIO.convert(\"Quality/example.fastq\", \"fastq\", handle, \"fasta\")\n",
      "    3\n",
      "    >>> print(handle.getvalue())\n",
      "    >EAS54_6_R1_2_1_413_324\n",
      "    CCCTTCTTGTCTTCAGCGTTTCTCC\n",
      "    >EAS54_6_R1_2_1_540_792\n",
      "    TTGGCAGGCCAAGGCCGATGGATCA\n",
      "    >EAS54_6_R1_2_1_443_348\n",
      "    GTTGCTTCTGGCGTGGGTGGGGGGG\n",
      "    <BLANKLINE>\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "help(SeqIO.convert)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# COUNT LENGTH"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15 records with total sequence length 5376\n"
     ]
    }
   ],
   "source": [
    "from Bio.SeqIO.FastaIO import SimpleFastaParser\n",
    "count = 0\n",
    "total_len = 0\n",
    "with open(\"DNA.fasta\") as in_handle:\n",
    "    for title, seq in SimpleFastaParser(in_handle):\n",
    "        count += 1\n",
    "        total_len += len(seq)\n",
    "print(\"%i records with total sequence length %i\" % (count, total_len))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1000 records with total sequence length 100000\n"
     ]
    }
   ],
   "source": [
    "from Bio.SeqIO.QualityIO import FastqGeneralIterator\n",
    "count = 0\n",
    "total_len = 0\n",
    "with open(\"AlternateDNA.fastq\") as in_handle:\n",
    "    for title, seq, qual in FastqGeneralIterator(in_handle):\n",
    "        count += 1\n",
    "        total_len += len(seq)\n",
    "print(\"%i records with total sequence length %i\" % (count, total_len))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# COUNTING & FIND PROBABILITY-Importing Time Counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.07080912590026855\n"
     ]
    }
   ],
   "source": [
    "import time\n",
    "\n",
    "def dCounter():\n",
    "    tic = time.time()\n",
    "    file = open('DNA.fa', 'r')\n",
    "    dMonoDict = {}\n",
    "    dDiDict = {}\n",
    "    seq = file.read().replace(\"\\n\", \"\").upper()\n",
    "    seq = seq.replace(\"N\", \"\")\n",
    "    \n",
    "    for i in range(5, len(seq)-1):\n",
    "        sMono = seq[i]\n",
    "        sDi = seq[i:i+2]\n",
    "        \n",
    "        # Mononucleotide\n",
    "        try:\n",
    "            if \"N\" not in sMono:\n",
    "                dMonoDict[sMono] += 1\n",
    "        except KeyError:\n",
    "            dMonoDict[sMono] = 1\n",
    "            \n",
    "        # Dinucleotides\n",
    "        try:\n",
    "            if \"N\" not in sDi:\n",
    "                dDiDict[sDi] += 1\n",
    "        except KeyError:\n",
    "            dDiDict[sDi] = 1\n",
    "    \n",
    "    file.close()\n",
    "    toc = time.time()\n",
    "    print(toc-tic)\n",
    "    \n",
    "    return dDiDict, dMonoDict\n",
    "\n",
    "dDiDict, dMonoDict = dCounter()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'R': 4,\n",
       " 'Z': 1,\n",
       " '0': 8,\n",
       " '1': 7,\n",
       " '9': 6,\n",
       " '3': 1,\n",
       " '2': 2,\n",
       " '.': 2,\n",
       " ' ': 11,\n",
       " 'Y': 1,\n",
       " 'E': 10,\n",
       " 'S': 6,\n",
       " 'I': 6,\n",
       " 'A': 1294,\n",
       " 'P': 1,\n",
       " 'T': 1689,\n",
       " 'B': 1,\n",
       " 'O': 5,\n",
       " 'V': 1,\n",
       " 'Q': 2,\n",
       " 'U': 3,\n",
       " '7': 3,\n",
       " 'C': 1159,\n",
       " 'G': 1257,\n",
       " '_': 2,\n",
       " '5': 1,\n",
       " ',': 1,\n",
       " 'W': 1,\n",
       " 'H': 2,\n",
       " 'L': 1,\n",
       " 'M': 1}"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dMonoDict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'RZ': 1,\n",
       " 'Z0': 1,\n",
       " '01': 3,\n",
       " '10': 1,\n",
       " '00': 4,\n",
       " '09': 1,\n",
       " '93': 1,\n",
       " '32': 1,\n",
       " '2.': 1,\n",
       " '.1': 1,\n",
       " '1 ': 2,\n",
       " ' Y': 1,\n",
       " 'YE': 1,\n",
       " 'ER': 1,\n",
       " 'RS': 1,\n",
       " 'SI': 1,\n",
       " 'II': 1,\n",
       " 'IA': 1,\n",
       " 'A ': 2,\n",
       " ' P': 1,\n",
       " 'PE': 1,\n",
       " 'ES': 1,\n",
       " 'ST': 2,\n",
       " 'TI': 3,\n",
       " 'IS': 1,\n",
       " 'S ': 1,\n",
       " ' B': 1,\n",
       " 'BI': 1,\n",
       " 'IO': 1,\n",
       " 'OV': 1,\n",
       " 'VA': 1,\n",
       " 'AR': 1,\n",
       " 'R ': 1,\n",
       " ' A': 1,\n",
       " 'AT': 384,\n",
       " 'IQ': 1,\n",
       " 'QU': 2,\n",
       " 'UA': 1,\n",
       " ' S': 3,\n",
       " 'TR': 1,\n",
       " 'R.': 1,\n",
       " '. ': 1,\n",
       " ' E': 1,\n",
       " 'E1': 2,\n",
       " '19': 3,\n",
       " '97': 2,\n",
       " '79': 2,\n",
       " '90': 2,\n",
       " ' C': 1,\n",
       " 'CO': 1,\n",
       " 'OT': 2,\n",
       " 'IG': 1,\n",
       " 'G_': 1,\n",
       " '_E': 1,\n",
       " '1_': 1,\n",
       " '_1': 1,\n",
       " '92': 1,\n",
       " '27': 1,\n",
       " '75': 1,\n",
       " '5,': 1,\n",
       " ', ': 1,\n",
       " ' W': 1,\n",
       " 'WH': 1,\n",
       " 'HO': 2,\n",
       " 'OL': 1,\n",
       " 'LE': 1,\n",
       " 'E ': 2,\n",
       " ' G': 1,\n",
       " 'GE': 1,\n",
       " 'EO': 1,\n",
       " 'OM': 1,\n",
       " 'ME': 1,\n",
       " 'SH': 1,\n",
       " 'TG': 481,\n",
       " 'GU': 1,\n",
       " 'U ': 1,\n",
       " 'SE': 1,\n",
       " 'EQ': 1,\n",
       " 'UE': 1,\n",
       " 'EC': 1,\n",
       " 'CE': 1,\n",
       " 'EG': 1,\n",
       " 'GA': 327,\n",
       " 'AG': 251,\n",
       " 'GT': 325,\n",
       " 'TT': 572,\n",
       " 'TA': 312,\n",
       " 'TC': 320,\n",
       " 'CG': 267,\n",
       " 'GC': 347,\n",
       " 'CT': 404,\n",
       " 'CC': 229,\n",
       " 'CA': 257,\n",
       " 'AC': 261,\n",
       " 'AA': 395,\n",
       " 'GG': 255}"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dDiDict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'R': Decimal('0.00073'),\n",
       " 'Z': Decimal('0.00018'),\n",
       " '0': Decimal('0.0015'),\n",
       " '1': Decimal('0.0013'),\n",
       " '9': Decimal('0.0011'),\n",
       " '3': Decimal('0.00018'),\n",
       " '2': Decimal('0.00036'),\n",
       " '.': Decimal('0.00036'),\n",
       " ' ': Decimal('0.0020'),\n",
       " 'Y': Decimal('0.00018'),\n",
       " 'E': Decimal('0.0018'),\n",
       " 'S': Decimal('0.0011'),\n",
       " 'I': Decimal('0.0011'),\n",
       " 'A': Decimal('0.24'),\n",
       " 'P': Decimal('0.00018'),\n",
       " 'T': Decimal('0.31'),\n",
       " 'B': Decimal('0.00018'),\n",
       " 'O': Decimal('0.00091'),\n",
       " 'V': Decimal('0.00018'),\n",
       " 'Q': Decimal('0.00036'),\n",
       " 'U': Decimal('0.00055'),\n",
       " '7': Decimal('0.00055'),\n",
       " 'C': Decimal('0.21'),\n",
       " 'G': Decimal('0.23'),\n",
       " '_': Decimal('0.00036'),\n",
       " '5': Decimal('0.00018'),\n",
       " ',': Decimal('0.00018'),\n",
       " 'W': Decimal('0.00018'),\n",
       " 'H': Decimal('0.00036'),\n",
       " 'L': Decimal('0.00018'),\n",
       " 'M': Decimal('0.00018')}"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from decimal import *\n",
    "\n",
    "getcontext().prec = 2\n",
    "monoSum = sum(dMonoDict.values())\n",
    "dMonoDictRatio = {key:Decimal(value)/monoSum for key, value in dMonoDict.items()}\n",
    "dMonoDictRatio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'RZ': Decimal('0.00018'),\n",
       " 'Z0': Decimal('0.00018'),\n",
       " '01': Decimal('0.00055'),\n",
       " '10': Decimal('0.00018'),\n",
       " '00': Decimal('0.00073'),\n",
       " '09': Decimal('0.00018'),\n",
       " '93': Decimal('0.00018'),\n",
       " '32': Decimal('0.00018'),\n",
       " '2.': Decimal('0.00018'),\n",
       " '.1': Decimal('0.00018'),\n",
       " '1 ': Decimal('0.00036'),\n",
       " ' Y': Decimal('0.00018'),\n",
       " 'YE': Decimal('0.00018'),\n",
       " 'ER': Decimal('0.00018'),\n",
       " 'RS': Decimal('0.00018'),\n",
       " 'SI': Decimal('0.00018'),\n",
       " 'II': Decimal('0.00018'),\n",
       " 'IA': Decimal('0.00018'),\n",
       " 'A ': Decimal('0.00036'),\n",
       " ' P': Decimal('0.00018'),\n",
       " 'PE': Decimal('0.00018'),\n",
       " 'ES': Decimal('0.00018'),\n",
       " 'ST': Decimal('0.00036'),\n",
       " 'TI': Decimal('0.00055'),\n",
       " 'IS': Decimal('0.00018'),\n",
       " 'S ': Decimal('0.00018'),\n",
       " ' B': Decimal('0.00018'),\n",
       " 'BI': Decimal('0.00018'),\n",
       " 'IO': Decimal('0.00018'),\n",
       " 'OV': Decimal('0.00018'),\n",
       " 'VA': Decimal('0.00018'),\n",
       " 'AR': Decimal('0.00018'),\n",
       " 'R ': Decimal('0.00018'),\n",
       " ' A': Decimal('0.00018'),\n",
       " 'AT': Decimal('0.070'),\n",
       " 'IQ': Decimal('0.00018'),\n",
       " 'QU': Decimal('0.00036'),\n",
       " 'UA': Decimal('0.00018'),\n",
       " ' S': Decimal('0.00055'),\n",
       " 'TR': Decimal('0.00018'),\n",
       " 'R.': Decimal('0.00018'),\n",
       " '. ': Decimal('0.00018'),\n",
       " ' E': Decimal('0.00018'),\n",
       " 'E1': Decimal('0.00036'),\n",
       " '19': Decimal('0.00055'),\n",
       " '97': Decimal('0.00036'),\n",
       " '79': Decimal('0.00036'),\n",
       " '90': Decimal('0.00036'),\n",
       " ' C': Decimal('0.00018'),\n",
       " 'CO': Decimal('0.00018'),\n",
       " 'OT': Decimal('0.00036'),\n",
       " 'IG': Decimal('0.00018'),\n",
       " 'G_': Decimal('0.00018'),\n",
       " '_E': Decimal('0.00018'),\n",
       " '1_': Decimal('0.00018'),\n",
       " '_1': Decimal('0.00018'),\n",
       " '92': Decimal('0.00018'),\n",
       " '27': Decimal('0.00018'),\n",
       " '75': Decimal('0.00018'),\n",
       " '5,': Decimal('0.00018'),\n",
       " ', ': Decimal('0.00018'),\n",
       " ' W': Decimal('0.00018'),\n",
       " 'WH': Decimal('0.00018'),\n",
       " 'HO': Decimal('0.00036'),\n",
       " 'OL': Decimal('0.00018'),\n",
       " 'LE': Decimal('0.00018'),\n",
       " 'E ': Decimal('0.00036'),\n",
       " ' G': Decimal('0.00018'),\n",
       " 'GE': Decimal('0.00018'),\n",
       " 'EO': Decimal('0.00018'),\n",
       " 'OM': Decimal('0.00018'),\n",
       " 'ME': Decimal('0.00018'),\n",
       " 'SH': Decimal('0.00018'),\n",
       " 'TG': Decimal('0.088'),\n",
       " 'GU': Decimal('0.00018'),\n",
       " 'U ': Decimal('0.00018'),\n",
       " 'SE': Decimal('0.00018'),\n",
       " 'EQ': Decimal('0.00018'),\n",
       " 'UE': Decimal('0.00018'),\n",
       " 'EC': Decimal('0.00018'),\n",
       " 'CE': Decimal('0.00018'),\n",
       " 'EG': Decimal('0.00018'),\n",
       " 'GA': Decimal('0.060'),\n",
       " 'AG': Decimal('0.046'),\n",
       " 'GT': Decimal('0.059'),\n",
       " 'TT': Decimal('0.10'),\n",
       " 'TA': Decimal('0.057'),\n",
       " 'TC': Decimal('0.058'),\n",
       " 'CG': Decimal('0.049'),\n",
       " 'GC': Decimal('0.063'),\n",
       " 'CT': Decimal('0.074'),\n",
       " 'CC': Decimal('0.042'),\n",
       " 'CA': Decimal('0.047'),\n",
       " 'AC': Decimal('0.048'),\n",
       " 'AA': Decimal('0.072'),\n",
       " 'GG': Decimal('0.046')}"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "getcontext().prec = 2\n",
    "diSum = sum(dDiDict.values())\n",
    "dDiDictRatio = {key:Decimal(value)/diSum for key, value in dDiDict.items()}\n",
    "dDiDictRatio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Decimal('0.032')"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dMonoDictRatio['C']*dMonoDictRatio['T']/2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SEQUENCE MANIPULATION--COMPLEMENT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import generic_dna\n",
    "dna_seq = Seq(\"acgtACGT\", generic_dna)\n",
    "dna_seq\n",
    "dna_seq.upper()\n",
    "dna_seq.lower()\n",
    "\"GTAC\" in dna_seq\n",
    "\"GTAC\" in dna_seq.upper()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG', IUPACUnambiguousDNA())"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import IUPAC\n",
    "dna_seq = Seq(\"ACGT\", IUPAC.unambiguous_dna)\n",
    "dna_seq\n",
    "dna_seq.lower()\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import IUPAC\n",
    "my_seq = Seq(\"GATCGATGGGCCTATATAGGATCGAAAATCGC\", IUPAC.unambiguous_dna)\n",
    "my_seq\n",
    "my_seq.complement()\n",
    "my_seq.reverse_complement()\n",
    "my_seq[::-1]\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import IUPAC\n",
    "protein_seq = Seq(\"EVRNAK\", IUPAC.protein)\n",
    "#protein_seq.complement()\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import IUPAC\n",
    "coding_dna = Seq(\"ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\", IUPAC.unambiguous_dna)\n",
    "coding_dna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG', IUPACUnambiguousDNA())"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "messenger_rna = coding_dna.transcribe()\n",
    "messenger_rna\n",
    "#template_dna.reverse_complement().transcribe()\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import IUPAC\n",
    "messenger_rna = Seq(\"AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG\", IUPAC.unambiguous_rna)\n",
    "messenger_rna\n",
    "messenger_rna.back_transcribe()\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import IUPAC\n",
    "messenger_rna = Seq(\"AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG\", IUPAC.unambiguous_rna)\n",
    "messenger_rna\n",
    "messenger_rna.translate()\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import IUPAC\n",
    "coding_dna = Seq(\"ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\", IUPAC.unambiguous_dna)\n",
    "coding_dna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "O23729\n",
      "CHS3_BROFI\n",
      "RecName: Full=Chalcone synthase 3; EC=2.3.1.74; AltName: Full=Naringenin-chalcone synthase 3;\n",
      "Seq('MAPAMEEIRQAQRAEGPAAVLAIGTSTPPNALYQADYPDYYFRITKSEHLTELK...GAE', ProteinAlphabet())\n",
      "Length 394\n",
      "['Acyltransferase', 'Flavonoid biosynthesis', 'Transferase']\n"
     ]
    }
   ],
   "source": [
    "from Bio import ExPASy\n",
    "from Bio import SeqIO\n",
    "with ExPASy.get_sprot_raw(\"O23729\") as handle:\n",
    "    seq_record = SeqIO.read(handle, \"swiss\")\n",
    "print(seq_record.id)\n",
    "print(seq_record.name)\n",
    "print(seq_record.description)\n",
    "print(repr(seq_record.seq))\n",
    "print(\"Length %i\" % len(seq_record))\n",
    "print(seq_record.annotations[\"keywords\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SEQUENCE MANIPULATION--REVERSE COMPLEMENT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NZ_ADRZ01000932.1 Yersinia pestis biovar Antiqua str. E1979001 Contig_E1979001_19275, whole genome shotgun sequence\n",
      "5386\n",
      "GAGTTTTATCGCTTCCATGACGCAGAAGTTAACACTTTCGGATATTTCTGATGAGTCGAAAAATTATCTTGATAAAGCAGGAATTACTACTGCTTGTTTACGAATTAAATCGAAGTGGACTGCTGGCGGAAAATGAGAAAATTCGACCTATCCTTGCGCAGCTCGAGAAGCTCTTACTTTGCGACCTTTCGCCATCAACTAACGATTCTGTCAAAAACTGACGCGTTGGATGAGGAGAAGTGGCTTAATATGCTTGGCACGTTCGTCAAGGACTGGTTTAGATATGAGTCACATTTTGTTCATGGTAGAGATTCTCTTGTTGACATTTTAAAAGAGCGTGGATTACTATCTGAGTCCGATGCTGTTCAACCACTAATAGGTAAGAAATCATGAGTCAAGTTACTGAACAATCCGTACGTTTCCAGACCGCTTTGGCCTCTATTAAGCTCATTCAGGCTTCTGCCGTTTTGGATTTAACCGAAGATGATTTCGATTTTCTGACGAGTAACAAAGTTTGGATTGCTACTGACCGCTCTCGTGCTCGTCGCTGCGTTGAGGCTTGCGTTTATGGTACGCTGGACTTTGTGGGATACCCTCGCTTTCCTGCTCCTGTTGAGTTTATTGCTGCCGTCATTGCTTATTATGTTCATCCCGTCAACATTCAAACGGCCTGTCTCATCATGGAAGGCGCTGAATTTACGGAAAACATTATTAATGGCGTCGAGCGTCCGGTTAAAGCCGCTGAATTGTTCGCGTTTACCTTGCGTGTACGCGCAGGAAACACTGACGTTCTTACTGACGCAGAAGAAAACGTGCGTCAAAAATTACGTGCGGAAGGAGTGATGTAATGTCTAAAGGTAAAAAACGTTCTGGCGCTCGCCCTGGTCGTCCGCAGCCGTTGCGAGGTACTAAAGGCAAGCGTAAAGGCGCTCGTCTTTGGTATGTAGGTGGTCAACAATTTTAATTGCAGGGGCTTCGGCCCCTTACTTGAGGATAAATTATGTCTAATATTCAAACTGGCGCCGAGCGTATGCCGCATGACCTTTCCCATCTTGGCTTCCTTGCTGGTCAGATTGGTCGTCTTATTACCATTTCAACTACTCCGGTTATCGCTGGCGACTCCTTCGAGATGGACGCCGTTGGCGCTCTCCGTCTTTCTCCATTGCGTCGTGGCCTTGCTATTGACTCTACTGTAGACATTTTTACTTTTTATGTCCCTCATCGTCACGTTTATGGTGAACAGTGGATTAAGTTCATGAAGGATGGTGTTAATGCCACTCCTCTCCCGACTGTTAACACTACTGGTTATATTGACCATGCCGCTTTTCTTGGCACGATTAACCCTGATACCAATAAAATCCCTAAGCATTTGTTTCAGGGTTATTTGAATATCTATAACAACTATTTTAAAGCGCCGTGGATGCCTGACCGTACCGAGGCTAACCCTAATGAGCTTAATCAAGATGATGCTCGTTATGGTTTCCGTTGCTGCCATCTCAAAAACATTTGGACTGCTCCGCTTCCTCCTGAGACTGAGCTTTCTCGCCAAATGACGACTTCTACCACATCTATTGACATTATGGGTCTGCAAGCTGCTTATGCTAATTTGCATACTGACCAAGAACGTGATTACTTCATGCAGCGTTACCATGATGTTATTTCTTCATTTGGAGGTAAAACCTCTTATGACGCTGACAACCGTCCTTTACTTGTCATGCGCTCTAATCTCTGGGCATCTGGCTATGATGTTGATGGAACTGACCAAACGTCGTTAGGCCAGTTTTCTGGTCGTGTTCAACAGACCTATAAACATTCTGTGCCGCGTTTCTTTGTTCCTGAGCATGGCACTATGTTTACTCTTGCGCTTGTTCGTTTTCCGCCTACTGCGACTAAAGAGATTCAGTACCTTAACGCTAAAGGTGCTTTGACTTATACCGATATTGCTGGCGACCCTGTTTTGTATGGCAACTTGCCGCCGCGTGAAATTTCTATGAAGGATGTTTTCCGTTCTGGTGATTCGTCTAAGAAGTTTAAGATTGCTGAGGGTCAGTGGTATCGTTATGCGCCTTCGTATGTTTCTCCTGCTTATCACCTTCTTGAAGGCTTCCCATTCATTCAGGAACCGCCTTCTGGTGATTTGCAAGAACGCGTACTTATTCGCCACCATGATTATGACCAGTGTTTCCAGTCCGTTCAGTTGTTGCAGTGGAATAGTCAGGTTAAATTTAATGTGACCGTTTATCGCAATCTGCCGACCACTCGCGATTCAATCATGACTTCGTGATAAAAGATTGAGTGTGAGGTTATAACGCCGAAGCGGTAAAAATTTTAATTTTTGCCGCTGAGGGGTTGACCAAGCGAAGCGCGGTAGGTTTTCTGCTTAGGAGTTTAATCATGTTTCAGACTTTTATTTCTCGCCATAATTCAAACTTTTTTTCTGATAAGCTGGTTCTCACTTCTGTTACTCCAGCTTCTTCGGCACCTGTTTTACAGACACCTAAAGCTACATCGTCAACGTTATATTTTGATAGTTTGACGGTTAATGCTGGTAATGGTGGTTTTCTTCATTGCATTCAGATGGATACATCTGTCAACGCCGCTAATCAGGTTGTTTCTGTTGGTGCTGATATTGCTTTTGATGCCGACCCTAAATTTTTTGCCTGTTTGGTTCGCTTTGAGTCTTCTTCGGTTCCGACTACCCTCCCGACTGCCTATGATGTTTATCCTTTGAATGGTCGCCATGATGGTGGTTATTATACCGTCAAGGACTGTGTGACTATTGACGTCCTTCCCCGTACGCCGGGCAATAACGTTTATGTTGGTTTCATGGTTTGGTCTAACTTTACCGCTACTAAATGCCGCGGATTGGTTTCGCTGAATCAGGTTATTAAAGAGATTATTTGTCTCCAGCCACTTAAGTGAGGTGATTTATGTTTGGTGCTATTGCTGGCGGTATTGCTTCTGCTCTTGCTGGTGGCGCCATGTCTAAATTGTTTGGAGGCGGTCAAAAAGCCGCCTCCGGTGGCATTCAAGGTGATGTGCTTGCTACCGATAACAATACTGTAGGCATGGGTGATGCTGGTATTAAATCTGCCATTCAAGGCTCTAATGTTCCTAACCCTGATGAGGCCGCCCCTAGTTTTGTTTCTGGTGCTATGGCTAAAGCTGGTAAAGGACTTCTTGAAGGTACGTTGCAGGCTGGCACTTCTGCCGTTTCTGATAAGTTGCTTGATTTGGTTGGACTTGGTGGCAAGTCTGCCGCTGATAAAGGAAAGGATACTCGTGATTATCTTGCTGCTGCATTTCCTGAGCTTAATGCTTGGGAGCGTGCTGGTGCTGATGCTTCCTCTGCTGGTATGGTTGACGCCGGATTTGAGAATCAAAAAGAGCTTACTAAAATGCAACTGGACAATCAGAAAGAGATTGCCGAGATGCAAAATGAGACTCAAAAAGAGATTGCTGGCATTCAGTCGGCGACTTCACGCCAGAATACGAAAGACCAGGTATATGCACAAAATGAGATGCTTGCTTATCAACAGAAGGAGTCTACTGCTCGCGTTGCGTCTATTATGGAAAACACCAATCTTTCCAAGCAACAGCAGGTTTCCGAGATTATGCGCCAAATGCTTACTCAAGCTCAAACGGCTGGTCAGTATTTTACCAATGACCAAATCAAAGAAATGACTCGCAAGGTTAGTGCTGAGGTTGACTTAGTTCATCAGCAAACGCAGAATCAGCGGTATGGCTCTTCTCATATTGGCGCTACTGCAAAGGATATTTCTAATGTCGTCACTGATGCTGCTTCTGGTGTGGTTGATATTTTTCATGGTATTGATAAAGCTGTTGCCGATACTTGGAACAATTTCTGGAAAGACGGTAAAGCTGATGGTATTGGCTCTAATTTGTCTAGGAAATAACCGTCAGGATTGACACCCTCCCAATTGTATGTTTTCATGCCTCCAAATCTTGGAGGCTTTTTTATGGTTCGTTCTTATTACCCTTCTGAATGTCACGCTGATTATTTTGACTTTGAGCGTATCGAGGCTCTTAAACCTGCTATTGAGGCTTGTGGCATTTCTACTCTTTCTCAATCCCCAATGCTTGGCTTCCATAAGCAGATGGATAACCGCATCAAGCTCTTGGAAGAGATTCTGTCTTTTCGTATGCAGGGCGTTGAGTTCGATAATGGTGATATGTATGTTGACGGCCATAAGGCTGCTTCTGACGTTCGTGATGAGTTTGTATCTGTTACTGAGAAGTTAATGGATGAATTGGCACAATGCTACAATGTGCTCCCCCAACTTGATATTAATAACACTATAGACCACCGCCCCGAAGGGGACGAAAAATGGTTTTTAGAGAACGAGAAGACGGTTACGCAGTTTTGCCGCAAGCTGGCTGCTGAACGCCCTCTTAAGGATATTCGCGATGAGTATAATTACCCCAAAAAGAAAGGTATTAAGGATGAGTGTTCAAGATTGCTGGAGGCCTCCACTATGAAATCGCGTAGAGGCTTTGCTATTCAGCGTTTGATGAATGCAATGCGACAGGCTCATGCTGATGGTTGGTTTATCGTTTTTGACACTCTCACGTTGGCTGACGACCGATTAGAGGCGTTTTATGATAATCCCAATGCTTTGCGTGACTATTTTCGTGATATTGGTCGTATGGTTCTTGCTGCCGAGGGTCGCAAGGCTAATGATTCACACGCCGACTGCTATCAGTATTTTTGTGTGCCTGAGTATGGTACAGCTAATGGCCGTCTTCATTTCCATGCGGTGCACTTTATGCGGACACTTCCTACAGGTAGCGTTGACCCTAATTTTGGTCGTCGGGTACGCAATCGCCGCCAGTTAAATAGCTTGCAAAATACGTGGCCTTATGGTTACAGTATGCCCATCGCAGTTCGCTACACGCAGGACGCTTTTTCACGTTCTGGTTGGTTGTGGCCTGTTGATGCTAAAGGTGAGCCGCTTAAAGCTACCAGTTATATGGCTGTTGGTTTCTATGTGGCTAAATACGTTAACAAAAAGTCAGATATGGACCTTGCTGCTAAAGGTCTAGGAGCTAAAGAATGGAACAACTCACTAAAAACCAAGCTGTCGCTACTTCCCAAGAAGCTGTTCAGAATCAGAATGAGCCGCAACTTCGGGATGAAAATGCTCACAATGACAAATCTGTCCACGGAGTGCTTAATCCAACTTACCAAGCTGGGTTACGACGCGACGCCGTTCAACCAGATATTGAAGCAGAACGCAAAAAGAGAGATGAGATTGAGGCTGGGAAAAGTTACTGTAGCCGACGTTTTGGCGGCGCAACCTGTGACGACAAATCTGCTCAAATTTATGCGCGCTTCGATAAAAATGATTGGCGTATCCAACCTGCA\n",
      "TGCAGGTTGGATACGCCAATCATTTTTATCGAAGCGCGCATAAATTTGAGCAGATTTGTCGTCACAGGTTGCGCCGCCAAAACGTCGGCTACAGTAACTTTTCCCAGCCTCAATCTCATCTCTCTTTTTGCGTTCTGCTTCAATATCTGGTTGAACGGCGTCGCGTCGTAACCCAGCTTGGTAAGTTGGATTAAGCACTCCGTGGACAGATTTGTCATTGTGAGCATTTTCATCCCGAAGTTGCGGCTCATTCTGATTCTGAACAGCTTCTTGGGAAGTAGCGACAGCTTGGTTTTTAGTGAGTTGTTCCATTCTTTAGCTCCTAGACCTTTAGCAGCAAGGTCCATATCTGACTTTTTGTTAACGTATTTAGCCACATAGAAACCAACAGCCATATAACTGGTAGCTTTAAGCGGCTCACCTTTAGCATCAACAGGCCACAACCAACCAGAACGTGAAAAAGCGTCCTGCGTGTAGCGAACTGCGATGGGCATACTGTAACCATAAGGCCACGTATTTTGCAAGCTATTTAACTGGCGGCGATTGCGTACCCGACGACCAAAATTAGGGTCAACGCTACCTGTAGGAAGTGTCCGCATAAAGTGCACCGCATGGAAATGAAGACGGCCATTAGCTGTACCATACTCAGGCACACAAAAATACTGATAGCAGTCGGCGTGTGAATCATTAGCCTTGCGACCCTCGGCAGCAAGAACCATACGACCAATATCACGAAAATAGTCACGCAAAGCATTGGGATTATCATAAAACGCCTCTAATCGGTCGTCAGCCAACGTGAGAGTGTCAAAAACGATAAACCAACCATCAGCATGAGCCTGTCGCATTGCATTCATCAAACGCTGAATAGCAAAGCCTCTACGCGATTTCATAGTGGAGGCCTCCAGCAATCTTGAACACTCATCCTTAATACCTTTCTTTTTGGGGTAATTATACTCATCGCGAATATCCTTAAGAGGGCGTTCAGCAGCCAGCTTGCGGCAAAACTGCGTAACCGTCTTCTCGTTCTCTAAAAACCATTTTTCGTCCCCTTCGGGGCGGTGGTCTATAGTGTTATTAATATCAAGTTGGGGGAGCACATTGTAGCATTGTGCCAATTCATCCATTAACTTCTCAGTAACAGATACAAACTCATCACGAACGTCAGAAGCAGCCTTATGGCCGTCAACATACATATCACCATTATCGAACTCAACGCCCTGCATACGAAAAGACAGAATCTCTTCCAAGAGCTTGATGCGGTTATCCATCTGCTTATGGAAGCCAAGCATTGGGGATTGAGAAAGAGTAGAAATGCCACAAGCCTCAATAGCAGGTTTAAGAGCCTCGATACGCTCAAAGTCAAAATAATCAGCGTGACATTCAGAAGGGTAATAAGAACGAACCATAAAAAAGCCTCCAAGATTTGGAGGCATGAAAACATACAATTGGGAGGGTGTCAATCCTGACGGTTATTTCCTAGACAAATTAGAGCCAATACCATCAGCTTTACCGTCTTTCCAGAAATTGTTCCAAGTATCGGCAACAGCTTTATCAATACCATGAAAAATATCAACCACACCAGAAGCAGCATCAGTGACGACATTAGAAATATCCTTTGCAGTAGCGCCAATATGAGAAGAGCCATACCGCTGATTCTGCGTTTGCTGATGAACTAAGTCAACCTCAGCACTAACCTTGCGAGTCATTTCTTTGATTTGGTCATTGGTAAAATACTGACCAGCCGTTTGAGCTTGAGTAAGCATTTGGCGCATAATCTCGGAAACCTGCTGTTGCTTGGAAAGATTGGTGTTTTCCATAATAGACGCAACGCGAGCAGTAGACTCCTTCTGTTGATAAGCAAGCATCTCATTTTGTGCATATACCTGGTCTTTCGTATTCTGGCGTGAAGTCGCCGACTGAATGCCAGCAATCTCTTTTTGAGTCTCATTTTGCATCTCGGCAATCTCTTTCTGATTGTCCAGTTGCATTTTAGTAAGCTCTTTTTGATTCTCAAATCCGGCGTCAACCATACCAGCAGAGGAAGCATCAGCACCAGCACGCTCCCAAGCATTAAGCTCAGGAAATGCAGCAGCAAGATAATCACGAGTATCCTTTCCTTTATCAGCGGCAGACTTGCCACCAAGTCCAACCAAATCAAGCAACTTATCAGAAACGGCAGAAGTGCCAGCCTGCAACGTACCTTCAAGAAGTCCTTTACCAGCTTTAGCCATAGCACCAGAAACAAAACTAGGGGCGGCCTCATCAGGGTTAGGAACATTAGAGCCTTGAATGGCAGATTTAATACCAGCATCACCCATGCCTACAGTATTGTTATCGGTAGCAAGCACATCACCTTGAATGCCACCGGAGGCGGCTTTTTGACCGCCTCCAAACAATTTAGACATGGCGCCACCAGCAAGAGCAGAAGCAATACCGCCAGCAATAGCACCAAACATAAATCACCTCACTTAAGTGGCTGGAGACAAATAATCTCTTTAATAACCTGATTCAGCGAAACCAATCCGCGGCATTTAGTAGCGGTAAAGTTAGACCAAACCATGAAACCAACATAAACGTTATTGCCCGGCGTACGGGGAAGGACGTCAATAGTCACACAGTCCTTGACGGTATAATAACCACCATCATGGCGACCATTCAAAGGATAAACATCATAGGCAGTCGGGAGGGTAGTCGGAACCGAAGAAGACTCAAAGCGAACCAAACAGGCAAAAAATTTAGGGTCGGCATCAAAAGCAATATCAGCACCAACAGAAACAACCTGATTAGCGGCGTTGACAGATGTATCCATCTGAATGCAATGAAGAAAACCACCATTACCAGCATTAACCGTCAAACTATCAAAATATAACGTTGACGATGTAGCTTTAGGTGTCTGTAAAACAGGTGCCGAAGAAGCTGGAGTAACAGAAGTGAGAACCAGCTTATCAGAAAAAAAGTTTGAATTATGGCGAGAAATAAAAGTCTGAAACATGATTAAACTCCTAAGCAGAAAACCTACCGCGCTTCGCTTGGTCAACCCCTCAGCGGCAAAAATTAAAATTTTTACCGCTTCGGCGTTATAACCTCACACTCAATCTTTTATCACGAAGTCATGATTGAATCGCGAGTGGTCGGCAGATTGCGATAAACGGTCACATTAAATTTAACCTGACTATTCCACTGCAACAACTGAACGGACTGGAAACACTGGTCATAATCATGGTGGCGAATAAGTACGCGTTCTTGCAAATCACCAGAAGGCGGTTCCTGAATGAATGGGAAGCCTTCAAGAAGGTGATAAGCAGGAGAAACATACGAAGGCGCATAACGATACCACTGACCCTCAGCAATCTTAAACTTCTTAGACGAATCACCAGAACGGAAAACATCCTTCATAGAAATTTCACGCGGCGGCAAGTTGCCATACAAAACAGGGTCGCCAGCAATATCGGTATAAGTCAAAGCACCTTTAGCGTTAAGGTACTGAATCTCTTTAGTCGCAGTAGGCGGAAAACGAACAAGCGCAAGAGTAAACATAGTGCCATGCTCAGGAACAAAGAAACGCGGCACAGAATGTTTATAGGTCTGTTGAACACGACCAGAAAACTGGCCTAACGACGTTTGGTCAGTTCCATCAACATCATAGCCAGATGCCCAGAGATTAGAGCGCATGACAAGTAAAGGACGGTTGTCAGCGTCATAAGAGGTTTTACCTCCAAATGAAGAAATAACATCATGGTAACGCTGCATGAAGTAATCACGTTCTTGGTCAGTATGCAAATTAGCATAAGCAGCTTGCAGACCCATAATGTCAATAGATGTGGTAGAAGTCGTCATTTGGCGAGAAAGCTCAGTCTCAGGAGGAAGCGGAGCAGTCCAAATGTTTTTGAGATGGCAGCAACGGAAACCATAACGAGCATCATCTTGATTAAGCTCATTAGGGTTAGCCTCGGTACGGTCAGGCATCCACGGCGCTTTAAAATAGTTGTTATAGATATTCAAATAACCCTGAAACAAATGCTTAGGGATTTTATTGGTATCAGGGTTAATCGTGCCAAGAAAAGCGGCATGGTCAATATAACCAGTAGTGTTAACAGTCGGGAGAGGAGTGGCATTAACACCATCCTTCATGAACTTAATCCACTGTTCACCATAAACGTGACGATGAGGGACATAAAAAGTAAAAATGTCTACAGTAGAGTCAATAGCAAGGCCACGACGCAATGGAGAAAGACGGAGAGCGCCAACGGCGTCCATCTCGAAGGAGTCGCCAGCGATAACCGGAGTAGTTGAAATGGTAATAAGACGACCAATCTGACCAGCAAGGAAGCCAAGATGGGAAAGGTCATGCGGCATACGCTCGGCGCCAGTTTGAATATTAGACATAATTTATCCTCAAGTAAGGGGCCGAAGCCCCTGCAATTAAAATTGTTGACCACCTACATACCAAAGACGAGCGCCTTTACGCTTGCCTTTAGTACCTCGCAACGGCTGCGGACGACCAGGGCGAGCGCCAGAACGTTTTTTACCTTTAGACATTACATCACTCCTTCCGCACGTAATTTTTGACGCACGTTTTCTTCTGCGTCAGTAAGAACGTCAGTGTTTCCTGCGCGTACACGCAAGGTAAACGCGAACAATTCAGCGGCTTTAACCGGACGCTCGACGCCATTAATAATGTTTTCCGTAAATTCAGCGCCTTCCATGATGAGACAGGCCGTTTGAATGTTGACGGGATGAACATAATAAGCAATGACGGCAGCAATAAACTCAACAGGAGCAGGAAAGCGAGGGTATCCCACAAAGTCCAGCGTACCATAAACGCAAGCCTCAACGCAGCGACGAGCACGAGAGCGGTCAGTAGCAATCCAAACTTTGTTACTCGTCAGAAAATCGAAATCATCTTCGGTTAAATCCAAAACGGCAGAAGCCTGAATGAGCTTAATAGAGGCCAAAGCGGTCTGGAAACGTACGGATTGTTCAGTAACTTGACTCATGATTTCTTACCTATTAGTGGTTGAACAGCATCGGACTCAGATAGTAATCCACGCTCTTTTAAAATGTCAACAAGAGAATCTCTACCATGAACAAAATGTGACTCATATCTAAACCAGTCCTTGACGAACGTGCCAAGCATATTAAGCCACTTCTCCTCATCCAACGCGTCAGTTTTTGACAGAATCGTTAGTTGATGGCGAAAGGTCGCAAAGTAAGAGCTTCTCGAGCTGCGCAAGGATAGGTCGAATTTTCTCATTTTCCGCCAGCAGTCCACTTCGATTTAATTCGTAAACAAGCAGTAGTAATTCCTGCTTTATCAAGATAATTTTTCGACTCATCAGAAATATCCGAAAGTGTTAACTTCTGCGTCATGGAAGCGATAAAACTC\n"
     ]
    }
   ],
   "source": [
    "#Finding Reverse Complement\n",
    "\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "genome =\"\"\n",
    "#This is compatible with multiple fasta files too\n",
    "for seq_record in SeqIO.parse(\"DNA.fa\", \"fasta\"):\n",
    "    print(seq_record.description)\n",
    "    genome = str(seq_record.seq)\n",
    "    print(len(seq_record))\n",
    "    print(genome)\n",
    "    \n",
    "\n",
    "#Making a seq object out of the genome enables seq manipulation functions\n",
    "rc_genome = Seq(genome).reverse_complement()\n",
    "print(rc_genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NP_001273979.1\n",
      "GCTGCTGGTGGTCAGGCCTTCCATCGGAATCGGGCCTTTCTGATATTCGCGGCGCGGTTTAAAGCTTTCGCGCGGCAGATAGCTATGATAAAACGGATAGTTTTCGGTATATTCGCTCAGCAGGCACGGTTTTTCGGTTTTATCATAAATTTTGGTCGGCAGATGCGGGCAATGATGGCGGCCGCAGCTGCACAGTTCGCAAATGCATTTGGTTTTCAT\n",
      "NP_001017978.1\n",
      "GGTGCTTTTGCGATGCGGGCTCGCGCCGCGAAAGCCTTTGCTCAGCAGGGTATGTTCCAGTTCCACCAGTTTGTTTTCCACCATGCTCAGGTTCACCAGAATGCGTTTCTGGCGCGCAATGCTATGCGGAAAGTTGTTCAGAATATCGCGGCTCAGATCATACACCGCCAGGTTGTTATCGGTGTTGCTGTTAATCAGGCCGCTGCTGCTCGGGCGCACCAGCGCCAGCGCGGTGCTGTTGCTGCTCATTTCGCCGGTGTTGCGCTGAAAGCGGCGATATTTCCAAAACACAATCAGCGCGCACAGAATGCTGCTCGCCAGCAGCAGATAAAAGTTCAT\n",
      "sp|Q5H943.1|KKLC1_HUMAN\n",
      "GGTGCTTTTGCGATGCGGGCTCGCGCCGCGAAAGCCTTTGCTCAGCAGGGTATGTTCCAGTTCCACCAGTTTGTTTTCCACCATGCTCAGGTTCACCAGAATGCGTTTCTGGCGCGCAATGCTATGCGGAAAGTTGTTCAGAATATCGCGGCTCAGATCATACACCGCCAGGTTGTTATCGGTGTTGCTGTTAATCAGGCCGCTGCTGCTCGGGCGCACCAGCGCCAGCGCGGTGCTGTTGCTGCTCATTTCGCCGGTGTTGCGCTGAAAGCGGCGATATTTCCAAAACACAATCAGCGCGCACAGAATGCTGCTCGCCAGCAGCAGATAAAAGTTCAT\n",
      "AAC35497.1\n",
      "CACTTCATACAGAATCACCGGAAAATCCACATGAATGCGCGGATGATCCAGTTTCACAATGCCCTGCGGAATCAGGTTTTTATGCACGCGGCGCAGTTTCGCGCGTTTCTGGCCGTTTTTGGTCACAATGGTTTCTTCATAAAATTCATGCGCCAGATCGCCATCTTCATCATAAAACATGCTGCCGCGGCGGGTAAACACAAACGGCGGCACCGCGCGGCCGCGCGGGCGCACCAGCGCCTGTTCCGCGCCCGCCGCTTCGCTGCCGCCGCCGCCCGCCGCGCTCGCAAACGGCCACAGGCCGCGCGCTTTGCTGCCGCTCGCGCCCAT\n",
      "XP_002699503.1\n",
      "CACTTCGGTCGGTTTATCGGTGCTTTTGCGGTTCACCAGCGCGCCGTTCAGGCCTTTGGTCACCAGAAACTGTTCCAGTTCCGCCAGTTTATAATCCACAATGCGCAGGTTCACCAGAATGCGGCGCTGCATCGCAATGCTATGCGGAAAGTTAATCAGAATATCGCGGCTCAGGCTGTTCACCAGCAGGCTTTTATCAATGTTGCTTTTGGTGCTCTGGGTGCTGCTGCTGCTGCGCACCAGCGCCAGGCTGGTGCTGTTGCTGCTCATTTCGCCGGTGTTGCGCTGAAACCAGGTTTTCCAAAACACAAACATAAACGCAAACAGAATGCTGCTCACCAGCAGAAACAGAAAGCTCAT\n",
      "sp|Q4R717.1|KKLC1_MACFA\n",
      "GGTGCTTTTGCGATGCGCGCTCGCGCTGCGAAAGCCTTTGCTCACCAGAATATGTTCCAGTTCCACCAGTTTGTTTTCCACGGTGGTCAGGTTCACCAGAATGCGTTTCTGCATCGCAATGCTATGCGGAAAGTTGTTCAGAATATCGCGGCTCAGATCATACACGCTCAGGTTGTTATCGGTGTTGCTGTTAATCAGGCCGGTGCTGCTCGGGCGCACCAGCGCCAGCGCGGTGCTGTTGCTGCTCATTTCGCCGGTGTTGCGCTGAAAGCGGCGATATTTCCAAAACACGGTCATCAGCGCGCACAGAATGCCGCTCGCCAGCAGCAGATACACGTTCAT\n",
      "AAN12271.1\n",
      "GCAGCGGCCCCAGCTCACGGTCGCGGTCGGCCAGCCGCCTTCGCCGCGATAGCCATGCGCCACGCTCAGGCCATGATCAAAAATGCGCACATGGCGAAACAGATGCAGCAGCACGCGATGCAGCGCGCTGCTCAGGCGCTGCAGAAACACCACATACAGCAGCACGTTGCGGCTCATATCCAGCAGAAATTTCACCAGCAGCACAATGCCGCCCGGCATCGCATCATGCGGCGGCACTTCGGTGGTAATATGGCACATAAAGCCAAACACTTCATCCACTTCCACCTGCGCCAGATGGCCGTTTTTCTGTTCAAAGCGGGTCAGCAGCAT\n",
      "NP_001270680.1\n",
      "GGTGCTTTTGCGATGCGCGCTCGCGCTGCGAAAGCCTTTGCTCACCAGAATATGTTCCAGTTCCACCAGTTTGTTTTCCACGGTGGTCAGGTTCACCAGAATGCGTTTCTGCATCGCAATGCTATGCGGAAAGTTGTTCAGAATATCGCGGCTCAGATCATACACGCTCAGGTTGTTATCGGTGTTGCTGTTAATCAGGCCGGTGCTGCTCGGGCGCACCAGCGCCAGCGCGGTGCTGTTGCTGCTCATTTCGCCGGTGTTGCGCTGAAAGCGGCGATATTTCCAAAACACGGTCATCAGCGCGCACAGAATGCCGCTCGCCAGCAGCAGATACACGTTCAT\n",
      "Sp|P63244.3|Rack1_Human\n",
      "GCGGGTGCCAATGGTCACCTGCCACACGCGCACCAGGTTATCGGTATAGCCCGCAAACAGGGTCTGGCCATCCGCGCTCCACGCCAGGCTGGTGCACTGCGGCGGTTCTTTGCTGCTGGTGCTAATCACTTCCTGTTTCAGTTCATCCACAATAATTTTGCCTTCCAGATCCCAAATTTTAATGCTCGGGCCGGTCGCCGCGCACAGCCAATAGCGGTTCGGGCTAAAGCACAGCGCGTTAATAATATCGCCGCCATCCAGGGTATACAGATGTTTGCCTTCGTTCAGATCCCACAGCATCGCCTGGCCATCTTTGCCGCCGCTCGCGCACAGGCTGCCATCCGGGCTCACGGTCACGGTGTTCAGATAGCCGGTATGGCCAATATGGTTGGTTTTCAGTTTGCAGTTCGCCAGGTTCCACACTTTCACCAGTTTATCCCAGCCGCAGCTCACAATAATCGGGTTGCTGCTGTTCGGGCTAAAGCGCACGCAGCTCACCCATTCGCTATGGCTTTCATCCTGCACGGTATATTTGCACACGCCCAGGGTGTTCCACAGTTTAATGGTTTTATCGCGGCTGCCGCTCACAATCTGGCGGTTATCGCTGCTAAACGCCACGCTCAGCACATCTTTGGTATGGCCCACAAAGCGGCGGGTGGTGGTGCCGGTGGTCAGATCCCACAGGCGCAGGGTGCCATCCCAGCTGCCGCTCAGCGCAAACTGGCCATCGCTGCTAATCACCACATCGCTCACAAAATGGCTATGGCCGCGCAGCGCGCGCTGCGGAATGCCATAGTTGGTTTCATCGCGGGTCAGTTTCCACATAATAATGGTTTTATCGCGGCTCGCGCTCAGAATCATATCCGGAAACTGCGGGGTGGTCGCAATCTGGGTCACCCAGCCGTTATGGCCTTTCAGGGTGCCGCGCAGGGTCATCTGTTCGGTCAT\n",
      "XP_027389897.1\n",
      "CACTTCGGTCGGTTTATCGGTGCTTTTGCGGTTCACCAGCGCGCCGTTCAGGCCTTTGGTCACCAGAAACTGTTCCAGTTCCGCCAGTTTATAATCCACAATGCGCAGGTTCACCAGAATGCGGCGCTGCATCGCAATGCTATGCGGAAAGTTAATCAGAATATCGCGGCTCAGGCTGTTCACCAGCAGGCTTTTATCAATGTTGCTTTTGGTGCTCTGGGTGCTGCTGCTGCTGCGCACCAGCGCCAGGCTGGTGCTGTTGCTGCTCATTTCGCCGGTGTTGCGCTGAAACCAGGTTTTCCAAAACACAAACATAAACGCAAACAGAATGCTGCTCACCAGCAGAAACAGAAAGCTCAT\n",
      "XP_007647427.1\n",
      "CTGGTTGCTGCCGCTATCGCTTTCCATCGGAATGCGCACGGTCGCGCGGCGGCGATGTTTGGTATGGCCCTGCGGGTTTTTGCTGCTCAGAAACTGTTCCAGTTTGGTCAGCTGGCATTCCATAATGTTCAGGTTCACCAGAATGCGTTTATGCAGCGCCAGGGTGCGCGGAAAGTTGTTCATCATATCCTGGTTATAGCTGTTGCCCGGAAAGCTCAGGCCGGTGTTGCGATAGCTCCAAAAGCTGTTCGCGCGCAGCAGGGTCGGGCTGGTCGGGTTGCTCGGGCGTTCATGATCCGGAATCTGAAACACCACTTTAAAAATGCTAAAAATGCTCGCAAAAATAATGCTGCCCACCAGAATAAACAGAATGTTCAT\n",
      "XP_014918346.1\n",
      "ATGGTTGCCTTCGCTTTCGTTGCATTCGCGCAGGCGATCGGTGCTTTTGCAGTTCGCCAGGGTGCCGTTCAGGCCTTTAATAATCAGAAAATATTCCAGTTCCGCCAGTTTATATTCCACAATGCGCAGGTTAATCAGAATGCGTTTCTGCATCGCAATGGTATGCGGGCTGTTAATCAGAATGCCGCGGCTAATGCTATCCACGCTCAGGTTTTTATCGGTGTTGCTTTTGGTGCTGCCGGTGCTGCTGGTCGCGCGCACCAGCGCCAGGCTGGTGCTGTTGCTGCTCATTTCGCCCACGCTGCTCTGAAAGCGGCTTTTCCACACCACCAGCAGCAGGGTAAACACAATAAAGCTCAGCAGCAGCAGCAGCATGCTCAT\n",
      "XP_0026997503.1\n",
      "CACTTCGGTCGGTTTATCGGTGCTTTTGCGGTTCACCAGCGCGCCGTTCAGGCCTTTGGTCACCAGAAACTGTTCCAGTTCCGCCAGTTTATAATCCACAATGCGCAGGTTCACCAGAATGCGGCGCTGCATCGCAATGCTATGCGGAAAGTTAATCAGAATATCGCGGCTCAGGCTGTTCACCAGCAGGCTTTTATCAATGTTGCTTTTGGTGCTCTGGGTGCTGCTGCTGCTGCGCACCAGCGCCAGGCTGGTGCTGTTGCTGCTCATTTCGCCGGTGTTGCGCTGAAACCAGGTTTTCCAAAACACAAACATAAACGCAAACAGAATGCTGCTCACCAGCAGAAACAGAAAGCTCAT\n",
      "XP_002699603.1\n",
      "TTTCACGCGCTGATGCCAAAAGTTAAAGTTAAATTTGCTCAGGCTCAGTTTATGGCCCACCACCACCACGCACGCGCGATAATGGCTGCTGGTGCTGTTGCTGCTCATTTCGCCGGTGTTGCGCTGAAACCAGGTTTTCCAAAACACAAACATAAACGCAAACAGAATGCTGCTCACCAGCAGAAACAGAAAGCTCAT\n",
      "XP_004699503.1\n",
      "GCTTTTGCGGTTCACCAGCGCGCCGTTCACTTCGGTCGGTTTATCGCTGTTGCTGCTCATTTCGCCGGTGTTGCGCTGAAACCAGGTTTTCCAAAACACAAACATAAACGCAAACAGAATGCTGCTCACCAGCAGAAACAGAAAGCTCAT\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "for record in SeqIO.parse(\"DNA.fasta\", \"fasta\"):\n",
    "   print(record.id)\n",
    "   print(record.seq.reverse_complement())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "15"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "records = [rec.reverse_complement(id=\"rc_\"+rec.id, description = \"reverse complement\") \n",
    " for rec in SeqIO.parse(\"DNA.fasta\", \"fasta\")]\n",
    "len(records)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">NP_001273979.1 stabilizer of axonemal microtubules 1 isoform c [Homo sapiens]\n",
      "ATGAAAACCAAATGCATTTGCGAACTGTGCAGCTGCGGCCGCCATCATTGCCCGCATCTG\n",
      "CCGACCAAAATTTATGATAAAACCGAAAAACCGTGCCTGCTGAGCGAATATACCGAAAAC\n",
      "TATCCGTTTTATCATAGCTATCTGCCGCGCGAAAGCTTTAAACCGCGCCGCGAATATCAG\n",
      "AAAGGCCCGATTCCGATGGAAGGCCTGACCACCAGCAGC\n",
      ">NP_001017978.1 kita-kyushu lung cancer antigen 1 [Homo sapiens]\n",
      "ATGAACTTTTATCTGCTGCTGGCGAGCAGCATTCTGTGCGCGCTGATTGTGTTTTGGAAA\n",
      "TATCGCCGCTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCACCGCGCTGGCGCTG\n",
      "GTGCGCCCGAGCAGCAGCGGCCTGATTAACAGCAACACCGATAACAACCTGGCGGTGTAT\n",
      "GATCTGAGCCGCGATATTCTGAACAACTTTCCGCATAGCATTGCGCGCCAGAAACGCATT\n",
      "CTGGTGAACCTGAGCATGGTGGAAAACAAACTGGTGGAACTGGAACATACCCTGCTGAGC\n",
      "AAAGGCTTTCGCGGCGCGAGCCCGCATCGCAAAAGCACC\n",
      ">sp|Q5H943.1|KKLC1_HUMAN RecName: Full=Kita-kyushu lung cancer antigen 1; Short=KK-LC-1; AltName: Full=Cancer/testis antigen 83\n",
      "ATGAACTTTTATCTGCTGCTGGCGAGCAGCATTCTGTGCGCGCTGATTGTGTTTTGGAAA\n",
      "TATCGCCGCTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCACCGCGCTGGCGCTG\n",
      "GTGCGCCCGAGCAGCAGCGGCCTGATTAACAGCAACACCGATAACAACCTGGCGGTGTAT\n",
      "GATCTGAGCCGCGATATTCTGAACAACTTTCCGCATAGCATTGCGCGCCAGAAACGCATT\n",
      "CTGGTGAACCTGAGCATGGTGGAAAACAAACTGGTGGAACTGGAACATACCCTGCTGAGC\n",
      "AAAGGCTTTCGCGGCGCGAGCCCGCATCGCAAAAGCACC\n",
      ">AAC35497.1 lung cancer candidate FUS1 [Homo sapiens]\n",
      "ATGGGCGCGAGCGGCAGCAAAGCGCGCGGCCTGTGGCCGTTTGCGAGCGCGGCGGGCGGC\n",
      "GGCGGCAGCGAAGCGGCGGGCGCGGAACAGGCGCTGGTGCGCCCGCGCGGCCGCGCGGTG\n",
      "CCGCCGTTTGTGTTTACCCGCCGCGGCAGCATGTTTTATGATGAAGATGGCGATCTGGCG\n",
      "CATGAATTTTATGAAGAAACCATTGTGACCAAAAACGGCCAGAAACGCGCGAAACTGCGC\n",
      "CGCGTGCATAAAAACCTGATTCCGCAGGGCATTGTGAAACTGGATCATCCGCGCATTCAT\n",
      "GTGGATTTTCCGGTGATTCTGTATGAAGTG\n",
      ">XP_002699503.1 kita-kyushu lung cancer antigen 1 [Bos taurus]\n",
      "ATGAGCTTTCTGTTTCTGCTGGTGAGCAGCATTCTGTTTGCGTTTATGTTTGTGTTTTGG\n",
      "AAAACCTGGTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCACCAGCCTGGCGCTG\n",
      "GTGCGCAGCAGCAGCAGCACCCAGAGCACCAAAAGCAACATTGATAAAAGCCTGCTGGTG\n",
      "AACAGCCTGAGCCGCGATATTCTGATTAACTTTCCGCATAGCATTGCGATGCAGCGCCGC\n",
      "ATTCTGGTGAACCTGCGCATTGTGGATTATAAACTGGCGGAACTGGAACAGTTTCTGGTG\n",
      "ACCAAAGGCCTGAACGGCGCGCTGGTGAACCGCAAAAGCACCGATAAACCGACCGAAGTG\n",
      ">sp|Q4R717.1|KKLC1_MACFA RecName: Full=Kita-kyushu lung cancer antigen 1 homolog\n",
      "ATGAACGTGTATCTGCTGCTGGCGAGCGGCATTCTGTGCGCGCTGATGACCGTGTTTTGG\n",
      "AAATATCGCCGCTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCACCGCGCTGGCG\n",
      "CTGGTGCGCCCGAGCAGCACCGGCCTGATTAACAGCAACACCGATAACAACCTGAGCGTG\n",
      "TATGATCTGAGCCGCGATATTCTGAACAACTTTCCGCATAGCATTGCGATGCAGAAACGC\n",
      "ATTCTGGTGAACCTGACCACCGTGGAAAACAAACTGGTGGAACTGGAACATATTCTGGTG\n",
      "AGCAAAGGCTTTCGCAGCGCGAGCGCGCATCGCAAAAGCACC\n",
      ">AAN12271.1 lung cancer oncogene 5 [Homo sapiens]\n",
      "ATGCTGCTGACCCGCTTTGAACAGAAAAACGGCCATCTGGCGCAGGTGGAAGTGGATGAA\n",
      "GTGTTTGGCTTTATGTGCCATATTACCACCGAAGTGCCGCCGCATGATGCGATGCCGGGC\n",
      "GGCATTGTGCTGCTGGTGAAATTTCTGCTGGATATGAGCCGCAACGTGCTGCTGTATGTG\n",
      "GTGTTTCTGCAGCGCCTGAGCAGCGCGCTGCATCGCGTGCTGCTGCATCTGTTTCGCCAT\n",
      "GTGCGCATTTTTGATCATGGCCTGAGCGTGGCGCATGGCTATCGCGGCGAAGGCGGCTGG\n",
      "CCGACCGCGACCGTGAGCTGGGGCCGCTGC\n",
      ">NP_001270680.1 kita-kyushu lung cancer antigen 1 homolog [Macaca fascicularis]\n",
      "ATGAACGTGTATCTGCTGCTGGCGAGCGGCATTCTGTGCGCGCTGATGACCGTGTTTTGG\n",
      "AAATATCGCCGCTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCACCGCGCTGGCG\n",
      "CTGGTGCGCCCGAGCAGCACCGGCCTGATTAACAGCAACACCGATAACAACCTGAGCGTG\n",
      "TATGATCTGAGCCGCGATATTCTGAACAACTTTCCGCATAGCATTGCGATGCAGAAACGC\n",
      "ATTCTGGTGAACCTGACCACCGTGGAAAACAAACTGGTGGAACTGGAACATATTCTGGTG\n",
      "AGCAAAGGCTTTCGCAGCGCGAGCGCGCATCGCAAAAGCACC\n",
      ">Sp|P63244.3|Rack1_Human Recname: Full=Receptor Of Activated Protein C Kinase 1; Altname: Full=Cell Proliferation-Inducing Gene 21 Protein; Altname: Full=Guanine Nucleotide-Binding Protein Subunit Beta-2-Like 1; Altname: Full=Guanine Nucleotide-Binding Protein Subunit Beta-Like Protein 12.3; Altname: Full=Human Lung Cancer Oncogene 7 Protein; Short=Hlc-7; Altname: Full=Receptor For Activated C Kinase; Altname: Full=Small Ribosomal Subunit Protein Rack1; Contains: Recname: Full=Receptor Of Activated Protein C Kinase 1, N-Terminally Processed; Altname: Full=Guanine Nucleotide-Binding Protein Subunit Beta-2-Like 1, N-Terminally Processed\n",
      "ATGACCGAACAGATGACCCTGCGCGGCACCCTGAAAGGCCATAACGGCTGGGTGACCCAG\n",
      "ATTGCGACCACCCCGCAGTTTCCGGATATGATTCTGAGCGCGAGCCGCGATAAAACCATT\n",
      "ATTATGTGGAAACTGACCCGCGATGAAACCAACTATGGCATTCCGCAGCGCGCGCTGCGC\n",
      "GGCCATAGCCATTTTGTGAGCGATGTGGTGATTAGCAGCGATGGCCAGTTTGCGCTGAGC\n",
      "GGCAGCTGGGATGGCACCCTGCGCCTGTGGGATCTGACCACCGGCACCACCACCCGCCGC\n",
      "TTTGTGGGCCATACCAAAGATGTGCTGAGCGTGGCGTTTAGCAGCGATAACCGCCAGATT\n",
      "GTGAGCGGCAGCCGCGATAAAACCATTAAACTGTGGAACACCCTGGGCGTGTGCAAATAT\n",
      "ACCGTGCAGGATGAAAGCCATAGCGAATGGGTGAGCTGCGTGCGCTTTAGCCCGAACAGC\n",
      "AGCAACCCGATTATTGTGAGCTGCGGCTGGGATAAACTGGTGAAAGTGTGGAACCTGGCG\n",
      "AACTGCAAACTGAAAACCAACCATATTGGCCATACCGGCTATCTGAACACCGTGACCGTG\n",
      "AGCCCGGATGGCAGCCTGTGCGCGAGCGGCGGCAAAGATGGCCAGGCGATGCTGTGGGAT\n",
      "CTGAACGAAGGCAAACATCTGTATACCCTGGATGGCGGCGATATTATTAACGCGCTGTGC\n",
      "TTTAGCCCGAACCGCTATTGGCTGTGCGCGGCGACCGGCCCGAGCATTAAAATTTGGGAT\n",
      "CTGGAAGGCAAAATTATTGTGGATGAACTGAAACAGGAAGTGATTAGCACCAGCAGCAAA\n",
      "GAACCGCCGCAGTGCACCAGCCTGGCGTGGAGCGCGGATGGCCAGACCCTGTTTGCGGGC\n",
      "TATACCGATAACCTGGTGCGCGTGTGGCAGGTGACCATTGGCACCCGC\n",
      ">XP_027389897.1 kita-kyushu lung cancer antigen 1 [Bos indicus x Bos taurus]\n",
      "ATGAGCTTTCTGTTTCTGCTGGTGAGCAGCATTCTGTTTGCGTTTATGTTTGTGTTTTGG\n",
      "AAAACCTGGTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCACCAGCCTGGCGCTG\n",
      "GTGCGCAGCAGCAGCAGCACCCAGAGCACCAAAAGCAACATTGATAAAAGCCTGCTGGTG\n",
      "AACAGCCTGAGCCGCGATATTCTGATTAACTTTCCGCATAGCATTGCGATGCAGCGCCGC\n",
      "ATTCTGGTGAACCTGCGCATTGTGGATTATAAACTGGCGGAACTGGAACAGTTTCTGGTG\n",
      "ACCAAAGGCCTGAACGGCGCGCTGGTGAACCGCAAAAGCACCGATAAACCGACCGAAGTG\n",
      ">XP_007647427.1 kita-kyushu lung cancer antigen 1 [Cricetulus griseus]\n",
      "ATGAACATTCTGTTTATTCTGGTGGGCAGCATTATTTTTGCGAGCATTTTTAGCATTTTT\n",
      "AAAGTGGTGTTTCAGATTCCGGATCATGAACGCCCGAGCAACCCGACCAGCCCGACCCTG\n",
      "CTGCGCGCGAACAGCTTTTGGAGCTATCGCAACACCGGCCTGAGCTTTCCGGGCAACAGC\n",
      "TATAACCAGGATATGATGAACAACTTTCCGCGCACCCTGGCGCTGCATAAACGCATTCTG\n",
      "GTGAACCTGAACATTATGGAATGCCAGCTGACCAAACTGGAACAGTTTCTGAGCAGCAAA\n",
      "AACCCGCAGGGCCATACCAAACATCGCCGCCGCGCGACCGTGCGCATTCCGATGGAAAGC\n",
      "GATAGCGGCAGCAACCAG\n",
      ">XP_014918346.1 kita-kyushu lung cancer antigen 1 [Acinonyx jubatus]\n",
      "ATGAGCATGCTGCTGCTGCTGCTGAGCTTTATTGTGTTTACCCTGCTGCTGGTGGTGTGG\n",
      "AAAAGCCGCTTTCAGAGCAGCGTGGGCGAAATGAGCAGCAACAGCACCAGCCTGGCGCTG\n",
      "GTGCGCGCGACCAGCAGCACCGGCAGCACCAAAAGCAACACCGATAAAAACCTGAGCGTG\n",
      "GATAGCATTAGCCGCGGCATTCTGATTAACAGCCCGCATACCATTGCGATGCAGAAACGC\n",
      "ATTCTGATTAACCTGCGCATTGTGGAATATAAACTGGCGGAACTGGAATATTTTCTGATT\n",
      "ATTAAAGGCCTGAACGGCACCCTGGCGAACTGCAAAAGCACCGATCGCCTGCGCGAATGC\n",
      "AACGAAAGCGAAGGCAACCAT\n",
      ">XP_0026997503.1 kita-kyushu lung cancer antigen 1 [Bos taurus]\n",
      "ATGAGCTTTCTGTTTCTGCTGGTGAGCAGCATTCTGTTTGCGTTTATGTTTGTGTTTTGG\n",
      "AAAACCTGGTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCACCAGCCTGGCGCTG\n",
      "GTGCGCAGCAGCAGCAGCACCCAGAGCACCAAAAGCAACATTGATAAAAGCCTGCTGGTG\n",
      "AACAGCCTGAGCCGCGATATTCTGATTAACTTTCCGCATAGCATTGCGATGCAGCGCCGC\n",
      "ATTCTGGTGAACCTGCGCATTGTGGATTATAAACTGGCGGAACTGGAACAGTTTCTGGTG\n",
      "ACCAAAGGCCTGAACGGCGCGCTGGTGAACCGCAAAAGCACCGATAAACCGACCGAAGTG\n",
      ">XP_002699603.1 kita-kyushu lung cancer antigen 1 [Bos taurus]\n",
      "ATGAGCTTTCTGTTTCTGCTGGTGAGCAGCATTCTGTTTGCGTTTATGTTTGTGTTTTGG\n",
      "AAAACCTGGTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCACCAGCAGCCATTAT\n",
      "CGCGCGTGCGTGGTGGTGGTGGGCCATAAACTGAGCCTGAGCAAATTTAACTTTAACTTT\n",
      "TGGCATCAGCGCGTGAAA\n",
      ">XP_004699503.1 kita-kyushu lung cancer antigen 1 [Bos taurus]\n",
      "ATGAGCTTTCTGTTTCTGCTGGTGAGCAGCATTCTGTTTGCGTTTATGTTTGTGTTTTGG\n",
      "AAAACCTGGTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCGATAAACCGACCGAA\n",
      "GTGAACGGCGCGCTGGTGAACCGCAAAAGC\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "from io import StringIO\n",
    "records = SeqIO.parse(\"DNA.fasta\", \"fasta\")\n",
    "out_handle = StringIO()\n",
    "SeqIO.write(records, out_handle, \"fasta\")\n",
    "fasta_data = out_handle.getvalue()\n",
    "print(fasta_data)\n",
    "\n",
    "def reverseComplement(s):\n",
    "    complement = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A'}\n",
    "    t = ''\n",
    "    for base in s:\n",
    "        t = complement[base] + t\n",
    "    return t\n",
    "\n",
    "def match(s1, s2):\n",
    "    if not len(s1) == len(s2):\n",
    "        return False\n",
    "    for i in range(0, len(s1)):\n",
    "        if not s1[i] == s2[i]:\n",
    "            return False\n",
    "    return True"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# NAIVE-MATCHING"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def readFastq(filename):\n",
    "    sequences = []\n",
    "    qualities = []\n",
    "    with open(filename) as fh:\n",
    "        while True:\n",
    "            fh.readline() # skip name line\n",
    "            seq = fh.readline().rstrip() # read base sequence\n",
    "            fh.readline() # skip placeholder line\n",
    "            qual = fh.readline().rstrip() #base quality line\n",
    "            if len(seq) == 0:\n",
    "                break\n",
    "            sequences.append(seq)\n",
    "            qualities.append(qual)\n",
    "    return sequences, qualities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def readGenome(filename):\n",
    "    genome = ''\n",
    "    with open(filename, 'r') as f:\n",
    "        for line in f:\n",
    "            # ignore header line with genome information\n",
    "            if not line[0] == '>':\n",
    "                genome += line.rstrip()\n",
    "    return genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def naive(p, t):\n",
    "    occurrences = []\n",
    "    for i in range(len(t) - len(p) + 1):\n",
    "        match = True\n",
    "        for j in range(len(p)):\n",
    "            if t[i+j] != p[j]:\n",
    "                match = False\n",
    "                break\n",
    "        if match:\n",
    "            occurrences.append(i)\n",
    "    return occurrences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "healthy_gene = readGenome('DNA.fa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[377, 855, 903, 945, 1671, 1917, 2216, 2300, 2369, 2605, 2881, 2921, 3031, 3184, 3501, 3600, 3690, 3702, 4453, 4804, 4963, 5056]\n",
      "22\n"
     ]
    }
   ],
   "source": [
    "print(naive('AGGT', healthy_gene))\n",
    "print(len(naive('AGGT', healthy_gene)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACCT\n",
      "14\n"
     ]
    }
   ],
   "source": [
    "reverse_complement_AGGT = reverseComplement('AGGT')\n",
    "print(reverse_complement_AGGT)\n",
    "print(len(naive(reverse_complement_AGGT,healthy_gene)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[28, 104, 244, 327, 441, 473, 711, 732, 960, 1247, 1268, 1292, 1337, 1406, 1454, 1907, 2030, 2219, 2225, 2328, 2388, 2539, 2887, 2914, 3084, 3312, 4047, 4259, 4308, 4413, 4458, 4853, 4974, 5022, 5192]\n",
      "35\n"
     ]
    }
   ],
   "source": [
    "print(naive('TTAA', healthy_gene))\n",
    "print(len(naive('TTAA', healthy_gene)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TTAA\n",
      "35\n"
     ]
    }
   ],
   "source": [
    "reverse_complement_TTAA = reverseComplement('TTAA')\n",
    "print(reverse_complement_TTAA)\n",
    "print(len(naive(reverse_complement_TTAA,healthy_gene)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[]\n",
      "[3708]\n"
     ]
    }
   ],
   "source": [
    "print(naive('ACTAAGT', healthy_gene))\n",
    "complement = reverseComplement('ACTAAGT')\n",
    "print(naive(complement, healthy_gene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[53]\n",
      "[]\n"
     ]
    }
   ],
   "source": [
    "print(naive('AGTCGA', healthy_gene))\n",
    "complement = reverseComplement('AGTCGA')\n",
    "print(naive(complement, healthy_gene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def hamming_distance(s1,s2):\n",
    "    count = 0\n",
    "    for i in range(len(s1)):\n",
    "        if s1[i]!=s2[i]:\n",
    "            count +=1\n",
    "    return count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0, 4]\n"
     ]
    }
   ],
   "source": [
    "def naive_2mm(p, t):\n",
    "    occurrences = []\n",
    "    for i in range(len(t) - len(p) + 1):\n",
    "        subset_dna = t[i:i+len(p)]\n",
    "        ham_dist = hamming_distance(subset_dna,p)\n",
    "        if ham_dist <= 2:\n",
    "            occurrences.append(i)\n",
    "    return occurrences\n",
    "print(naive_2mm('ACTTTA', 'ACTTACTTGATAAAGT'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32\n"
     ]
    }
   ],
   "source": [
    "print(len(naive_2mm('TTCAAGCC',healthy_gene)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[945, 968, 1259, 1448, 1649, 1668, 1742, 1892, 1917, 1994, 2023, 2024, 2297, 2366, 2382, 2443, 2552, 2602, 2605, 2713, 2744, 2878, 2890, 2921, 2998, 3384, 3417, 3450, 3540, 3597, 3699, 3771, 3967, 4143, 4167, 4233, 4368, 4416, 4611, 4917, 5062, 5267]\n"
     ]
    }
   ],
   "source": [
    "print(naive_2mm('AGGAGGTT',healthy_gene))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# NAIVE-MATCHING ON FASTA FILE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[18, 25, 28, 31, 33, 34, 36, 37, 48, 50, 59, 91, 100, 102, 143, 145, 147, 161, 163, 166, 168, 183, 201, 214, 233, 239, 241, 244, 253, 255, 257, 280, 283, 292, 303, 310, 313, 321, 324, 326, 332, 338, 340, 342, 346, 349, 352, 354, 355, 357, 389, 392, 403, 406, 408, 428, 440, 442, 451, 461, 469, 476, 491, 494, 500, 514, 522, 526, 528, 529, 531, 533, 535, 537, 539, 544, 552, 572, 578, 580, 583, 592, 594, 596, 619, 622, 631, 642, 649, 652, 660, 663, 665, 671, 677, 679, 681, 685, 688, 691, 693, 694, 696, 728, 731, 742, 745, 747, 767, 779, 781, 790, 800, 808, 815, 830, 833, 839, 853, 861, 865, 867, 868, 870, 872, 874, 876, 878, 883, 891, 898, 900, 902, 904, 906, 907, 909, 910, 917, 919, 921, 922, 924, 929, 931, 932, 938, 940, 942, 944, 945, 947, 949, 951, 952, 954, 955, 957, 958, 960, 961, 963, 966, 968, 969, 971, 973, 975, 977, 980, 986, 992, 994, 996, 998, 1000, 1002, 1003, 1005, 1006, 1008, 1010, 1013, 1016, 1019, 1033, 1036, 1038, 1039, 1041, 1042, 1065, 1073, 1113, 1120, 1122, 1124, 1132, 1135, 1137, 1139, 1157, 1162, 1164, 1178, 1184, 1186, 1208, 1228, 1247, 1249, 1252, 1265, 1300, 1311, 1318, 1321, 1329, 1340, 1346, 1348, 1351, 1354, 1357, 1360, 1362, 1369, 1371, 1403, 1414, 1417, 1419, 1439, 1451, 1454, 1457, 1459, 1462, 1472, 1480, 1502, 1505, 1511, 1523, 1533, 1542, 1544, 1550, 1555, 1563, 1601, 1607, 1609, 1611, 1612, 1614, 1621, 1623, 1625, 1651, 1654, 1663, 1674, 1681, 1684, 1692, 1695, 1697, 1703, 1709, 1711, 1713, 1717, 1720, 1722, 1728, 1759, 1761, 1774, 1777, 1779, 1799, 1811, 1814, 1822, 1832, 1862, 1865, 1871, 1883, 1885, 1893, 1897, 1900, 1902, 1904, 1906, 1908, 1910, 1915, 1923, 1931, 1937, 1939, 1959, 1967, 1973, 1995, 2002, 2021, 2024, 2027, 2034, 2036, 2039, 2042, 2044, 2046, 2047, 2049, 2054, 2057, 2063, 2083, 2086, 2093, 2096, 2108, 2117, 2119, 2125, 2128, 2130, 2132, 2135, 2140, 2142, 2144, 2147, 2150, 2161, 2168, 2170, 2187, 2191, 2193, 2196, 2198, 2205, 2209, 2211, 2212, 2214, 2220, 2221, 2223, 2227, 2228, 2232, 2234, 2242, 2248, 2249, 2250, 2251, 2254, 2273, 2279, 2281, 2283, 2284, 2286, 2293, 2295, 2297, 2323, 2326, 2335, 2346, 2353, 2356, 2364, 2367, 2369, 2375, 2381, 2383, 2385, 2389, 2392, 2394, 2400, 2431, 2433, 2446, 2449, 2451, 2471, 2483, 2486, 2494, 2504, 2534, 2537, 2543, 2555, 2557, 2565, 2569, 2572, 2574, 2576, 2578, 2580, 2582, 2587, 2595, 2615, 2620, 2622, 2623, 2625, 2637, 2646, 2651, 2654, 2663, 2672, 2695, 2697, 2699, 2701, 2704, 2706, 2735, 2737, 2739, 2757, 2762, 2765, 2767, 2769, 2771, 2774, 2776, 2778, 2779, 2781, 2797, 2799, 2807, 2815, 2817, 2823, 2831, 2836, 2838, 2839, 2841, 2842, 2853, 2860, 2870, 2883, 2893, 2896, 2905, 2906, 2907, 2921, 2926, 2928, 2931, 2933, 2941, 2943, 2950, 2962, 2964, 2965, 2967, 2968, 2971, 2973, 2995, 3004, 3006, 3010, 3023, 3042, 3050, 3052, 3055, 3057, 3059, 3061, 3069, 3079, 3097, 3100, 3102, 3103, 3105, 3119, 3130, 3137, 3168, 3177, 3199, 3201, 3210, 3211, 3217, 3219, 3221, 3223, 3225, 3226, 3228, 3229, 3231, 3240, 3245, 3248, 3257, 3270, 3294, 3295, 3297, 3309, 3311, 3316, 3324, 3331, 3340, 3343, 3345, 3347, 3348, 3350, 3357, 3361, 3377, 3387, 3426, 3433, 3443, 3446, 3451, 3453, 3464, 3469, 3471, 3473, 3480, 3485, 3494, 3496, 3498, 3515, 3517, 3519, 3524, 3526, 3530, 3540, 3544, 3550, 3569, 3571, 3574, 3587, 3622, 3633, 3640, 3643, 3651, 3662, 3668, 3670, 3673, 3676, 3679, 3682, 3684, 3691, 3693, 3725, 3736, 3739, 3741, 3761, 3773, 3776, 3779, 3781, 3784, 3794, 3802, 3824, 3827, 3833, 3845, 3855, 3864, 3866, 3872, 3877, 3885, 3929, 3931, 3933, 3934, 3947, 3949, 3974, 3989, 3997, 3999, 4003, 4017, 4022, 4028, 4030, 4032, 4034, 4048, 4054, 4065, 4069, 4078, 4080, 4115, 4117, 4119, 4127, 4130, 4138, 4148, 4187, 4198, 4201, 4211, 4216, 4217, 4218, 4231, 4234, 4237, 4239, 4241, 4247, 4249, 4266, 4272, 4273, 4275, 4276, 4288, 4295, 4298, 4301, 4304, 4309, 4331, 4337, 4340, 4351, 4360, 4363, 4365, 4369, 4371, 4378, 4381, 4389, 4400, 4406, 4408, 4410, 4412, 4420, 4422, 4428, 4429, 4431, 4459, 4461, 4477, 4479, 4480, 4482, 4497, 4499, 4511, 4514, 4522, 4540, 4562, 4565, 4593, 4602, 4610, 4623, 4630, 4636, 4638, 4653, 4659, 4669, 4688, 4690, 4693, 4706, 4741, 4752, 4759, 4762, 4770, 4781, 4787, 4789, 4792, 4795, 4798, 4801, 4803, 4810, 4812, 4844, 4855, 4858, 4860, 4880, 4892, 4895, 4898, 4900, 4903, 4913, 4921, 4943, 4946, 4952, 4964, 4974, 4983, 4985, 4991, 4996, 5004, 5029, 5048, 5050, 5053, 5066, 5101, 5112, 5119, 5122, 5130, 5137, 5146, 5148, 5150, 5152, 5154, 5159, 5162, 5165, 5167, 5168, 5169, 5179, 5185, 5209, 5215, 5217, 5227, 5246, 5248, 5251, 5264, 5299, 5310, 5317, 5320, 5328, 5352, 5354, 5360, 5365]\n",
      "792\n"
     ]
    }
   ],
   "source": [
    "#function for parsing fasta file\n",
    "def readGenome(filename):\n",
    "    genome = ''\n",
    "    with open(filename, 'r') as f:\n",
    "        for line in f:\n",
    "            # ignore header line with genome information\n",
    "            if not line[0] == '>':\n",
    "                genome += line.rstrip()\n",
    "    return genome\n",
    "genome = readGenome('DNA.fasta')\n",
    "genome[:100]\n",
    "p = 'GGCGC'\n",
    "t = genome\n",
    "print(naive_2mm(p,t))\n",
    "print(len(naive_2mm(p,t)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wall time: 0 ns\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "p = 'shake spea'\n",
    "t = 'Shakespear'\n",
    "naive_2mm(p,t)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CREATING GENERIC SEQUENCE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'Bio.Seq.Seq'>\n",
      "['alphabet', 'back_transcribe', 'complement', 'count', 'count_overlap', 'endswith', 'find', 'lower', 'lstrip', 'reverse_complement', 'rfind', 'rsplit', 'rstrip', 'split', 'startswith', 'strip', 'tomutable', 'tostring', 'transcribe', 'translate', 'ungap', 'upper']\n"
     ]
    }
   ],
   "source": [
    "#let's make a generic sequence\n",
    "\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "my_seq = Seq(\"CCCGGAGAGA\")\n",
    "print(type(my_seq))\n",
    "#let's see what attributes this object has\n",
    "attributes = [a for a in dir(my_seq) if not a.startswith(\"_\")]\n",
    "print(attributes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alphabet()\n"
     ]
    }
   ],
   "source": [
    "#right now it has just a generic alphabet\n",
    "print(my_seq.alphabet)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Alphabet import generic_dna, generic_protein, generic_rna\n",
    "\n",
    "my_dna = Seq(\"CCCGGAGAG\", generic_dna)\n",
    "my_rna = Seq(\"ACCCGUUGU\", generic_rna)\n",
    "my_protein = Seq(\"AKKKGGGUUULL\", generic_protein)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACUAGCAGCGGA\n",
      "RNAAlphabet()\n",
      "TSSG\n",
      "ExtendedIUPACProtein()\n"
     ]
    }
   ],
   "source": [
    "my_gene = Seq(\"ACTAGCAGCGGA\", generic_dna)\n",
    "\n",
    "#get the mRNA\n",
    "\n",
    "my_transcript = my_gene.transcribe()\n",
    "print(my_transcript)\n",
    "print(my_transcript.alphabet)\n",
    "\n",
    "\n",
    "#get the protein from the mRNA\n",
    "my_protein = my_transcript.translate()\n",
    "print(my_protein)\n",
    "print(my_protein.alphabet)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MAIVMGR*KGAR*\n"
     ]
    }
   ],
   "source": [
    "coding_dna = Seq(\"ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\", generic_dna)\n",
    "myprot = coding_dna.translate()\n",
    "print(myprot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('MAIVM', ExtendedIUPACProtein())"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene = Seq(\"ATGGCCATTGTAATGTAG\", generic_dna)\n",
    "gene.translate(cds=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('AAACGGAGGAGAT', DNAAlphabet())"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq1 = Seq(\"AAACGGA\", generic_dna)\n",
    "seq2 = Seq(\"GGAGAT\", generic_dna)\n",
    "\n",
    "seq1 + seq2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "MutableSeq('AAACGGA', DNAAlphabet())"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mut_seq = seq1.tomutable()\n",
    "mut_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GAACGGA\n"
     ]
    }
   ],
   "source": [
    "mut_seq[0] = \"G\"\n",
    "print(mut_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3\n",
      "2\n"
     ]
    }
   ],
   "source": [
    "myseq = Seq(\"CCAGAAACCCGGAA\", generic_dna)\n",
    "\n",
    "#find the first occurence of the pattern\n",
    "print(myseq.find(\"GAA\"))\n",
    "\n",
    "#find the number of non-overlapping occurrences of a pattern\n",
    "print(myseq.count(\"GAA\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Blast import NCBIWWW\n",
    "\n",
    "result_handle = NCBIWWW.qblast(\"blastn\", \"nt\", Seq(\"AAAAGGAGAGAGAGTTTATA\", generic_dna))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Blast import NCBIXML\n",
    "blast_records = NCBIXML.parse(result_handle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "****Alignment****\n",
      "sequence: gi|1395234831|gb|CP026251.1| Scophthalmus maximus chromosome 9\n",
      "length: 25242470\n",
      "e value: 3.49484\n",
      "AAAAGGAGAGAGAGTTTATA...\n",
      "||||||||||||||||||||...\n",
      "AAAAGGAGAGAGAGTTTATA...\n"
     ]
    }
   ],
   "source": [
    "for b in blast_records:\n",
    "    for alignment in b.alignments:\n",
    "        for hsp in alignment.hsps:\n",
    "            print('****Alignment****')\n",
    "            print('sequence:', alignment.title)\n",
    "            print('length:', alignment.length)\n",
    "            print('e value:', hsp.expect)\n",
    "            print(hsp.query[0:75] + '...')\n",
    "            print(hsp.match[0:75] + '...')\n",
    "            print(hsp.sbjct[0:75] + '...')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Length of Sequence A: 21\n",
      "\n",
      "Length of Sequence B: 21\n",
      "\n",
      "TGGAGGCAATGGCGGCCAGCA\n",
      "||||||||| |||||||||  \n",
      "GACTCCTCCTCCTCCTGCTCA\n"
     ]
    }
   ],
   "source": [
    "seq_a = \"TGGAGGCAATGGCGGCCAGCA\"\n",
    "seq_b = \"GACTCCTCCTCCTCCTGCTCA\"    \n",
    "len_a = len(seq_a)    \n",
    "len_b = len(seq_b)    \n",
    "print(\"Length of Sequence A: \" + str(len_a))    \n",
    "print()    \n",
    "print(\"Length of Sequence B: \" + str(len_b))\n",
    "print()\n",
    "\n",
    "def sequence_compare(seq_a, seq_b):\n",
    "        len1 = len(seq_a)\n",
    "        len2 = len(seq_b)\n",
    "        mismatches = []\n",
    "        for pos in range (0, min(len1, len2)) :\n",
    "              if seq_a[pos] != seq_b[pos]:\n",
    "                       mismatches.append('|')\n",
    "              else:\n",
    "                       mismatches.append(' ')\n",
    "        print (seq_a)\n",
    "        print (\"\".join(mismatches))\n",
    "        print (seq_b)\n",
    "sequence_compare(seq_a,seq_b)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# HISTOGRAM--VISUALISATION"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ATGAAAACCAAATGCATTTGCGAACTGTGCAGCTGCGGCCGCCATCATTGCCCGCATCTGCCGACCAAAATTTATGATAAAACCGAAAAACCGTGCCTGCTGAGCGAATATACCGAAAACTATCCGTTTTATCATAGCTATCTGCCGCGCGAAAGCTTTAAACCGCGCCGCGAATATCAGAAAGGCCCGATTCCGATGGAAGGCCTGACCACCAGCAGCATGAACTTTTATCTGCTGCTGGCGAGCAGCATTCTGTGCGCGCTGATTGTGTTTTGGAAATATCGCCGCTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCACCGCGCTGGCGCTGGTGCGCCCGAGCAGCAGCGGCCTGATTAACAGCAACACCGATAACAACCTGGCGGTGTATGATCTGAGCCGCGATATTCTGAACAACTTTCCGCATAGCATTGCGCGCCAGAAACGCATTCTGGTGAACCTGAGCATGGTGGAAAACAAACTGGTGGAACTGGAACATACCCTGCTGAGCAAAGGCTTTCGCGGCGCGAGCCCGCATCGCAAAAGCACCATGAACTTTTATCTGCTGCTGGCGAGCAGCATTCTGTGCGCGCTGATTGTGTTTTGGAAATATCGCCGCTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCACCGCGCTGGCGCTGGTGCGCCCGAGCAGCAGCGGCCTGATTAACAGCAACACCGATAACAACCTGGCGGTGTATGATCTGAGCCGCGATATTCTGAACAACTTTCCGCATAGCATTGCGCGCCAGAAACGCATTCTGGTGAACCTGAGCATGGTGGAAAACAAACTGGTGGAACTGGAACATACCCTGCTGAGCAAAGGCTTTCGCGGCGCGAGCCCGCATCGCAAAAGCACCATGGGCGCGAGCGGCAGCAAAGCGCGCGGCCTGTGGCCGTTTGCGAGCGCGGCGGGCGGCGGCGGCAGCGAAGCGGCGGGCGCGGAACAGGCGCTGGTGCGCCCGCGCGGCCGCGCGGTGCCGCCGTTTGTGTTTACCCGCCGCGGCAGCATGTTTTATGATGAAGATGGCGATCTGGCGCATGAATTTTATGAAGAAACCATTGTGACCAAAAACGGCCAGAAACGCGCGAAACTGCGCCGCGTGCATAAAAACCTGATTCCGCAGGGCATTGTGAAACTGGATCATCCGCGCATTCATGTGGATTTTCCGGTGATTCTGTATGAAGTGATGAGCTTTCTGTTTCTGCTGGTGAGCAGCATTCTGTTTGCGTTTATGTTTGTGTTTTGGAAAACCTGGTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCACCAGCCTGGCGCTGGTGCGCAGCAGCAGCAGCACCCAGAGCACCAAAAGCAACATTGATAAAAGCCTGCTGGTGAACAGCCTGAGCCGCGATATTCTGATTAACTTTCCGCATAGCATTGCGATGCAGCGCCGCATTCTGGTGAACCTGCGCATTGTGGATTATAAACTGGCGGAACTGGAACAGTTTCTGGTGACCAAAGGCCTGAACGGCGCGCTGGTGAACCGCAAAAGCACCGATAAACCGACCGAAGTGATGAACGTGTATCTGCTGCTGGCGAGCGGCATTCTGTGCGCGCTGATGACCGTGTTTTGGAAATATCGCCGCTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCACCGCGCTGGCGCTGGTGCGCCCGAGCAGCACCGGCCTGATTAACAGCAACACCGATAACAACCTGAGCGTGTATGATCTGAGCCGCGATATTCTGAACAACTTTCCGCATAGCATTGCGATGCAGAAACGCATTCTGGTGAACCTGACCACCGTGGAAAACAAACTGGTGGAACTGGAACATATTCTGGTGAGCAAAGGCTTTCGCAGCGCGAGCGCGCATCGCAAAAGCACCATGCTGCTGACCCGCTTTGAACAGAAAAACGGCCATCTGGCGCAGGTGGAAGTGGATGAAGTGTTTGGCTTTATGTGCCATATTACCACCGAAGTGCCGCCGCATGATGCGATGCCGGGCGGCATTGTGCTGCTGGTGAAATTTCTGCTGGATATGAGCCGCAACGTGCTGCTGTATGTGGTGTTTCTGCAGCGCCTGAGCAGCGCGCTGCATCGCGTGCTGCTGCATCTGTTTCGCCATGTGCGCATTTTTGATCATGGCCTGAGCGTGGCGCATGGCTATCGCGGCGAAGGCGGCTGGCCGACCGCGACCGTGAGCTGGGGCCGCTGCATGAACGTGTATCTGCTGCTGGCGAGCGGCATTCTGTGCGCGCTGATGACCGTGTTTTGGAAATATCGCCGCTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCACCGCGCTGGCGCTGGTGCGCCCGAGCAGCACCGGCCTGATTAACAGCAACACCGATAACAACCTGAGCGTGTATGATCTGAGCCGCGATATTCTGAACAACTTTCCGCATAGCATTGCGATGCAGAAACGCATTCTGGTGAACCTGACCACCGTGGAAAACAAACTGGTGGAACTGGAACATATTCTGGTGAGCAAAGGCTTTCGCAGCGCGAGCGCGCATCGCAAAAGCACCATGACCGAACAGATGACCCTGCGCGGCACCCTGAAAGGCCATAACGGCTGGGTGACCCAGATTGCGACCACCCCGCAGTTTCCGGATATGATTCTGAGCGCGAGCCGCGATAAAACCATTATTATGTGGAAACTGACCCGCGATGAAACCAACTATGGCATTCCGCAGCGCGCGCTGCGCGGCCATAGCCATTTTGTGAGCGATGTGGTGATTAGCAGCGATGGCCAGTTTGCGCTGAGCGGCAGCTGGGATGGCACCCTGCGCCTGTGGGATCTGACCACCGGCACCACCACCCGCCGCTTTGTGGGCCATACCAAAGATGTGCTGAGCGTGGCGTTTAGCAGCGATAACCGCCAGATTGTGAGCGGCAGCCGCGATAAAACCATTAAACTGTGGAACACCCTGGGCGTGTGCAAATATACCGTGCAGGATGAAAGCCATAGCGAATGGGTGAGCTGCGTGCGCTTTAGCCCGAACAGCAGCAACCCGATTATTGTGAGCTGCGGCTGGGATAAACTGGTGAAAGTGTGGAACCTGGCGAACTGCAAACTGAAAACCAACCATATTGGCCATACCGGCTATCTGAACACCGTGACCGTGAGCCCGGATGGCAGCCTGTGCGCGAGCGGCGGCAAAGATGGCCAGGCGATGCTGTGGGATCTGAACGAAGGCAAACATCTGTATACCCTGGATGGCGGCGATATTATTAACGCGCTGTGCTTTAGCCCGAACCGCTATTGGCTGTGCGCGGCGACCGGCCCGAGCATTAAAATTTGGGATCTGGAAGGCAAAATTATTGTGGATGAACTGAAACAGGAAGTGATTAGCACCAGCAGCAAAGAACCGCCGCAGTGCACCAGCCTGGCGTGGAGCGCGGATGGCCAGACCCTGTTTGCGGGCTATACCGATAACCTGGTGCGCGTGTGGCAGGTGACCATTGGCACCCGCATGAGCTTTCTGTTTCTGCTGGTGAGCAGCATTCTGTTTGCGTTTATGTTTGTGTTTTGGAAAACCTGGTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCACCAGCCTGGCGCTGGTGCGCAGCAGCAGCAGCACCCAGAGCACCAAAAGCAACATTGATAAAAGCCTGCTGGTGAACAGCCTGAGCCGCGATATTCTGATTAACTTTCCGCATAGCATTGCGATGCAGCGCCGCATTCTGGTGAACCTGCGCATTGTGGATTATAAACTGGCGGAACTGGAACAGTTTCTGGTGACCAAAGGCCTGAACGGCGCGCTGGTGAACCGCAAAAGCACCGATAAACCGACCGAAGTGATGAACATTCTGTTTATTCTGGTGGGCAGCATTATTTTTGCGAGCATTTTTAGCATTTTTAAAGTGGTGTTTCAGATTCCGGATCATGAACGCCCGAGCAACCCGACCAGCCCGACCCTGCTGCGCGCGAACAGCTTTTGGAGCTATCGCAACACCGGCCTGAGCTTTCCGGGCAACAGCTATAACCAGGATATGATGAACAACTTTCCGCGCACCCTGGCGCTGCATAAACGCATTCTGGTGAACCTGAACATTATGGAATGCCAGCTGACCAAACTGGAACAGTTTCTGAGCAGCAAAAACCCGCAGGGCCATACCAAACATCGCCGCCGCGCGACCGTGCGCATTCCGATGGAAAGCGATAGCGGCAGCAACCAGATGAGCATGCTGCTGCTGCTGCTGAGCTTTATTGTGTTTACCCTGCTGCTGGTGGTGTGGAAAAGCCGCTTTCAGAGCAGCGTGGGCGAAATGAGCAGCAACAGCACCAGCCTGGCGCTGGTGCGCGCGACCAGCAGCACCGGCAGCACCAAAAGCAACACCGATAAAAACCTGAGCGTGGATAGCATTAGCCGCGGCATTCTGATTAACAGCCCGCATACCATTGCGATGCAGAAACGCATTCTGATTAACCTGCGCATTGTGGAATATAAACTGGCGGAACTGGAATATTTTCTGATTATTAAAGGCCTGAACGGCACCCTGGCGAACTGCAAAAGCACCGATCGCCTGCGCGAATGCAACGAAAGCGAAGGCAACCATATGAGCTTTCTGTTTCTGCTGGTGAGCAGCATTCTGTTTGCGTTTATGTTTGTGTTTTGGAAAACCTGGTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCACCAGCCTGGCGCTGGTGCGCAGCAGCAGCAGCACCCAGAGCACCAAAAGCAACATTGATAAAAGCCTGCTGGTGAACAGCCTGAGCCGCGATATTCTGATTAACTTTCCGCATAGCATTGCGATGCAGCGCCGCATTCTGGTGAACCTGCGCATTGTGGATTATAAACTGGCGGAACTGGAACAGTTTCTGGTGACCAAAGGCCTGAACGGCGCGCTGGTGAACCGCAAAAGCACCGATAAACCGACCGAAGTGATGAGCTTTCTGTTTCTGCTGGTGAGCAGCATTCTGTTTGCGTTTATGTTTGTGTTTTGGAAAACCTGGTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCACCAGCAGCCATTATCGCGCGTGCGTGGTGGTGGTGGGCCATAAACTGAGCCTGAGCAAATTTAACTTTAACTTTTGGCATCAGCGCGTGAAAATGAGCTTTCTGTTTCTGCTGGTGAGCAGCATTCTGTTTGCGTTTATGTTTGTGTTTTGGAAAACCTGGTTTCAGCGCAACACCGGCGAAATGAGCAGCAACAGCGATAAACCGACCGAAGTGAACGGCGCGCTGGTGAACCGCAAAAGC'"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def readGenome(filename):\n",
    "    genome = ''\n",
    "    with open(filename, 'r') as f:\n",
    "        for line in f:\n",
    "            # ignore header line with genome information\n",
    "            if not line[0] == '>':\n",
    "                genome += line.rstrip()\n",
    "    return genome\n",
    "genome = readGenome('DNA.fasta')\n",
    "genome[0:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'A': 1321, 'C': 1411, 'G': 1522, 'T': 1122}\n"
     ]
    }
   ],
   "source": [
    "# Count the number of occurences of each base\n",
    "counts = {'A': 0, 'C': 0, 'G': 0, 'T': 0}\n",
    "for base in genome:\n",
    "    counts[base] += 1\n",
    "print(counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({'A': 1321, 'T': 1122, 'G': 1522, 'C': 1411})"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import collections\n",
    "collections.Counter(genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "def readFastq(filename):\n",
    "    sequences = []\n",
    "    qualities = []\n",
    "    with open(filename) as fh:\n",
    "        while True:\n",
    "            fh.readline() # skip name line\n",
    "            seq = fh.readline().rstrip() # read base sequence\n",
    "            fh.readline() # skip placeholder line\n",
    "            qual = fh.readline().rstrip() #base quality line\n",
    "            if len(seq) == 0:\n",
    "                break\n",
    "            sequences.append(seq)\n",
    "            qualities.append(qual)\n",
    "    return sequences, qualities\n",
    "seqs, quals = readFastq('DNA.fastq')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def phred33ToQ(qual):\n",
    "    return ord(qual) - 33"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0, 0, 6178, 0, 0, 54, 108, 574, 345, 83, 193, 124, 79, 165, 49, 236, 184, 327, 514, 238, 531, 254, 313, 798, 992, 888, 1396, 1488, 993, 1752, 3387, 4487, 3248, 5476, 8375, 11814, 4243, 7827, 6579, 8179, 9349, 8180, 0, 0, 0, 0, 0, 0, 0, 0]\n"
     ]
    }
   ],
   "source": [
    "def createHist(qualities):\n",
    "    # Create a histogram of quality scores\n",
    "    hist = [0]*50\n",
    "    for qual in qualities:\n",
    "        for phred in qual:\n",
    "            q = phred33ToQ(phred)\n",
    "            hist[q] += 1\n",
    "    return hist\n",
    "h = createHist(quals)\n",
    "print(h)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYcAAAD8CAYAAACcjGjIAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvnQurowAAIABJREFUeJzt3Xl829WZ6P/PI8mSLdmOnXjJTgJJyAKEJWyl0JJQkrYMobdlLp3ONDNDy/3N0NIO099A5y5M6Y9p+5t22mHa0ttbuEPvpSyXaYeUUkIaCHS6QJwAIQtJTDY78ZZ4kTfZWs79Q9+vrdjaLMmWZT3v1ysvS0dfSUet8aNznnPOI8YYlFJKqViOfHdAKaXU9KPBQSml1DgaHJRSSo2jwUEppdQ4GhyUUkqNo8FBKaXUOBoclFJKjaPBQSml1DgaHJRSSo3jyncHMlVTU2OWLFmS724opVRB2b179xljTG2q6wo2OCxZsoSGhoZ8d0MppQqKiJxI5zqdVlJKKTVOyuAgIo+JSLuI7Itp+wcReVdE9orIz0SkKuaxL4tIo4gcEpGNMe2brLZGEbk/pn2piLwuIkdE5GkRcefyAyqllJq4dEYO/wJsGtO2HbjIGHMJcBj4MoCIrAbuANZYz/m+iDhFxAl8D/gwsBr4pHUtwDeAbxtjlgNdwJ1ZfSKllFJZSxkcjDGvAZ1j2l4yxoSsu78HFlq3NwNPGWOGjDHHgEbgKutfozHmqDFmGHgK2CwiAqwHnrWe/zhwW5afSSmlVJZykXP4c+CX1u0FQFPMY81WW6L2OUB3TKCx25VSSuVRVsFBRP4zEAKesJviXGYyaE/0fneJSIOINHR0dEy0u0oppdKUcXAQkS3ALcCnzGg5uWZgUcxlC4HTSdrPAFUi4hrTHpcx5ofGmHXGmHW1tSmX6SqllMpQRsFBRDYB9wG3GmMGYh7aCtwhIh4RWQosB94AdgHLrZVJbqJJ661WUHkF+IT1/C3Ac5l9FKXUdNDVP8zWtxN+x1MFIp2lrE8CvwMuFJFmEbkT+C5QAWwXkbdE5AcAxpj9wDPAAeBF4G5jTNjKKXwO2AYcBJ6xroVokLlXRBqJ5iAezeknVEpNqWcamrjnyTfp6h/Od1dUFlLukDbGfDJOc8I/4MaYh4CH4rS/ALwQp/0o0dVMSqkZ4HT3IAC9gRDVPt22VKh0h7RSKqda/QEA+oZCKa5U05kGB6VUTrX2RIPDwLAGh0KmwUEplVMtPTpymAk0OCilciYYjtDRNwTAwHA4z71R2dDgoJTKmY7eIexdTzpyKGwaHJRSOWMnowEGNDgUNA0OSqmcsZPRAP06rVTQNDgopXKmJTY46MihoBVsmVCl1PTT5g/gcTnwuBwaHAqcBgelVM609ASYO6uUYChC35BOKxUynVZSSuVMa88gcytL8XlcugmuwGlwUErlTKs/wLxZpXg9Ll3KWuA0OCilcsIYQ1vPEPWzSin3OHUTXIHT4KCUyonO/mGGwxHmVZbidbs0IV3gNDgopXLCXsY6d1Yp5R4X/ZpzKGgaHJRSOdE6EhzK8Lqd9OtqpYKmwUEplRP20Rnz7JGDTisVNA0OSqmcaO0J4HQINeUefB4XQ6EIoXAk391SGdLgoJTKiZaeAHUVHpwOwet2AujUUgHT4KCUyok2f3R3NEC5J3r4gialC5cGB6VUTrRYu6MBvHZw0LxDwdLgoJTKidae2JGDNa2kG+EKlgYHpVTWegNB+ofDoyMHt44cCp0GB6VU1lpjNsBBTM5Bg0PB0uCglMra6B6HMoDR1UqakC5YKYODiDwmIu0isi+mbbaIbBeRI9bPaqtdRORhEWkUkb0icnnMc7ZY1x8RkS0x7VeIyDvWcx4WEcn1h1RKTa6RozMqx44cNOdQqNIZOfwLsGlM2/3ADmPMcmCHdR/gw8By699dwCMQDSbAA8DVwFXAA3ZAsa65K+Z5Y99LKTXN2dNKdZUeAHw6rVTwUgYHY8xrQOeY5s3A49btx4HbYtp/bKJ+D1SJyDxgI7DdGNNpjOkCtgObrMcqjTG/M8YY4Mcxr6WUKhCt/gBzfG5KS6LTSWUl9iY4DQ6FKtOcQ70xpgXA+llntS8AmmKua7bakrU3x2lXShWQ1p4A9daUEoDDIfjcTl3KWsBynZCOly8wGbTHf3GRu0SkQUQaOjo6MuyiUirXWnqiFeBiefXwvYKWaXBos6aEsH62W+3NwKKY6xYCp1O0L4zTHpcx5ofGmHXGmHW1tbUZdl0plWtt/gD1Y4JDtKaDjhwKVabBYStgrzjaAjwX0/5pa9XSNUCPNe20DbhZRKqtRPTNwDbrsV4RucZapfTpmNdSShWAQDBMZ/8w8yrHjBzcTh05FDBXqgtE5Engg0CNiDQTXXX0deAZEbkTOAncbl3+AvARoBEYAP4MwBjTKSJfBXZZ1z1ojLGT3H9BdEVUGfBL659SqkC0+c/dAGfz6bRSQUsZHIwxn0zw0IY41xrg7gSv8xjwWJz2BuCiVP1QSk1PY3dH23xuJx19Q/noksoB3SGtlMpKbAW4WD6PiwHdBFewNDgopbISWzs6ls/tok+nlQqWBgelVFZaegKUe1wjR2bYNOdQ2DQ4KKWyElvHIVa5x8lAMEwkknDrkprGNDgopbLS6g+MHLgXy+txYQwMBqcu7xCJGI529E3Z+81kGhyUUllJNHLw5aGO9NMNTWz4x1c50tY7Ze85U2lwUEplLBSO0N47/ugMiC5lhak9tvsXe1swBn65r3XK3nOm0uCglMrYmb5hIoZzDt2zTfWx3d0Dw/zu6FkAtu3X4JAtDQ5KqYy19AwC4/c4QHQpK0xdcNhxsJ1wxPDRi+ex/7Sfps6BKXnfmUqDg1IqY/Yeh/gjh+i00sAUHb63bX8r82aV8qWNFwLw0oG2KXnfmUqDg1IqY4l2R8PotNJUbIQbGA7x2pEObl5dz9IaHyvnVrBN8w5Z0eCglMpYa08At9PBbJ973GN2cBiYgtVKrx3uIBCMsHHNXABuXjOXXSc6OaNnO2VMg4NSKmOt/gD1szxET9w/V7nbHjlM/rTStv1tVHlLuGrpbAA2rqnHGPiVTi1lTIODUipjLT0B5lWWxX3M65maOtLDoQg7DrZx06p6XM7on7TV8ypZNLuMF3XVUsY0OCilMpZoAxxAidOB2+WY9E1wvz96Fn8gNDKlBCAibFw9l982nqU3EJzU95+pNDgopTJijIkenZEgOEB0I9xkjxy27W/F63Zy/fKac9o3XjSX4XCEVw5pvflMaHBQSmWkayDIcCgS91wl22TXdIhEDNsPtPGBFbWUljjPeezyxdXUlHt0Q1yGNDgopTJib4BLPnKY3JoObzZ10947xKaL5o57zOkQPrS6np3vthOYwsP/ZgoNDkqpjCSqHR3L53FO6ia4bftbKXEKN66si/v4xjX19A+H+U3jmUnrw0ylwUEplZGWnsQb4Gw+z+SNHIwxbNvfyrUX1FBZWhL3mvddUEOFx6VTSxnQ4KCUykhrTwCHQG25J+E1Prdr0jbBHWrr5cTZATauqU94jdvlYP2qOrYfaCMUjkxKP2YqDQ5KqYy09gSoKfeM7C2Ix+txTtqR3dv2tSECH1qdODgAbFwzl66BILuOd01KP2YqDQ5KqYw0dw2ysDr+Bjhb+SROK724v5UrFldTV5F4WgvgAytqcbscOrU0QRoclFIZaeoaYGG1N+k1Ps/kTCs1dQ5wsMV/zsa3ZH24YXktL+1vxRitZ52urIKDiPyViOwXkX0i8qSIlIrIUhF5XUSOiMjTIuK2rvVY9xutx5fEvM6XrfZDIrIxu4+klJpsoXCElp4Ai2YnHzn43E6CYcNQKLdTS/YoIJ3gEL2untM9Ad451ZPTfsxkGQcHEVkA3AOsM8ZcBDiBO4BvAN82xiwHuoA7rafcCXQZY5YB37auQ0RWW89bA2wCvi8i5+5mUUpNKy09AcIRw6I0Rg5AzjfC/abxDBfU+lg8J/n72644rxqAxva+nPZjJst2WskFlImIC/ACLcB64Fnr8ceB26zbm637WI9vkOhRjpuBp4wxQ8aYY0AjcFWW/VJKTaKmrmiVtUWzUwQHd+5rOkQiht0nukZOYE1H+cjx4boZLl0ZBwdjzCngm8BJokGhB9gNdBtj7N+EZmCBdXsB0GQ9N2RdPye2Pc5zlFLTUHNndHd02iOHHP5RPtLehz8Q4orz0g8OZW67Kt3UlCydCbKZVqom+q1/KTAf8AEfjnOpnQEaf+B79LFE7fHe8y4RaRCRho4OPUxLqXxp6hrAITCvKvlKIfvY7lyOHBpOdAKwzpoqSofXrSOHicpmWukm4JgxpsMYEwR+CrwPqLKmmQAWAqet283AIgDr8VlAZ2x7nOecwxjzQ2PMOmPMutra2iy6rpTKRlPnAPNmlVGSZI8DxE7n5DA4HO+iptzDeWnmGyB6zlJpiUODwwRkExxOAteIiNfKHWwADgCvAJ+wrtkCPGfd3mrdx3r8ZRNdV7YVuMNazbQUWA68kUW/lFKTLJ09DgBed+qCP8YY7nt2L78/ejat92440cm686rjVp9L3pfJ2609E2WTc3idaGJ5D/CO9Vo/BO4D7hWRRqI5hUetpzwKzLHa7wXut15nP/AM0cDyInC3MUbDu1LTWFPXQMpkNIyOHJKVCu0ZDPJ0QxOP/fuxlK/X5g/Q1DnIuiXpTynZvG7npB4fPtO4Ul+SmDHmAeCBMc1HibPayBgTAG5P8DoPAQ9l0xel1NQIBMO0+YfSGjn40phWarVOd/31kTMEguFxdRliNVhHYKxbkn4y2uZ1Oye9Kt1MojuklVITcqo7vZVKkN5S1jb/EACDwXDKqaWGE52UljhYM78y3e6OiE4r6cghXRoclFIT0txlBYc0ppVKSxw4JPkmuDbr6G8RePnd9qSv13C8i0sXVaVMhMcz2bUlZhoNDkqpCWnqtDfApZ5WEpGU1eDsokHXL69lx8H2hOcf9Q+FONDiZ90E9jfEKitxTXo965lEg4NSakKaugYocQr1KU5DtaU6fK/VH6DaW8JHLprLqe5BDrfFP+Li7aZuwhGTUTI62g8ng1ouNG0aHJRSE9LcOciCqjIcjvSWkqaq6dDmH6K+snSk1OevDrbFvW7X8S5E4PIJbH47px/uyastMRNpcFBKTUi6y1ht5R5X0lVCbf4A9ZWl1FeWcvGCWQnzDg0nOrmwviJhSdBUvG4Xg7paKW0aHJRSExLdAJd+cIh+Y08VHKKlRjesqmPPyS46+4fPuSYcMbx5sjvjKSWIHh8+EAwTiWhNh3RocFBKpa1/KERn/3BayWhbtBpc/OmcUDjCmb4h5lZG8xcbVtZjDOw8dO7o4d1WP31DoYyT0QBlbhfGQCDHtSVmKg0OSqm0jRzVPYGRQ7KE9Jm+YSIG6qzgsGZ+JXUVHnaMmVrafSK6+e2KDPMN0X7YJ7NqcEiHBgelVNqarKO609kdbfO6Ey8htXdH2yMHh0NYv7KO1w51MByKjFy363gXcytLJ/S+8foBuS88NFNpcFBKpa05zSI/scqTrFay9zjUV44ui12/so7eoRANxztH2nYf7+SKJRM/bC+WfQjgQFCT0unQ4KCUSltT5yBlJU7m+NxpP8frdjEYDBOOkwgeCQ6zPCNt1y2rwe1yjEwtneoe5HRPgCuzmFKK9sM+IVZHDunQ4KCUSlt0GWvZhL7BJ6vp0OYP4HQIc3yjwcHncXHt+XNGlrTaI4hMDtuLNVrwR0cO6dDgoJRKW1PnwISWscJoNbh439hbe4aoq/DgHLOh7qZVdRw70897HX3sPtGF1+1k5dyKzDtOzLSSJqTTosFBKZUWYwynugZZNMGksD1yiLcRrr03MLJSKZa9W/rlg+3sOt7F5YurcWVw2F6sdI4PV6M0OCil0tIzGKR3KDShZDSMTufEW7HU2hNgbqVnXPvCai8r51bw3NunONTqz2oJ62g/NOcwERoclFJpGV3GOrHgYO8viHcya5s/MLKMdaz1K+vYd8pPxMCVWeYbYDQ4DOq0Ulo0OCil0jKyAW4Cu6NhtODP2P0Fg8Nh/IFQ3GkliB6lAeAQuHRx1US7O87ICEanldKSVZlQpVTxsOs4THzkEP+PctuYDXBjXbqomtk+N/NmlY7kLbLhdAgel0NHDmnS4KCUSktz1yCVpS5mlU3sVNSRhPSYkUNrnA1wsZwO4f//+CUjwSUXfClOiFWjNDgopdIy0aO6baNLWROMHGaNT0jbblpdP+H3S6asxKnHZ6RJcw5KqbQ0dQ5M6MA9my/BXL8dHBLlHCaD1pFOnwYHpVRKxhiauwYnnIyG6PRQaYkjzshhCK/bSUUOp41S8bp1WildGhyUUil19A0xFIpMOBlti1aDG59zqK8szeowvYnyunXkkC4NDilEIoazfUP57oZSeWXvcchk5ADxj+1uj6kAN1W8bpcGhzRlFRxEpEpEnhWRd0XkoIhcKyKzRWS7iByxflZb14qIPCwijSKyV0Quj3mdLdb1R0RkS7YfKpde2NfCdd94GX8gmO+uKJU3zRkU+Ynl84wPDvbIYSpFRw46rZSObEcO/wS8aIxZCawFDgL3AzuMMcuBHdZ9gA8Dy61/dwGPAIjIbOAB4GrgKuABO6BMByc7BwgEI5zp1dGDKl6Z7nGw+dzn1nQwxtDmH0q4x2GyaEI6fRkHBxGpBG4AHgUwxgwbY7qBzcDj1mWPA7dZtzcDPzZRvweqRGQesBHYbozpNMZ0AduBTZn2K9f8g9FvGf6AfttQxaupc5Cacjdl1hEUEzV2f0H3QJDhUGRKVyqBNa2UoCqdOlc2I4fzgQ7gf4rImyLyIxHxAfXGmBYA62eddf0CoCnm+c1WW6L2aaHXmk7q1WklVcSauyd+VHes8jHTSm29yXdHTxav28lAMIwx4wsPqXNlExxcwOXAI8aYy4B+RqeQ4om3JMEkaR//AiJ3iUiDiDR0dHRMtL8Z6bVGDL06clBFrKlzMKMNcDbvmGml1h57d/TUJ6SNgUAwkvriIpdNcGgGmo0xr1v3nyUaLNqs6SKsn+0x1y+Kef5C4HSS9nGMMT80xqwzxqyrra3NouvpsxPR/kEdOajiFI4YTndPvI5DrLHTSu3+aA5vqhPS9gmxutchtYyDgzGmFWgSkQutpg3AAWArYK842gI8Z93eCnzaWrV0DdBjTTttA24WkWorEX2z1TYt6MhBFbuWnkFCEZPVyMHncdI/FBqZzmkd2R09tSOHshI9tjtd2W5N/DzwhIi4gaPAnxENOM+IyJ3ASeB269oXgI8AjcCAdS3GmE4R+Sqwy7ruQWNMZ5b9yhl7xKA5B1WsmrvsOg7ZjRwiBoZCEUpLnLT5A8z2ufG4MktwZ9MP0JFDOrIKDsaYt4B1cR7aEOdaA9yd4HUeAx7Lpi+TxR4x6GolVazsZayZ7nGA0fOV+oZCI8GhrmJqRw3AyGorrQaXmu6QTsEeMegmOFWsmroGEYH5VdmNHGD0ZNY2/xBzZ01tvgFGg5ROK6WmwSGJUDgych6Mvd9BqWLT3DnAvMpS3K7M/1z4xnxjb/UHqK+Y+uAwUkdap5VS0uCQRGzNW805qGLV3DXIwiyS0XDuXH8oHOFM3xD1eRg52MFBj9BITYNDErErlHS1kipWx8/2szjr4DBa8Kejbwhjpn4DXLQfVj1rnVZKSYNDEj3WSiWf26k5B1WUegaCtPcOsbyuPKvX8cWUCm0b2eOQv4S0VoNLTYNDEvZoYUF1mY4cVFE63N4LwIr6iqxeJ7Ya3Oju6DxMK5XY00oaHFLR4JCEPVpYUFVGbyCo57GoonOkrQ+AZTkbOYRo781fcHA5HXhcDs05pGHq6vMVoNiRQ8RA/3CY8iksaahUvh1u68XrdrIgi2WsMJpzGBgO0z8UwuUQ5vjcuejihHndTl2tlAYdOSTROzJy8J5zX6licaS9l+V15Tgc2ZXydDsduBxC31CINv8QdRWerF8zU1oNLj0aHJKw9zbMryo9575SxeJIWx/L6rLLNwCICD5PtJZCmz+Ql2WsNp/HqQnpNGhwSKI3EMTrdlLldY/cV6pY2CuVVtRnl2+w+dxO+obC0eCQhw1wtjK3S6eV0qDBIQl/IEhlaQmVpdE8g65YUsUkVyuVbHYd6VZ/IC9HZ4z0w+3U4zPSoMEhid5AiIpSFxWlJYCer6SKS65WKtm8Hhdn+oboDYSm/Kjuc/rhdo4ci6MS0+CQhB0cKsuiIwc9mVUVk1ytVLKVe5wcO9MP5Gd3tM3rdjGo00opaXBIwh8IUllWQqU9ctBqcKqI5Gqlks3ndnG2fxjIzx6HkX54dOSQDg0OSURHDiV4XA5KnKI5B1VUcrVSyeaL2SOUz+BQVhJdNaWS0+CQhH8wSGWpCxGhsrREVyupopHrlUowuhEO8nOuUmw/BoJhPfEgBQ0OSdgjB4CKUpfmHFTRyPVKJRg9X8nndo78d5UPZW4nxkAgGMlbHwqBBocEAsEww+EIFdYy1soyHTmo4pHrlUowOq2UzyklGA1Ser5SchocErCXrVaWjY4cNOegCt17HX189OFfc6p7MOl1uV6pBKOFdvIdHEYL/mhSOhkNDgnYgcDeAFfhKdHVSqrgPbOrif2n/fxi7+mk1+V6pRIwcmhlPvMNEF3KCloqNBUNDgmMBofoyKGyTEcOqrAZY3h+bwsAvzrYnvTaXK9UgugmOCCv5ypF+6Ejh3RocEjAHiXYOYeK0hLdIa0K2ltN3ZzqHmRpjY/dJ7roHhiOe91krFSC6CY4IK/nKkFMzkEP30tKg0MC9ighdrXSwHCYUFhXOKjC9PzeFtxOB1+5dQ3hiOHVwx1xr5uMlUowOp2Tz3OVov2w6lnrtFJSWQcHEXGKyJsi8rx1f6mIvC4iR0TkaRFxW+0e636j9fiSmNf4stV+SEQ2ZtunXBhNSFurlawg0aebZ1QBikQML7zTwg0ranj/shpqyj0Jp5YmY6USwKWLqvjM+5fy/uU1OX3dibKDgx6+l1wuRg5fAA7G3P8G8G1jzHKgC7jTar8T6DLGLAO+bV2HiKwG7gDWAJuA74uIkzyzl63GjhxAazqowrTnZBctPQFuuWQ+DoewfmUtOw+1E4wzEp6MlUoApSVO/sstq0e+aOWLJqTTk1VwEJGFwEeBH1n3BVgPPGtd8jhwm3V7s3Uf6/EN1vWbgaeMMUPGmGNAI3BVNv3Khd5ACIdEN+zA6JJWzTuoQvT83hbcLgcbVtUBsH5lPb2BEA3Hu8ZdOxkrlaYTOyGtI4fksh05fAf4G8D++jEH6DbG2CG5GVhg3V4ANAFYj/dY14+0x3lO3vgHg1SUlhCNX6MjB12xpApN2JpSuvHC2pGR8PXLa3A7Hew42Dbu+slYqTSdeEusnIMmpJPKODiIyC1AuzFmd2xznEtNiseSPWfse94lIg0i0tDRET+Zliv2cd22Sq3poApUw/FO2nuH+Ogl80fafB4X114whx3vnpt3mKyVStOJy+nA7XLoDukUshk5XAfcKiLHgaeITid9B6gSEfuv6kLA3m3TDCwCsB6fBXTGtsd5zjmMMT80xqwzxqyrra3Nouup2VXgbPZtHTmoQvP83hZKSxxsWFl3TvuGVXUcO9PP0Y6+kbbJWqk03fjcTt3nkELGwcEY82VjzEJjzBKiCeWXjTGfAl4BPmFdtgV4zrq91bqP9fjLJnos4lbgDms101JgOfBGpv3KFf+YkcNoQlpHDqpwhCOGX+5rYf3KunOOzAZYbwWLHTGrliZrpdJ049U60ilNxj6H+4B7RaSRaE7hUav9UWCO1X4vcD+AMWY/8AxwAHgRuNsYk/eQHnsiK2jOQRWm14+d5UzfMLfETCnZFlZ7WTm3gl/F5B0ma6XSdON1O3UTXAqu1JekZozZCey0bh8lzmojY0wAuD3B8x8CHspFX3LFPxhk1bzRobXL6cDrdurJrKqgPL+3hbISJzdeWBf38ZtW1fPIq+/RMxBklrdkxq9Usnk9LgaCGhyS0R3SCfSOyTmAXdNBg4MqDKFwhBf3tbJhVR1l7vhbh9avqiMcMew8HJ1amukrlWzeEqdWg0tBg0MckYihd+jcnANgVYPTXyhVGH539Cyd/fGnlGyXLqyiptzNjoPtRbFSyebzaEI6FQ0OcfQPhzAGHTmogvaLvS343E4+eGHilX0Oh3DjhXXsPNTOgRY/MPNXKkE0Ia1LWZPT4BDH6KF7Y0YOZTpyUIUhGI7w4v5WPrS6ntKS5KfRbFhVhz8Q4qldJ4GZv1IJognpfh05JKXBIY6xVeBsFTqtpArEbxrP0D0QPGfjWyLXL6/F7XTw/N6WolipBNGRgx6fkZwGhzgSjRwqSl26z0EVhJ2HOigrcXLDitQnoPo8Lq65YA7hiCmKlUpgjxxCRLdaqXg0OMQx9kRWmyakVaHYc7KLSxdV4XGld8DxTdaBfMWwUgmih+8ZA0Mhrc+SiAaHOOxjuSvjjByGwxECuj5aTWMDwyH2n/ZzxXnVaT9n/co6HAJr5ldOYs+mD7saXL8uZ00oJ5vgZprEIwfrCI1AMGWST6l82dvcQzhiJhQcFlZ7+cU917O0xjeJPZs+7H0fA8Nh5uS5L9OVjhzi8CdZrQR6hIaa3nafiNZouGxx1YSet2peZdF86RmpI61J6YQ0OMThDwRxuxzj/kPRw/dUIdh9ootldeVUed357sq0ZRf80cP3EtPgEEdvIDQu3wB6bLea/iIRw56TXVyxOP0ppWJkF/zRw/cS0+AQh39w/LlKMJqD0OCgpqujZ/rpHghOKN9QjOzjy3WXdGIaHOIYWwXOVhGTkFZqOtpj5Rsu1+CQVGxCWsWnwSEOfyA4bqUSxCakNTio6Wn3iS6qvCVcUFscq44ypQnp1DQ4xNEbCFFZNn7k4HM7cYhOK6npa7eVbxCZ+bucs2EnpHVaKTENDnH0BoJUeMaPHESEitISXa2kpqXugWEa2/t0SikNdkK6XxPSCWlwiMM/GD/nANG8g44c1HT05sluAE1Gp8HldOB2ORgI6n/LiWhwGCMYjjAYDI87kdVWUVqiCWk1LTWc6MTpENYunNjmt2Ll0zrSSWlwGCPzyseLAAAUTUlEQVTRiay2ylLXyA5qpaaT3Se6WDO/MmFJUHUur9ulm+CS0OAwhr0SKd4+B9CaDmp6CoYjvN3Uw+W6+S1tXrdTazokocFhjJQjhzKt6aCmn3dbehkMhjXfMAFaDS45DQ5j2H/44+1zALumgwYHNb3sPtEJwLolGhzSFa0Gp7MAiWhwGMPOJ8Tb5wDREUXfUIhIRCtIqelj98lu5s8qZd6smV/iM1d8HqcuZU1Cg8MY/hQ5h8rSEiJGT3NU08ueE126v2GCytwu3QSXRMbBQUQWicgrInJQRPaLyBes9tkisl1Ejlg/q612EZGHRaRRRPaKyOUxr7XFuv6IiGzJ/mNlLlXOwW7XpLSaLlp6BjnVPaj5hgnyuZ16fEYS2YwcQsBfG2NWAdcAd4vIauB+YIcxZjmww7oP8GFgufXvLuARiAYT4AHgauAq4AE7oOSDnU8o9yQKDtERhe51UNPFnhO6+S0TZRockso4OBhjWowxe6zbvcBBYAGwGXjcuuxx4Dbr9mbgxybq90CViMwDNgLbjTGdxpguYDuwKdN+Zcs/GMLnduJyxv+fxs5F6MhBTRcNJzopLXGwal5x1H/OFZ81rWSM5g/jyUnOQUSWAJcBrwP1xpgWiAYQoM66bAHQFPO0ZqstUXu897lLRBpEpKGjoyMXXR+nNxBMuDsaYms66MhBTQ97TnSxdmEVJQm+0Kj4vB4nEQNDoUi+uzItZf3bJCLlwL8CXzTG+JNdGqfNJGkf32jMD40x64wx62prayfe2TQkquVgqxwpFaojB5V/g8Nh9p/265RSBkYP39P/luPJKjiISAnRwPCEMeanVnObNV2E9bPdam8GFsU8fSFwOkl7XiSq5WDTkYOaTvY2dxOKGN3fkAGvR2s6JJPNaiUBHgUOGmP+MeahrYC94mgL8FxM+6etVUvXAD3WtNM24GYRqbYS0TdbbXmRqH60bbQanH7bUPm3+2S08ttlizQ4TJQW/Eku8V/B1K4D/gR4R0Testr+Fvg68IyI3AmcBG63HnsB+AjQCAwAfwZgjOkUka8Cu6zrHjTGdGbRr6z4A0GW1iSuolVa4sTtcuhqJTUt7DzUwQW1Pqp97nx3peB4rQMKdc9SfBkHB2PMvxM/XwCwIc71Brg7wWs9BjyWaV9yKVEVuFiVWtNBTQO7T3TyxrFO/vNHVuW7KwXJDg56+F58urwhhjEmWgUuSc4Boruk9fA9lW8P72hkts/Np65ZnO+uFCSvNa2kCen4NDjECAQjBMMm6Wol0GpwKv/eburm1cMdfOb6pSN/5NTE2HWkB4M6cohHf6tipKrlYKvQk1lVDg2HIrxzqpvfH+1k36ke7r5xGRctmJX0Of/88hFmlZXw6WuXTE0nZyDfyMhBg0M8Ghxi2EnmVCOHyjIXrf7AVHRJzUDGGPac7Oa3jWd4/VgnDSc6CQSjG7E8Lgf7Tvfw/OevZ1aCzZj7T/fwq4Pt/NVNKxIe86JSsyvm6eF78elvVoyR47pTjRw8OnJQmfvBq0f5xovvArBybgV3XLmYa86fzVVL53DsTD//8b//jvue3csjf3w50RXj5/ruy41UeFz86XVLprjnM4t3JDjoyCEeDQ4xelPUcrBFq8Hptw01cbtPdPHNlw7x4Yvm8vcfu3jcEtTZPjf3bVrJQy8c5PHfHudPr1t6zuOHWnv55b5WPnfjsoQjC5WeEqcDt9OhS1kT0IR0jFRV4GwVpSUMBsMEw3omi0pfz2CQe558k3mzSvnGJy5JuDfhM9cv5aZVdTz0wkH2Nnef89h3X2nE63Zy5/uXxn2umhivx8mA5hzi0uAQozfdaSUrJ9GnK5ZUmowx/O1P36HVH+DhT16W9HdMRPjm7Wupqyjl7p/socf60vJeRx/P7z3Nn1x7nm56y5HoyawaHOLR4BAj7YS01nRQE/TUriZ+8U4Lf33zCi5fnPqoiyqvm4c/eRkt3QHu/9e9GGP43iuNeFwOPnv9+VPQ4+IQremgX/Li0eAQozcQxOmQkURVIloNTk3E4bZevvLz/bx/WQ3/zw0XpP28K86r5r5NK/nlvlb+/oWDPPfWaf7oqvOoKfdMYm+Li1aDS0wT0jHs47rjrRCJpdXglM0Yw+vHOukZDHLtBXPGTRcFgmE+95M9lHtc/ON/XIvDkfx3a6zPXL+U14+d5X/8+hhul4P/9AEdNeSSV+tIJ6TBIYZ/MJhySglGVzPpiqXi1nC8k3/YdojXj0XPiXQ6hMsXV/GBFbV8YEUda+ZX8tXnD3C4rY/H//wq6ipKJ/wedv7h9h/8jo1r5lJfOfHXUIl53U5aevRLXjwaHGL0BkJUeFIvD6zUmg5Fbd+pHr710iFeOdRBTbmHr9y6hpVzK3jtSAevHT7DN186zDdfOky1t4SugSB33XA+H1iReXGqKq+bbV+8YcKjDpWa1+PS4zMS0OAQI50TWSE2Ia0jh2JytKOPb20/zC/2tjCrrIT7Nq1ky/vOGznb6Orz5/D/boQzfUP8+kgHrx7qIGLgSzdfmPV7a2CYHN4Spx68l4AGhxj+QJBFs70prysfSUjryKFYHGnrZfP3fgPA59cv4zPXn59wE1pNuYePXbaQj122cCq7qDLg9Tj1yO4ENDjEiFaBSz2t5HQIPrdTVysVif6hEH/xxB68bifPfe79LKgqy3eXVI743C76h0MYY1IuRCk2upQ1RroJaYDKMq3pUAyMMXz5p+9wtKOPh++4TAPDDFPmdhIxMBTS0w7G0uBgiUQMfcPJ60fHmoqaDn1DIZ576xQBTZjlzf9+/SRb3z7NvR9awfuW1eS7OyrHfHr4XkIaHCx9wyGMiY4I0lFZWkLv0OSNHCIRw+d/socvPPUWG7/zGq8d7pi09yoUQ6EwW98+TUvPYFrXG2N48o2TfOXn++nLIOn4dlM3X/35AT54YS1/+cFlE36+mv68Hq0Gl4jmHCyjh+6lP3I40zc8af35zq8O88qhDv7suiW8eqiDTz/2BrdcMo//dstq6opsrftQKMwzDc18/5VGWnoCzCor4Zu3r+VDq+sTPicQDPO3P3uHn+45BcCrhzr47h9dzur5lWm9Z/fAMH/5xB5qKzx8+w8v1dVCM5Qe252Yjhws9hRRqhNZbZNZDW7b/lYefrmRP1y3kP92y2p++cXrufdDK3jpQBsbvvUqj//2OOGImZT3nk6GQxGeeP0EN/7DTv7rv+1j3qxS/umOS1k0u4zP/riBrz5/gOE4c8VNnQN8/JHf8tM9p/jiTct58rPX0D8c4rbv/4b//fsTGJP8f7tIxHDvM2/T3hvge5+6XA+5m8HsanC6S3o8HTlY7JFDOquVwKrpMAk5h8b2Xv76mbdZu3AWD26+CBHB43Jyz4bl3Lp2Pv/1uX08sHU/z+5u5h9uv4SVc9P7JlxIAsEwP3vzFN99uZFT3YNcuqiKr338Em5YXoOIsOmiuXzthXd59N+Pset4J9/95OUsnhNdgvza4Q7ueepNwhHDo1vWsWFVdHTxwj3X81fPvM1/+bd9/O7oWb7+Hy5O+EXgkVff4+V323lw8xouXVQ1ZZ9bTb0yHTkkpMHBMjpySHdaKTpyyOUSOH8gyF3/azelJQ4e+eMrKC059wDAJTU+fvznV/H83ha+8vMD3PrPv+FvNl3In1+3tOCnPYwxvHOqh2camnjurdP0BkKsXVTFQx+7iA+sqD3nf2OPy8nf3bqGa86fzd88u5ePPvxrvv7xSzh+tp9vvnSIFXUV/Pc/uYIlNb6R58wp9/Avf3olP3jtPb710mH2nerhax+7mGDE0NjeR2N7H++19/FeRx9n+4f5g7Xz+ZNrzsvH/xRqCo2OHDQ4jKXBwWInlyeSkA6GDUOhyLg/4pmIRAz3Pv02J88O8MRnrmZ+giWTIsIfrJ3P+y6Yw33/+g7/3y8OsvNQB9/6w7VJz90JRwzGGFzOqZ1JbO0J8INX36OxvY/Fc7wsmeNl8WwfS2q8nDfbRyAY5t/eOsXTu5p4t7UXj8vBRy6ex+3rFnLt+XOSBt5NF81jzfxZfO7JN7n7J3sA+IO18/nGxy8e2bUcy+EQ/vKDy7hyyWzuefJN/uhHr488NqushGV15dy0qp6V86KlO3Xd+8zn9Wgd6USmTXAQkU3APwFO4EfGmK9P5fvbh+hNJCEdfV4wJ8Hhn19u5FcH2/i7P1jN1efPSXn9nHIP/+PTV/DUriYe/PkBNn7nNb72sYv58MXzRq7pGQiy83A7r7zbzs7DHQwMhVleX86qeZWsnlfJ6vmVrJpXmXa5ydPdg+w63sn+035W1FewYWVdwvn4Nn+A77/SyJNvNBExhlXzKvnlOy10DZybp3E6hHDEcMnCWXz1tou4de38CZW/XDTby//5T9fyg1ffo9rn5o+vTv1H/cols3nhnut59XAH82aVckFdOXN8bg0GRchOSPdrNbhxpkVwEBEn8D3gQ0AzsEtEthpjDkxVH3rTLPRjGwkOgRB1E5z2j0QMLf4Ax8/0c+xMP4fbevnx707wHy5fwJb3LUn7dUSET161mKuXzuavnn6Lv3hiD7dfsZBldeXseLed3Se6CEcMc3xuNqysp6bczYEWPzsPtfPs7uaR15lbWcrC6jIWVJdFf1Z5WVBdRmWpi3dO9bDreBe7j3dyuicAjP5BdzqEK5dUc/PquXxodT2LZntp9wf4/s73+MkbJ4lEDJ+4YiF337hs5FiSnoEgJzr7OX52gJNn+xkMhvnoxfPTXkUUj9vl4J4Nyyf0nGqfm9suW5Dxe6qZwasJ6YSmRXAArgIajTFHAUTkKWAzMGXBwR8I4XE58LjSGwXY00+JajoYY+joHeKYFQCOne0fCQYnzg6csyPT43KwYWUdf/+xizP69np+bTnP/sX7+KdfHeH7OxuJGFg9r5K//OAF3LiyjrULq3COyUm09wY42NLLgdN+Gtv7ONU9wO4TXfxibwuhMSuh5laWsm5JNXedV826JbO5cG4FB1v8vLS/jZcOtPLg8wd48PkDXFhfwfGz/YQiho9fvoDP3bh8JFFsm+Ut4RJvFZcs1ESvyj9dyprYdAkOC4CmmPvNwNWT8UafeXwXJ84OjGtv7x1KexkrMLKT+vM/eXNc5biIMbT2BOiP+YUrcQqLZ3tZWuPjAytqWVLjY+kcH0tqfMytLM06oVzidPCljRfyh+sW4XJKwpyFra6ilLqK0nFHSYcjhjZ/gFPdg3T1D7N6fiULqsrGBa1LFkb/wH9p44UcP9PP9gNtvPxuO2sXzeLuG5dx3hwfSk13JU4HbqeD//mbY/z87dP57k7anr/n/Wl/kc3UdAkO8f4yjluMLiJ3AXcBLF68OKM3Wjzbh9s1Pim7vL6cK5fMTvt11syfxR1XLko4crh+eS1La6J//M+v8TG/qmzct/fJMPab+kQ5HdHAkiq4xFpS4+OzN5zPZ2/QKmWq8HzhpuXsP92T725MiMT9k5nj90i1IWgqiMi1wN8ZYzZa978MYIz5WqLnrFu3zjQ0NExRD5VSamYQkd3GmHWprpsuO6R3ActFZKmIuIE7gK157pNSShWtaTGtZIwJicjngG1El7I+ZozZn+duKaVU0ZoWwQHAGPMC8EK++6GUUmr6TCsppZSaRjQ4KKWUGkeDg1JKqXE0OCillBpHg4NSSqlxpsUmuEyISAdwIsOn1wBnctidQqGfu7jo5y4u6X7u84wxtakuKtjgkA0RaUhnh+BMo5+7uOjnLi65/tw6raSUUmocDQ5KKaXGKdbg8MN8dyBP9HMXF/3cxSWnn7socw5KKaWSK9aRg1JKqSSKKjiIyCYROSQijSJyf777M5lE5DERaReRfTFts0Vku4gcsX5W57OPk0FEFonIKyJyUET2i8gXrPYZ/dlFpFRE3hCRt63P/RWrfamIvG597qetI/FnHBFxisibIvK8dX/Gf24ROS4i74jIWyLSYLXl7Pe8aIKDiDiB7wEfBlYDnxSR1fnt1aT6F2DTmLb7gR3GmOXADuv+TBMC/toYswq4Brjb+v95pn/2IWC9MWYtcCmwSUSuAb4BfNv63F3AnXns42T6AnAw5n6xfO4bjTGXxixhzdnvedEEB+AqoNEYc9QYMww8BWzOc58mjTHmNaBzTPNm4HHr9uPAbVPaqSlgjGkxxuyxbvcS/YOxgBn+2U1Un3W3xPpngPXAs1b7jPvcACKyEPgo8CPrvlAEnzuBnP2eF1NwWAA0xdxvttqKSb0xpgWif0SBujz3Z1KJyBLgMuB1iuCzW1MrbwHtwHbgPaDbGBOyLpmpv/PfAf4GiFj351Acn9sAL4nIbhG5y2rL2e/5tCn2MwXiVeTWpVozlIiUA/8KfNEY449+mZzZjDFh4FIRqQJ+BqyKd9nU9mpyicgtQLsxZreIfNBujnPpjPrcluuMMadFpA7YLiLv5vLFi2nk0Awsirm/EDidp77kS5uIzAOwfrbnuT+TQkRKiAaGJ4wxP7Wai+KzAxhjuoGdRHMuVSJifwmcib/z1wG3ishxolPF64mOJGb658YYc9r62U70y8BV5PD3vJiCwy5gubWKwQ3cAWzNc5+m2lZgi3V7C/BcHvsyKaz55keBg8aYf4x5aEZ/dhGptUYMiEgZcBPRfMsrwCesy2bc5zbGfNkYs9AYs4Tof9MvG2M+xQz/3CLiE5EK+zZwM7CPHP6eF9UmOBH5CNFvFU7gMWPMQ3nu0qQRkSeBDxI9qbENeAD4N+AZYDFwErjdGDM2aV3QROT9wK+Bdxidg/5bonmHGfvZReQSoglIJ9Evfc8YYx4UkfOJfqOeDbwJ/LExZih/PZ081rTSl4wxt8z0z219vp9Zd13AT4wxD4nIHHL0e15UwUEppVR6imlaSSmlVJo0OCillBpHg4NSSqlxNDgopZQaR4ODUkqpcTQ4KKWUGkeDg1JKqXE0OCillBrn/wLwA+rhI/jXggAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Plot the histogram\n",
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "plt.plot(range(len(h)), h)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX0AAAD8CAYAAACb4nSYAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvnQurowAAIABJREFUeJzsvXmUJHd15/u9uW+VtXd39d5qtZAaSWhpZAkQCNuA8GBhGw/bGKPxYGwzGt4ZH9sjzfPDM7I9x88+Zvw81vE8YLBlc3js2MIIJGGzCBCMWkIgdbd6Vau7equ9KvfMyPy9PyJ+Eb/YMiOysqqys+/nnD5dmZUZFRkZceP+vncjIQQYhmGYK4PIRu8AwzAMs36w0WcYhrmCYKPPMAxzBcFGn2EY5gqCjT7DMMwVBBt9hmGYKwg2+gzDMFcQbPQZhmGuINjoMwzDXEHENnoHnExMTIjdu3dv9G4wDMNcVjzzzDNzQojJTq/rO6O/e/duHDx4cKN3g2EY5rKCiF4O8jqWdxiGYa4g2OgzDMNcQbDRZxiGuYJgo88wDHMFwUafYRjmCoKNPsMwzBUEG32GYZgrCDb6TEceP3QRl1aqG70bDMP0ADb6TFu0Zgu/+aln8OkfntnoXWEYpgew0WfaUm400RJAua5t9K4wDNMD2OgzbSnXmgCAaqO1wXvCMEwvYKPPtEV6+NVGc4P3hGGYXsBGn2lLuW54+hp7+gwzCAQy+kR0NxEdJaITRHS/z2veSUSHiegQEX3aeO4mInrKeO4nRPSuXu48s/aYRp89fYYZCDq2ViaiKICHALwJwDSAp4noESHEYeU1+wA8AOC1QohFItpk/KoM4FeFEMeJaCuAZ4joMSHEUs8/CbMmsLzDMINFEE//NgAnhBCnhBB1AJ8B8HbHa34dwENCiEUAEELMGP8fE0IcN34+D2AGQMcm/0z/ID39GgdyGWYgCGL0twE4qzyeNp5TuQbANUT0PSL6ARHd7dwIEd0GIAHgZLc7y6w/0uhX2NNnmIEgyOQs8nhOeGxnH4C7AGwH8CQRXS9lHCKaAvD3AN4vhHC5jET0QQAfBICdO3cG3nlm7amwvMMwA0UQT38awA7l8XYA5z1e849CiIYQ4iUAR6HfBEBEeQBfBfD7QogfeP0BIcTHhBAHhBAHJidZ/eknSmb2Dht9hhkEghj9pwHsI6I9RJQA8G4Ajzhe8w8A3ggARDQBXe45Zbz+ywD+Tgjx+d7tNrNeWNk7rOkzzCDQ0egLITQA9wF4DMARAJ8TQhwiogeJ6B7jZY8BmCeiwwC+CeB3hRDzAN4J4PUA7iWi54x/N63JJ2HWBJZ3GGawCKLpQwjxKIBHHc99RPlZAPht45/6mk8B+NTqd5PZKEqcvcMwAwVX5DJtqRhGv95sodlyxu8ZhrncYKPPtKVUs7prssTDMJc/bPSZtqj5+Wz0Gebyh40+0xaZvQNw0zWGGQTY6DNtYXmHYQYLNvpMWyqNJtLxKAA2+gwzCLDRZ9pSrjcxlk0A4AIthhkE2OgzbSnXNIxm4wCAGnv6DHPZw0af8UUIgXKjibFsEgB32mSYQYCNPuNLTWtBCGAso3v6LO8wzOUPG33GF5m5Iz19DuQyzOUPG33GF5mjP54zArncXplhLnvY6DO+SKM/muHsHYbxotFs4Y+/ehjzxdpG70pg2Ogzvsih6FbKJnv63XB2oYxGk2+Yg8jxS0V8/MmX8K2jsxu9K4Fho8/4IjtsjmTiIOKUzW4oVBv4mY9+G196dnqjd4VZA6TkWag2NnhPgsNGn/FF9tLPJmJIxiKcstkFs4Ua6loL04uVjd4VZg2QcyZWqlqHV/YPbPQZX6S8k05EkYpHWdPvgsVyHQCwVL58PEEmONLTX6lcPt8vG33GFynvZJNRpONR1vS7YKGkGwNp/JnBQkqeKyzvMIOAlHcy8Zju6XNr5dAslnRjv3wZeYJMcGrGNbFSGTB5h4juJqKjRHSCiO73ec07iegwER0iok8rz7+fiI4b/97fqx1n1p6KIu8kYxH29LtgwfDw2dPvDZ948hTOLfVPfKR6GXr6HQejE1EUwEMA3gRgGsDTRPSIEOKw8pp9AB4A8FohxCIRbTKeHwPwBwAOABAAnjHeu9j7j8KoNFsC86UaNg2lut5Gud5EPEpIxCKGps9GPyzS2C+WLh+j0K8slur4o68eQbXRxH0/va9n261rLRRrmpmaHAYZ5yoMWCD3NgAnhBCnhBB1AJ8B8HbHa34dwEPSmAshZozn3wLgCSHEgvG7JwDc3ZtdZ9rxmafP4PV/+k3MraJopFy3eumn4hEzU4EJDss7vaNotAWZK/Z21fTxJ0/hZz/6bdS6qDi/HD39IEZ/G4CzyuNp4zmVawBcQ0TfI6IfENHdId7LrAHPnVlCtdHCPx+51PU2ynUNmYS+GEzFo5yy2QUykFusaahzTGRVlAy5cbbH1a9n5stYKNXx7MtLod9rafqDZfTJ4znheBwDsA/AXQDeA+ATRDQS8L0gog8S0UEiOjg7e/lUtvUzx2aKAIAnDndv9Ev1JjJJ3dPvh+yd7x6fQ7PlOn36GlXLZ29/dcgGgLOF3hr9pYr+HT15PLztsTx9DUJcHudmEKM/DWCH8ng7gPMer/lHIURDCPESgKPQbwJB3gshxMeEEAeEEAcmJyfD7D/jgRACJy4VECHgO8fnbHNuw1CpN5FJSHknuqEN145fKuBX/tcPL7vK1sVSHbGI7vsscTB3VRRr+vm3GsnSC1lD8eTxudDvlZp+syXMXlX9ThCj/zSAfUS0h4gSAN4N4BHHa/4BwBsBgIgmoMs9pwA8BuDNRDRKRKMA3mw8x6wh55YqKNWb+Fc3bkVda3XlwQCGvBOX8k5kQ4uzpI77nS4uzI1koVzHzrEMAGCxzwu0tGYLb/+r7+KxQxc3elc8KUtNv8eevlyBvXB+GQulcDdmNQ6g6vqtlsCb//u3cc3/+TX93+9/DZ/532d6s8OrpKPRF0JoAO6DbqyPAPicEOIQET1IRPcYL3sMwDwRHQbwTQC/K4SYF0IsAPhD6DeOpwE8aDzHrCHHL+nSzntv24nhdByPdynxlBV5JxnbWHlH9jb57vFZtC4TiUdrtrBcaeCqySyAtfP0LyxX8IVnVr8CurBcxY+nl/HDU/15icpA7kpV6yro6sdSuYF9m3IQAvjeiXBOheoIqbn6i+U6jl0q4va94/h3d+5BlAgvnF/u2T6vhkB5+kKIR4UQ1wgh9goh/th47iNCiEeMn4UQ4reFEPuFEDcIIT6jvPeTQoirjX9/szYfg1E5dqkAANg/lcfPXLsJ//LiDLQuujyWHfLORmbvyAt+sdzAofMrG7Yf7XDKaMuVBoQA9kxIo782nv7nD07jdz7/Y7OCulvOLpYBABdX+icPXkU9vvM9zOBZqtRx575J5FOx0Kviqo+nL1cMv3zrdvynu6/FaCbeN21MuCJ3ADl2qYhNQ0kMZ+J48ys3Y6ncwNOnw5dG6Jq+Je/Um60NC6SqedBPnui/YP/J2SJu/K+P47mzVgaIDOLumcgBsAKGvUb+ndVmV00v6Mb+wnJ11fu0FpSUm1qvgrnVRhPVRgvjuQRee/UEvnt8LlRAtmbz9C2jL+XICSP3v5/qXNjoDyDHZwq4ZvMQAOD110wiGYvg8cPhddpSXbN5+sDG9dSX8s7eySyePNZ/uv6RCytotgQOKUt4ma65fTSNWITWTNNfNra72u/G9PT71egrnn6vgrlSzx9Ox3HnvkmcX67i5Gwp8PtrWhOjxgxpL09/zJg6l+yjhoVs9AeMVkvg+KUirt6ke5eZRAyvu3oCjx+6FDqlrFxvIp2wUjaBDTT6NQ2JWAQ/c91mHHx5wewA2i+cNbzkc0oLZemBj2UTGMkk1kzekYZrtZ7+2QXd6M8Uan2ZGrsWRl9+JyOZOO7cNwHAO3Wz1RJ44Zxbk682mmbVu6rpL5T0/RszPf1IT+MQq4GN/oBxbqmCSqNpevoAcNcrJnFuqRKqZ4nWbKGutZBV5B0AG9Z0rVDVkE/FcOe+CTSaAj98qb+CjdJLVo+xrMbVjX58zQK5S5XeePqy53+zJXqeFtkLirUmxg0j2quqXPmdjKQT2DGWwe7xDL7rkSH25Ik5vO1/fBcnZ4u256uNFiaG9H1SB6nI/RszRo2mNjgRQoWN/oAhg7jXbM6Zz40aF0qYPOKycYKuVt75+6dO4wMPH1z1uMBCVUMuGcOrd48hGYusicTzpWen8duffa6r90ovWfX0ZbO10UwCo5n4mjVdW+6R0T+7WDY9037U9ct1vT/OUDLWM01f3jBHDInmdfsm8NSpeVfig/x7zpTOmtbEUDKOVDxiG6SyUKpjJBNHLKqbWL+U5z/52ov462+d7MlnCQob/QHjmJGuuU/x9FOx8AZbZoJIeSfZxTYA4Psn5/GNI5fw0SeOhXqfk0K1gaFUHKl4FLftGeu69qAd3z4223WOuvSSzzs8/XQ8inQiuqbyjtxupd79jbXaaOLSSg0Hdo0CAC4u918GT7GmIZuMYWIo2bNWDKqmDwBXTeRQrjfNbDGJlJacjlO10UIqHkE+FbcFcudLNXNVAvgHcr91dAbfPDrjen4tYaM/YByfKWBzPmmexIBluMOk9MmT3CXvhAxGyeDW//z2Sc9lc1CKVQ1DKX1f7tw3geMzxZ4HHBdKdZTqzdB6dqslcG6xgggBF1eq5qpmodQwPeeRdHxNjL4QwjQ2q/H0pSx1254xAGvr6XdbZ1Gqacgmo5jMJXtWoLVctnv6OeMccxp9+bjiiCVVG02k4lHk03FbIHe+WMd4Nmk+9qtorzSa5j6sF2z0B4zjl4o2PR/oTo8vOzx9Ke+EHY6+UtFw+1Vj2DuZw3/83HOY79JDKyhG/3VX6606vn+ytxKPXLqXQgaJZwo11JstXL9tGC1hZb8slusYzerGZDSbWJOUzWqjhbpxk1lNIFeuVG7cPoJENLJmGTxff+Eibv2jJ1xGNQjlehPZRAwTQ4neBXIrdUQjhFxSP7fk/0E9/ZrW0o1+KmYL5M6X6hjPqZ6+t7xTqTfXvScTG/0BotUSODFjZe5IupFmKn6afsgMhJVqA5vzKfzlu2/GcqWB+7/0fKj3SwrVBnJJ3YDu25xDNEI4FSK1LgjS6BdD9kaXQdzbrxoHYHnNC6U6Ro1A3nBaL87x+w4+f/BsVysh9UayGqMvYxI7xzLYMpzCxZW1MfonZ4tYLDdw5EL4ArtiTY/rTOSSPQzkNjCSjoNI74+UNYy+s9DOX95pIhmLuDz9hVLd1p/fr6K90mCjz6yC6UV35g7QXRBWnuSyOEumbIbVjfWsmzj2b83j1167B984cqmrdMtCzfL049EIto6kcMYwVL1ACGEZ/ZBeqDSYt1+lSyMymLtYtoy+/N9P4vnLfzmOh586HXa3bQYj7CpM5exiGYloBJuGktgynFozeUce2xe7MPqlmoZMMoqJXBLLlUZPUiCXKg0MZywpNGe0HZHN3STysSqRCiFQ01pIxqM2Tb/ZElgs1zGes8s7XhXttUYLlUZzXdM52egPEF6ZO4Al0XQTyLU8fanpB9+G1Jvzad1Y37JzBEIARy8WAm8D0FcwxZqesinZNZbFyz00+uV60+yNHnYKkszRP7DbMPqKp29q+oZh8cvgKVQ1W8pfUNSbyKrknYUKto2mEYkQpoZTaybvSGfi8IVw5wCgV+RmDU8f6E0rhmXD05fI1aRztefl6cvzJRWPIJ+Omdk7i+U6hIAjkOuuaNealjS3nt4+G/0B4tiMfiFdvcnh6cfCB2HlyZ1VhqgA4eSdSqMJrSWQT+kX0nVTeQDAkZAXfKmuQQgryAYAO8czODPfO3lHTcUL7ekvlrE5n0Q+FcfkUBLnFitoNFsoVDXTw29n9IUQKFa1rkbuqcZiNRWfZxfL2D6aBgBd3lmurkl/eNPTvxjO028YdSO5RAyTQ7rR74Wuv1SpYyRjGees4em75B1jdareWKXnnoxZnr4QwrwZ2TV9IyamXD9qjG09g7ls9PuYF84t46W54Ibt+ellbB9N2zJ3gO7knbIyFB1Q0z6DGxYZ2Mob+7N9NI1cMhb6gpeGYihlfa6dYxkslhs9G1NnM/ohje/0YhnbR/X2ydtG0ji3VFGqcY1ArmFYvC7umtaC1hJdfRbV6K82kLvDaAG9JZ9CvdkK3WY4CNKYHr1YCJXFY8qNyRgmcrJAqwdG3+XpB8/ekQ6Q7unHobUEKo0m5h3VuIC346VKRezpMwCA3/7cc/jTr78Y+PU/OrOEW3aOup6XRj+MUSg75J1kF/KONGLS0yciXLtlKHQQT3rAQzZ5RzdQZ+Z7I/HYPf1wF+DZhQp2GF7ytlHd6EvZZdQl77i3LY9TV56+sb1ohLruslmqaVgo1U1Pf2pYbyuwFrp+ydDGy/VmqJiMbLaWMzR9AJgr9EbeUTX9MIFceS2kDE8f0B0d6elPODR99T3On9eqhsMLNvp9zPmlauCT4cJyBRdXqrh554jrd9EIIRENNwTFTNmMy+KsCIjCBQtlYEs11tdN5fHihUIo6UBq3dILA3R5BwBe7pHRn1eMfhjj22i2cGHZ8pK3G56+9ELHnIFcj7RNubIodDFyb7nSQDRCGMsmug4GyuyjHcZqZcuwbvzXQtcv1jTzBhjm5m/WjSQteWe1BVqNZguFmoaRtOWRx6MRJGMRj5RN42bVcGv6SUPTB/RzdUFpvyHpZPTZ02dQrGnmvyA8d0Zv6Xuzh6cP6CdmWHknHY8iYoz6IyK9f0iIXH/T01eWz9dODaFQ00L1AbI8fWs7u8b1HvUvL/RG11/sUtO/sFRFS1gGc+tIGnWthRPGjOJRpbVuMhbxvInLv9fNyL2lSh35VAzpeLRrT18GouWNS3r6a5G2WappuGnHCCIEHAkR0JfHKJuIIRWPIteDVgwrjhYMklwy1kbe8fb05bm5Um1gvlQHkXWjB3SnSX+PIu+w0WdUpJcV1AD96OwSEtEIrpsa8vx9Kh4N5QmqA1SsbURCGRZT03d4+kC4YK40+up2cskYxrMJM11ytcyX6ohHCel4NNRMYeklbx8z5J0R/f/np/WOjKq3N5pJ2G4uEnVlEVbXX65oGMkkjMH19huyEAIzhc6Ge9r09PV9n8glEY3Qmnj6pZqG8WwSeyayoTz9suFpS/llcii5ak3f2XdHkk3GbOeAEEKRdxRNvyGzd6LmuanLOzWMZhKIGg6TfA1gD+Sq19ISG31mZiWk0T+ziFduy5uFWE7CeoLqqERJ2EEQXp7+K4wagjB52tIoqtk7gC7x9EreWSjV9GZeKbeX145phzSyzTCczxtteFWDMpKJe17cqtEPq+svVxrIp+NIJaKumM1TJ+dx+3/7547G9exCBel41LxBRSOEzUPJNdH09QKrKK6dyocK6JuevnFOTuRWX5UrV13OxAenpy8D7YCPp28EcgH9nHcWZgFqTMzb019ho8/IpXWQTJJGs4WfTC/j5h3e0g4QfrC5OhTd2kZIecdD088mY9g1nsGRUBe83I794tw11kuj38BoJoFcKhbK8J5dqCBq5LYDltE/PlNELhmz3YT92iurBibsxb9crmMkHUcqFnEZ/fPLuvT0tecvtP8Mi2XsGEubVamAkbbZ47GJQggz137/VB5nFyqBaxOcvaCCVuX+/VOn8ekfeg8kXzbiK2rKJuA2+qrX75Wnn7QFchtG3x37Nr1SntXrca3abnvBRr8HNJqt0AOVO3FpRfdiKo3ODcCOXiygprU8g7gSv4ZPfqgDVCTJWLi4wEpVQyoeca0+rtuiB3ODUqhqIAKyjv3ZOZbBheUK6j3o8b9QqmE8p7ftDePpn10sY2o4ZbbQzafiGErF0GwJs++OZNSn02ZRMXzdePrD6TjSiagryC6liMcPX2r/GRbK5kpFMjWc7rmnX9P04qRsMoZrt+grvqCFevKzSHlHN/qWp//CuWVPKetzB6fxxWe9h8abA1TSTnknagZuASuIS+STvROPmI7NSlXTO2zmHEY/5u5dJd+fS8b6T9MnoruJ6CgRnSCi+z1+fy8RzRLRc8a/Dyi/+1MiOkRER4joL0l1JwaETzz5Ev7NJ36IU44BC6vhkhJE62SEfnRGn3/byeiHlXeyq5V3Kg3TA1K5dmoIL82XArdjkL30nafOzvEsWsKSWFaDviRPIpeKhcrT9zKYUtcfdXiQI5m4Z8rmajT9pUoDI5k4UjG3vGMVQhV8U1uF0DuEyiCuZC0KtOT+5JIxJbYTbMUn2yDkFKO/VG6grrVwbqmCX/rr7+Ov/uWE630r1YZvjGap7K/pq9ec/Hk8m7QdY8voR81A/UpFD+SqHTb11/jLO5vzyf4y+kQUBfAQgLcC2A/gPUS03+OlnxVC3GT8+4Tx3tcAeC2AGwFcD+DVAN7Qq53vBxrNFv7uqdMA3AMWVkM4o7+EiVzSNDZehJVmyvUm0i55J+LZP8SPlWrDpudLrpvKQwir938nZP8eJ7tk2mYPgrkLpTrGMnFkE2E9/Qp2jNmPu8x3dxv9BJYrdZchtck7IW44rZbe5kJ6+k6jX1a8Vb8ZySsVDYWaZu6zZGo4hXK9GWp/OqGmXU4Np5BPxQJn8JRqGiJkGU+ZtjlfquHPHz+KutbybMuwUmn4rp6WKg0QuWVDZ1xHVuNODiVtjool7xirvHQcC6U6lsoNl6bvlbIpnbAtw6m+C+TeBuCEEOKUEKIO4DMA3h5w+wJACkACQBJAHED7teY68t3jczh4enVj977+wkVzGdyr6lDAni7XKZvkubNLuHnniMsTVknFIqFy7Ct1zeXpp0NKRHLEoZPrtoTz8vQBKu7tyAKt1WbwNJotrFQ109MPKrFUG03MFmpmNa5E3nydF/5IOo5GU5iFRpJCTcNQUmZ/BD+HCjUNLaEHIlMe2TtFo0ndtVuGfCWeOaN6VC0kAnRDBNidj9ViefpREBGum8oHPgdKdQ3ZhLXak1W5Tx6bw5d/dA6A+/oTQmClqvm2yl4u15FPxW1ZNoAeNyh5ePqTQ0lUGy2zklga8KRh0POpmBljmnDKO15G3/h5Sz7dd4HcbQDOKo+njeecvIOIfkJEXyCiHQAghHgKwDcBXDD+PSaEOOJ8IxF9kIgOEtHB2dneT0Ty4z9/+Xn86dePrmobf/O9l8wlp9pPe7XMrNTME6edEVos1XFqrtRW2gHCSzMlz5RNf4morrVc2vpKxdvTN9sxGBe8bKjmh5R3nEwOJZGOR1cdzDVbJoTU9GUPeqenv83H05ePnWmbhaqG8VwCiWgklKYvDYVu9COo1t2afjYRw5v3b8bB0wueK1FzcpRD4tiS731VbsmRdnndVD5wO4aSMTVLMmF4+v/ta0cwnI7j5p0jLomkbAzEKfoUvUlpzEk2GTPfC1grpknjxiiNtdpwDdA9/ZeMflBjfvKOco3UGk0QAZvyulS1Fr2OvAhi9L3cR+fefQXAbiHEjQC+AeBhACCiqwFcB2A79BvFTxPR610bE+JjQogDQogDk5OTYfa/a5bLDZxZKIcqEnLy3NklPHtmCfe+ZjeA3nn6rZbApZUqrprUu2W28/SfmzaKstpk7gBGymbILptuecff07/v08/iP33xJ7bnVnxkmUiE8IotQ3j+3DI+f/As3vIX38Ftf/wNzxx2wPJYnRARdvYgg8esoDSyd4q1YJWxcoXh1vT1x2OOQK40rE7jVDRGQQ6lYqHOIXXUn9cqrGTEZd78yi1oCeCfj7i9fee4QIn09Hs5NlGVdwDguqkhlOtNnArQX6pUs8eYpAFeKjdw3xuvxo7RjMtblsdSawnTQKs4++5I5LkmVwhyv+UAdHkdVQ2jnVCC+LJgzC+Q6/T0U7Eoho2+PWEL87oliNGfBrBDebwdwHn1BUKIeSGEDKV/HMCtxs+/COAHQoiiEKII4GsAbl/dLveGQ+f1POqLK1XXEOSgSC//1163B0Dvcm0XynVoLWEOQ2nnef7ozBIiBNy4fbjtNsOkbAohdC/RFcj138bp+ZJrqa62VXZy3dQQnj2zhN/9wk9QqGoo15t4wfhOnMj5uF7sHM/gzCqrcheKVtl8LhlHsyVsn/PYpQL+yyOHXFlUx82upvZW1qann/Xx9B3peXI4SD4dD+XpW4FIvTir0RS2AfTSO37l1jy2Dqc8JZ5lnwyWTUMpEOmtQHqFGsgFgJ/aow+dCTIBrej09A2jv20kjffdsQv5tDsDRj2WXteQ3ks/4Xre2X/HlHekp1+3PH29PYnuF6urWmfKZsSjFUqloWfIyWO/Xrp+EKP/NIB9RLSHiBIA3g3gEfUFRDSlPLwHgJRwzgB4AxHFiCgOPYjrknc2Alk802yJrsrNLy5X8dWfXMA7D+zAWDaBVDzSs6CXrITca3j67bJJzi6UsXUkbbsgvAgj79S0FloCHimb/tsoVjWb/qvrqd7ZOwDwS7dsx9tunMLDv3YbHv0/7gQA3zTOQlVzFWZJdo5lcGahvKql8ULZaoUr/05Babr22AsX8bffP42X5uyB52OXipgcSrryvK/dMoRfumUbXr/PvmodNTx9Z9qmHAU5lIqFchyWbfKO25Ms13SJjojwpv2b8eTxWZc85+fpJ2IRTOVTPat4Btye/q7xDHaMpfFkgIlhUqqSpBNR/Oodu/An77gBScNbXnHIOOqx9LqGZI2DE6fRNz19w+hLj1zOx5Wo8StnPAdwt0Kp1FtIx6PmsV+v9sodjb4QQgNwH4DHoBvszwkhDhHRg0R0j/GyDxtpmT8G8GEA9xrPfwHASQDPA/gxgB8LIb7S48/QFdLoA9akozB8/YUL0FoC77tjFwDYJuesFplvvHdS7y/TztNfUqYztUNvw9AKpJ86PTLbNnw8/WJNw2K5YZ7U1UYLjabw1PQB4Jado/ir996CN1wzibFsApvzSd+gXsFH3gF0w1FttDCzij4sUt4ZzSTMgKpqJGQzNufgj+OXCq6BNYB+nD76zptcaZDDptF3a/q5VAz5VDzUIBXZvG3YqMgF7CmBpboVC7nr2k2oNlp47uySfRs+VakAcNVkDid7mIZsnleG8SYi3LlvEk+dnLc/6XrSAAAgAElEQVStULzf23Q5Ng++/XrcadxY86m4q3eRKpX5efpemr6cniVXCsW6hkTMaqomM3iqhjwjkavRCLkLvgB3K5Sq1kQqHvGV/daKQHn6QohHhRDXCCH2CiH+2HjuI0KIR4yfHxBCvFII8SohxBuFEC8azzeFEL8hhLhOCLFfCPHba/dRwnHo/Ip5wXaj65+cLWEoGcNuI23QOSNzNVxc1g2Y6em3MfqLZe8T14nV+6OzxCOLXrxyjZ3TfwBjCIixj1LTlMfCz1g7uW4q75m+V9OaqGst3xXDzrHVd9uUqX6jmbhnP3V5PNTWEUIIHJ8pYt8m715HXshujs5c/aKRvTMUshpYGgk9T9/d+rpU08xxlzKjSPZ6V7eRS8bM4jKVvZNZnJot9SzAaAVyLUN559UTKNY0183I/V633KginQvVcKqJFc5rqNUSWK54a/pyepbc35Ihv8kYl03eiVvHTd4UnH13JE55tGoUQJqevkcH1rXgiqzIXak28NJcCW955RYA3Xn6p+dL2D2RtfS8VKxn2TsXV6og0oNpqbi7zavKcqXh6VU4CTPuUPYpd6adpePuplGArk3K+4CUyuSqx89YO7l2Sx4nZgquDCCz746PfCW7ba5mXu5iuY6RTByxaMSUd7yMvroSObdUQbnexD4PT9+PREyv3FSzaOQNcygV11eLbRyHEzMFPHncym5brjSQiEWQikdNKU4N1peUAju/zKGlSt3TyweAvZtyKNS0rrpZnpkv4zvH7Jl4pbqGZCxiu8G8Zu8EIoSOEo8eY/J3IORnUI+fzdN33Ez1Ntbw0fTlnFwp7+jHUWaz2eSdmCrv6PvgDOJKUg55VAZy5fXbV57+5U6zJWzeyqFz+sV7665RTOQSXXn6L82VsGciaz7upac/s1LFRC6JeDTi2eZVZdFHl3QSZtyhNHIyLc65DacurF5QFx01C37yjpPrpobQaAqccujmctt+KwbpwZ5fRRbWfKlu9r3Peck7xkrgRWUlctwoLHMOoe+Es32AbLORS3X29P/gkUP40KeeNVda6nzXtIemX6pZOrhcDS6UHBkuRnGXF3KleaILied/fuck7vv0s7bnpMesMpyJ48btI7abmRdFj/eqSIOr6uKq3OrM1ZfSmLen7w7kZhMxy+grEmbK5unr2/LS8wF3XE0GcuXxX69BKleE0f+5/+dJ/Pnjx8zHMnPn+m3D5ni7MNS0Js4vVbBbMfpDPdT0L65UsTmvG9xc0r8tQNNYoo4GkHfSPgbbC+nZTbqMvjvXGNA1d8kl09N3t0Nuh19ZvlcvfZVELIKhZMx34LjkAw8/jT/4xxc8f7dQtLoiDnl4+vOlOiKk56xLPd4cQh9C3gH0rA61crSorGTy6TjK9aanvr1cbuAHpxZQqGmmzr6sGGznDbll6NsZw4DFoxHkU+7jtBzA6J+cDZ8dNbNSxUpVs60KSzXN1bkVAF6/bwI/Prvk6+lqzRaqjZarbkTF8vS9q5udN1O/FgyAFcgtKoHcXDJmraYUTT8Zcwdyxx2FbhKnvFOp64HgbCKKaITY0+8VQgicmC3ib79/2gySPX9uGVPDKb11wWg6tLxzdqGMlgD2TFiBunwq1rPsnUsrNbM4JpeK+ebpF6oNCOEdNHLi1fvDj7lizTSm9m14z9pVb0qm0Q/p6V81kUUiGnFl8HhNzXIyko375vhLDp1f8e3hv1i2jL5T09eaLSyW63jVDr34TW7j2KUiNg0lXUVNnZjIJW26urxhyuwdwDvT5JtHZ0wPX/ZaWlLiOdZKTv9+pcyTU4zsWDbhKtBaahMT2pxPIpuI4uRMeE9fOg6q91qsNW0ZOJI7r5lESwBP+aRuls3P0sbTT7srmlcqDSswX3N6+v5G33kOyNRXGR8p+2r6hrzTztNXboI1Tc/eISKMpONs9HtFtaEHHos1DZ8/qHfbe/7cMl65Vc9rl55+mGDVKcPz2T1ul3d0I+y/nZfnS4GyZy6tVLHJMPrZRMzmSasstvFWnCRDyDuzxRomc0lXW4ekR4EJYL+gZHfQsJp+LBrBvs05HHZ6+rX28g5gDCdpszQWQmC+WHcFMSXzSv9zM2XTMLwL5TqE0AOOgLUSOT5TCC3tALreq7YEVuf/5lNuXVry+OGLmBxKYjgdx4+MKWl2T1+/lKWnbw4SV4zsSCYRytMnoq4zeORnVFc1XvIOANy0YwS5ZAzfMXT9Zkvg+CVrpKYz1dMLU96p2DX9zcN6vYHTcZIrtuG0R2plLIJYhGzyTi4Z89b04x6aftbb09dTnp2efsTYD+9ZC2vBwBt9NQXu4adOm0HcG7ZZRr+mtQL15pacNkqtbZp+Su+r4udJzxdr+Jk//za+2qG3eU1rYqFUNz39oTZdH+WJGyRl09R8A8g7c8W6K4gL+K8WTMOVjFmB3A5avBfXTeVturm67XY3j5FMom0/8kJNQ73Z8mxDIITAomL0k7EoElEreC6N1rVTeYxnE3jx4gpaLYETM8VQQVzJeC6JxXLdLAi05J24eaycUkS10cS3js7iTfs346YdI2amixygArg1fdnfJ9vJ06802q5WZAZPGIQQ5vxa9e+VfIKx8WgEt181ju8cm8Wnf3gGP/vRb+NN//07+N6Jef195g3MX96xWhvbs3fyqRhyCXesZLmNp09Etk6bMpAr50Sr2Tuq0Z/MJZGIRbB7IuPaJiAbFjo0feP9w5neycOdGHyjb3xxb96/GS/Pl/HQv5yAEMAN23UNeZtRQh9G139prozRTNwmq5jLS59g7lxRr7J9eb79BTRjeMpS088mY74No8wc6xApm4E8/ULNpeer23A2bpMXx95NOXPi10q1gaSRWRKUa7cMYbZQswU6TXmnrafv3bJYMielhkrDlW66UtWgtYQt+JZNRk1jbAa1c0mjQVjBzNzpxtOfyCUghLVKkwNicsmYNWfVcfE/dXIe5XoTb9q/GTfvHMHRSwUUa5qRcqjvdzrhMPqOoSOAu59/taGnw/p5+oCu659bqvjGgj7x5Ck89E17S+NCTTOzsBaUm3G7YOzrr5nA9GIF//nLz5s3Khl7K9U6yzsxI+nB6enn03FPibRdfYL8W055h4iQiUcteafRNDtsAvp1+K3fuQtvu3Gr5za9ArmyvmI4HedAbq+QF+87bt2OrcMpfOK7LwEArlfkHSBc2ubpuZItiAvANjnHcz+Mi7vTikIWZm2Wmn6bQK7MQAhWnBVO03d2XQQUb1Jzavr6Z9s7mcPFFb0H+0pFC6znS2QwV9X1CwFWDKMesoWKLK4Swl0YZfbdUYy+7L8DWJ7+eC6Ba7cM4dilgrka8SrM6oQ8rlJqUldEluNg/74fP3wR2UQUr9k7jpt3jkII4JmXF1GsaZa8E7OnbHpJImPZuM3z7mT4AP1GDsCVVSX50rPnXENK1BTPBeUG3i7X/u2v2oZfe+0e/H+/fju+ct/rMDmUNGWlIPKO/Bxq2rSc5+CVAbdU1usT4h71CYB+3ZWMHkxqkVs6EbN672j27B0A2DqS9szRB4xArnEzbLUE6oamD4A1/V4ijcZIOo733bEbzZbApqGkqZnLPinnloLneZ+eL2HPuMPoe+QJq8ghELMd5nrKwqzNSiDXL2Vz0Ui/C5K949XwyYtWS2ChVPc0+laGiLuFLwDs3ZRFtdHCSkUzWjAEl3YAmNOU1AyeYk2fvuV3cQK60S9UNd+qznnlmDvljQXD+NqMftLqgaN6+tdO5VHTWnj8kN6b/uqQmTuAFeSTtRDFDpp+qyXwxOEZ3HXtJiRjUdy0XQ8of/uonuIo5Qlnnn7ZlHcUTz+bQKXRNL12U+Lw0LUlVxlV4X4ZPOeWKq5hK3MF7+Nd8qiqlQxn4vjIz+/HHXvHQUTYO5k1/2bRY9XihbNhXaGqIZ+OuYaiAPrNv93NLpuMoljTzBoUud+ZRNSWvZPymUnthZqnLx0nU95Je4/SXAsG3uiby+dUDO+5bQdS8Yip5wP6wR5KxgJ7+pV6ExeWqx6efvv2yvLi7lToIrNfzOydRAw1j7bFgH7ieg2B8MKreMeLxXIdzZbooOm7+8EnYhGzr/ylQtW3rXI7xnNJbBpK2ubnFqoNs0LSDzmW0G95PKusrtxGX3+PGnzT2ytbK7N4lJBPxXDdlG7kv/7CRWzOJ9saDT/GHZ5+UfFipdFX9ecfnV3CXLGGN+/fDEA3jldNZvGtYzP6Y2MfkjH7Ss4ylIqm72j45td3R2X3eBZE8MzgkRJTud60JRuoq1kp7zg95k5cNZnDiZmi0fzPHZ/wIq94y2rvJ69h92rGlhf6jaJp+34A3eiX600IIfSUzXhwEyrlHSGEeeNNKUa/UNMCJXqsloE3+mogcCSTwN/cexse+LnrbK/ZNho8V/9lo6Ojy+h38PTlEnWug6d/aaWKRCxienBSy/ZK21wyMi/8lpMqlqffXt6RF6yzMAvwjwsUq3obAXmjurhc9W2r3Ampm0tWfAaxqMjYip+nFMTTV+fZ2uWdGsazeibT1ZtyiEUIhZrWlZ4PWJ0a5XEuVBtIx6N6IZ7pOFjn0OOHLyIWIdz1ik3mczfvGDWDq9JgE5GRBy49fSP46fD01WMgj1e77K9UPIodoxnPDB7VUbqo9N2X5/hoxpKTdEPZWaKR7J3MYbnSwEKp7tsLyoku7+jHTu395CWRLpQbri6oKkNGHMCKJ+jnvpxQprUEWgLhPP14BC0BNJrCdL6sQK4e6wk7I7kbBt7oO0+YO/aOu1vhjqTNgRideMm42K4KqelLT2iug6cvC7NkuqSzUERl0acfuBfJgG0YzMIsL3nH58ZRrOkNw2Tw+dJKFYUuPH1An597YqZgSjXFqn+zNYmUt/yCufNFfUUE2AOLgKX3q56+aiTmS3Wzj3oyFjULlsL03FHJp2OIRcg0jPLYAUA0Qsgl7Zkmz768iJt2jNi8cXVgjhrETyuBQtNYJVRNP7ynD/hn8KiS6AWH0Y+QbrhlTCSoLq/+TQA4NVeysnc6vFdtemjWiaTiejKE09M3xmP6IadnOQPi0tNX5+MGRXWazPcrgVzAitOtJQNv9M3eLW0MRxhPX07GcXr6VspYe3nHWaXo5NJKFZuHUtZ22xj9pXI9UGEWADPdrJPR92vBAPjfOIrGZCsZh7i0Uu1K0weA/VN5NJoCLxhdUAvVRtvvDvDvUy+ZK9bMxmwLjkD6YqmOVDxiayOtevpzhqcvudaQeLoJ4gK6Rz6eS5irj0JVsxXBOXXpk7MlV2roTTsUo68YbHWymWUorc8lj5P0vv2mZjnZO5nDqbmiS3qwe/rWz7OFGsaySUwY6amAfVRiEMxq4JmimX6a6WBgZXtlQKkTScf0G6mH0W/n6WeNG7/TaUzHpdG3T80KQlJJq5XvVwO5wPr03xl4ox8kELhtJI1CVQvUO+f0XAkTuaRn2+FkLOLr6atpl14DnCWXVmrYPGwZ/bbyTsAOm4Cx/G/TD1+iBi6dyBuHM2WzYKTipYze4BdXqlipaIFiDU7uumYTEtEIHvmxPqdHN4qdNH3vZmKS+aJe9zCUipmevfm7Ut1VTKOOTJwv1m0NtGSG0b4u5R3AqMo15R37rAC1vfJiqY6FUt00gJJrtwyZxkZd6anT0Ur1JhIx+3k/5jhOy5UGImRfDXixd1MO1UYL5x1TtKaXKohH9SWU09OfHEpiLGfVBZgdNjv8Lcm2kTSSsQhOzhaNHkJRRDrImPm0/r1pzZbN01czcQB9tGehppkxDi9yRqq0dNbkKiOdiKFStxy3ZKhArv5d1Bot83syi7N8Zi2sBQNv9PX5qu2NhpnBE0DiOT1XtrVfUGnXdE1dsvsFc4UQuLhs9/TlctirKnepEqyXviTI9KzZYg0Jo0+LE/PG4Qgql5R+91vyKbw8X0a92fKdmtWO4UwcP3PdJnzlx+ehNVu+oxJVOsk7cyU9BXUs607tnFmpuVY1uWTM0IRbrvTVn3/VVrzv9l22ZICwjOeSmCtZHrD6+YaUbq0yTdJp9GPRCG40snjyDk9ffr/SUKoMp+Mg0vVsQDcw+XS8ozGVUqYzg+fcYgVbR9KYyCVtmv6sUdw3ZlRKqzOQgwZyIxFZDVwyevZ0fp8aCDd7Pxl5+i1hJTGYRY1tPH35HhlwlyuUjMPTDxvIBXRPX67I2NNfA/RRe+1PmDC5+i/Nl2ztF1TatVdWPXW/YO5KRU8R2zriIe94yEZLpeCePhBsepYszHK2YJBkElGX1KQW3WzKJ80OlN0EcgHgF27ehrliHU8enzOmSrXfTjoeRSIWaRPI1b11r4rU88sVbFOON2Ctri4uV1HTWrZMpm0jafzhL1yPRKz7S2cimzBjO1Iak+TTcXNq18kZI3406T7f7nrFJHaNZ2yevBrILdU1WwsGQI8ZDKfjNk8/SExI5uo7M3jOLVWwbSSNqeGU3dMv6G08xrIJNFt6Fk1YTR/QP/ep2WLHDpsStb2y5enHXJ1TFwJUssv9lG1Fsqanr0toq9L0bZ6+XdNno98DgpwwVq5+e6NfNHqL7/G4CAGj06Zvnr5meqR+Rv/Civ73twy7PX2nvNNo6kvUdjnWToIMR/drwSAZzSZcMkpRkSi25FNmK4ZuArkA8MZXbMJIJo4vPDttC3T6QURGVa7b6Ne1FpYrDUzkkq4Ol0IInF+qYGo4bXuPPF/kYBa/XirdMp5LYL5UgxDClZKqevonZ4tIRK1UWJXffP1ePPEf32B7Lp2wburlWtPzvB/LJEyj167vjm1/swkMp+OuFsvnFnWjv2U4ZaYayxYME0NJUxabL9VNeTOM0d87mcOZhTKWyo2O6ZqAfZCKpem7B+OYk9Ky/p9devbyc9lSNhtNRd4J4+nLLrXWTUPGkryGwKwVg2/0A2R/TGT1nhleRv/YpQI++d2XsFiq4/Sc0XPHz9NP+/fPKNY0c+CHX1WuXCJPeWj6Tu9anhztTlwnyXi0c8pmwbsaVzKeTbh0cV3T1/dDBnOB4G2VnSRiEbztxik89sLFwNsZzSRcveIB6wIfzyVclbtL5QaqjRa2jtiNvjxfZI8lv6EY3TKRS6LaaJn57UM+mv7J2SL2TGQ9U3IjEXKtNlIxVdP3bmM8mrX6FPkNBndCRLh2yxAOn7fqJ2paEzOFGraN2j192YJhMpe0BY7DyjuAnsHTEno7BueqxQvT069otkpnp9GXunnbPP2Ew9NXsnf01Yu+re48fcXoG8+l4lGk4hE2+r2gUO3s6UcipHfb9JB3/vCfDuPBfzqM1/zJv+CPvnoYgDtzR9KuvXKxqmEip89g9dP0pdHfonie8mTzqigEOqfbqaTikbaZQ4B/CwaJmnkCWOMMpeFSg9DdevoA8Is3b4dmZIsEado26tN0TR39OJbTb1gyoCcDk1uHHfKOcQOTfZLaHY9uGDdz9Wvemr4x4PvUbAl7N3mfa16kElFbGwav8169ObYboOLkhm3DOHJhxWwUd2FJP1elp68XaWmmbDUxlDCN6kKprsg7wY2kjGUsGi0TOiFjSNLTT8YiSMaiVtpz1e7ptw3kpqTRryIdj5o33rRxPcpzLZTRV1KencVZwPpV5QYy+kR0NxEdJaITRHS/x+/vJaJZInrO+PcB5Xc7iehxIjpCRIeJaHfvdr8zQeQBANg+msaLF1ds5eQzK1V878Qc3nHLdvzcDVN45uVFJKIR7Br3D+T6DbaWMtPkUNK3FcOFZX1M4iYlsBiNkK6j+wyBCBPITXfQ9FstgflS3bPZmmQ8m3SV1gOWB7fF5ul3b/Rv2Tlizh/uFIgH9BWPl7wjjf7kkB5YrGstMwXwvGG4phyevjRML83p8k7vjb7+nZ1ZKEMIuDT9ZkvvXfTyQtkVxG1HWhlcX6o1PbtSjimzB5YCTl0D9IFDNa1lSjxyVbxtNG0ryjOL+3KWvKN7+uGydwB7LCOILKS2sZDN1gD3YBz5+dulO+eSltFX/7Y8pvLG2ZW802ii4kjZBPR2GH3h6RNRFMBDAN4KYD+A9xDRfo+XflYIcZPx7xPK838H4M+EENcBuA3ATA/2OzAFoxS7Ez93wxROzpbww5cWzOf+8bnzaAngQ2/ciz9/56vwnd97I774W6/xXWrqxSGaZ0992alvIpf0LdC6uGyNSVTxahi12IXRT3XQ9Nu1YJDoGTANj9bAhqeftwxkN9k7EiLCL9y8DUAwT3/E0UFSYjZMyyZdKYsXpKfvCOTKvyc9/XYyQDfIwjcpF6qBavm3nz+3jGZLhDL6qXjEJu94GdjRrK7py8HgQT39641speen9foJuSrePpIxY1C60bdSflV5Rx/S3jntUiWTiJmrMGcmkheWvNMw2yoD7qEoC+U6hpKxtsF4aejnijVbbYE0+l15+mr2TsMdE1ivTptBblO3ATghhDglhKgD+AyAtwfZuHFziAkhngAAIURRCNH9BOuQyKHTQZaGv3jzNoxm4vik0YUTAL70o3N41fZh88KbGk7jhu3+qXr5dAz1Zgs1jz45BWPFMTGUaBPIrdr0fIl3l8DOJfROOqVstmvBIJE3BHnTKSi9jYDeefoA8N6f2om33TiFG9scc8loRh9C4Swgkil347mELbAI6J5+PEqYyDpTNg15Z6GM4XR8VZk6Xsj9OG0Eip15+gDw3Fl9OpZX5o4faaU4q1z3bm4mVzszhRpaIrg8eNVEFtlEFIcMXX96qQIiPelABsIvLFdtozblKEBp9MMEcSUycyjIezPK2EHV03cmQ3QqzAKsrLmWo3WE9MwXTaMfpjjLGjdaa+gDVNSb4A3bh20zOtaKIHu8DcBZ5fG08ZyTdxDRT4joC0S0w3juGgBLRPQlIvoREf2ZsXKwQUQfJKKDRHRwdrb9gOQwlOt6h7wg8k4qHsV7btuJJ45cwtmFMl68uIIjF1bwizd7fVRv/Fox1I2GabmE4en7BnIrNqMp8eq02W7Gpx+dUjbbFWZJxrKOhmHKABVA16ujRpAxjBfkxaahFP7qvbcEqjoezejpgc7eJXPFOpIxvde65Xnq+35+qYItwymX9ynPl7rW6nkQF7BWDnIl4azIBWAOSrkqpLxT1Zqms+MXyAWAl4xVRtBxj5EIYf/WPJ4/Z3n6m4b0BAhT3lmpmi0Y5LEeNdJkgzpfTqTDFcToE5FRlduw9X4yh9OYnn77vjvOv2eXd/SfZYfbcL13rHkUFcfULQD4v962H3/yjhsDb69bghh9r/WYU7/4CoDdQogbAXwDwMPG8zEAdwL4HQCvBnAVgHtdGxPiY0KIA0KIA5OTkwF3vTPSUAad3vS+O3YhSoSHv38aX372HGIRws+/ynsgghd+Tdekh5FL6UZ/udLwDKheWPb39F29Q8p1xIxeLUEJavTbafpmcK7oKK9X+sdM5pKr9vLD4teKQQamichMvZR67IXlCrY60jUBvQBHlik4VwG9IBmLYigVMw2vzdNPS09/CVvyqVDfbzIehRC6syOdDCfyOMnMpDCJANdvG8bh8ytotgTOLZXN+pZ0IoqRTBwXliuYK+otGGTgc9ww+u166bdD9uAJIu8AejLFckWz9X5KxiKIRsh0UDr13QH0VYM8B9TvIG1q+vp5Fqo4S2lvXqk3bXr+ehJkj6cB7FAebwdwXn2BEGJeCCE1i48DuFV5748MaUgD8A8AblndLgenUA2XJjY1nMZbb5jCZ58+iy//6BzecM2k72R7L6SGuOwo0FLT1aRBdbZiKNU0FKqaLXNHkk26x70tVfTCLL8iKi9SsfYpm3Jp3s7Tl/KOWlEq91GyOZ9clZ7fDTJ11Wn01TYKYzmnp191pWsCulcrDaZsttZrJnNJnFnQ5R17yqbUkuuhMncAS3qQBsmrinXMOE7yhhM0kAvoGTyVRhOnZot6YZZSP7Aln8LF5SpmC/Y6jzHT6HsPRe9EGE8fsDptqr2fiMjmOC0EkHeIyNxfr0CuKe+E8PTjUUKErOKsfjb6TwPYR0R7iCgB4N0AHlFfQERTysN7ABxR3jtKRNJ9/2kAh1e3y8GRmTRh5rT+2mt3o1DTMFOo4RdvCS7tAP6evmr0J5R0PRVZ0OTl6Q/5aPph+7mrFZtetGvBILE8fathmNxHyR17J3Bg12iofVstVntl+7FXU1CzCX3+7XxJD1hf9ImhAJb33evCLMl4LoFGU18w27J3lBVSmCAuYMkHMjvMyzuWnn5YeQewgrk/nl7GxeWq6ekD+nkr5R11pbhaeeeVW4excyxj9jzqhOyp75zcpjZdWywHa18i99crkLtYriMRjYQKTOvtr6Nmw7XVyp/d0vFbEEJoRHQfgMcARAF8UghxiIgeBHBQCPEIgA8T0T0ANAALMCQcIUSTiH4HwD+T7pI+A30lsC5Yxjb4iX3zzlHctGMEJ2eK+NnrNof6e36aviqByBPJZfTNHH23EfJqDbtUboTK3AF0T1BrCTSaLc8GdHOGl9Zu9TCSSSBCljfplHcA4P63Xhtqv3rBmKODpGS+WMd+w2AQkZ59VKpjtlBDsyU8PX3AuuDXQtMHnENb1Owd62dn++5OpBP6dypXkZ6BXIemH6aie+9kDul4FN98cQaNpjAr2QH9vH3+3DKSsahtv6W8E4tSV4Hc4Uwc3/m9NwZ+fT4Vx6nZkt77KWU3+sWqhmqjiXK9GSgjS8pR6gpFyjuLpUaodE1Jyoi7VBtNW2fX9STQtyCEeBTAo47nPqL8/ACAB3ze+wSAtY9OeKCOogvD/3jPzZgv1UPfif1mnHp6+gW7cbrgUY0r8QrkLpYbNk8rCGrKmKfRL3oPRFeJRgijmYQl71Q1RAgbtlSVeGn6QgjMl2o2iU7KDed90jUlpqff4xx9iSobqTdMvSMsodEUZuZKUOR3IIvnvDT0fCqOCAFnjMyhMKvFqBHM/eZRPet6u3L+bcmnMVfUvV81+2ssm0Sl0cR8sd6V0Q9LPmius7sAABnLSURBVB03U3FViTGX0rtmhqlvkTd+r0Cu3lAwfNwqFYuY8k6YzJ9eMtAVuYVaOE1fsmMsY+tZHhRfT7/q1vSdBVqyH/lmr+ydZAyNprAFf5fL9VCZO0Dn4eizHVowSMZzCVsgN5eMhYotrAVDqRgiZJd3VioaGk3h0pjnS3WcN4qLnH13JPKcmVxjT1+mGUqIyPT2w8o7sl+7TEn10tAjxk273mwhEY2ENjzXb82b4wtVT186K/Vmy3G89c+inydr7xjk03p3TMAulcn++GY1boD2JVlT3nGnbALh0jWt90Qvi0DuZUuhS0+/W1JxXTP21fRTes95r1YMF5arGM3EPVcXzi6BgO7pBxmI7tw/wH+QSqcWDBLdcFqTn7rpm99rIhHCiKO3zlzJHZiW8o5sI+An7wyttadvGEYvhySfiiGTiHqm77ZDGpE509P3Pu9lEHM4ZCIAYOn6AGwrTVWWtB9v6+f18PTVlYvqicu42GKADpsSL08/qvQ86kaTl/2vqh4pm+vFQBt9Z7XoepBPu9srO9vKTgwlPTV9r8wdwF1RKCv6gk7NkrQz+rIFQ5BslfFs0vQmna2BN5KRjL2i0azG9fL0lyvIJqK+QWtT0+9xNa5EGkavGpJ8Wh9+HiZICFjfr/zcXm0YACv+0c1gd1mcOJKJ24yhKktODtlvspJusnfCkveobgasAkfL0w9j9O3HUR7X7jR9vf9VdQOzd/rjal0jijV96HSszdSsXpP3aK8sVxzypJ/Iuaty/XL0AfecXNmfI7y8Y59x22wJfPEZvX1xVWui2RKes3Gd6E3XFHlnnVZSnRjL2PvlexWbjWUTKFQ1nJkvY2ok7evpyuB/u+rk1SBXEF6rpA/ddTViIQ0+oGj65uAPP09f/5th0jUlV0/mkIxFXPEkf09fMfrrpOmbP3vIO4sBBqio7wHcN6tMPIolNLry1OX0Oq/irPWiP67WNcI5im49GPJoryynGEntdiKXxLFLBdtrLq5UcdNO7ziC2TCqaqWcAeEyLwB7P28AOHh6Ab/3xZ+YvycCrtnSeQzgWFZvDGX19N94eQfQM4umF60uHzKg6fT0AeDQ+ZW2n/UVW3K4aiJrS0XtJXKfvLZ/9/Vbutpm2unp++y7PAbdePqxaAR37pt0OShDxljCYk1rY/TX3sjZ5R1nILdpVsMHOWel7XDerGTWTXeafgRzRX1YUl9n71yuFGraml20fni1Vy46+o5M5JL4/sl583G10cRCqY4pHw3Xrx94WE0/7ZB3po2mWV+573XYOZYJnFYnvdTFUh3FagPbR8NlEa0Vo5k4Xjhn3XDninUQ2VvoSiN0caWKu17hX/39rlfvxLtevXPN9lVW+vZSGpNGSBo2v0HiUhYMk6Ov8on3H/B8fnM+ifKcZjP0+VQMsQhBa4l1kQHztkI3u6YPANOLej+lIKt/K0/f4ekbnn+Y+bgS2fRwI/P0B17TX68griSfjqPg8PQLDglEtmKoG43ZZoxBDV45+oBb3rGarXWn6cumXDK1bd/mHIYdGm07pM49bxTdrPeN1Y9Rxwzc+VINo5mE7QJXDZJf5s56kE/HEI9ST8/PlNkioNa2o+VqNP12TA2nbS0YAGOqmXHM1zOQ6+z9JP/29EIlcNdUWdzmXKGsztOPmvIsZ++sARuhN+uavjuQq3oLZisGQ3uVxtfP6Dv7gXfTbA1Q5R39ZnNuqYrxbCK0x2Ea/WK97wK5Nc0aUDFXqLsCsepjvxz99YCI8As3bcPr9k30bJvSiLQE2k6aGl2FvNOOt1y/BT//qinX8/KYr4unb3wmZ+8naQfOLpYDr5AP7B7Da/aOuzK8ZCA3TAsGSSpuzXJOb1Cefn9crWtEodrARG7tW5Wq5NMxd8qmwzCa/WsKdUwNp9u2YAA8WsN20UsfcGfvXFiuYKoLwyf16LliDaV6s68CuYDeL31bIm0UZtmP0ajN6G+sLPVn//pVPd1ePKo3Fmu2RNuc+LFVBHLb8b7bd/n8vfXz9KWxd/Z+ksfj4krVrNDuxPXbhvHpX7/d9byZvdOF0U7Gomb7DZZ31gBd3lnfIGM+FUdda9nSIp19RybMAi3d2F/wGJOoIrs+ykDuUqVuLF/DfX0uo79U7UrikLnXsmFY/3j61pCUVkvg/FLVVXcwmkmY3RP9brKXM9Lbb+vpr1LTD4sl76y9kUvEIkjHo25P38jGEiJY5k470vHVafrmdjYokDvQRr/QZZOn1eDVdM1p9CcdrRguLlcxlIz57mskonf8kxXGSyW9MCtsYY3T6J9fqrjmwwZhJK2X8sv2vOsdN/FDLtuXyg184runcG6pgjdcYw/WRiNkergb7emvBdIRaGdg92/N41du34nXXd27NubtWE95B9C9fGeLBPVvr3YSminvdJOyqThq7On3mFZLuIZOrwcye0At0Co5YguTQ0nkkjH8v985iXNLFVxYrvjq+RLZGvbp0wt49IUL2DUWXrZKxaw2DIVqA4Wa1pXhi0T0xmWyf0uYhnZrifTgvn1sBn/69aO4+5Vb8Mu3bvd83VgXsYzLAfmZ2kkpyVgUf/QLN3Tss9Qrrtk8hM355LoFLq/elHM1q1ONflhZ1ElmlYFcCRdn9Zhyowkh1t8LlR6GjNDLKUbqRZiKR/G/3n8AH3j4IH75r7+PaIQ6jknLpWL43y8t4B+fO49tI2l89F3h9eBYVG/mVW00rQZvXXq749mkOe5vPZbtQZCB7Y8/+RK2DqfwJ++4wXM1NJFL+larXu5IQ7Ie1a9Bee9tO/GuV+9Yt/5MD//b21x/S3W6gvTdaUfarMjtpjiLPf01w2rBsL5eqOyXIht61bQWGk13jvJPXTWOz/7GHWg0BaYXKx315WwyhtPzZVyzeQif/807sF0ZYBGGVEzPEz5n7F838g6gL5FlxWv/yDu6Bxch4C/efbNvSuv9b70Wf/Dzr1zPXVs3LE+/f25qkQh5dnVdK2JGQFtFPR7s6Q8ocoDKemeW7BzTjbEMcsqMGy/DuH9rHl/8rTvw7z/9LG6/arztdl+9axSTuQT+4t03r0oblQ2fOjUc64SaFdMv8k48GsFr9o7jrldM4rY9Y76vu2Xn+g54WU+CBHKvRJIxvRlivdlataafNo5td5q+GsjllM2eIoOe6104lDV65suh1+Y4QZ+LcNd4Fv/0H+7suN3ff9v+nuxfOhFBrdHE+aUKIgRs6lLXVfPd+yVlE4Bnit2VRMqnoIjRz9MgoxI7ISudu224Zv3M8k5P6XaASi/YNZ7By4be7TVZaiOR8s755Qo251NdN6NTW+b2S8omY2nG65ETf7khb4Rjq5R30qvI3kn2gbwzuEZ/A43trrGMKe9sRHvndsghDhd8hoIHxS7v9MdnYyyD1E+B3H4hl9RTjbuZeKWyqjYMSvCXPf0eY2r6G2CQdoxlcHGlimqjaRuV2A/ow9FbOL/cOXjcDinvOCc/MRuLNCrs6bvJJaMYTsdXfb5uH0mDSB8RGZbLRt4horuJ6CgRnSCi+z1+fy8RzRLRc8a/Dzh+nyeic0T0V73a8U5YU7PWP8i4azwDIfQuln0n7xhd/i4sr9bT732XSGb1WJ4+a/pOhlLxVev5ALBv8xCe/f03Yf/WYO0cVKShT8TcGUbrRccrloiiAB4C8CYA0wCeJqJHhBCHHS/9rBDiPp/N/CGAb69qT0OykR72rnGZwVPqQ08/ihcvFlDXWl2nawJWVWO/3MwYHWlU/HrpX8l86K69Zv3Maun25iG/n43S84Fg2Tu3ATghhDgFAET0GQBvB+A0+p4Q0a0ANgP4OgDvRtxrQKGqbZj0sNOoln15voxGU+9o2U9GX87n7bYwC7CaxvVLW2VGR8oH6zGE/HLjwG7/NN71Qn4/3cQDekWQv7wNwFnl8bTxnJN3ENFPiOgLRLQDAIgoAuDPAfzuqvc0JBvZ8ncil0AmEcXL82UUqxqI/OeVrjdqO9etq+gnn0/p2ih7+v0F5+n3NzLmspGefhCj7+UqC8fjrwDYLYS4EcA3ADxsPP8hAI8KIc6iDUT0QSI6SEQHZ2dnA+xSZzai746EiLDTyOAp1DTkErF1K0HvhBo8Wk0/+UiEMJpJ9M0KhtGR3y9/L/1J0vT0+1vemQawQ3m8HcB59QVCiHnl4ccB/N/Gz3cAuJOIPgQgByBBREUhxP2O938MwMcA4MCBA84bSlfo06o2rlJ013gGJ2aKmMgl+sobTimFJautTHz3q3fgqsn1nVfAtOf2q8bx1uu3YLPP6E1mYzE9/Q1c+QexRk8D2EdEewCcA/BuAO9VX0BEU0KIC8bDewAcAQAhxL9RXnMvgANOg79WFKqNDdWbd45l8M2js7hm81Bfpc/J4p2p4dSqVx+/85ZX9GKXmB7yii1D+OtfuXWjd4PxIRIhJKKR/g7kCiE0IroPwGMAogA+KYQ4REQPAjgohHgEwIeJ6B4AGoAFAPeu4T4HoljVzOZnG8HO8SzqWgsnZ4t9tdSWZfqD2EueYS4HkvFI38s7EEI8CuBRx3MfUX5+AMADHbbxtwD+NvQedolzcMl6s8tovHZytoTX7G3fTG09kcvLjRwKzjBXMql4tO8DuZclher6D0VXkbn6zZboq5J46WFs5FBwhrmSuXbLEK7elNuwv98/1qiHmFOzNtDT3zqStoZU91EgV7ZzZXmHYTaGv/93P7Whf38gPf2KMQN2IwOo8WgE2wzD2leavinvsKfPMFciA2n0y3Xd6G90QZQcqNJPRv/6bcO4ddcobtw+stG7wjDMBjCgRl/vd5PeYC19p6Hr95O8s2Msgy/+1mtWnaPPMMzlyYAafUPe2WBPX2bw9FOePsMwVzYDbfQ3suoNsDJ4uCkZwzD9woAafV3e2eimU6/cOox4lLB7glsVMAzTHwykC9ovgdwdYxk8/1/esqHVdwzDMCoD7ulvvLFlg88wTD8xoEZ/4/P0GYZh+pGBNPqVPgnkMgzD9BsDafRLNUPTZ2mFYRjGxkAa/XJDQyIWQSw6kB+PYRimawbSKlbqzb4I4jIMw/QbA2n0S7UmSzsMwzAeDKTRrzQ0ZDhzh2EYxsVAGv1SjeUdhmEYLwbS6LOmzzAM400go09EdxPRUSI6QUT3e/z+XiKaJaLnjH8fMJ6/iYieIqJDRPQTInpXrz+AF+WGtuF9dxiGYfqRjpaRiKIAHgLwJgDTAJ4mokeEEIcdL/2sEOI+x3NlAL8qhDhORFsBPENEjwkhlnqx836Ua02kx9nTZxiGcRLE078NwAkhxCkhRB3AZwC8PcjGhRDHhBDHjZ/PA5gBMNntzgalXG9ueC99hmGYfiSI0d8G4KzyeNp4zsk7DAnnC0S0w/lLIroNQALAya72NATlOss7DMMwXgQx+uTxnHA8/gqA3UKIGwF8A8DDtg0QTQH4ewD/VgjRcv0Bog8S0UEiOjg7Oxtsz9tQ5kAuwzCMJ0GM/jQA1XPfDuC8+gIhxLwQomY8/DiAW+XviCgP4KsAfl8I8QOvPyCE+JgQ4oAQ4sDk5OrUn7rWgtYSbPQZhmE8CGL0nwawj4j2EFECwLsBPKK+wPDkJfcAOGI8nwDwZQB/J4T4fG92uT1Wh02WdxiGYZx0tIxCCI2I7gPwGIAogE8KIQ4R0YMADgohHgHwYSK6B4AGYAHAvcbb3wng9QDGiUg+d68Q4rnefgyLkjFAhQO5DMMwbgK5w0KIRwE86njuI8rPDwB4wON9nwLwqVXuYyj6ZSg6wzBMPzJwFbn9MhSdYRimHxlAo2+MSmRPn2EYxsXAGX0elcgwDOPPwBl9M5DLrZUZhmFcDJzRNwO5PESFYRjGxcAZfSnvcHEWwzCMm4Ez+izvMAzD+DNwRr9Sb4IISMYG7qMxDMOsmoGzjHpb5RiIvPrEMQzDXNkMoNHXOF2TYRjGhwE0+txWmWEYxo+BM/qlWpNbMDAMw/gwcEa/0tDY02cYhvFh4Iw+yzsMwzD+DJ7Rr7HRZxiG8WPwjH6Dh6IzDMP4MXBGv8LyDsMwjC8DZ/RLLO8wDMP4MlBGv9USqDSaPBSdYRjGh0BGn4juJqKjRHSCiO73+P29RDRLRM8Z/z6g/O79RHTc+Pf+Xu68k0qDp2YxDMO0o6NLTERRAA8BeBOAaQBPE9EjQojDjpd+Vghxn+O9YwD+AMABAALAM8Z7F3uy9w7K3FaZYRimLUE8/dsAnBBCnBJC1AF8BsDbA27/LQCeEEIsGIb+CQB3d7ernbF66bO8wzAM40UQo78NwFnl8bTxnJN3ENFPiOgLRLQj5Ht7guylz54+wzCMN0GMvlePYuF4/BUAu4UQNwL4BoCHQ7wXRPRBIjpIRAdnZ2cD7JI3ZR6KzjAM05YgRn8awA7l8XYA59UXCCHmhRA14+HHAdwa9L3G+z8mhDgghDgwOTkZdN9dSHmHp2YxDMN4E8ToPw1gHxHtIaIEgHcDeER9ARFNKQ/vAXDE+PkxAG8molEiGgXwZuO5NUHKOzwUnWEYxpuOLrEQQiOi+6Ab6yiATwohDhHRgwAOCiEeAfBhIroHgAZgAcC9xnsXiOgPod84AOBBIcTCGnwOADwUnWEYphOBdBAhxKMAHnU89xHl5wcAPODz3k8C+OQq9jEwZZZ3GIZh2jJQFbllKe+wp88wDOPJgBl9Q95hTZ9hGMaTgTL6pbqGRCyCWHSgPhbDMEzPGCjryG2VGYZh2jNQRr9cbyLLLRgYhmF8GTCjr3EQl2EYpg0DZvRZ3mEYhmkHG32GYZgriAEz+jwUnWEYph0DZvTZ02cYhmnHQBl9TtlkGIZpz0AZ/VKN5R2GYZh2DJTRrzTY02cYhmnHwBj9utZCoynY6DMMw7RhYIw+D0VnGIbpzMAYfQD4VzdOYe+m3EbvBsMwTN8yMG7xcCaOh957y0bvBsMwTF8zUJ4+wzAM0x42+gzDMFcQbPQZhmGuIAIZfSK6m4iOEtEJIrq/zet+mYgEER0wHseJ6GEiep6IjhCR5/B0hmEYZn3oaPSJKArgIQBvBbAfwHuIaL/H64YAfBjAD5Wn/zWApBDiBgC3AvgNItq9+t1mGIZhuiGIp38bgBNCiFNCiPr/3969hUhZxnEc//7SDh4KtezggVKQDIQylrADIdZFWWgXXUhBQoI3BRZGB4KguiokKxAhLLOITiYlIkWYEgQdditMU7Kz2pYrpR0g7PDr4nlWpm1nd3bdafR5/x8YZt5n39n3+fMf/jv7zPv+B3gBmN/Lfg8CDwO/14wZGCVpODACOAT8fGRTDiGEMFiNFP2JwO6a7T157DBJM4HJtjf0eO5a4DegE/gWWGb7x54HkLRYUruk9q6uroHMP4QQwgA0UvTVy5gP/1A6DlgOLO1lv4uAv4AJwBRgqaSp//ll9hO222y3jR8/vqGJhxBCGLhGLs7aA0yu2Z4EfFezfTIwA9giCeBMYL2kecANwOu2/wD2SXoHaAO+rHewjo6O/ZK+GVAU/3YasP8Inn8sqmLMUM24qxgzVDPugcZ8diM7yXbfO6T1+M+AK4C9wAfADba319l/C3CH7XZJdwHTgZuBkfm5C2xvbTCIAZPUbrutWb//aFTFmKGacVcxZqhm3M2Kud/lHdt/ArcCbwA7gJdsb5f0QH4335cVwGhgG6ngr25mwQ8hhNC3hnrv2N4IbOwxdl+dfWfXPP6VdNpmCCGEo0CJV+Q+0eoJtEAVY4Zqxl3FmKGacTcl5n7X9EMIIZSjxHf6IYQQ6iim6DfaH+hYJ2mypM25l9F2SUvy+DhJb0rale/HtnquQ03SMEkfSdqQt6dIei/H/KKkE1o9x6EmaYyktZJ25pxfXHquJd2eX9vbJD0v6aQScy3pKUn7JG2rGes1t0oez/Vtq6RBf3lIEUW/0f5AhfgTWGr7PGAWcEuO9W5gk+1pwKa8XZolpDPIuj0ELM8x/wQsasmsmusx0rUu04HzSfEXm2tJE0k9vNpszwCGAQsoM9dPA1f1GKuX26uBafm2GFg52IMWUfRpvD/QMc92p+0P8+NfSEVgIineNXm3NcB1rZlhc0iaBFwDrMrbAuaQWn1AmTGfAlwOPAlg+5DtAxSea9JZhSPyNUIjSW1cisu17beBnm1p6uV2PvCMk3eBMZLOGsxxSyn6/fYHKlHuWDqT1Nn0DNudkP4wAKe3bmZN8ShwJ/B33j4VOJCvI4Eycz4V6AJW52WtVZJGUXCube8FlpF6dXUCB4EOys91t3q5HbIaV0rR77M/UIkkjQZeAW6zXXTnUknXAvtsd9QO97JraTkfDlwIrLQ9k9S8sJilnN7kNez5pF5dE4BRpKWNnkrLdX+G7PVeStHvrz9QUSQdTyr4z9lel4d/6P53L9/va9X8muBSYJ6kr0lLd3NI7/zH5CUAKDPne4A9tru/o2It6Y9Aybm+EvjKdlfu2bUOuITyc92tXm6HrMaVUvQ/AKblT/hPIH3ws77Fc2qKvJb9JLDD9iM1P1oPLMyPFwKv/d9zaxbb99ieZPscUm7fsn0jsBm4Pu9WVMwAtr8Hdks6Nw9dAXxKwbkmLevMkjQyv9a7Yy461zXq5XY9cFM+i2cWcLB7GWjAbBdxA+aSGsN9Adzb6vk0Mc7LSP/WbQU+zre5pDXuTcCufD+u1XNtUvyzgQ358VTgfeBz4GXSt7S1fI5DHO8FQHvO96vA2NJzDdwP7CT17HoWOLHEXAPPkz63+IP0Tn5RvdySlndW5Pr2CenspkEdN67IDSGECilleSeEEEIDouiHEEKFRNEPIYQKiaIfQggVEkU/hBAqJIp+CCFUSBT9EEKokCj6IYRQIf8ApQNcdxoBaD0AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "def findGCByPos(reads):\n",
    "    ''' Find the GC ratio at each position in the read '''\n",
    "    # Keep track of the number of G/C bases and the total number of bases at each position\n",
    "    gc = [0] * 100\n",
    "    totals = [0] * 100\n",
    "    for read in reads:\n",
    "        for i in range(len(read)):\n",
    "            if read[i] == 'C' or read[i] == 'G':\n",
    "                gc[i] += 1\n",
    "            totals[i] += 1\n",
    "    # Divide G/C counts by total counts to get the average at each position\n",
    "    for i in range(len(gc)):\n",
    "        if totals[i] > 0:\n",
    "            gc[i] /= float(totals[i])\n",
    "    return gc\n",
    "\n",
    "gc = findGCByPos(seqs)\n",
    "plt.plot(range(len(gc)), gc)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({'T': 21836, 'A': 21132, 'C': 28272, 'G': 28742, 'N': 18})"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import collections\n",
    "count = collections.Counter()\n",
    "for seq in seqs:\n",
    "    count.update(seq)\n",
    "count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "def readGenome(filename):\n",
    "    genome = ''\n",
    "    with open(filename, 'r') as f:\n",
    "        for line in f:\n",
    "            # ignore header line with genome information\n",
    "            if not line[0] == '>':\n",
    "                genome += line.rstrip()\n",
    "    return genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome = readGenome('DNA.fa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "def naive(p, t):\n",
    "    occurrences = []\n",
    "    for i in range(len(t) - len(p) + 1):\n",
    "        match = True\n",
    "        for j in range(len(p)):\n",
    "            if t[i+j] != p[j]:\n",
    "                match = False\n",
    "                break\n",
    "        if match:\n",
    "            occurrences.append(i)\n",
    "    return occurrences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0, 5, 9]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t = 'AGCTTAGATAGC'\n",
    "p = 'AG'\n",
    "naive(p, t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "def generateReads(genome, numReads, readLen):\n",
    "    ''' Generate reads from random positions in the given genome. '''\n",
    "    reads = []\n",
    "    for _ in range(numReads):\n",
    "        start = random.randint(0, len(genome)-readLen) - 1\n",
    "        reads.append(genome[start : start+readLen])\n",
    "    return reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100 / 100 reads matched the genome exactly!\n"
     ]
    }
   ],
   "source": [
    "# Generate 100 reads of length 100\n",
    "reads = generateReads(genome, 100, 100)\n",
    "\n",
    "# Count how many reads match the genome exactly\n",
    "numMatched = 0\n",
    "for r in reads:\n",
    "    matches = naive(r, genome)\n",
    "    if len(matches) > 0:\n",
    "        numMatched += 1\n",
    "print('%d / %d reads matched the genome exactly!' % (numMatched, len(reads)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "def readFastq(filename):\n",
    "    sequences = []\n",
    "    with open(filename) as fh:\n",
    "        while True:\n",
    "            fh.readline() # skip name line\n",
    "            seq = fh.readline().rstrip() # read base sequence\n",
    "            fh.readline() # skip placeholder line\n",
    "            fh.readline() # skip base quality line\n",
    "            if len(seq) == 0:\n",
    "                break\n",
    "            sequences.append(seq)\n",
    "    return sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({'T': 30531, 'A': 28426, 'C': 21890, 'G': 19147, 'N': 6})"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import collections\n",
    "phix_reads = readFastq('AlternateDNA.fastq')\n",
    "count = collections.Counter()\n",
    "for read in phix_reads:\n",
    "    count.update(read)\n",
    "count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7 / 1000 reads matched the genome exactly!\n"
     ]
    }
   ],
   "source": [
    "numMatched = 0\n",
    "n = 0\n",
    "for r in phix_reads:\n",
    "    matches = naive(r, genome)\n",
    "    n += 1\n",
    "    if len(matches) > 0:\n",
    "        numMatched += 1\n",
    "print('%d / %d reads matched the genome exactly!' % (numMatched, n))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "459 / 1000 reads matched the genome exactly!\n"
     ]
    }
   ],
   "source": [
    "# Now let's try matching just the first 30 bases of each read\n",
    "numMatched = 0\n",
    "n = 0\n",
    "for r in phix_reads:\n",
    "    r = r[:30]  # just taking the first 30 bases\n",
    "    matches = naive(r, genome)\n",
    "    n += 1\n",
    "    if len(matches) > 0:\n",
    "        numMatched += 1\n",
    "print('%d / %d reads matched the genome exactly!' % (numMatched, n))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "def reverseComplement(s):\n",
    "    complement = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A', 'N': 'N'}\n",
    "    t = ''\n",
    "    for base in s:\n",
    "        t = complement[base] + t\n",
    "    return t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "932 / 1000 reads matched the genome exactly!\n"
     ]
    }
   ],
   "source": [
    "numMatched = 0\n",
    "n = 0\n",
    "for r in phix_reads:\n",
    "    r = r[:30]  # just taking the first 30 bases\n",
    "    matches = naive(r, genome)\n",
    "    matches.extend(naive(reverseComplement(r), genome))\n",
    "    n += 1\n",
    "    if len(matches) > 0:\n",
    "        numMatched += 1\n",
    "print('%d / %d reads matched the genome exactly!' % (numMatched, n))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PREPROCESSING--Z ALGO & BOYER-MOORE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "import string\n",
    "\n",
    "def z_array(s):\n",
    "    \"\"\" Use Z algorithm (Gusfield theorem 1.4.1) to preprocess s \"\"\"\n",
    "    assert len(s) > 1\n",
    "    z = [len(s)] + [0] * (len(s)-1)\n",
    "    # Initial comparison of s[1:] with prefix\n",
    "    for i in range(1, len(s)):\n",
    "        if s[i] == s[i-1]:\n",
    "            z[1] += 1\n",
    "        else:\n",
    "            break\n",
    "    r, l = 0, 0\n",
    "    if z[1] > 0:\n",
    "        r, l = z[1], 1\n",
    "    for k in range(2, len(s)):\n",
    "        assert z[k] == 0\n",
    "        if k > r:\n",
    "            # Case 1\n",
    "            for i in range(k, len(s)):\n",
    "                if s[i] == s[i-k]:\n",
    "                    z[k] += 1\n",
    "                else:\n",
    "                    break\n",
    "            r, l = k + z[k] - 1, k\n",
    "        else:\n",
    "            # Case 2\n",
    "            # Calculate length of beta\n",
    "            nbeta = r - k + 1\n",
    "            zkp = z[k - l]\n",
    "            if nbeta > zkp:\n",
    "                # Case 2a: Zkp wins\n",
    "                z[k] = zkp\n",
    "            else:\n",
    "                # Case 2b: Compare characters just past r\n",
    "                nmatch = 0\n",
    "                for i in range(r+1, len(s)):\n",
    "                    if s[i] == s[i - k]:\n",
    "                        nmatch += 1\n",
    "                    else:\n",
    "                        break\n",
    "                l, r = k, r + nmatch\n",
    "                z[k] = r - k + 1\n",
    "    return z\n",
    "\n",
    "\n",
    "def n_array(s):\n",
    "    \"\"\" Compile the N array (Gusfield theorem 2.2.2) from the Z array \"\"\"\n",
    "    return z_array(s[::-1])[::-1]\n",
    "\n",
    "\n",
    "def big_l_prime_array(p, n):\n",
    "    \"\"\" Compile L' array (Gusfield theorem 2.2.2) using p and N array.\n",
    "        L'[i] = largest index j less than n such that N[j] = |P[i:]| \"\"\"\n",
    "    lp = [0] * len(p)\n",
    "    for j in range(len(p)-1):\n",
    "        i = len(p) - n[j]\n",
    "        if i < len(p):\n",
    "            lp[i] = j + 1\n",
    "    return lp\n",
    "\n",
    "\n",
    "def big_l_array(p, lp):\n",
    "    \"\"\" Compile L array (Gusfield theorem 2.2.2) using p and L' array.\n",
    "        L[i] = largest index j less than n such that N[j] >= |P[i:]| \"\"\"\n",
    "    l = [0] * len(p)\n",
    "    l[1] = lp[1]\n",
    "    for i in range(2, len(p)):\n",
    "        l[i] = max(l[i-1], lp[i])\n",
    "    return l\n",
    "\n",
    "\n",
    "def small_l_prime_array(n):\n",
    "    \"\"\" Compile lp' array (Gusfield theorem 2.2.4) using N array. \"\"\"\n",
    "    small_lp = [0] * len(n)\n",
    "    for i in range(len(n)):\n",
    "        if n[i] == i+1:  # prefix matching a suffix\n",
    "            small_lp[len(n)-i-1] = i+1\n",
    "    for i in range(len(n)-2, -1, -1):  # \"smear\" them out to the left\n",
    "        if small_lp[i] == 0:\n",
    "            small_lp[i] = small_lp[i+1]\n",
    "    return small_lp\n",
    "\n",
    "\n",
    "def good_suffix_table(p):\n",
    "    \"\"\" Return tables needed to apply good suffix rule. \"\"\"\n",
    "    n = n_array(p)\n",
    "    lp = big_l_prime_array(p, n)\n",
    "    return lp, big_l_array(p, lp), small_l_prime_array(n)\n",
    "\n",
    "\n",
    "def good_suffix_mismatch(i, big_l_prime, small_l_prime):\n",
    "    \"\"\" Given a mismatch at offset i, and given L/L' and l' arrays,\n",
    "        return amount to shift as determined by good suffix rule. \"\"\"\n",
    "    length = len(big_l_prime)\n",
    "    assert i < length\n",
    "    if i == length - 1:\n",
    "        return 0\n",
    "    i += 1  # i points to leftmost matching position of P\n",
    "    if big_l_prime[i] > 0:\n",
    "        return length - big_l_prime[i]\n",
    "    return length - small_l_prime[i]\n",
    "\n",
    "\n",
    "def good_suffix_match(small_l_prime):\n",
    "    \"\"\" Given a full match of P to T, return amount to shift as\n",
    "        determined by good suffix rule. \"\"\"\n",
    "    return len(small_l_prime) - small_l_prime[1]\n",
    "\n",
    "\n",
    "def dense_bad_char_tab(p, amap):\n",
    "    \"\"\" Given pattern string and list with ordered alphabet characters, create\n",
    "        and return a dense bad character table.  Table is indexed by offset\n",
    "        then by character. \"\"\"\n",
    "    tab = []\n",
    "    nxt = [0] * len(amap)\n",
    "    for i in range(0, len(p)):\n",
    "        c = p[i]\n",
    "        assert c in amap\n",
    "        tab.append(nxt[:])\n",
    "        nxt[amap[c]] = i+1\n",
    "    return tab\n",
    "\n",
    "\n",
    "class BoyerMoore(object):\n",
    "    \"\"\" Encapsulates pattern and associated Boyer-Moore preprocessing. \"\"\"\n",
    "    \n",
    "    def __init__(self, p, alphabet='ACGT'):\n",
    "        self.p = p\n",
    "        self.alphabet = alphabet\n",
    "        # Create map from alphabet characters to integers\n",
    "        self.amap = {}\n",
    "        for i in range(len(self.alphabet)):\n",
    "            self.amap[self.alphabet[i]] = i\n",
    "        # Make bad character rule table\n",
    "        self.bad_char = dense_bad_char_tab(p, self.amap)\n",
    "        # Create good suffix rule table\n",
    "        _, self.big_l, self.small_l_prime = good_suffix_table(p)\n",
    "    \n",
    "    def bad_character_rule(self, i, c):\n",
    "        \"\"\" Return # skips given by bad character rule at offset i \"\"\"\n",
    "        assert c in self.amap\n",
    "        ci = self.amap[c]\n",
    "        assert i > (self.bad_char[i][ci]-1)\n",
    "        return i - (self.bad_char[i][ci]-1)\n",
    "    \n",
    "    def good_suffix_rule(self, i):\n",
    "        \"\"\" Given a mismatch at offset i, return amount to shift\n",
    "            as determined by (weak) good suffix rule. \"\"\"\n",
    "        length = len(self.big_l)\n",
    "        assert i < length\n",
    "        if i == length - 1:\n",
    "            return 0\n",
    "        i += 1  # i points to leftmost matching position of P\n",
    "        if self.big_l[i] > 0:\n",
    "            return length - self.big_l[i]\n",
    "        return length - self.small_l_prime[i]\n",
    "    \n",
    "    def match_skip(self):\n",
    "        \"\"\" Return amount to shift in case where P matches T \"\"\"\n",
    "        return len(self.small_l_prime) - self.small_l_prime[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# GCTAGCTCTACGAGTCTA\n",
    "p = 'TCAA'\n",
    "p_bm = BoyerMoore(p, alphabet='ACGT')\n",
    "p_bm.bad_character_rule(2, 'T')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# GCTAGCTCTACGAGTCTA\n",
    "# ACTA\n",
    "p = 'ACTA'\n",
    "p_bm = BoyerMoore(p, alphabet='ACGT')\n",
    "p_bm.good_suffix_rule(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ACACGCTCTACGAGTCTA\n",
    "# ACAC\n",
    "p = 'ACAC'\n",
    "p_bm = BoyerMoore(p, alphabet='ACGT')\n",
    "p_bm.match_skip()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "def boyer_moore(p, p_bm, t):\n",
    "    \"\"\" Do Boyer-Moore matching \"\"\"\n",
    "    i = 0\n",
    "    occurrences = []\n",
    "    while i < len(t) - len(p) + 1:\n",
    "        shift = 1\n",
    "        mismatched = False\n",
    "        for j in range(len(p)-1, -1, -1):\n",
    "            if p[j] != t[i+j]:\n",
    "                skip_bc = p_bm.bad_character_rule(j, t[i+j])\n",
    "                skip_gs = p_bm.good_suffix_rule(j)\n",
    "                shift = max(shift, skip_bc, skip_gs)\n",
    "                mismatched = True\n",
    "                break\n",
    "        if not mismatched:\n",
    "            occurrences.append(i)\n",
    "            skip_gs = p_bm.match_skip()\n",
    "            shift = max(shift, skip_gs)\n",
    "        i += shift\n",
    "    return occurrences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "t = 'GCTAGCTCTACGAGTCTA'\n",
    "p = 'TCTA'\n",
    "p_bm = BoyerMoore(p, alphabet='ACGT')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[6, 14]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "boyer_moore(p, p_bm, t)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# K-MER INDEXING"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "import bisect\n",
    "import sys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Index(object):\n",
    "    def __init__(self, t, k):\n",
    "        ''' Create index from all substrings of size 'length' '''\n",
    "        self.k = k  # k-mer length (k)\n",
    "        self.index = []\n",
    "        for i in range(len(t) - k + 1):  # for each k-mer\n",
    "            self.index.append((t[i:i+k], i))  # add (k-mer, offset) pair\n",
    "        self.index.sort()  # alphabetize by k-mer\n",
    "    \n",
    "    def query(self, p):\n",
    "        ''' Return index hits for first k-mer of P '''\n",
    "        kmer = p[:self.k]  # query with first k-mer\n",
    "        i = bisect.bisect_left(self.index, (kmer, -1))  # binary search\n",
    "        hits = []\n",
    "        while i < len(self.index):  # collect matching index entries\n",
    "            if self.index[i][0] != kmer:\n",
    "                break\n",
    "            hits.append(self.index[i][1])\n",
    "            i += 1\n",
    "        return hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "def queryIndex(p, t, index):\n",
    "    k = index.k\n",
    "    offsets = []\n",
    "    for i in index.query(p):\n",
    "        if p[k:] == t[i+k:i+len(p)]:  # verify that rest of P matches\n",
    "            offsets.append(i)\n",
    "    return offsets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "t = 'ACTTGGAGATCTTTGAGGCTAGGTATTCGGGATCGAAGCTCATTTCGGGGATCGATTACGATATGGTGGGTATTCGGGA'#p found at 21st and 68th Start-position\n",
    "p = 'GGTATTCGGGA'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[21, 68]\n"
     ]
    }
   ],
   "source": [
    "index = Index(t, 4)\n",
    "print(queryIndex(p, t, index))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BOYER-MOORE & APPROXIMATE MATCHING"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "import string\n",
    "\n",
    "def z_array(s):\n",
    "    \"\"\" Use Z algorithm (Gusfield theorem 1.4.1) to preprocess s \"\"\"\n",
    "    assert len(s) > 1\n",
    "    z = [len(s)] + [0] * (len(s)-1)\n",
    "    # Initial comparison of s[1:] with prefix\n",
    "    for i in range(1, len(s)):\n",
    "        if s[i] == s[i-1]:\n",
    "            z[1] += 1\n",
    "        else:\n",
    "            break\n",
    "    r, l = 0, 0\n",
    "    if z[1] > 0:\n",
    "        r, l = z[1], 1\n",
    "    for k in range(2, len(s)):\n",
    "        assert z[k] == 0\n",
    "        if k > r:\n",
    "            # Case 1\n",
    "            for i in range(k, len(s)):\n",
    "                if s[i] == s[i-k]:\n",
    "                    z[k] += 1\n",
    "                else:\n",
    "                    break\n",
    "            r, l = k + z[k] - 1, k\n",
    "        else:\n",
    "            # Case 2\n",
    "            # Calculate length of beta\n",
    "            nbeta = r - k + 1\n",
    "            zkp = z[k - l]\n",
    "            if nbeta > zkp:\n",
    "                # Case 2a: Zkp wins\n",
    "                z[k] = zkp\n",
    "            else:\n",
    "                # Case 2b: Compare characters just past r\n",
    "                nmatch = 0\n",
    "                for i in range(r+1, len(s)):\n",
    "                    if s[i] == s[i - k]:\n",
    "                        nmatch += 1\n",
    "                    else:\n",
    "                        break\n",
    "                l, r = k, r + nmatch\n",
    "                z[k] = r - k + 1\n",
    "    return z\n",
    "\n",
    "\n",
    "def n_array(s):\n",
    "    \"\"\" Compile the N array (Gusfield theorem 2.2.2) from the Z array \"\"\"\n",
    "    return z_array(s[::-1])[::-1]\n",
    "\n",
    "\n",
    "def big_l_prime_array(p, n):\n",
    "    \"\"\" Compile L' array (Gusfield theorem 2.2.2) using p and N array.\n",
    "        L'[i] = largest index j less than n such that N[j] = |P[i:]| \"\"\"\n",
    "    lp = [0] * len(p)\n",
    "    for j in range(len(p)-1):\n",
    "        i = len(p) - n[j]\n",
    "        if i < len(p):\n",
    "            lp[i] = j + 1\n",
    "    return lp\n",
    "\n",
    "\n",
    "def big_l_array(p, lp):\n",
    "    \"\"\" Compile L array (Gusfield theorem 2.2.2) using p and L' array.\n",
    "        L[i] = largest index j less than n such that N[j] >= |P[i:]| \"\"\"\n",
    "    l = [0] * len(p)\n",
    "    l[1] = lp[1]\n",
    "    for i in range(2, len(p)):\n",
    "        l[i] = max(l[i-1], lp[i])\n",
    "    return l\n",
    "\n",
    "\n",
    "def small_l_prime_array(n):\n",
    "    \"\"\" Compile lp' array (Gusfield theorem 2.2.4) using N array. \"\"\"\n",
    "    small_lp = [0] * len(n)\n",
    "    for i in range(len(n)):\n",
    "        if n[i] == i+1:  # prefix matching a suffix\n",
    "            small_lp[len(n)-i-1] = i+1\n",
    "    for i in range(len(n)-2, -1, -1):  # \"smear\" them out to the left\n",
    "        if small_lp[i] == 0:\n",
    "            small_lp[i] = small_lp[i+1]\n",
    "    return small_lp\n",
    "\n",
    "\n",
    "def good_suffix_table(p):\n",
    "    \"\"\" Return tables needed to apply good suffix rule. \"\"\"\n",
    "    n = n_array(p)\n",
    "    lp = big_l_prime_array(p, n)\n",
    "    return lp, big_l_array(p, lp), small_l_prime_array(n)\n",
    "\n",
    "\n",
    "def good_suffix_mismatch(i, big_l_prime, small_l_prime):\n",
    "    \"\"\" Given a mismatch at offset i, and given L/L' and l' arrays,\n",
    "        return amount to shift as determined by good suffix rule. \"\"\"\n",
    "    length = len(big_l_prime)\n",
    "    assert i < length\n",
    "    if i == length - 1:\n",
    "        return 0\n",
    "    i += 1  # i points to leftmost matching position of P\n",
    "    if big_l_prime[i] > 0:\n",
    "        return length - big_l_prime[i]\n",
    "    return length - small_l_prime[i]\n",
    "\n",
    "\n",
    "def good_suffix_match(small_l_prime):\n",
    "    \"\"\" Given a full match of P to T, return amount to shift as\n",
    "        determined by good suffix rule. \"\"\"\n",
    "    return len(small_l_prime) - small_l_prime[1]\n",
    "\n",
    "\n",
    "def dense_bad_char_tab(p, amap):\n",
    "    \"\"\" Given pattern string and list with ordered alphabet characters, create\n",
    "        and return a dense bad character table.  Table is indexed by offset\n",
    "        then by character. \"\"\"\n",
    "    tab = []\n",
    "    nxt = [0] * len(amap)\n",
    "    for i in range(0, len(p)):\n",
    "        c = p[i]\n",
    "        assert c in amap\n",
    "        tab.append(nxt[:])\n",
    "        nxt[amap[c]] = i+1\n",
    "    return tab\n",
    "\n",
    "\n",
    "class BoyerMoore(object):\n",
    "    \"\"\" Encapsulates pattern and associated Boyer-Moore preprocessing. \"\"\"\n",
    "    \n",
    "    def __init__(self, p, alphabet='ACGT'):\n",
    "        self.p = p\n",
    "        self.alphabet = alphabet\n",
    "        # Create map from alphabet characters to integers\n",
    "        self.amap = {}\n",
    "        for i in range(len(self.alphabet)):\n",
    "            self.amap[self.alphabet[i]] = i\n",
    "        # Make bad character rule table\n",
    "        self.bad_char = dense_bad_char_tab(p, self.amap)\n",
    "        # Create good suffix rule table\n",
    "        _, self.big_l, self.small_l_prime = good_suffix_table(p)\n",
    "    \n",
    "    def bad_character_rule(self, i, c):\n",
    "        \"\"\" Return # skips given by bad character rule at offset i \"\"\"\n",
    "        assert c in self.amap\n",
    "        ci = self.amap[c]\n",
    "        assert i > (self.bad_char[i][ci]-1)\n",
    "        return i - (self.bad_char[i][ci]-1)\n",
    "    \n",
    "    def good_suffix_rule(self, i):\n",
    "        \"\"\" Given a mismatch at offset i, return amount to shift\n",
    "            as determined by (weak) good suffix rule. \"\"\"\n",
    "        length = len(self.big_l)\n",
    "        assert i < length\n",
    "        if i == length - 1:\n",
    "            return 0\n",
    "        i += 1  # i points to leftmost matching position of P\n",
    "        if self.big_l[i] > 0:\n",
    "            return length - self.big_l[i]\n",
    "        return length - self.small_l_prime[i]\n",
    "    \n",
    "    def match_skip(self):\n",
    "        \"\"\" Return amount to shift in case where P matches T \"\"\"\n",
    "        return len(self.small_l_prime) - self.small_l_prime[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "def boyer_moore(p, p_bm, t):\n",
    "    \"\"\" Do Boyer-Moore matching \"\"\"\n",
    "    i = 0\n",
    "    occurrences = []\n",
    "    while i < len(t) - len(p) + 1:\n",
    "        shift = 1\n",
    "        mismatched = False\n",
    "        for j in range(len(p)-1, -1, -1):\n",
    "            if p[j] != t[i+j]:\n",
    "                skip_bc = p_bm.bad_character_rule(j, t[i+j])\n",
    "                skip_gs = p_bm.good_suffix_rule(j)\n",
    "                shift = max(shift, skip_bc, skip_gs)\n",
    "                mismatched = True\n",
    "                break\n",
    "        if not mismatched:\n",
    "            occurrences.append(i)\n",
    "            skip_gs = p_bm.match_skip()\n",
    "            shift = max(shift, skip_gs)\n",
    "        i += shift\n",
    "    return occurrences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "def approximate_match(p, t, n):\n",
    "    segment_length = int(round(len(p) / (n+1)))\n",
    "    all_matches = set()\n",
    "    for i in range(n+1):\n",
    "        start = i*segment_length\n",
    "        end = min((i+1)*segment_length, len(p))\n",
    "        p_bm = BoyerMoore(p[start:end], alphabet='ACGT')\n",
    "        matches = boyer_moore(p[start:end], p_bm, t)\n",
    "        # Extend matching segments to see if whole p matches\n",
    "        for m in matches:\n",
    "            if m < start or m-start+len(p) > len(t):\n",
    "                continue\n",
    "            mismatches = 0\n",
    "            for j in range(0, start):\n",
    "                if not p[j] == t[m-start+j]:\n",
    "                    mismatches += 1\n",
    "                    if mismatches > n:\n",
    "                        break\n",
    "            for j in range(end, len(p)):\n",
    "                if not p[j] == t[m-start+j]:\n",
    "                    mismatches += 1\n",
    "                    if mismatches > n:\n",
    "                        break\n",
    "            if mismatches <= n:\n",
    "                all_matches.add(m - start)\n",
    "    return list(all_matches)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0, 5]\n"
     ]
    }
   ],
   "source": [
    "p = 'AACTTG'\n",
    "t = 'CACTTAATTTG'\n",
    "print(approximate_match(p, t, 2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "def editDistRecursive(x, y):\n",
    "    # This implementation is very slow\n",
    "    if len(x) == 0:\n",
    "        return len(y)\n",
    "    elif len(y) == 0:\n",
    "        return len(x)\n",
    "    else:\n",
    "        distHor = editDistRecursive(x[:-1], y) + 1\n",
    "        distVer = editDistRecursive(x, y[:-1]) + 1\n",
    "        if x[-1] == y[-1]:\n",
    "            distDiag = editDistRecursive(x[:-1], y[:-1])\n",
    "        else:\n",
    "            distDiag = editDistRecursive(x[:-1], y[:-1]) + 1\n",
    "        return min(distHor, distVer, distDiag)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "def editDistance(x, y):\n",
    "    # Create distance matrix\n",
    "    D = []\n",
    "    for i in range(len(x)+1):\n",
    "        D.append([0]*(len(y)+1))\n",
    "    # Initialize first row and column of matrix\n",
    "    for i in range(len(x)+1):\n",
    "        D[i][0] = i\n",
    "    for i in range(len(y)+1):\n",
    "        D[0][i] = i\n",
    "    # Fill in the rest of the matrix\n",
    "    for i in range(1, len(x)+1):\n",
    "        for j in range(1, len(y)+1):\n",
    "            distHor = D[i][j-1] + 1\n",
    "            distVer = D[i-1][j] + 1\n",
    "            if x[i-1] == y[j-1]:\n",
    "                distDiag = D[i-1][j-1]\n",
    "            else:\n",
    "                distDiag = D[i-1][j-1] + 1\n",
    "            D[i][j] = min(distHor, distVer, distDiag)\n",
    "    # Edit distance is the value in the bottom right corner of the matrix\n",
    "    return D[-1][-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wall time: 12.2 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "x = 'shake spea'\n",
    "y = 'Shakespear'\n",
    "editDistRecursive(x, y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wall time: 0 ns\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "x = 'shake spea'\n",
    "y = 'Shakespear'\n",
    "editDistance(x, y)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PATTERN-MATCHING:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Test dataset results - Occurences and leftmost offset: \n",
      "12646 21 \n",
      "\n",
      "Q1: The 'AGGT' or 'ACCT' occurs 36 times \n",
      "\n",
      "Q2: The 'TTAA' occurs 35 times \n",
      "\n",
      "Q3: The offset of the leftmost occurrence of ACTAAGT is 3708 \n",
      "\n",
      "Q4: The offset of the leftmost occurrence of AGTCGA is 53 \n",
      "\n",
      "Q5: The 'TTCAAGCC' occurs 32 times with mismatch less than 2 \n",
      "\n",
      "Q6: The offset of the leftmost occurrence of AGGAGGTT with mismatch less than 2 is 945 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "class findPattern ():\n",
    "    \"\"\"\n",
    "    This class finds the occurance and position of a given pattern in a given \n",
    "    genomic sequence in a file.\n",
    "    \"\"\"\n",
    "    def __init__(self, pattern, filename):\n",
    "        # initiate parameters\n",
    "        self.pattern = pattern\n",
    "        self.filename = filename\n",
    "        \n",
    "    def readGenome (self):\n",
    "        \"\"\"\n",
    "        read genomic DNA sequence to a string\n",
    "        \"\"\"\n",
    "        genome = \"\"\n",
    "        with open (self.filename, \"r\") as f:\n",
    "            for line in f:\n",
    "                # skip header\n",
    "                if not line[0] == \">\":\n",
    "                    genome += line.rstrip()\n",
    "            f.close()\n",
    "        return genome\n",
    "        \n",
    "    \n",
    "    def reverseComplement (self):\n",
    "        \"\"\"\n",
    "        generate reverse complement sequence for a given dna sequence\n",
    "        \"\"\"\n",
    "        complement = {\"A\": \"T\", \"C\": \"G\", \"T\": \"A\", \"G\": \"C\"}\n",
    "        revComPattern = \"\" # reversed compliment pattern\n",
    "        for nt in self.pattern:\n",
    "            revComPattern = complement[nt] + revComPattern\n",
    "        \n",
    "        return revComPattern\n",
    "        \n",
    "    def match(self, string1, string2, numOfMismatch):\n",
    "        \"\"\"\n",
    "        return True or False for matching results of two strings under the offset\n",
    "        of numOfMismatch\n",
    "        \"\"\"\n",
    "        counter = 0\n",
    "        if len(string1) != len(string2):\n",
    "            return False\n",
    "        # loop over string to compare character\n",
    "        for i in range(len(string1)):\n",
    "            if string1[i] != string2[i]:\n",
    "                counter += 1\n",
    "        if counter > numOfMismatch:\n",
    "            return False\n",
    "        return True\n",
    "        \n",
    "    def patternIdentifier (self, numOfMismatch):\n",
    "        \"\"\"\n",
    "        find positions of a given pattern and the reversed complement pattern \n",
    "        in a given genome\n",
    "        \"\"\"\n",
    "        patternLength = len(self.pattern)\n",
    "        genome = self.readGenome()\n",
    "        revComPattern = self.reverseComplement()\n",
    "        occurances = []\n",
    "        \n",
    "        for i in range (patternLength): # loop over pattern index\n",
    "            # loop over genome patterns\n",
    "            for j in range (i, len(genome), patternLength): \n",
    "                genomeMotif = genome[j: j+patternLength]\n",
    "                # compare genomic motif and patterns\n",
    "                if numOfMismatch == 0:\n",
    "                    if (self.match(genomeMotif, self.pattern, 0) or \\\n",
    "                        self.match(genomeMotif, revComPattern, 0))\\\n",
    "                       and j not in occurances: # avoid duplicate records\n",
    "                        occurances.append(j)\n",
    "                else:\n",
    "                    if self.match(genomeMotif, self.pattern, numOfMismatch)\\\n",
    "                       and j not in occurances: # avoid duplicate records:\n",
    "                        occurances.append(j)\n",
    "        return occurances\n",
    "    \n",
    "class checkQuality ():\n",
    "    \"\"\"\n",
    "    The checkQuality class exams quality of sequencing for each cycle\n",
    "    \"\"\"\n",
    "    def __init__ (self, filename):\n",
    "        self.filename = filename\n",
    "    \n",
    "    def readFastq(self):\n",
    "        \"\"\"\n",
    "        read dna sequence and quality base from a fastq sequencing file to lists\n",
    "        \"\"\"\n",
    "        with open (self.filename, \"r\") as f:\n",
    "             sequences = []\n",
    "             qualities = []\n",
    "             while True:\n",
    "                 f.readline() # skip name line\n",
    "                 seq = f.readline().rstrip() # read sequence line\n",
    "                 f.readline() # skip strand line\n",
    "                 qual = f.readline().rstrip() # read quality line\n",
    "                 if len(seq) == 0: #finish read\n",
    "                     break\n",
    "                 # add seqence and quality information to list\n",
    "                 sequences.append(seq)\n",
    "                 qualities.append(qual)\n",
    "             f.close()\n",
    "        return sequences, qualities\n",
    "        \n",
    "    def phre33ToQ (self, qualString):\n",
    "        \"\"\"\n",
    "        transform quality string to quality base score\n",
    "        \"\"\"\n",
    "        qScore = []\n",
    "        for qual in qualString:\n",
    "            qScore.append(ord(qual) - 33)\n",
    "        return qScore\n",
    "    def findPoorQuality(self):\n",
    "        \"\"\"\n",
    "        find the index of poorest q score in each sequencing\n",
    "        \"\"\"\n",
    "        _, qualities = self.readFastq()\n",
    "        lowestQScoreIndex = []\n",
    "        for qualString in qualities:\n",
    "            qScore = self.phre33ToQ(qualString)\n",
    "            lowestQScoreIndex.append(qScore.index(min(qScore)))\n",
    "        return lowestQScoreIndex\n",
    "        \n",
    "    def countPoorQuality(self):    \n",
    "        \"\"\"\n",
    "        count number of poorest q score in each cycle\n",
    "        \"\"\"\n",
    "        import collections\n",
    "        return collections.Counter(self.findPoorQuality())\n",
    "        \n",
    "    def plotHist(self):\n",
    "        \"\"\"\n",
    "         show the distribution of poorest q score\n",
    "        \"\"\"\n",
    "        import matplotlib.pyplot as plt\n",
    "        data = self.countPoorQuality()\n",
    "        plt.bar(data.keys(), data.values())\n",
    "        plt.show()\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    #Test\n",
    "    filename = \"ConcatDNA.fastq\"\n",
    "    pattern = \"ATTA\"\n",
    "    patterns = findPattern(pattern,filename)\n",
    "    print(\"Test dataset results - Occurences and leftmost offset: \")\n",
    "    print(len(patterns.patternIdentifier(0)), min(patterns.patternIdentifier(0)), \"\\n\")\n",
    "    \n",
    "    #Questions 1-6\n",
    "    filename1 = \"DNA.fa\"\n",
    "    #Q1: How many times does AGGT or its reverse complement (ACCT) occur in the \n",
    "    #lambda virus genome? E.g. if AGGT occurs 10 times and ACCT occurs 12 times, \n",
    "    #you should report 22.\n",
    "    pattern = \"AGGT\"\n",
    "    patterns = findPattern(pattern,filename1)\n",
    "    print(\"Q1: The 'AGGT' or 'ACCT' occurs %d times \\n\" \\\n",
    "           %len(patterns.patternIdentifier(0)))\n",
    "    \n",
    "    #Q2: How many times does TTAA or its reverse complement occur in the lambda \n",
    "    #virus genome? Hint: TTAA and its reverse complement are equal, so remember \n",
    "    #not to double count.    \n",
    "    pattern = \"TTAA\"\n",
    "    patterns = findPattern(pattern,filename1)\n",
    "    print(\"Q2: The 'TTAA' occurs %d times \\n\" \\\n",
    "           %len(patterns.patternIdentifier(0)))\n",
    "    \n",
    "    #Q3: What is the offset of the leftmost occurrence of ACTAAGT or its reverse\n",
    "    #complement in the Lambda virus genome? E.g. if the leftmost occurrence of \n",
    "    #ACTAAGT is at offset 40 (0-based) and the leftmost occurrence of its reverse \n",
    "    #complement ACTTAGT is at offset 29, then report 29.\n",
    "    pattern = \"ACTAAGT\"\n",
    "    patterns = findPattern(pattern,filename1)\n",
    "    print(\"Q3: The offset of the leftmost occurrence of ACTAAGT is %d \\n\" \\\n",
    "           %min(patterns.patternIdentifier(0)))\n",
    "    \n",
    "    #Q4: What is the offset of the leftmost occurrence of AGTCGA or its reverse \n",
    "    #complement in the Lambda virus genome?\n",
    "    pattern = \"AGTCGA\"\n",
    "    patterns = findPattern(pattern,filename1)\n",
    "    print(\"Q4: The offset of the leftmost occurrence of AGTCGA is %d \\n\" \\\n",
    "           %min(patterns.patternIdentifier(0)))\n",
    "           \n",
    "    #Q5: How many times does TTCAAGCC occur in the Lambda virus genome when \n",
    "    #allowing up to 2 mismatches?    \n",
    "    pattern = \"TTCAAGCC\"\n",
    "    patterns = findPattern(pattern,filename1)\n",
    "    print(\"Q5: The 'TTCAAGCC' occurs %d times with mismatch less than 2 \\n\" \\\n",
    "           %len(patterns.patternIdentifier(2)))\n",
    "       \n",
    "    #Q6: What is the offset of the leftmost occurrence of AGGAGGTT in the \n",
    "    #Lambda virus genome when allowing up to 2 mismatches?   \n",
    "    pattern = \"AGGAGGTT\"\n",
    "    patterns = findPattern(pattern,filename1)\n",
    "    print(\"Q6: The offset of the leftmost occurrence of AGGAGGTT with mismatch less than 2 is %d \\n\" \\\n",
    "          %min(patterns.patternIdentifier(2)))      \n",
    "      \n",
    "    #Q7: Report which sequencing cycle has the problem.    \n",
    "    filename2 = \"DNA.fastq\"\n",
    "    qualities = checkQuality(filename2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Q1: The alignments for naive match algorithm is 5330\n",
      "\n",
      "Q2: The characters comparisons for naive match algorithm is 7366\n",
      "\n",
      "Q3: The Alignments for Boyer-Moore algorithm is 119\n",
      "\n",
      "Q4: Within 2 mismatches, the string occurs 0 times\n",
      "\n",
      "Q5: Within 2 mismatches, the total index hits are 1 \n",
      "\n",
      "Q6: Within 2 mismatches, the hits are 0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import bisect\n",
    "\n",
    "class findPatternV2 ():\n",
    "    \"\"\"\n",
    "    This class finds the occurence and position of a given pattern in a given \n",
    "    genomic sequence in a file.\n",
    "    \"\"\"\n",
    "    def __init__(self, pattern, filename = False):\n",
    "        # initiate parameters\n",
    "        self.pattern = pattern\n",
    "        self.filename = filename\n",
    "        \n",
    "    def readGenome (self):\n",
    "        \"\"\"\n",
    "        read genomic DNA sequence to a string\n",
    "        \"\"\"\n",
    "        genome = \"\"\n",
    "        with open (self.filename, \"r\") as f:\n",
    "            for line in f:\n",
    "                # skip header\n",
    "                if not line[0] == \">\":\n",
    "                    genome += line.rstrip()\n",
    "            f.close()\n",
    "        return genome\n",
    "        \n",
    "    def naiveMatch (self, numberOfMismatch, text = False):\n",
    "        \"\"\"\n",
    "        this is naive match to find the index of matched patterns in a genome\n",
    "        and calculate number of total character comparisons and alignments\n",
    "        \"\"\"\n",
    "        if text != False: #for test cases\n",
    "            genome = text\n",
    "        else:\n",
    "            genome = self.readGenome()\n",
    "        pattern = self.pattern\n",
    "        occurences = []\n",
    "        alignments = 0\n",
    "        comparisons = 0\n",
    "        for i in range(len(genome) - len(pattern) + 1):\n",
    "            match = True\n",
    "            counter = 0\n",
    "            for j in range(len(pattern)):\n",
    "                comparisons += 1\n",
    "                if pattern[j] != genome[i+j]:\n",
    "                    counter += 1\n",
    "                if counter > numberOfMismatch:\n",
    "                    match = False\n",
    "                    break\n",
    "            if match:\n",
    "                occurences.append(i)\n",
    "            alignments += 1\n",
    "        return occurences, alignments, comparisons\n",
    "        \n",
    "    def boyerMoore (self, numberOfMismatch, bm, text = False):\n",
    "        \"\"\"\n",
    "        this is naive match to find the index of matched patterns in a genome\n",
    "        and calculate number of total character comparisons and alignments\n",
    "        \"\"\"\n",
    "        i = 0\n",
    "        if text != False: #for test cases\n",
    "            genome = text\n",
    "        else:\n",
    "            genome = self.readGenome()\n",
    "        pattern = self.pattern\n",
    "        occurences = []\n",
    "        alignments = 0\n",
    "        comparisons = 0\n",
    "        while i < len(genome) - len(pattern) + 1:\n",
    "            shift = 1\n",
    "            match = True\n",
    "            for j in range(len(pattern) - 1, -1, -1):\n",
    "                comparisons += 1\n",
    "                if pattern[j] != genome[i+j]:\n",
    "                    badCharacterSkip = bm.bad_character_rule(j, genome[i+j])\n",
    "                    goodSuffixSkip = bm.good_suffix_rule(j)\n",
    "                    shift = max(shift, badCharacterSkip, goodSuffixSkip)\n",
    "                    match = False\n",
    "                    break\n",
    "            if match:\n",
    "                occurences.append(i)\n",
    "                goodSuffixSkip = bm.match_skip()\n",
    "                shift = max(shift, goodSuffixSkip)\n",
    "            i += shift\n",
    "            alignments += 1\n",
    "        return occurences, alignments, comparisons\n",
    "        \n",
    "    def matchedIndex(self, index, k_mer, pattern, isSubseqIndex):\n",
    "        \"\"\"\n",
    "        find number of hits, occurences and time of occurence for a given pattern\n",
    "        using string index in genome\n",
    "        \"\"\"\n",
    "        genome = self.readGenome()\n",
    "        occurences_match = []\n",
    "        hit_index = []\n",
    "        occurence_genome = []\n",
    "        counter = 0\n",
    "        if not isSubseqIndex:\n",
    "            length = len(pattern)-k_mer + 1\n",
    "        else:\n",
    "            length = isSubseqIndex\n",
    "        for i in range(length): # loop over to generate kmers\n",
    "            if not isSubseqIndex:                \n",
    "                pattern_q = pattern[i:i+k_mer]\n",
    "            else:\n",
    "                pattern_q = pattern[i:]\n",
    "            hits = index.query(pattern_q) # query each kmer\n",
    "            for hit in hits:\n",
    "                counter += 1 #count total number of hits\n",
    "                text = genome[hit-i : hit+len(pattern) -i]\n",
    "                if hit-i not in hit_index: #avoid duplicated counts\n",
    "                    hit_index.append(hit-i)\n",
    "                    occurence, _, _ = self.naiveMatch(2,text)\n",
    "                    occurences_match.extend(occurence)\n",
    "                if len(occurence) != 0 and hit-i not in occurence_genome:\n",
    "                    occurence_genome.append(hit-i) \n",
    "        return occurence_genome, len(occurences_match), counter\n",
    "        \n",
    "class Index(object):\n",
    "    def __init__(self, t, k):\n",
    "        ''' Create index from all substrings of size 'length' '''\n",
    "        self.k = k  # k-mer length (k)\n",
    "        self.index = []\n",
    "        for i in range(len(t) - k + 1):  # for each k-mer\n",
    "            self.index.append((t[i:i+k], i))  # add (k-mer, offset) pair\n",
    "        self.index.sort()  # alphabetize by k-mer\n",
    "    \n",
    "    def query(self, p):\n",
    "        ''' Return index hits for first k-mer of P '''\n",
    "        kmer = p[:self.k]  # query with first k-mer\n",
    "        i = bisect.bisect_left(self.index, (kmer, -1))  # binary search\n",
    "        hits = []\n",
    "        while i < len(self.index):  # collect matching index entries\n",
    "            if self.index[i][0] != kmer:\n",
    "                break\n",
    "            hits.append(self.index[i][1])\n",
    "            i += 1\n",
    "        return hits\n",
    "    def genome_index(self):\n",
    "         return self.index       \n",
    "\n",
    "class SubseqIndex(object):\n",
    "    \"\"\" Holds a subsequence index for a text T \"\"\"\n",
    "    \n",
    "    def __init__(self, t, k, ival):\n",
    "        \"\"\" Create index from all subsequences consisting of k characters\n",
    "            spaced ival positions apart.  E.g., SubseqIndex(\"ATAT\", 2, 2)\n",
    "            extracts (\"AA\", 0) and (\"TT\", 1). \"\"\"\n",
    "        self.k = k  # num characters per subsequence extracted\n",
    "        self.ival = ival  # space between them; 1=adjacent, 2=every other, etc\n",
    "        self.index = []\n",
    "        self.span = 1 + ival * (k - 1)\n",
    "        for i in range(len(t) - self.span + 1):  # for each subseq\n",
    "            self.index.append((t[i:i+self.span:ival], i))  # add (subseq, offset)\n",
    "        self.index.sort()  # alphabetize by subseq\n",
    "    \n",
    "    def query(self, p):\n",
    "        \"\"\" Return index hits for first subseq of p \"\"\"\n",
    "        subseq = p[:self.span:self.ival]  # query with first subseq\n",
    "        i = bisect.bisect_left(self.index, (subseq, -1))  # binary search\n",
    "        hits = []\n",
    "        while i < len(self.index):  # collect matching index entries\n",
    "            if self.index[i][0] != subseq:\n",
    "                break\n",
    "            hits.append(self.index[i][1])\n",
    "            i += 1\n",
    "        return hits\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    from bm_preproc import BoyerMoore\n",
    "    #Questions 1-3\n",
    "    filename = (\"DNA.fasta\")\n",
    "    #Q1: How many alignments does the naive exact matching algorithm try when \n",
    "    #matching the string GGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGG \n",
    "    #(derived from human Alu sequences) to the excerpt of human chromosome 1? \n",
    "    #(Don't consider reverse complements.)\n",
    "    pattern = \"GGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGG\"\n",
    "    patterns = findPatternV2(pattern, filename)\n",
    "    print(\"Q1: The alignments for naive match algorithm is %d\\n\"%patterns.naiveMatch(0)[1])\n",
    "    patterns = findPatternV2(\"GGCGCGGTGGCTCACGCCTGTAAT\", filename)    \n",
    "    \n",
    "    #Q2: How many character comparisons does the naive exact matching algorithm \n",
    "    #try when matching the string GGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGG \n",
    "    #(derived from human Alu sequences) to the excerpt of human chromosome 1? \n",
    "    #(Don't consider reverse complements.)\n",
    "    print(\"Q2: The characters comparisons for naive match algorithm is %d\\n\"%patterns.naiveMatch(0)[2])\n",
    "    \n",
    "    #How many alignments does Boyer-Moore try when matching the string \n",
    "    #GGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGG \n",
    "    #(derived from human Alu sequences) to the excerpt of human chromosome 1? \n",
    "    #(Don't consider reverse complements.)\n",
    "    print(\"Q3: The Alignments for Boyer-Moore algorithm is %d\\n\"%patterns.boyerMoore(0, \\\n",
    "          BoyerMoore(pattern, \"ACGT\"))[1])\n",
    "    \n",
    "    #Q4: How many times does the string GGCGCGGTGGCTCACGCCTGTAAT, \n",
    "    #which is derived from a human Alu sequence, occur with up to 2 \n",
    "    #substitutions in the excerpt of human chromosome 1? \n",
    "    #(Don't consider reverse complements here.)\n",
    "    k_mer = 8      \n",
    "    pattern = \"GGCGCGGTGGCTCACGCCTGTAAT\"\n",
    "    genome = patterns.readGenome()\n",
    "    index = Index(genome, k_mer)\n",
    "    occurences, numberOfOccurs, numberOfhits = patterns.matchedIndex(index, \\\n",
    "                                               k_mer, pattern, isSubseqIndex = False)\n",
    "    print(\"Q4: Within 2 mismatches, the string occurs %d times\\n\" %numberOfOccurs)\n",
    "    \n",
    "    #Q5:Using the instructions given in Question 4, how many total index hits \n",
    "    #are there when searching for occurrences of GGCGCGGTGGCTCACGCCTGTAAT with \n",
    "    #up to 2 substitutions in the excerpt of human chromosome 1?\n",
    "    print(\"Q5: Within 2 mismatches, the total index hits are %d \\n\" %numberOfhits)\n",
    "    \n",
    "    #Q6: When using this function, how many total index hits are there when \n",
    "    #searching for GGCGCGGTGGCTCACGCCTGTAAT with up to 2 substitutions in the \n",
    "    #excerpt of human chromosome 1? (Again, don't consider reverse complements.)\n",
    "    pattern = \"GGCGCGGTGGCTCACGCCTGTAAT\"\n",
    "    k_mer = 8\n",
    "    vial = 3\n",
    "    index = SubseqIndex(genome, k_mer, vial)\n",
    "    occurences, numberOfOccurs, numberOfhits = patterns.matchedIndex(index, \\\n",
    "                                               k_mer, pattern, isSubseqIndex = vial)\n",
    "\n",
    "    print(\"Q6: Within 2 mismatches, the hits are %d\\n\" %numberOfhits)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Q1: the edit distance of the best match between pattern and the genome is 4\n",
      "\n",
      "Q2: the edit distance of the best match between pattern and the genome is 5\n",
      "\n",
      "Running time for naive overlap mapping: 2 sec\n",
      "\n",
      "Running time for phrase reads: 0 sec\n",
      "\n",
      "Running time for optimized algorithm: 0 sec\n",
      "\n",
      "Q3: the total edges are 3\n",
      "\n",
      "Q4: the total nodes are 2\n"
     ]
    }
   ],
   "source": [
    "from itertools import permutations\n",
    "\n",
    "class findPatternV3 ():\n",
    "    \"\"\"\n",
    "    This class finds the edit distance of a given pattern in a given genomic \n",
    "    sequence and constructs overlap graphs.\n",
    "    \"\"\"\n",
    "    def __init__(self, filename, pattern = False):\n",
    "        # initiate parameters\n",
    "        self.pattern = pattern\n",
    "        self.filename = filename\n",
    "        \n",
    "    def readGenome (self):\n",
    "        \"\"\"\n",
    "        read genomic DNA sequence to a string\n",
    "        \"\"\"\n",
    "        genome = \"\"\n",
    "        with open (self.filename, \"r\") as f:\n",
    "            for line in f:\n",
    "                # skip header\n",
    "                if not line[0] == \">\":\n",
    "                    genome += line.rstrip()\n",
    "            f.close()\n",
    "        return genome\n",
    "        \n",
    "    def readFastq(self):\n",
    "        \"\"\"\n",
    "        read dna sequence and quality base from a fastq sequencing file to lists\n",
    "        \"\"\"\n",
    "        with open (self.filename, \"r\") as f:\n",
    "             sequences = []\n",
    "             qualities = []\n",
    "             while True:\n",
    "                 f.readline() # skip name line\n",
    "                 seq = f.readline().rstrip() # read sequence line\n",
    "                 f.readline() # skip strand line\n",
    "                 qual = f.readline().rstrip() # read quality line\n",
    "                 if len(seq) == 0: #finish read\n",
    "                     break\n",
    "                 # add seqence and quality information to list\n",
    "                 sequences.append(seq)\n",
    "                 qualities.append(qual)\n",
    "             f.close()\n",
    "        return sequences, qualities\n",
    "        \n",
    "    def editDistance(self):\n",
    "        \"\"\"\n",
    "        Implement dynamic algorithm to calculate edit distance between a given\n",
    "        pattern and a given genome\n",
    "        \"\"\"\n",
    "        pattern = self.pattern\n",
    "        genome = self.readGenome()\n",
    "        pattern_length = len(pattern) + 1\n",
    "        genome_length = len(genome) + 1\n",
    "        #generate matrix\n",
    "        matrix = [[0]*genome_length for i in range(pattern_length)]\n",
    "        # initiate the first column\n",
    "        for i in range(pattern_length):\n",
    "            matrix[i][0] = i\n",
    "        for i in range(1, pattern_length):\n",
    "            for j in range(1, genome_length):\n",
    "                dist_hor = matrix[i][j-1] + 1\n",
    "                dist_vel = matrix[i-1][j] + 1\n",
    "                dist_diag = matrix[i-1][j-1] + 1 if pattern[i-1] != genome[j-1]\\\n",
    "                            else matrix[i-1][j-1]\n",
    "                matrix[i][j] = min(dist_hor, dist_vel, dist_diag)\n",
    "        return min(matrix[-1])                \n",
    "            \n",
    "    def phraseReads(self, k_mer):\n",
    "        \"\"\"\n",
    "        construct the prefix and suffix of a read to a dictornary with read as\n",
    "        key and pre-,suffix as values\n",
    "        \"\"\"\n",
    "        reads, _ = self.readFastq()\n",
    "        reads_dict = {}\n",
    "        for read in reads:\n",
    "            for i in range(len(read) - k_mer + 1):\n",
    "                substring = read[i:i+k_mer]\n",
    "                if substring not in reads_dict:\n",
    "                    reads_dict[substring] = set([read])\n",
    "                else:\n",
    "                    reads_dict[substring].add(read)\n",
    "        return reads_dict\n",
    "    def overlap(self, read1, read2, k_mer):\n",
    "        \"\"\"\n",
    "        find overlaped leftmost offset\n",
    "        \"\"\"\n",
    "        start = 0\n",
    "        while True:\n",
    "            start = read1.find(read2[:k_mer], start)\n",
    "            if start == -1:\n",
    "                return 0 # without overlap\n",
    "            if read2.startswith(read1[start:]):\n",
    "                return len(read1) - start\n",
    "            start += 1\n",
    "            \n",
    "    def overlapGraph(self, k_mer):\n",
    "        \"\"\"\n",
    "        construct graph with key as a read (node) and values as all other \n",
    "        reads overlapped with the previous read (node)\n",
    "        \"\"\"\n",
    "        reads_dict = self.phraseReads(k_mer)\n",
    "        reads, _= self.readFastq()\n",
    "        graph = {}\n",
    "        for read1 in reads:\n",
    "            k_mer_string = read1[len(read1) - k_mer:]\n",
    "            if k_mer_string in reads_dict:\n",
    "                edges = set([])\n",
    "                reads_set = reads_dict[k_mer_string]\n",
    "                for read2 in reads_set:\n",
    "                    if read1 != read2: #skip self comparison\n",
    "                        offset = self.overlap(read1, read2, k_mer)           \n",
    "                        if offset > 0: # skip non-overlapped pairs\n",
    "                            edges.add(read2) #add overlapped reads to be values\n",
    "                            graph[read1] = edges           \n",
    "        return graph\n",
    "                        \n",
    "    def naive_overlap_map(self, k_mer):\n",
    "        \"\"\"\n",
    "        construct graph with key as a pair of reads with overlap and values as\n",
    "        the leftmost offset of the overlap\n",
    "        \"\"\"\n",
    "        graph = {}\n",
    "        reads, _ = self.readFastq()\n",
    "        for read1, read2 in permutations(reads, 2):\n",
    "            #skip non-overlapped reads\n",
    "            if read1[len(read1) - k_mer:] in read2:\n",
    "                offset = self.overlap(read1, read2, k_mer)\n",
    "                    # check if reads[i] overlapped with reads[j]\n",
    "                if offset != 0:\n",
    "                    graph[(read1,read2)] = offset  \n",
    "        return graph\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    \n",
    "    #Q1: What is the edit distance of the best match between pattern \n",
    "    #GCTGATCGATCGTACG and the excerpt of human chromosome 1? \n",
    "    #(Don't consider reverse complements.)\n",
    "    pattern = \"GCTGATCGATCGTACG\"\n",
    "    filename = \"DNA.fasta\"\n",
    "    patterns = findPatternV3 (filename, pattern)\n",
    "    edit_dist = patterns.editDistance()\n",
    "    print(\"Q1: the edit distance of the best match between pattern and the genome is %d\\n\"\\\n",
    "           %edit_dist)\n",
    "           \n",
    "    #Q2: What is the edit distance of the best match between pattern \n",
    "    #GATTTACCAGATTGAG and the excerpt of human chromosome 1? \n",
    "    #(Don't consider reverse complements.)\n",
    "    pattern = \"GATTTACCAGATTGAG\"\n",
    "    filename = \"DNA.fasta\"\n",
    "    patterns = findPatternV3 (filename, pattern)\n",
    "    edit_dist = patterns.editDistance()\n",
    "    print(\"Q2: the edit distance of the best match between pattern and the genome is %d\\n\"\\\n",
    "           %edit_dist)\n",
    "    \n",
    "    #Q3: Picture the overlap graph corresponding to the overlaps just calculated. \n",
    "    # How many edges are in the graph? In other words, how many distinct pairs \n",
    "    # of reads overlap?\n",
    "    #Q4: Picture the overlap graph corresponding to the overlaps computed for \n",
    "    #the previous question. How many nodes in this graph have at least one \n",
    "    #outgoing edge? (In other words, how many reads have a suffix involved in \n",
    "    #an overlap?)\n",
    "    import time\n",
    "    t1 = time.time()\n",
    "    filename = \"ConcatDNA.fastq\"\n",
    "    patterns = findPatternV3 (filename)\n",
    "    k_mer = 30\n",
    "    graph = patterns.naive_overlap_map(k_mer)\n",
    "    t2 = time.time()\n",
    "    print(\"Running time for naive overlap mapping: %d sec\\n\"%(t2 - t1))\n",
    "    \n",
    "    reads = patterns.phraseReads(k_mer)\n",
    "    t3 = time.time()\n",
    "    print(\"Running time for phrase reads: %d sec\\n\"%(t3 - t2))\n",
    "    graph = patterns.overlapGraph(k_mer)\n",
    "    t4 = time.time()\n",
    "    print(\"Running time for optimized algorithm: %d sec\\n\"%(t4 - t2))\n",
    "    numberOfNodes = len(graph)\n",
    "    numberOfEdges = sum([len(edges) for edges in graph.values()])\n",
    "    print(\"Q3: the total edges are %d\\n\"%numberOfEdges)\n",
    "    print(\"Q4: the total nodes are %d\"%numberOfNodes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DNA Sequencing Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Q1: How many records are in the multi-FASTA file: 17 \n",
      "\n",
      "Q2: The length of the longest sequence: 4773 \n",
      " The number of longest sequence: 1 \n",
      "\n",
      "Q3: The length of the shortest sequence: 37 \n",
      " The number of shortest sequence: 1 \n",
      "\n",
      "Q4: The length of longest ORF in frame2: 1479\n",
      "\n",
      "Q5: The start position of longest ORF in frame1: 22\n",
      "\n",
      "Q6: The longest ORF of all frames and sequences: 1479\n",
      "\n",
      "Q7: The length of longest ORF for  >gi|142022655|gb|EQ086233.1|129 marine metagenome JCVI_SCAF_1096627390048 genomic scaffold, whole genome shotgun sequence is: 885 \n",
      "\n",
      "Q8: The most frequently repeats occur: 55 times \n",
      "\n",
      "Q10: The following repeats occured most frequently: \n",
      " ['CGCGGCG']\n",
      "Q9: The number of different 10-base sequences occur max times: 2 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "class dna_tool_sets ():\n",
    "    \"\"\" \n",
    "    This class includes a set of tools for dna sequence analysis:\n",
    "    a. count_records\n",
    "\n",
    "    b. check_length\n",
    "    \n",
    "    c. orf_identifier\n",
    "       \n",
    "    d. repeats_identifier\n",
    "    \n",
    "    parameter: file name including path directory\n",
    "    Note: the file should be in FASTA format\n",
    "    \n",
    "    \"\"\"\n",
    "    def __init__(self, file_name):\n",
    "        \"\"\" this function open input file and transform each record to dictornary\n",
    "            the key and value of the dict are header and dna sequence of the record,\n",
    "            respectively\n",
    "        \"\"\"\n",
    "        self.file_name = file_name # initiate loading file \n",
    "        self.dict = {} # initiate an empty dictornary\n",
    "        f_reader = open (self.file_name)\n",
    "        for line in f_reader:\n",
    "            line = line.strip(\"\\n\") # remove \"\\n\" in line\n",
    "            if \">\" in line: # check if this line is a header\n",
    "                header = line # assign line to header\n",
    "                self.dict[header] = \"\" #add header to dictornary with an initial empty string\n",
    "            else:\n",
    "                self.dict[header] += line # add dna sequence to the header accordingly\n",
    "        f_reader.close()\n",
    "    \n",
    "    def count_records (self):\n",
    "        \"\"\" this function count number of records in the file\"\"\"\n",
    "        number_of_records = len(self.dict) # check records in dictornary\n",
    "        # Q1: How many records are in the multi-FASTA file?\n",
    "        print(\"Q1: How many records are in the multi-FASTA file: %d \\n\"\\\n",
    "               %number_of_records)\n",
    "        #return number_of_records\n",
    "        \n",
    "    def check_length(self):\n",
    "        \"\"\"\n",
    "        this function check length of each record and the according header \n",
    "        of the record\n",
    "        \"\"\"\n",
    "        length_dict = {} # creat a dictornary to record length of each record\n",
    "        for key, value in self.dict.items():\n",
    "            length_dict[key] = len(value)\n",
    "            \n",
    "        lengths = length_dict.values() # count length for each sequence\n",
    "        \n",
    "        max_length = max(lengths) # max length of sequences \n",
    "        min_length = min(lengths) # min length of sequences\n",
    "        # identifier of sequences with max length\n",
    "        record_max_length = [item for item in length_dict if length_dict[item] == max_length]\n",
    "        # identifier of sequences with min length        \n",
    "        record_min_length = [item for item in length_dict if length_dict[item] == min_length]\n",
    "        \n",
    "        # Q2: What is the length of the longest sequence in the file?\n",
    "        print(\"Q2: The length of the longest sequence: %d \\n\"%max_length, \\\n",
    "              \"The number of longest sequence: %d \\n\"%len(record_max_length))\n",
    "        # Q3: What is the length of the shortest sequence in the file?\n",
    "        print(\"Q3: The length of the shortest sequence: %d \\n\"%min_length, \\\n",
    "              \"The number of shortest sequence: %d \\n\"%len(record_min_length))\n",
    "        # uncomment the return code to return the whole length information for \n",
    "        # each record\n",
    "        #return length_dict\n",
    "               \n",
    "    def find_pos(self, dna):\n",
    "        \"\"\" \n",
    "        help function for orf_identifier: \n",
    "        find start position and length for each read frame for a given sequence\n",
    "        dna: sequence, string\n",
    "        \"\"\"\n",
    "        start_code = \"ATG\"\n",
    "        stop_codes = [\"TAA\", \"TAG\", \"TGA\"]\n",
    "\n",
    "        pos_dict = {} #record orfs of each frame to dictornary\n",
    "        \n",
    "        for i in range(3): # walk through 3 different frames\n",
    "            # record strat position and length for each frame to a list\n",
    "            pos = []\n",
    "            # generate frames\n",
    "            if i == 0:\n",
    "                frame = [dna[j:j+3] for j in range(i, len(dna), 3)]\n",
    "            else:\n",
    "                frame = [dna[:i]] + [dna[j:j+3] for j in range(i, len(dna), 3)]\n",
    "            # find all possible start postions and stop positions\n",
    "            start_pos = []\n",
    "            stop_pos = []\n",
    "            try:\n",
    "                index_start_pos = [m for m, y in enumerate(frame) if \\\n",
    "                                  y == start_code]\n",
    "                start_pos += index_start_pos # find possible positions for start code \"ATG\"\n",
    "            except ValueError:\n",
    "                pos.append((-1, 0)) # return -1 as start position if no ATG\n",
    "                continue\n",
    " \n",
    "            for stop_code in stop_codes:\n",
    "                try:\n",
    "                    # find all possible positions of stop codes\n",
    "                    index_stop_code = [n for n, x in enumerate(frame) if \\\n",
    "                                       x == stop_code and n > min(start_pos)]\n",
    "                    stop_pos += index_stop_code\n",
    "                except ValueError:\n",
    "                    continue\n",
    "            if len(stop_pos) == 0: # add -1 as start position when no stop code find\n",
    "                 pos.append((-1, 0))\n",
    "            else:\n",
    "                #find the closest paired code\n",
    "                 while len(start_pos) != 0:\n",
    "                     start = min(start_pos)\n",
    "                     try:\n",
    "                         end = min([stop for stop in stop_pos if stop > start])\n",
    "                     except ValueError:\n",
    "                         break     \n",
    "                 # add start position and length\n",
    "                     s_pos = len(\"\".join(frame[:start])) + 1\n",
    "                     pos.append((s_pos, (end - start + 1)*3))\n",
    "                     start_pos.remove(start) \n",
    "            pos_dict[\"frame%d\"%(i+1)] = pos \n",
    "            \n",
    "        return pos_dict\n",
    "        \n",
    "    def revs_complement(dna):\n",
    "        \"\"\"\n",
    "        help function for orf_identifier:\n",
    "        to transform a sequence to reverse complementary sequence\n",
    "        \"\"\"\n",
    "        pairs = {\"A\": \"T\", \"C\": \"G\", \"G\": \"C\", \"T\": \"A\"} # complementary code\n",
    "        c_dna = [pairs[s] for s in dna] # complementary replace\n",
    "        return \"\".join(c_dna)[::-1] # reverse\n",
    "        \n",
    "    \n",
    "    def orf_identifier (self):\n",
    "        \"\"\"\n",
    "        This function return all the orf information with start posiotion and\n",
    "        length of orf in read frame 1, 2 and 3\n",
    "        the values for frame 1, 2, and are represented as pairs of tuple in a list\n",
    "        for example: {\"header1\": {\"frame1\":[(0, 100)], \"frame2\":[(20, 400)], \"frame3\":[(-1, 0)]},...} \n",
    "        represents for header1:\n",
    "        frame1- start position: 0, length of orf: 100\n",
    "        frame2- start position: 20, length of orf: 400\n",
    "        frame3- No ORF detected\n",
    "        \"\"\"\n",
    "        orf = {}\n",
    "        for header, dna_seq in self.dict.items(): # generate orf for the whole file\n",
    "            pos = self.find_pos(dna_seq)\n",
    "            orf[header] = pos\n",
    "        # find header for question 7\n",
    "        id_key = [key for key in orf if \"gi|142022655|gb|EQ086233.1|129\" in key]\n",
    "        idx = id_key[0]  \n",
    "        # generate list of frames for questions 4 to 7\n",
    "        frame1, frame2, all_frames, id_frames = [], [], [], []\n",
    "        for key, dict_value in orf.items():\n",
    "            frame1 += dict_value[\"frame1\"]\n",
    "            frame2 += dict_value[\"frame2\"]\n",
    "            frames = dict_value[\"frame1\"] + dict_value[\"frame2\"] + dict_value[\"frame3\"]\n",
    "            all_frames += frames\n",
    "            if key == idx:\n",
    "                id_frames = dict_value[\"frame1\"] + dict_value[\"frame2\"] + dict_value[\"frame3\"]\n",
    "            \n",
    "        #Q4: What is the length of the longest ORF appearing in reading\n",
    "        #frame 2 of any of the sequences?\n",
    "        frame2_max_length = max(frame2, key = lambda x: x[1])\n",
    "        print(\"Q4: The length of longest ORF in frame2: %d\\n\"%frame2_max_length[1])\n",
    "        #Q5: What is the starting position of the longest ORF in reading frame 1 \n",
    "        #in any of the sequences? \n",
    "        \n",
    "        frame1_max_length_pos = max(frame1, key = lambda x: x[1])\n",
    "        print(\"Q5: The start position of longest ORF in frame1: %d\\n\"%frame1_max_length_pos[0])\n",
    "        #Q6: What is the length of the longest ORF appearing in any sequence and \n",
    "        #in any forward reading frame?\n",
    "        max_length = max(all_frames, key = lambda x: x[1])\n",
    "        print(\"Q6: The longest ORF of all frames and sequences: %d\\n\"%max_length[1])\n",
    "        #Q7: What is the length of the longest forward ORF that appears in the \n",
    "        #sequence with the identifier gi|142022655|gb|EQ086233.1|129?\n",
    "\n",
    "        max_length_id = max(id_frames, key = lambda x: x[1])\n",
    "        print(\"Q7: The length of longest ORF for \", idx, \"is: %d \\n\" %max_length_id[1])\n",
    "        \n",
    "        # uncomment the return code to return the whole orf information \n",
    "        #including starting position and length of each frame for each sequence\n",
    "        \n",
    "        #return orf\n",
    "            \n",
    "    def find_repeats(self, dna, n):\n",
    "        \"\"\"\n",
    "        This help function for repeats_identifier find and count repeats for \n",
    "        each dna sequence\n",
    "        dna: sequence, string\n",
    "        n: number of repeats, int\n",
    "        \"\"\"\n",
    "        repeats = {}\n",
    "        for i in range(0, len(dna)):\n",
    "            repeat = dna[i:i+n] # generate possible repeats\n",
    "            if len(repeat) == n:\n",
    "                if repeat not in repeats:\n",
    "                    repeats [repeat] = 1 # initiate record\n",
    "                else:\n",
    "                    # count repeated repeats\n",
    "                    repeats[repeat] = repeats.get(repeat) + 1\n",
    "        return repeats\n",
    "    \n",
    "    def repeats_identifier(self, n):\n",
    "        \"\"\"\n",
    "        This function generates repeats with counts for each record \n",
    "        (repeats_set) and for the whole file (combined_repeats)\n",
    "        n: number of repeats, int\n",
    "        \"\"\"\n",
    "        # record the repeats with counts for each record \n",
    "        repeats_set = {}\n",
    "        for header, dna_seq in self.dict.items():\n",
    "            repeats = self.find_repeats(dna_seq, n)\n",
    "            repeats_set[header] = repeats \n",
    "        # record the repeats with counts for the whole file\n",
    "        combined_repeats = {}\n",
    "        for dict_value in repeats_set.values():\n",
    "            for key in dict_value:\n",
    "                if key not in combined_repeats:\n",
    "                    combined_repeats[key] = dict_value[key]\n",
    "                else:\n",
    "                    combined_repeats[key] = combined_repeats.get(key) \\\n",
    "                                            + dict_value[key]\n",
    "        # Q8:Find the most frequently occurring repeat of length 7 in all \n",
    "        # sequences. How many times does it occur in all?\n",
    "        if n == 7:\n",
    "            most_freq_7 = max (combined_repeats.values())\n",
    "            print(\"Q8: The most frequently repeats occur: %d times \\n\"%most_freq_7)\n",
    "        # Q10:Which one of the following repeats of length 7 has a maximum \n",
    "        #number of occurrences?\n",
    "            most_freq_7_seq = [key for key in combined_repeats if \\\n",
    "                       combined_repeats[key] == max(combined_repeats.values())]\n",
    "            print(\"Q10: The following repeats occured most frequently: \\n\", most_freq_7_seq)\n",
    "        # Q9:Find all repeats of length 10 in the input file. Let's use Max to \n",
    "        #specify the number of copies of the most frequent repeat of length 10. \n",
    "        #How many different 10-base sequences occur Max times?\n",
    "        if n == 10:\n",
    "            # answer question 9\n",
    "            count_most_freq_10 = len([value for value in combined_repeats.values()\\\n",
    "                             if value == max(combined_repeats.values())])\n",
    "        \n",
    "        \n",
    "            print(\"Q9: The number of different 10-base sequences occur max times: %d \\n\"\\\n",
    "                  %count_most_freq_10)\n",
    "        \n",
    "        # uncomment the return code to return the repeats with counts for each record \n",
    "        # (repeats_set) and for the whole file (combined_repeats)\n",
    "        #return repeats_set, combined_repeats\n",
    "        \n",
    "if __name__ == \"__main__\":\n",
    "    \n",
    "    file_name = \"dna3.fasta\"\n",
    "    dna_tools = dna_tool_sets (file_name)\n",
    "    # Q1: How many records are in the multi-FASTA file?\n",
    "    dna_tools.count_records()\n",
    "    \n",
    "    # Q2: What is the length of the longest sequence in the file?\n",
    "    # Q3: What is the length of the shortest sequence in the file?\n",
    "    dna_tools.check_length()\n",
    "    \n",
    "    #Q4: What is the length of the longest ORF appearing in reading\n",
    "    #frame 2 of any of the sequences?\n",
    "    #Q5: What is the starting position of the longest ORF in reading frame 1 \n",
    "    #in any of the sequences? \n",
    "    #Q6: What is the length of the longest ORF appearing in any sequence and \n",
    "    #in any forward reading frame?\n",
    "    #Q7: What is the length of the longest forward ORF that appears in the \n",
    "    #sequence with the identifier gi|142022655|gb|EQ086233.1|129?\n",
    "    dna_tools.orf_identifier()\n",
    "    \n",
    "    # Q8:Find the most frequently occurring repeat of length 7 in all \n",
    "    # sequences. How many times does it occur in all?\n",
    "    # Q10:Which one of the following repeats of length 7 has a maximum \n",
    "    #number of occurrences?\n",
    "    dna_tools.repeats_identifier(7)\n",
    "    \n",
    "    # Q9:Find all repeats of length 10 in the input file. Let's use Max to \n",
    "    #specify the number of copies of the most frequent repeat of length 10. \n",
    "    #How many different 10-base sequences occur Max times?\n",
    "    dna_tools.repeats_identifier(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SEQUENCES ALIGNMENT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [],
   "source": [
    "from utils import cost, addgap, print_table, deletion, substitution, insertion\n",
    "from alignment import align_nw_2by2, align_star_multiple\n",
    "from numpy.random import choice, randint\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.random.seed(42)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [],
   "source": [
    "alphabet = [\"A\", \"C\", \"G\", \"T\"] # nucleotides to consider for DNA sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The cost of (G, T) equals -1.\n",
      "The cost of (A, G) equals -1.\n",
      "The cost of (G, T) equals -1.\n",
      "The cost of (A, A) equals 2.\n",
      "The cost of (G, C) equals -1.\n"
     ]
    }
   ],
   "source": [
    "for i in range(5):\n",
    "    l1 = choice(alphabet)\n",
    "    l2 = choice(alphabet)\n",
    "    print(\"The cost of ({}, {}) equals {}.\".format(l1, l2, cost(l1, l2)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The initial sequence equals: GGGGTAT and the gaplist equals: [7 7 2 5 4].\n",
      "gap introduced in 7th position -> new sequence equals GGGGTAT_.\n",
      "gap introduced in 7th position -> new sequence equals GGGGTAT__.\n",
      "gap introduced in 2th position -> new sequence equals GG_GGTAT__.\n",
      "gap introduced in 5th position -> new sequence equals GG_GG_TAT__.\n",
      "gap introduced in 4th position -> new sequence equals GG_G_G_TAT__.\n",
      "The final sequence equals:   GG_G_G_TAT__.\n"
     ]
    }
   ],
   "source": [
    "seq = ''.join(choice(alphabet, 7))\n",
    "gaplist = randint(7+5, size=5)\n",
    "print(\"The initial sequence equals: {} and the gaplist equals: {}.\".format(seq, gaplist))\n",
    "newseq = addgap(seq, gaplist, verbose=1)\n",
    "print(\"The final sequence equals:   {}.\".format(newseq))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The initial sequences are:\n",
      "ATGAGAT AGGAGAGT\n",
      "The aligned sequences are:\n",
      "['ATGAGA_T', 'AGGAGAGT']\n",
      "The positions of the introduced gaps are:\n",
      "[]\n",
      "The score matrix equals:\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAWQAAADuCAYAAAAOR30qAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvnQurowAAIABJREFUeJzt3XtQVGeax/Ffc1F3BLkYe5RGAgiB5tAXbQVhXAxOqSE6lQHMBtbbRqitNeNWrC1rZ8qqhMkku4OXlDAu/5igVSZmsjWuiopircYLEhMSBLwSLQWWBgZ15CIQTbf97B9ZesEEOdDNed9Mnk8VVUJ3nf56+vRj24dzjo6IwBhjTDwf0QGMMca+xQOZMcYkwQOZMcYkwQOZMcYkwQOZMcYkwQOZMcYkwQOZMcYkwQOZMcYkwQOZMcYk4TeaO/v6+pLL5RqvljHx8fEBN42Mm9STsYub1JGx6f8QEY34BnhUA9nlckG2Q611Oh03qcBN6snYxU3qyNgEADqdTqfmfvyRBWOMSYIHMmOMSYIHMmOMSYIHMmOMSYIHMmOMSYIHMmOMSYIHMmOMSULzgVxRUYG4uDjExMSgsLBQ64f/XmVlZTCbzbBarZg7dy7Onz8vOgkAcObMGVitViiKgoULF4rOQWdnJzIzM2E2m5GUlIQrV64I6WhoaEBKSgomTpyI7du3u3/e0tKC9PR0JCQkQFEUFBcXS9EFAJGRkTCZTO5tTIamHTt2QFEUJCYmIjc3Fw8fPtSkad++fTCbzTCZTEhNTUV9fb37tnXr1kGv1yMxMVGTFjVNAPD48WPMnj0by5cvH98QIlL99e3dx87pdFJ0dDTdunWLHj16RGazma5everRMj1tIiJ68OABuVwuIiKqr6+nuLg44U2dnZ1kNBqpubmZiIg6OjqEN23atIl++9vfEhHR9evXadGiRUKaOjo6qLq6mjZv3kzbtm1z/7ytrY1qamqIiKinp4diY2PHtH15u4uI6Nlnn6W7d++Oabnj0WS32ykyMpL6+/uJiOjll1+mPXv2aNJUVVVF9+/fJyKiY8eOUVJSkvu2s2fPUk1NDSmKMqZlj0cTEdG7775Lubm5tGzZMk+6Rpyxmr5Drq6uRkxMDKKjozFhwgTk5OSgrKxMy4TvFRAQgIEDafr6+qDyoJpx9dFHHyErKwsREREAAL1eL7gIuHbtGhYtWgQAiI+PR1NTEzo6OjTv0Ov1mDdvHvz9/Yf8fMaMGZgzZw4AIDAwEEajEa2trcK7RHpak9PpxNdffw2n04n+/n6EhYVp0pSamoqQkBAAwPz582G32923paWlITQ0VJMOtU12ux3l5eXIz88f9w5NB3Jraytmzpzp/j48PFzTF8zTHDx4EPHx8Vi2bBl2794tOgc3btxAZ2cnnn/+edhsNuzdu1d0EiwWCw4cOADg239cm5ubh2y4MmlqakJtbS2Sk5NFpwD49pDeJUuWwGazYdeuXaJzYDAYsGnTJkRERGDGjBkICgrCkiVLNO8oLS1FRkaG5o/7NE82bdy4EVu3boWPz/iPS96p938yMzPR0NCAQ4cO4Y033hCdA6fTiZqaGpSXl+PEiRN4++23cePGDaFNv/nNb9DV1QWr1YqdO3di9uzZ8PX1Fdr0fXp7e5GdnY2ioiJMmTJFdA4A4Pz587h48SKOHz+OkpISnDt3TmhPZ2cnysrK0NjYiLa2NvT19eHDDz/UtOH06dMoLS3Fli1bNH3cp3my6ejRo9Dr9bDZbJo8vqYD2WAwoKWlxf293W6HwWDQMsGtpKQEVqsVVqsVbW1t7p+npaXh9u3buHfvntCmsLAwLF26FJMnT8YzzzyDtLS07+xo0Lqpt7cXe/bsQV1dHfbu3Yu7d+8iOjpa847Bz9eTHA4HsrOzsXLlSmRlZUnTNbCd6/V6ZGZmorq6WmjTyZMnERUVhWnTpsHf3x9ZWVn49NNPNWu6dOkS8vPzUVZWhqlTp47b43raVFVVhcOHDyMyMhI5OTn45JNPsGrVqvGLUvNBM3lpp57D4aCoqCi6ffu2e6felStXPFqmp01ERDdv3nTv1KupqaGwsDD396Karl27RosWLSKHw0F9fX2kKApdvnxZaFNnZyc9evSIiIh27dpFq1ev9mh5njYVFBQM2VHlcrlo9erV9Prrr0vV1dvbSz09Pe4/p6Sk0PHjx4U2ffbZZ5SQkEB9fX3kcrlozZo19Ic//EGTpubmZpo1axZVVVV97+2NjY2a79QbqYmI6PTp0+O+U0/TgUxEVF5eTrGxsRQdHU3vvPOOx8vzRlNhYSElJCSQxWKh+fPnU2VlpfAmIqKtW7eS0WgkRVFox44dwps+/fRTio2Npeeee44yMzPde6W1bmpvbyeDwUCBgYEUFBREBoOBuru7qbKykgCQyWQii8VCFouFysvLhXfdunWLzGYzmc1mSkhIGNN27+0mIqI333yT4uLiSFEUWrVqFT18+FCTpry8PAoODnY/RzabzX1bTk4OTZ8+nfz8/MhgMND7778vvGmAFgNZR6M4d6hOp6PR3F8LMp7/lJvUkbEJkLOLm9SRsQlwd43461u8U48xxiTBA5kxxiTBA5kxxiTBA5kxxiTBA5kxxiTBA5kxxiTBA5kxxiTBA5kxxiTBA5kxxiThN5o7+/j4SHGu4MEmTZrETSpwk3oydnGTOjI2AVDdNKqB7HK5pDssUcZDJblJHRmbADm7uEkdGZsA9QOZP7JgjDFJ8EBmjDFJ8EBmjDFJ8EBmjDFJ8EBmjDFJ8EBmjDFJ8EBmjDFJaD6Qy8rKYDabYbVaMXfuXJw/f17rhCFaWlqQnp6OhIQEKIqC4uJioT0DGhoakJKSgokTJ2L79u2ic4b44osv4Ofnh/3794tOAQBs27bNffXgxMRE+Pr64v79+5p3PO05q6ioQFxcHGJiYlBYWCi86eHDh0hKSoLFYoGiKCgoKBDeBABdXV1YsWIF4uPjYTQaceHCBeFNxcXFSExMhKIoKCoqGt8QNRfeG/iCFy6U+eDBA/cVnevr6ykuLs6j5Xna1NbWRjU1NURE1NPTQ7GxsXT16lWhTUREHR0dVF1dTZs3bx5ytWCRTURETqeT0tPTKSMjg/70pz9J0TTY4cOHKT093aNljLVruOfM6XRSdHQ03bp1y3219dFuY95ucrlc9ODBAyIi+uabbygpKYkuXLggtImIaM2aNfTee+8REdGjR4+os7NTaNPly5dJURTq6+sjh8NBP//5z+nmzZujXj5UXuRU83fIAQEB7qNW+vr6hB/mOGPGDMyZMwcAEBgYCKPRiNbWVqFNAKDX6zFv3jz4+/uLThli586dyM7Ohl6vF53yvf74xz8iNzdXyGMP95xVV1cjJiYG0dHRmDBhAnJyclBWVia0SafTISAgAADgcDjgcDg0ey0O19Td3Y1z584hLy8PADBhwgQEBwcLbbp+/TqSk5Pxk5/8BH5+fli4cCEOHDgwbh1CPkM+ePAg4uPjsWzZMuzevVtEwvdqampCbW0tkpOTRadIqbW1FQcPHsT69etFp3yv/v5+VFRUIDs7W3TKEK2trZg5c6b7+/DwcCn+0X/8+DGsViv0ej0WL14sfLtvbGzEtGnT8Oqrr2L27NnIz89HX1+f0KbExERUVlbiL3/5C/r7+3Hs2DG0tLSM2+MJGciZmZloaGjAoUOH8MYbb4hI+I7e3l5kZ2ejqKgIU6ZMEZ0jpY0bN2LLli3w8ZFzX/CRI0fws5/9DKGhoaJTfhB8fX1RV1cHu92O6upqXLlyRWiP0+nExYsXsX79etTW1mLy5Mmaft7+fYxGI379619jyZIleOGFF2C1WuHr6ztuj6fJK6ukpMS906Wtrc3987S0NNy+fRv37t3TImNYDocD2dnZWLlyJbKysoR1DLeeRBrc9OWXXyInJweRkZHYv38/XnvtNRw6dEh418C6+vjjjzX/uELNc2YwGIa8q7Lb7TAYDEKbBgsODkZ6ejoqKiqENoWHhyM8PNz9Tn3FihW4ePGi0CYAyMvLQ01NDc6dO4eQkBA899xz49ak+U69mzdvunfq1dTUUFhYmPv7sfC0yeVy0erVq+n111/3aDmDeWM9DSgoKJBqp96AtWvXSrVTr6uri0JCQqi3t9fjZXna9eRz5nA4KCoqim7fvu3eqXflyhWhTXfu3HHvMOvv76cFCxbQkSNHhDYRES1YsIAaGhrct2/atEl4U0dHBxERNTc3U1xc3Kh3NA7qGnnGqrkTeXEgFxYWUkJCAlksFpo/fz5VVlZ6tDxPmyorKwkAmUwmslgsZLFYqLy8XGgTEVF7ezsZDAYKDAykoKAgMhgM1N3dLbRpMNkG8p49e+iVV17xyrLG2vW056y8vJxiY2MpOjqa3nnnHeFN9fX1ZLVayWQykaIo9NZbbwlvIiKqra0lm81GJpOJXnrpJbp//77wpgULFpDRaCSz2UwnT54c0/LVDmQdjeLcoTqdjkZzfy3IeP5TblJHxiZAzi5uUkfGJsDdNeKvsci5d4Yxxn6EeCAzxpgkeCAzxpgkeCAzxpgkeCAzxpgkeCAzxpgkeCAzxpgkeCAzxpgkeCAzxpgk/EZzZx8fH+HnL37SpEmTuEkFblJPxi5uUkfGJgCqm0Y1kF0ul3SHJcp4qCQ3qSNjEyBnFzepI2MToH4g80cWjDEmCR7IjDEmCR7IjDEmCR7IjDEmCR7IjDEmCR7IjDEmCR7IjDEmCR7IjDEmCSED+cyZM7BarVAUBQsXLhSRMERDQwNSUlIwceJEbN++XXQOAGDdunXQ6/VITEwUneK2b98+mM1mmEwmpKamor6+XnQSgG+3p6CgIPcl3X/3u9+JTkJ3dzd+8YtfwGKxQFEU7NmzR0jH07ZtUdvYSK+3x48fY/bs2Vi+fPmPrmlUR+p5Q1dXF1577TVUVFQgIiICd+7c0TrhO0JDQ/GHP/wBhw4dEp3i9g//8A/YsGED1qxZIzrFLSoqCmfPnkVISAiOHz+Of/zHf8Tnn38uOgsA8Ld/+7c4evSo6Ay3kpISJCQk4MiRI7h79y7i4uKwcuVKTJgwQdOOp23boraxkV5vxcXFMBqN6Onp+dE1af4O+aOPPkJWVhYiIiIAAHq9XuuE79Dr9Zg3bx78/f1Fp7ilpaUhNDRUdMYQqampCAkJAQDMnz8fdrtdcJG8dDodHjx4ACJCb28vQkND4een+fufp27boraxpzXZ7XaUl5cjPz//R9mk+UC+ceMGOjs78fzzz8Nms2Hv3r1aJzAvKC0tRUZGhugMtwsXLsBisSAjIwNXr14VnYMNGzbg+vXrCAsLg8lkQnFxMXx8eJfNSDZu3IitW7dKta60bNL8b+10OlFTU4Py8nKcOHECb7/9Nm7cuKF1BvPA6dOnUVpaii1btohOAQDMmTMHzc3NqK+vxz//8z/jl7/8pegknDhxAlarFW1tbairq8OGDRs0/S/4D9HRo0eh1+ths9lEp7hp3aTJQC4pKXHvcAkLC8PSpUsxefJkPPPMM0hLSxOyc2hwU1tbm+aP/0Px5Hq6dOkS8vPzUVZWhqlTp0rR1dvbi4CAAADAiy++CIfDgXv37gltKikpQVZWFnQ6HWJiYhAVFYWGhgbNO2TZttU0VVVV4fDhw4iMjEROTg4++eQTrFq16kfVBCJS/fXt3T1z7do1WrRoETkcDurr6yNFUejy5ctjXp43mgYUFBTQtm3bPF6Ot5oaGxtJURSvLMsbTc3NzTRr1iyqqqryQpH31lN7ezu5XC4iIvr8889p5syZ7u9Fdf3TP/0TFRQUEBHRn//8ZwoLC6O7d+8Kaxpu2/ZkGxuvJiKi06dP07Jly/4qmojcXSPPWDV3Ii8OZCKirVu3ktFoJEVRaMeOHR4tyxtN7e3tZDAYKDAwkIKCgshgMFB3d7fQppycHJo+fTr5+fmRwWCg999/36PleaMpLy+PgoODyWKxkMViIZvNJryJiGjnzp2UkJBAZrOZkpOTPf4Hwxtdra2ttHjxYkpMTCRFUeiDDz4Q0vS0bdvTbWw8mgZoPZDHs2lQ14gzVkejOJmzTqej0dxfCzKekJqb1JGxCZCzi5vUkbEJcHeNeJZ6eXZlMsbYjxwPZMYYkwQPZMYYkwQPZMYYkwQPZMYYkwQPZMYYkwQPZMYYkwQPZMYYk8Sozgfo4+MDnW7E323W1KRJk7hJBW5ST8YublJHxiYAqptGNZBdLpd0R8HIeGQON6kjYxMgZxc3qSNjE6B+IPNHFowxJgkeyIwxJgkeyIwxJgkeyIwxJgkeyIwxJgkeyIwxJgkeyIwxJgkeyIwxJgnNB3JnZycyMzNhNpuRlJSEK1euaJ0wrC+++AJ+fn7Yv3+/6BSsW7cOer0eiYmJolPcHj58iKSkJFgsFiiKgoKCAtFJAICKigrExcUhJiYGhYWFonMAAGVlZTCbzbBarZg7dy7Onz8vOgkAcObMGVitViiKgoULF4rOkWYeNDQ0ICUlBRMnTsT27dvdP29paUF6ejoSEhKgKAqKi4vHN0TNhfcGvuCFiz9u2rSJfvvb3xIR0fXr12nRokUeLc8bTURETqeT0tPTKSMjg/70pz8Jbzp79izV1NRIddVpl8tFDx48ICKib775hpKSkujChQtCm5xOJ0VHR9OtW7fo0aNHZDab6erVqx4t0xtdDx48cF/5ur6+nuLi4oQ3dXZ2ktFopObmZiIi6ujoEN4kyzzo6Oig6upq2rx585CrTre1tVFNTQ0REfX09FBsbOyYti+ovMip5u+Qr127hkWLFgEA4uPj0dTUhI6ODq0zvmPnzp3Izs6GXq8XnQIASEtLQ2hoqOiMIXQ6HQICAgAADocDDodD+HkDqqurERMTg+joaEyYMAE5OTkoKysT2gQAAQEB7nXT19cnfD0BwEcffYSsrCxEREQAgBTbuizzQK/XY968efD39x/y8xkzZmDOnDkAgMDAQBiNRrS2to5bh+YD2WKx4MCBAwC+fTE1NzfDbrdrnTFEa2srDh48iPXr1wvt+CF4/PgxrFYr9Ho9Fi9ejOTkZKE9ra2tmDlzpvv78PDwcX3BjMbBgwcRHx+PZcuWYffu3aJzcOPGDXR2duL555+HzWbD3r17RSdJOQ+G09TUhNra2nHd5jUfyL/5zW/Q1dUFq9WKnTt3Yvbs2fD19dU6Y4iNGzdiy5Yt8PHhfZwj8fX1RV1dHex2O6qrq6XaByCbzMxMNDQ04NChQ3jjjTdE58DpdKKmpgbl5eU4ceIE3n77bdy4cUNok4zz4Pv09vYiOzsbRUVFmDJlyrg9jiYTqKSkBFarFVarFb29vdizZw/q6uqwd+9e3L17F9HR0VpkDNv05ZdfIicnB5GRkdi/fz9ee+01HDp0SPOmH5Lg4GCkp6ejoqJCaIfBYEBLS4v7e7vdDoPBIKRl8DbV1tbm/nlaWhpu376Ne/fuCW0KCwvD0qVLMXnyZDzzzDNIS0tDfX290CaR82C45+tJDocD2dnZWLlyJbKyssY3Ss0HzeTFnXqdnZ306NEjIiLatWsXrV692qPleaNpsLVr10qxU4+IqLGxUaqdenfu3KHOzk4iIurv76cFCxbQkSNHhDY5HA6Kioqi27dvu3fqXblyxaNleqPr5s2b7p16NTU1FBYW5v5eVNO1a9do0aJF5HA4qK+vjxRFocuXLwttkm0eFBQUDNmp53K5aPXq1fT66697o2vEGTuq8yF7w/Xr17F27VrodDooioLS0lKtE34QcnNzcebMGdy7dw/h4eF46623kJeXJ7Spvb0da9euxePHj+FyufB3f/d3WL58udAmPz8//Md//AeWLl2Kx48fY926dVAURWgTAPzXf/0X9u7dC39/f/zN3/wN/vM//1P4jj2j0YgXXngBZrMZPj4+yM/PF/5rlbLMgz//+c+YO3cuenp64OPjg6KiIly7dg2XLl3CBx98AJPJBKvVCgD493//d7z44ovj0qGjUZzMWafT0WjurwUZT0jNTerI2ATI2cVN6sjYBLi7RvwXmfdiMcaYJHggM8aYJHggM8aYJHggM8aYJHggM8aYJHggM8aYJHggM8aYJHggM8aYJHggM8aYJEZ16LSPj4/wwz+fNGnSJG5SgZvUk7GLm9SRsQmA6qZRDWSXyyXdYYkyHirJTerI2ATI2cVN6sjYBKgfyPyRBWOMSYIHMmOMSYIHMmOMSYIHMmOMSYIHMmOMSYIHMmOMSYIHMmOMSUKTgdzQ0ICUlBRMnDgR27dvd/+8paUF6enpSEhIgKIoKC4u1iLnO7Zt2+a++mxiYiJ8fX1x//59IS0D9u3bB7PZDJPJhNTUVCFXB35SRUUF4uLiEBMTg8LCQtE5bpGRke5rns2dO1d0jjTb9ZOGex3K4IsvvoCfnx/2798vOkXsPFBzJdSBL4zxiq4dHR1UXV1NmzdvHnJF17a2NqqpqSEiop6eHoqNjaWrV6+OatljbRrO4cOHKT093aNleKOpqqqK7t+/T0REx44do6SkJKFNTqeToqOj6datW+6rO4/2ufJ204Bnn32W7t6965VlEXne5Y3t2ttNRMO/DkU2EX27baWnp1NGRoY0V3wf4I15QKT+qtOavEPW6/WYN28e/P39h/x8xowZmDNnDgAgMDAQRqMRra2tWiQN649//CNyc3OFNgBAamoqQkJCAADz58+H3W4X2lNdXY2YmBhER0djwoQJyMnJQVlZmdAmWcm4XQPDvw5F27lzJ7Kzs6HX60WnfIfW80Caz5CbmppQW1uL5ORkYQ39/f2oqKhAdna2sIbvU1paioyMDKENra2tmDlzpvv78PBwKYYM8O1hqUuWLIHNZsOuXbtE5wwhw3Yts9bWVhw8eBDr168XnfIdIubBqM5lMV56e3uRnZ2NoqIiTJkyRVjHkSNH8LOf/QyhoaHCGp50+vRplJaW4vz586JTpHX+/HkYDAbcuXMHixcvRnx8PNLS0kRnSbNdy2zjxo3YsmULfHykeW/oJmIejNtaKCkpcX8w3tbWNuz9HA4HsrOzsXLlSmRlZY1Xjqq+jz/+WOjHFU82Xbp0Cfn5+SgrK8PUqVOFdQGAwWBAS0uL+3u73Q6DwSCw6P8NdOj1emRmZqK6ulpwkbjt+klqX4daGtz05ZdfIicnB5GRkdi/fz9ee+01HDp0SGiT0Hmg5oNm8nCn3oCCgoIhOxNcLhetXr2aXn/99TEv09OmAV1dXRQSEkK9vb0eL8sbTc3NzTRr1iyqqqryeFlEnjc5HA6Kioqi27dvu3fqXblyRWgTEVFvby/19PS4/5ySkkLHjx8X2uWN7fpJ3trOib77OhwrbzYREa1du1aanXrenAdE6nfqaTKQ29vbyWAwUGBgIAUFBZHBYKDu7m6qrKwkAGQymchisZDFYqHy8vKx/EU9tmfPHnrllVe8sixvNOXl5VFwcLB7vdhsNuFN5eXlFBsbS9HR0fTOO+94vDxvNN26dYvMZjOZzWZKSEiQossb27W3m4iGfx2KbBpMpoHszXlApH4g62gU5w7V6XQ0mvtrQcbzn3KTOjI2AXJ2cZM6MjYB7q4RT4os3yfpjDH2I8UDmTHGJMEDmTHGJMEDmTHGJMEDmTHGJMEDmTHGJMEDmTHGJMEDmTHGJMEDmTHGJDGqs735+PhApxvxYBNNTZo0iZtU4Cb1ZOziJnVkbAKgumlUA9nlckl3WKKMh0pykzoyNgFydnGTOjI2AeoHMn9kwRhjkuCBzBhjkuCBzBhjkuCBzBhjkuCBzBhjkuCBzBhjkuCBzBhjktBkIDc0NCAlJQUTJ07E9u3bh9wWGRkJk8kEq9WKuXPnapEzYlNFRQXi4uIQExODwsJCzZoGO3PmDIKCgtxXwv3d734npGMwGdbLk1paWpCeno6EhAQoioLi4mLRSfjqq6/cz5vVasWUKVNQVFQkOgsA8PjxY8yePRvLly8XnQIAWLduHfR6PRITE0WnuO3btw9msxkmkwmpqamor6/X7sHVXHhv4AtjvIBgR0cHVVdX0+bNm79ztdtnn32W7t69O6blEo39oobDNTmdToqOjqZbt265r6589epVTZoGO336NC1btszj5QzwtMkb68XbTUREbW1tVFNTQ0REPT09FBsbK0XXAKfTST/96U+pqanJo+V4q+ndd9+l3Nxcr2xb3mg6e/Ys1dTUkKIoHi+LyDtNVVVVdP/+fSIiOnbsGCUlJXm8TKi8yKkm75D1ej3mzZsHf39/LR5OleGaqqurERMTg+joaEyYMAE5OTkoKysTVCkPWdfLjBkzMGfOHABAYGAgjEYjWltbBVf9v1OnTmHWrFl49tlnRafAbrejvLwc+fn5olPc0tLSEBoaKjpjiNTUVISEhAAA5s+fD7vdrtljC/8MWafTYcmSJbDZbNi1a5foHLS2tmLmzJnu78PDw4W9wC9cuACLxYKMjAxcvXpVSMMAmdbLcJqamlBbW4vk5GTRKW4ff/wxcnNzRWcAADZu3IitW7fCx0f4y/4Ho7S0FBkZGZo93qjOZTEezp8/D4PBgDt37mDx4sWIj49HWlqa6Czh5syZg+bmZgQEBODYsWP45S9/iZs3b4rOklZvby+ys7NRVFSEKVOmiM4BAHzzzTc4fPgwfv/734tOwdGjR6HX62Gz2XDmzBnROT8Ip0+fRmlpKc6fP6/ZY47bP5UlJSXunRptbW3D3s9gMAD49iOEzMxMVFdXj1eSqiaDwYCWlhb393a73d043gb39fb2IiAgAADw4osvwuFw4N69e5p0fB+R62UkDocD2dnZWLlyJbKyskTnuB0/fhxz5szBT3/6U9EpqKqqwuHDhxEZGYmcnBx88sknWLVqlegsaTw5Gy5duoT8/HyUlZVh6tSp2oWo+aCZPNypN6CgoGDIDrTe3l7q6elx/zklJYWOHz8+qmV6u8nhcFBUVBTdvn3bvfPqypUrmjYREbW3t5PL5SIios8//5xmzpzp/n4sPG3yxnrxdhMRkcvlotWrV9Prr7/u8bIGeKOLiOiVV16h3bt3e2VZ3moi8t4OY281NTY2SrVTr7m5mWbNmkVVVVVeKPoWVO7U02Qgt7e3k8FgoMDAQAoKCiKDwUDd3d1VfXkhAAAMFElEQVR069YtMpvNZDabKSEhgd55552x/kW91kREVF5eTrGxsRQdHa1p02A7d+6khIQEMpvNlJyc7PHG4Y0mT9fLeDRVVlYSADKZTGSxWMhisVB5ebnwrt7eXgoNDaWuri6Pl0X01zuQc3JyaPr06eTn50cGg4Hef/994U15eXkUHBzs3p5sNpvHy1Q7kHU0inOH6nQ6Gs39tSDj+U+5SR0ZmwA5u7hJHRmbAHfXiCdF5t2tjDEmCR7IjDEmCR7IjDEmCR7IjDEmCR7IjDEmCR7IjDEmCR7IjDEmCR7IjDEmCR7IjDEmiVGd7c3Hxwc63YgHm2hq0qRJ3KQCN6knYxc3qSNjEwDVTaMayC6XS7rDEmU8VJKb1JGxCZCzi5vUkbEJUD+Q+SMLxhiTBA9kxhiTBA9kxhiTBA9kxhiTBA9kxhiTBA9kxhiTBA9kxhiThCYDuaGhASkpKZg4cSK2b98+5LYdO3ZAURQkJiYiNzcXDx8+1CJp2KaHDx8iKSkJFosFiqKgoKBAk54ndXd34xe/+IW7Y8+ePUI6BisrK4PZbIbVasXcuXM1vTz6cFpaWpCeno6EhAQoioLi4mLRSW6PHz/G7NmzsXz5ctEpAIDi4mIkJiZCURQUFRWJzgEAdHV1YcWKFYiPj4fRaMSFCxdEJ4mdAWouvDfwhTFeQLCjo4Oqq6tp8+bNQ67wbLfbKTIykvr7+4mI6OWXX6Y9e/aMatnebnK5XPTgwQMiIvrmm28oKSmJLly4oEnTYP/2b/9G//qv/0pERHfu3KGQkBB69OjRmJfnjaYHDx64r3xdX19PcXFxHi3PG01tbW1UU1NDREQ9PT0UGxtLV69eFd5FRPTuu+9Sbm6uFBcUvXz5MimKQn19feRwOOjnP/853bx5U2gTEdGaNWvovffeIyKiR48eUWdnp/Amb8yAYbpGnLGavEPW6/WYN28e/P39v3Ob0+nE119/DafTif7+foSFhWmRNGyTTqdDQEAAAMDhcMDhcAg5FFOn0+HBgwcgIvT29iI0NBR+fqM6sNLrAgIC3Ouir69PikNUZ8yYgTlz5gAAAgMDYTQa0draKrgKsNvtKC8vR35+vugUAMD169eRnJyMn/zkJ/Dz88PChQtx4MABoU3d3d04d+4c8vLyAAATJkxAcHCw0CZA7AwQ+hmywWDApk2bEBERgRkzZiAoKAhLliwRmQTg2/9qWq1W6PV6LF68GMnJyZo3bNiwAdevX0dYWBhMJhOKi4vh4yP+I/+DBw8iPj4ey5Ytw+7du0XnDNHU1ITa2lohz9eTNm7ciK1bt0rxnAFAYmIiKisr8Ze//AX9/f04duwYWlpahDY1NjZi2rRpePXVVzF79mzk5+ejr69PaNMAUTNA6NbS2dmJsrIyNDY2oq2tDX19ffjwww9FJgEAfH19UVdXB7vdjurqaly5ckXzhhMnTsBqtaKtrQ11dXXYsGEDenp6NO94UmZmJhoaGnDo0CG88cYbonPcent7kZ2djaKiIkyZMkVoy9GjR6HX62Gz2YR2DGY0GvHrX/8aS5YswQsvvACr1QpfX1+hTU6nExcvXsT69etRW1uLyZMno7CwUGjTAFEzYNwGcklJCaxWq3uofJ+TJ08iKioK06ZNg7+/P7KysvDpp5+OV5KqpsGCg4ORnp6OioqKcWsabHBfSUkJsrKyoNPpEBMTg6ioKDQ0NGjSMVzT4HWWlpaG27dv4969e5o3PcnhcCA7OxsrV65EVlaW6BxUVVXh8OHDiIyMRE5ODj755BOsWrVKdBby8vJQU1ODc+fOISQkBM8995zQnvDwcISHh7vffa5YsQIXL14U2vQkrWfAuA3kX/3qV6irq0NdXd2wnwtHRETgs88+Q39/P4gIp06dgtFoHK8kVU13795FV1cXAODrr7/Gf//3fyM+Pn7cmobri4+Px6lTpwAAHR0d+OqrrxAdHa1Jx3BNA88TAFy8eBGPHj3C1KlTNW8ajIiQl5cHo9GIf/mXfxHaMuD3v/897HY7mpqa8PHHH2PRokVS/M/vzp07AID/+Z//wYEDB/D3f//3QnumT5+OmTNn4quvvgIAnDp1CgkJCUKbALEzQJPfsmhvbyeDwUCBgYEUFBREBoOBuru7iYjozTffpLi4OFIUhVatWkUPHz4cy95LrzXV19eT1Wolk8lEiqLQW2+9Neplj7VpsNbWVlq8eDElJiaSoij0wQcfeLQ8bzQVFhZSQkICWSwWmj9/PlVWVgpvqqysJABkMpnIYrGQxWKh8vJy4V0DTp8+LcVvWRARLViwgIxGI5nNZjp58qQUTbW1tWSz2chkMtFLL71E9+/fF97kjRkwTNeIM1ZHozh3qE6no9HcXwsynv+Um9SRsQmQs4ub1JGxCXB3jfirGnLsAmaMMcYDmTHGZMEDmTHGJMEDmTHGJMEDmTHGJMEDmTHGJMEDmTHGJMEDmTHGJMEDmTHGJDGqE+z6+PhIcQ7cwSZNmsRNKnCTejJ2cZM6MjYBUN00qoHscrmkOyxRxkMluUkdGZsAObu4SR0ZmwD1A5k/smCMMUnwQGaMMUnwQGaMMUnwQGaMMUnwQGaMMUnwQGaMMUnwQGaMMUnwQGaMMUloMpD37dsHs9kMk8mE1NRU1NfXu29bt24d9Ho9EhMTtUhxa2hoQEpKCiZOnIjt27cPua2rqwsrVqxAfHw8jEYjLly4ILxJ1Hp60pkzZ2C1WqEoChYuXCi0ZcDT1psIsjxXTxK1XT/Njh07oCgKEhMTkZubi4cPH4pOEtuk5kqoA18Y4xVdq6qq3FeTPXbsGCUlJblvO3v2LNXU1JCiKGNa9libOjo6qLq6mjZv3kzbtm0bctuaNWvovffeIyKiR48eUWdnp/AmUetpsM7OTjIajdTc3Ozu9YQ3mgY6hltvY+Fpl6fP1ffxxrrydLv2dpPdbqfIyEjq7+8nIqKXX36Z9uzZ81fXNKhrxBmryTvk1NRUhISEAADmz58Pu93uvi0tLQ2hoaFaZAyh1+sxb948+Pv7D/l5d3c3zp07h7y8PADAhAkTEBwcLLQJELeeBvvoo4+QlZWFiIgIAN/2yuBp600EGZ6rJ4ncrp/G6XTi66+/htPpRH9/P8LCwkQnCW3S/DPk0tJSZGRkaP2wqjU2NmLatGl49dVXMXv2bOTn56Ovr090lhRu3LiBzs5OPP/887DZbNi7d6/oJKaSjNu1wWDApk2bEBERgRkzZiAoKAhLliz5UTdpOpBPnz6N0tJSbNmyRcuHHRWn04mLFy9i/fr1qK2txeTJk1FYWCg6SwpOpxM1NTUoLy/HiRMn8Pbbb+PGjRuis5gKMm7XnZ2dKCsrQ2NjI9ra2tDX14cPP/zwR900bgO5pKQEVqsVVqsVbW1tuHTpEvLz81FWVoapU6eO18OOqun7hIeHIzw8HMnJyQCAFStW4OLFi0KbRBrcFxYWhqVLl2Ly5Ml45plnkJaWNmQHraguGdebbLTertU4efIkoqKiMG3aNPj7+yMrKwuffvrpj7pp3Abyr371K9TV1aGurg5OpxNZWVn44IMP8Nxzz43XQ46qabjPhaZPn46ZM2fiq6++AgCcOnUKCQkJQptEGtyXmZmJ8+fPuz9b+/zzz2E0GoV3ybjeZKP1dq1GREQEPvvsM/T394OIcOrUKWHbkzRNavb8kYe/ZZGXl0fBwcFksVjIYrGQzWZz35aTk0PTp08nPz8/MhgM9P77749l7+Wotbe3k8FgoMDAQAoKCiKDwUDd3d1ERFRbW0s2m41MJhO99NJL7t8QEdkkaj09aevWrWQ0GklRFNqxY4dHy/JW09PWm4guT5+r8Wgi8ny7Ho+mN998k+Li4khRFFq1ahU9fPjwr65pUNeIM1ZHoziZs06no9HcXwsynpCam9SRsQmQs4ub1JGxCXB3jXiWej5SjzHGJMEDmTHGJMEDmTHGJMEDmTHGJMEDmTHGJMEDmTHGJMEDmTHGJMEDmTHGJOE3yvuTTqcb8ZebtaTT6SBZEjepJGMTIGcXN6kjY9P/UXW0yqiO1GOMMTZ++CMLxhiTBA9kxhiTBA9kxhiTBA9kxhiTBA9kxhiTBA9kxhiTBA9kxhiTBA9kxhiTBA9kxhiTxP8ChtjfB57ndqgAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "None\n",
      "The path matrix equals:\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAWQAAADuCAYAAAAOR30qAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvnQurowAAFB5JREFUeJzt3T9IPPn9x/HXqAQhKmKxKOaa4ySu4zKLkLWT28ImhBQuBMEiQct0FmlFCCGl9nYpkuaKFJIFK5EgCIvK6TUpTvBPCoszKihO3PevuGi+f7yfn836nXkrzwdscbCMT97z5c3e7s5sZGYCAOSvI+8AAMD3WMgA4AQLGQCcYCEDgBMsZABwgoUMAE6wkAHACRYyADjBQgYAJ7paeXJnZ6c1m81P1fI/6ejoEE3Poymcxy6awnhs+g8zs2dfALe0kJvNprxdah1FEU0BaArnsYumMB6bJCmKoijkebxlAQBOsJABwAkWMgA4wUIGACdYyADgBAsZAJxgIQOAE5kv5Hq9rp/+9Kf64osv9Mc//jHrP/+R+fl5FQoFjY+P553yHo9dNIWhKZzHrjybMl3I9/f3+u1vf6u//e1v+uabb/TnP/9Z33zzTZYJH/nNb36jer2ea8NTPHbRFIamcB678mzKdCHv7Ozoiy++0Oeff64f/ehHmp2d1V//+tcsEz4yNTWlgYGBXBue4rGLpjA0hfPYlWdTpgv59PRUn3322eN//+QnP9Hp6WmWCQDgFh/qAYATmS7k4eFhHR8fP/73ycmJhoeHs0wAALcyXcg/+9nP9I9//EPffvut7u7u9Je//EW//OUvs0wAAL/MLPjx/dPbs76+biMjI/b555/b73//+7aP127T7OysDQ4OWldXlw0PD9va2lruTZ+ii6bsumjKr8tj0ztdz+7YyFq4d2gURdbK87Pg8f6nNIXx2CT57KIpjMcm6bHr2Xsi86EeADjBQgYAJ1jIAOAECxkAnGAhA4ATLGQAcIKFDABOsJABwAkWMgA40dXKkzs6OhRFz15skqnu7m6aAtAUzmMXTWE8NkkKbmppITebTXeXJXq8VJKmMB6bJJ9dNIXx2CSFL2TesgAAJ1jIAOAECxkAnGAhA4ATLGQAcIKFDABOsJABwIlMF/Lt7a0qlYqSJFEcx1paWsryzz/p+PhY1WpVY2NjiuNYq6ureSdJkubn51UoFDQ+Pp53yiOPTR7PH3MK53FWuTaF/PDew0Nt/oBgs9m0q6srMzO7u7uzSqVi29vbbR2z3aazszNrNBpmZnZ5eWkjIyN2eHiYa5OZ2ebmpjUaDYvjuO1jmb3dJo/n76XnZMa/81Aem8zCf+Q001fIURSpp6dHkpSmqdI0zf0yx6GhIU1MTEiSent7VSwWdXp6mmuTJE1NTWlgYCDvjPd4bPJ4/phTOI+zyrMp8/eQ7+/vVS6XVSgUND09rcnJyawTftDR0ZF2d3ddNSEc5y8Mc/Ir84Xc2dmpvb09nZycaGdnRwcHB1knPOn6+lq1Wk0rKyvq6+vLOwct4vyFYU6+5fYti/7+flWrVdXr9bwSHqVpqlqtprm5Oc3MzOSdgxZx/sIwJ/8yXcjn5+e6uLiQJN3c3GhjY0Ojo6NZJnzEzLSwsKBisajFxcVcW9A6zl8Y5vRKhHzyZy/0LYv9/X0rl8tWKpUsjmNbXl5u63hm7X+qurW1ZZKsVCpZkiSWJImtr6/n2mRmNjs7a4ODg9bV1WXDw8O2trZG0xM8nr+XntNLNHmck5nPf1Of8Pw9u2Mja+HeoVEUWSvPz4LH+5/SFMZjk+Szi6YwHpukx65nv1LGlXoA4AQLGQCcYCEDgBMsZABwgoUMAE6wkAHACRYyADjBQgYAJ1jIAOBEVytP7ujoyP3+xR/q7u6mKQBN4Tx20RTGY5Ok4KaWFnKz2XR3WaLHSyVpCuOxSfLZRVMYj01S+ELmLQsAcIKFDABOsJABwAkWMgA4wUIGACdYyADgBAsZAJxgIQOAE5ku5NvbW1UqFSVJojiOtbS0lOWffzVNx8fHqlarGhsbUxzHWl1dzTvJZZPks4smmv5nIb+E+vBQm7/o2mw27erqyszM7u7urFKp2Pb2dlvHfItNZ2dn1mg0zMzs8vLSRkZG7PDwkKZX0kUTTT/Q9eyOzfQVchRF6unpkSSlaao0TXO/7txj09DQkCYmJiRJvb29KhaLOj09pekJHrtooul/FrK17YVeIZuZ/fvf/7YkSezHP/6x/e53v2v7eG+16cG3335rn332mf3rX/9q6zhvvcnMZxdNYd5yk1n4K+TMF/KD7777zr788kv7+uuv2zrOW266urqyiYkJ++qrr9o+1ltuMvPZRVOYt95k5vQti3f19/erWq2qXq/nlfART01pmqpWq2lubk4zMzN550jy2ST57KIpDE3vy3Qhn5+f6+LiQpJ0c3OjjY0NjY6OZpnwKprMTAsLCyoWi1pcXMy15YHHJslnF01haPqBgNCH2vzfgf39fSuXy1YqlSyOY1teXm7reGbt/y+Kx6atrS2TZKVSyZIksSRJbH19naZX0kUTTT/Q9eyOjayFmzlHUWStPD8LHm9ITVMYj02Szy6awnhskh67nv36FlfqAYATLGQAcIKFDABOsJABwAkWMgA4wUIGACdYyADgBAsZAJzoauXJHR0dud+a8kPd3d00BaApnMcumsJ4bJIU3NTSQm42m+6ugvF4ZQ5NYTw2ST67aArjsUkKX8i8ZQEATrCQAcAJFjIAOMFCBgAnWMgA4AQLGQCcYCEDgBMsZABwItOFfHt7q0qloiRJFMexlpaWsvzzr6bp+PhY1WpVY2NjiuNYq6ureSe5bJKk+fl5FQoFjY+P553yyOOsmFO4XGcV8sN7Dw+1+QOCzWbTrq6uzMzs7u7OKpWKbW9vt3XMt9h0dnZmjUbDzMwuLy9tZGTEDg8PaXrC5uamNRoNi+O47WM9eIuzYk7hPuGsnt2xmb5CjqJIPT09kqQ0TZWmae7XnXtsGhoa0sTEhCSpt7dXxWJRp6enND1hampKAwMDeWe8x+OsmFO4PGeV+XvI9/f3KpfLKhQKmp6e1uTkZNYJr6LpwdHRkXZ3d2l6pZhVGOb0vcwXcmdnp/b29nRycqKdnR0dHBxknfAqmiTp+vpatVpNKysr6uvryztHks8mr5hVGOb0X7l9y6K/v1/ValX1ej2vhI94akrTVLVaTXNzc5qZmck7R5LPJq+YVRjm9L5MF/L5+bkuLi4kSTc3N9rY2NDo6GiWCa+iycy0sLCgYrGoxcXFXFseeGzyilmFYU5PCPnk7+GhNj/B3N/ft3K5bKVSyeI4tuXl5baOZ9b+p6oem7a2tkySlUolS5LEkiSx9fV1mp4wOztrg4OD1tXVZcPDw7a2ttb2Md/irJhTuE84q2d3bGQt3Mw5iiJr5flZ8HhDaprCeGySfHbRFMZjk/TY9ezXt7hSDwCcYCEDgBMsZABwgoUMAE6wkAHACRYyADjBQgYAJ1jIAOAECxkAnOhq5ckdHR253yv4Q93d3TQFoCmcxy6awnhskhTc1NJCbjab7i5L9HipJE1hPDZJPrtoCuOxSQpfyLxlAQBOsJABwAkWMgA4wUIGACdYyADgBAsZAJxgIQOAE5ku5NvbW1UqFSVJojiOtbS0lOWff9Lx8bGq1arGxsYUx7FWV1fzTmJOLfDYNT8/r0KhoPHx8bxTHnmck+RzVrk2hfzw3sNDbf6AYLPZtKurKzMzu7u7s0qlYtvb220ds92ms7MzazQaZmZ2eXlpIyMjdnh4mGsTc3rdXZubm9ZoNCyO47aO8663ev5eelYem8zCf+Q001fIURSpp6dHkpSmqdI0zf0yx6GhIU1MTEiSent7VSwWdXp6mmsTcwrnsWtqakoDAwO5NnzI45wkn7PKsynz95Dv7+9VLpdVKBQ0PT2tycnJrBN+0NHRkXZ3d100MafWee3yhjn5lflC7uzs1N7enk5OTrSzs6ODg4OsE550fX2tWq2mlZUV9fX15Z3DnFrktcsb5uRbbt+y6O/vV7VaVb1ezyvhUZqmqtVqmpub08zMTN4572FOz/Pa5Q1z8i/ThXx+fq6LiwtJ0s3NjTY2NjQ6OpplwkfMTAsLCyoWi1pcXMy15QFzCue1yxvm9EqEfPL38FCbn2Du7+9buVy2UqlkcRzb8vJyW8cza/9T1a2tLZNkpVLJkiSxJElsfX091ybm9Lq7ZmdnbXBw0Lq6umx4eNjW1tbaOt5LNHmck9nLz8pj0ztdz+7YyFq4d2gURdbK87Pg8f6nNIXx2CT57KIpjMcm6bHr2a9KcaUeADjBQgYAJ1jIAOAECxkAnGAhA4ATLGQAcIKFDABOsJABwAkWMgA40dXKkzs6OnK/L++Huru7aQpAUziPXTSF8dgkKbippYXcbDbdXZbo8VJJmsJ4bJJ8dtEUxmOTFL6QecsCAJxgIQOAEyxkAHCChQwATrCQAcAJFjIAOMFCBgAnMl3It7e3qlQqSpJEcRxraWkpyz//apqOj49VrVY1NjamOI61urqad5LLJonzF2p+fl6FQkHj4+N5pzzyOCcp51mF/PDew0Nt/oBgs9m0q6srMzO7u7uzSqVi29vbbR3zLTadnZ1Zo9EwM7PLy0sbGRmxw8NDmp7A+QuzublpjUbD4jhu6zjveotzMvuks3p2x2b6CjmKIvX09EiS0jRVmqa5X+bosWloaEgTExOSpN7eXhWLRZ2entL0BM5fmKmpKQ0MDOTa8CGPc5LynVXm7yHf39+rXC6rUChoenpak5OTWSe8iqYHR0dH2t3dpen/wfl7/ZjT9zJfyJ2dndrb29PJyYl2dnZ0cHCQdcKraJKk6+tr1Wo1raysqK+vL+8cST6bOH+vG3P6r9y+ZdHf369qtap6vZ5Xwkc8NaVpqlqtprm5Oc3MzOSdI8ln07s4f68Pc3pfpgv5/PxcFxcXkqSbmxttbGxodHQ0y4RX0WRmWlhYULFY1OLiYq4tDzw2SZy/14w5PSHkk7+Hh9r8BHN/f9/K5bKVSiWL49iWl5fbOp5Z+5+qemza2toySVYqlSxJEkuSxNbX12l6AucvzOzsrA0ODlpXV5cNDw/b2tpaW8d7iSaPczL7pLN6dsdG1sK9Q6MoslaenwWP9z+lKYzHJslnF01hPDZJj13Pfv2HK/UAwAkWMgA4wUIGACdYyADgBAsZAJxgIQOAEyxkAHCChQwATrCQAcCJrlae3NHRkfu9Zj/U3d1NUwCawnnsoimMxyZJwU0tLeRms+nuskSPl0rSFMZjk+Szi6YwHpuk8IXMWxYA4AQLGQCcYCEDgBMsZABwgoUMAE6wkAHACRYyADiR6UK+vb1VpVJRkiSK41hLS0tZ/vknHR8fq1qtamxsTHEca3V1Ne8kl3OSfHZx/sJ4nJMkzc/Pq1AoaHx8PO+UR7k2hfzw3sNDbf6AYLPZtKurKzMzu7u7s0qlYtvb220ds92ms7MzazQaZmZ2eXlpIyMjdnh4mGuTxzl9iq6XaOL8hfE4JzOzzc1NazQaFsdx28cy89lkFv4jp5m+Qo6iSD09PZKkNE2VpmnulzkODQ1pYmJCktTb26tisajT09NcmzzOSfLZxfkL43FOkjQ1NaWBgYG8M96TZ1Pm7yHf39+rXC6rUChoenpak5OTWSf8oKOjI+3u7rpo8jonr10S5y+UpznhfZkv5M7OTu3t7enk5EQ7Ozs6ODjIOuFJ19fXqtVqWllZUV9fX945bufktYvzF8bbnPC+3L5l0d/fr2q1qnq9nlfCozRNVavVNDc3p5mZmbxz3uNpTu/y1MX5C+N5Tvhepgv5/PxcFxcXkqSbmxttbGxodHQ0y4SPmJkWFhZULBa1uLiYa8sDj3OSfHZx/sJ4nBOeEPLJn73Qtyz29/etXC5bqVSyOI5teXm5reOZtf+p6tbWlkmyUqlkSZJYkiS2vr6ea5PHOX2Krpdo4vyF8TgnM7PZ2VkbHBy0rq4uGx4etrW1tTfX9E7Xszs2shbuHRpFkbXy/Cx4vP8pTWE8Nkk+u2gK47FJeux69qs2XKkHAE6wkAHACRYyADjBQgYAJ1jIAOAECxkAnGAhA4ATLGQAcIKFDABOdLXy5I6Ojtzv6/qh7u5umgLQFM5jF01hPDZJCm5qaSE3m013lyV6vFSSpjAemySfXTSF8dgkhS9k3rIAACdYyADgBAsZAJxgIQOAEyxkAHCChQwATrCQAcAJFjIAOJHpQr69vVWlUlGSJIrjWEtLS1n+eZra5LHLY9Px8bGq1arGxsYUx7FWV1fzTmJOr6Up5JdQHx5q8xddm82mXV1dmZnZ3d2dVSoV297ebuuYNGXT9Cm6PDa9RNfZ2Zk1Gg0zM7u8vLSRkRE7PDzMtYk55df0TtezOzbTV8hRFKmnp0eSlKap0jTN/bpzmsJ57PLYNDQ0pImJCUlSb2+visWiTk9Pc21iTq+jKfP3kO/v71Uul1UoFDQ9Pa3JycmsE2hqg8cuj00Pjo6OtLu766KJObUml6aQl9H2Qm9ZvOu7776zL7/80r7++uu2jkNTmJdsMnuZLo9NZi/XdXV1ZRMTE/bVV1+1fay3/G/K45xessnM6VsW7+rv71e1WlW9Xs8r4SM0hfPY5akpTVPVajXNzc1pZmYm75z3MKf/X55NmS7k8/NzXVxcSJJubm60sbGh0dHRLBNoaoPHLo9NZqaFhQUVi0UtLi7m2vKAOYXJu6ml+yG365///Kd+/etf6/7+Xs1mU7/61a/0i1/8IssEmt5Yl8emv//97/rTn/6kUqmkcrksSfrDH/6gn//857k1MafX0RRZCzdzjqLIWnl+FjzekJqmMB6bJJ9dNIXx2CQ9dj37tRau1AMAJ1jIAOAECxkAnGAhA4ATLGQAcIKFDABOsJABwAkWMgA40eqVehblfc++D0RRlPttBD9EUxiPTZLPLprCeGz6j6CrVVq6Ug8A8OnwlgUAOMFCBgAnWMgA4AQLGQCcYCEDgBMsZABwgoUMAE6wkAHACRYyADjxfxp9LT0MMaGVAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "None\n"
     ]
    }
   ],
   "source": [
    "s1 = \"ATGAGAT\"\n",
    "s2 = \"AGGAGAGT\"\n",
    "res2by2 = align_nw_2by2(s1, s2)\n",
    "print(\"The initial sequences are:\")\n",
    "print(s1, s2)\n",
    "print(\"The aligned sequences are:\")\n",
    "print(res2by2[2])\n",
    "print(\"The positions of the introduced gaps are:\")\n",
    "print(res2by2[3])\n",
    "print(\"The score matrix equals:\")\n",
    "print(print_table(res2by2[0]))\n",
    "print(\"The path matrix equals:\")\n",
    "print(print_table(res2by2[1]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pairs = [(0, 1), (0, 2), (0, 3), (0, 4), (1, 2), (1, 3), (1, 4), (2, 3), (2, 4), (3, 4)]\n",
      "scores2a2 = [8.0, -2.0, 3.0, 1.0, 1.0, 11.0, -2.0, 1.0, 1.0, -2.0]\n",
      "global scores = [10.0, 18.0, 1.0, 13.0, -2.0]\n",
      "pivot sequence =  AGGAGAGT\n",
      "initial seqs =\n",
      "ATGAGAT\n",
      "AGGAGAGT\n",
      "GGAGG\n",
      "AGGGAGT\n",
      "AGAAC\n",
      "res aligned to pivot =\n",
      "ATGAGA_T\n",
      "AGGAGAGT\n",
      "_GGAG_G_\n",
      "AGG_GAGT\n",
      "A_GA_A_C\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['ATGAGA_T', 'AGGAGAGT', '_GGAG_G_', 'AGG_GAGT', 'A_GA_A_C']"
      ]
     },
     "execution_count": 162,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "align_star_multiple([\"ATGAGAT\", \"AGGAGAGT\", \"GGAGG\", \"AGGGAGT\", \"AGAAC\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pairs = [(0, 1), (0, 2), (0, 3), (0, 4), (0, 5), (0, 6), (0, 7), (0, 8), (0, 9), (1, 2), (1, 3), (1, 4), (1, 5), (1, 6), (1, 7), (1, 8), (1, 9), (2, 3), (2, 4), (2, 5), (2, 6), (2, 7), (2, 8), (2, 9), (3, 4), (3, 5), (3, 6), (3, 7), (3, 8), (3, 9), (4, 5), (4, 6), (4, 7), (4, 8), (4, 9), (5, 6), (5, 7), (5, 8), (5, 9), (6, 7), (6, 8), (6, 9), (7, 8), (7, 9), (8, 9)]\n",
      "scores2a2 = [17.0, 15.0, 17.0, 17.0, 15.0, 17.0, 17.0, 15.0, 17.0, 15.0, 14.0, 14.0, 12.0, 14.0, 14.0, 12.0, 14.0, 12.0, 12.0, 12.0, 12.0, 12.0, 12.0, 12.0, 14.0, 12.0, 14.0, 14.0, 12.0, 16.0, 15.0, 14.0, 14.0, 12.0, 14.0, 12.0, 12.0, 10.0, 12.0, 14.0, 12.0, 14.0, 15.0, 14.0, 12.0]\n",
      "global scores = [147.0, 126.0, 114.0, 125.0, 126.0, 112.0, 123.0, 126.0, 112.0, 125.0]\n",
      "pivot sequence =  CTTCCCTTAA\n",
      "initial seqs =\n",
      "CTTCCCTTAA\n",
      "CTTCCGTTAA\n",
      "CTTCCTTAA\n",
      "CTTCCCCTTAA\n",
      "ATTCCCTTAA\n",
      "TTCCCTTAA\n",
      "TCTTCCCTTAA\n",
      "CTTCCCTTAC\n",
      "CTTCCCTTA\n",
      "CTTCCCTTAAA\n",
      "res aligned to pivot =\n",
      "_CTT_CCCTT_AA\n",
      "_CTT_CCGTT_AA\n",
      "_CTT__CCTT_AA\n",
      "_CTTCCCCTT_AA\n",
      "_ATT_CCCTT_AA\n",
      "__TT_CCCTT_AA\n",
      "TCTT_CCCTT_AA\n",
      "_CTT_CCCTT_AC\n",
      "_CTT_CCCTT__A\n",
      "_CTT_CCCTTAAA\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['_CTT_CCCTT_AA',\n",
       " '_CTT_CCGTT_AA',\n",
       " '_CTT__CCTT_AA',\n",
       " '_CTTCCCCTT_AA',\n",
       " '_ATT_CCCTT_AA',\n",
       " '__TT_CCCTT_AA',\n",
       " 'TCTT_CCCTT_AA',\n",
       " '_CTT_CCCTT_AC',\n",
       " '_CTT_CCCTT__A',\n",
       " '_CTT_CCCTTAAA']"
      ]
     },
     "execution_count": 163,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n = 10 # length of the initial sequence\n",
    "seq0 = choice(alphabet, n)\n",
    "seq0 = ''.join(seq0)\n",
    "\n",
    "# seq 1 = substitution middle\n",
    "seq1 = substitution(seq0, int(n/2))\n",
    "# seq 2 = deletion middle\n",
    "seq2 = deletion(seq0, int(n/2)-1)\n",
    "# seq 3 = insertion middle\n",
    "seq3 = insertion(seq0, int(n/2))\n",
    "\n",
    "# seq 4 = substitution begining\n",
    "seq4 = substitution(seq0, 0)\n",
    "# seq 5 = deletion begining\n",
    "seq5 = deletion(seq0, 0)\n",
    "# seq 6 = insertion begining\n",
    "seq6 = insertion(seq0, 0)\n",
    "\n",
    "# seq 7 = substitution end\n",
    "seq7 = substitution(seq0, -1)\n",
    "# seq 8 = deletion end\n",
    "seq8 = deletion(seq0, -1)\n",
    "# seq 9 = insertion end\n",
    "seq9 = insertion(seq0, -1)\n",
    "\n",
    "# global alignement\n",
    "align_star_multiple([seq0, seq1, seq2, seq3, seq4, seq5, seq6, seq7, seq8, seq9])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pairs = [(0, 1), (0, 2), (0, 3), (0, 4), (0, 5), (0, 6), (0, 7), (0, 8), (0, 9), (1, 2), (1, 3), (1, 4), (1, 5), (1, 6), (1, 7), (1, 8), (1, 9), (2, 3), (2, 4), (2, 5), (2, 6), (2, 7), (2, 8), (2, 9), (3, 4), (3, 5), (3, 6), (3, 7), (3, 8), (3, 9), (4, 5), (4, 6), (4, 7), (4, 8), (4, 9), (5, 6), (5, 7), (5, 8), (5, 9), (6, 7), (6, 8), (6, 9), (7, 8), (7, 9), (8, 9)]\n",
      "scores2a2 = [14.0, 10.0, 14.0, 14.0, 10.0, 14.0, 14.0, 10.0, 14.0, 10.0, 11.0, 8.0, 4.0, 8.0, 8.0, 4.0, 8.0, 4.0, 4.0, 7.0, 4.0, 4.0, 10.0, 4.0, 8.0, 4.0, 10.0, 8.0, 4.0, 9.0, 10.0, 8.0, 8.0, 4.0, 8.0, 4.0, 4.0, 1.0, 4.0, 8.0, 4.0, 8.0, 10.0, 8.0, 4.0]\n",
      "global scores = [114.0, 75.0, 57.0, 72.0, 72.0, 48.0, 68.0, 72.0, 51.0, 67.0]\n",
      "pivot sequence =  GGGCTTTTGC\n",
      "initial seqs =\n",
      "GGGCTTTTGC\n",
      "GGGCTCGTGC\n",
      "GGGCTTGC\n",
      "GGGCTGCTTTGC\n",
      "TAGCTTTTGC\n",
      "GCTTTTGC\n",
      "AGGGGCTTTTGC\n",
      "GGGCTTTTTT\n",
      "GGGCTTTT\n",
      "GGGCTTTTGAAC\n",
      "res aligned to pivot =\n",
      "GGGCT__TTTGC\n",
      "GGGCT__CGTGC\n",
      "GGGC____TTGC\n",
      "GGGCTGCTTTGC\n",
      "TAGCT__TTTGC\n",
      "__GCT__TTTGC\n",
      "AGGGGCTTTTGC\n",
      "GGGCT__TTTTT\n",
      "GGGCT__TTT__\n",
      "GGGCTTTTGAAC\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['GGGCT__TTTGC',\n",
       " 'GGGCT__CGTGC',\n",
       " 'GGGC____TTGC',\n",
       " 'GGGCTGCTTTGC',\n",
       " 'TAGCT__TTTGC',\n",
       " '__GCT__TTTGC',\n",
       " 'AGGGGCTTTTGC',\n",
       " 'GGGCT__TTTTT',\n",
       " 'GGGCT__TTT__',\n",
       " 'GGGCTTTTGAAC']"
      ]
     },
     "execution_count": 164,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n = 10 # length of the initial sequence\n",
    "seq0 = choice(alphabet, n)\n",
    "seq0 = ''.join(seq0)\n",
    "\n",
    "# seq 1 = substitution middle\n",
    "seq1 = substitution(seq0, int(n/2))\n",
    "seq1 = substitution(seq1, int(n/2)+1)\n",
    "# seq 2 = deletion middle\n",
    "seq2 = deletion(seq0, int(n/2)-1)\n",
    "seq2 = deletion(seq2, int(n/2)-1)\n",
    "# seq 3 = insertion middle\n",
    "seq3 = insertion(seq0, int(n/2))\n",
    "seq3 = insertion(seq3, int(n/2))\n",
    "\n",
    "\n",
    "# seq 4 = substitution begining\n",
    "seq4 = substitution(seq0, 0)\n",
    "seq4 = substitution(seq4, 1)\n",
    "# seq 5 = deletion begining\n",
    "seq5 = deletion(seq0, 0)\n",
    "seq5 = deletion(seq5, 0)\n",
    "# seq 6 = insertion begining\n",
    "seq6 = insertion(seq0, 0)\n",
    "seq6 = insertion(seq6, 0)\n",
    "\n",
    "# seq 7 = substitution end\n",
    "seq7 = substitution(seq0, -1)\n",
    "seq7 = substitution(seq7, -2)\n",
    "# seq 8 = deletion end\n",
    "seq8 = deletion(seq0, -1)\n",
    "seq8 = deletion(seq8, -1)\n",
    "# seq 9 = insertion end\n",
    "seq9 = insertion(seq0, -1)\n",
    "seq9 = insertion(seq9, -1)\n",
    "\n",
    "# global alignement\n",
    "align_star_multiple([seq0, seq1, seq2, seq3, seq4, seq5, seq6, seq7, seq8, seq9])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TEST ADDGAP FUNCTION\n",
      "The initial sequence equals: GCTATCC and the gaplist equals: [1, 4, 7, 0, 0]. The final sequence equals: __G_CT_AT_CC.\n",
      "TEST COST FUNCTION\n",
      "The cost of (C, A) equals -1.\n",
      "The cost of (C, A) equals -1.\n",
      "The cost of (C, T) equals -1.\n",
      "The cost of (T, G) equals -1.\n",
      "The cost of (T, G) equals -1.\n"
     ]
    }
   ],
   "source": [
    "# -*- coding: utf-8 -*-\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from numpy.random import choice\n",
    "\n",
    "\n",
    "def cost(letter1, letter2):\n",
    "    \"\"\" Calculate the cost or score of the alignment of 2 letters.\n",
    "    Parameters:\n",
    "    ----------\n",
    "    letter1: character\n",
    "    letter2: character\n",
    "    Return:\n",
    "    ------\n",
    "    score: int\n",
    "        score of the alignment (-1 if the 2 letters are different, 2 if they are similar)\n",
    "    \"\"\"\n",
    "    if letter1 == letter2:\n",
    "        return 2\n",
    "    else:\n",
    "        return -1\n",
    "\n",
    "\n",
    "def addgap(seq, gaplist, verbose=0):\n",
    "    \"\"\" Insert gaps into sequence.\n",
    "    Parameters:\n",
    "    ----------\n",
    "    seq: string of characters\n",
    "        sequence of nucleotides\n",
    "    gaplist: list of int\n",
    "        list of gaps to insert successively in the sequence seq\n",
    "    Return:\n",
    "    ------\n",
    "    newseq: string of characters\n",
    "        the modified sequence with inserted gaps\n",
    "    \"\"\"\n",
    "    newseq = seq[:]\n",
    "    for g in gaplist:\n",
    "        if g > len(newseq):\n",
    "            print(\"gap postion bigger than sequence length -> gap inserted at the end of the sequence\")\n",
    "        newseq = newseq[:g] + '_' + newseq[g:]\n",
    "        if verbose > 0:\n",
    "            print(\"gap introduced in {}th position -> new sequence equals {}.\".format(g, newseq))\n",
    "    return newseq\n",
    "\n",
    "\n",
    "def substitution(seq, pos, verbose=0, value=None, alphabet=[\"A\", \"C\", \"G\", \"T\"]):\n",
    "    \"\"\" Induce a mutation of the sequence by substituting a letter.\n",
    "    Parameters:\n",
    "    ----------\n",
    "    seq: string of characters\n",
    "        sequence of nucleotides\n",
    "    pos: int in [-len(seq), len(seq)]\n",
    "        position of the mutation\n",
    "    verbose: int (default=0)\n",
    "        level of verbosity\n",
    "    value: None or alphabet item (default=None)\n",
    "        new letter induced by the mutation.\n",
    "        if None, an item different from the initial one is randomly chosen in alphabet.\n",
    "    alphabet: list of characters (default=[\"A\", \"C\", \"G\", \"T\"])\n",
    "        list of nucleotides to consider in sequences\n",
    "    Return:\n",
    "    ------\n",
    "    seqr: string\n",
    "        the modified sequence containing the mutation\n",
    "    \"\"\"\n",
    "    seqr = list(seq[:])\n",
    "    alphabis = alphabet.copy()\n",
    "    alphabis.remove(seqr[pos])\n",
    "    if value == None:\n",
    "        seqr[pos] = choice(alphabis, 1)[0]\n",
    "    elif value in alphabet:\n",
    "        seqr[pos] = value\n",
    "    else:\n",
    "        return \"error\"\n",
    "    seqr = \"\".join(seqr)\n",
    "    if verbose > 0:\n",
    "        print(seqr)\n",
    "    return seqr\n",
    "\n",
    "\n",
    "def insertion(seq, pos, verbose=0, value=None, alphabet=[\"A\", \"T\", \"G\", \"C\"]):\n",
    "    \"\"\" Induce a mutation of the sequence by inserting a new letter.\n",
    "    Parameters:\n",
    "    ----------\n",
    "    seq: string of characters\n",
    "        sequence of nucleotides\n",
    "    pos: int in [-len(seq), len(seq)]\n",
    "        position of the mutation\n",
    "    verbose: int (default=0)\n",
    "        level of verbosity\n",
    "    value: None or alphabet item (default=None)\n",
    "        new letter induced by the mutation.\n",
    "        if None, an item different from the initial one is randomly chosen in alphabet.\n",
    "    alphabet: list of characters (default=[\"A\", \"C\", \"G\", \"T\"])\n",
    "        list of nucleotides to consider in sequences\n",
    "    Return:\n",
    "    ------\n",
    "    seqr: string\n",
    "        the modified sequence containing the mutation\n",
    "    \"\"\"\n",
    "    seqr = seq[:]\n",
    "    if value is None:\n",
    "        value = choice(alphabet, 1)[0]\n",
    "    seqr = seqr[:pos] + value + seqr[pos:]\n",
    "    seqr = \"\".join(seqr)\n",
    "    if verbose > 0:\n",
    "        print(seqr)\n",
    "    return seqr\n",
    "\n",
    "\n",
    "def deletion(seq, pos, verbose=0):\n",
    "    \"\"\" Induce a mutation of the sequence by deleting a letter.\n",
    "    Parameters:\n",
    "    ----------\n",
    "    seq: string of characters\n",
    "        sequence of nucleotides\n",
    "    pos: int in [-len(seq), len(seq)]\n",
    "        position of the mutation\n",
    "    verbose: int (default=0)\n",
    "        level of verbosity\n",
    "    Return:\n",
    "    ------\n",
    "    seqr: string\n",
    "        the modified sequence containing the mutation\n",
    "    \"\"\"\n",
    "    seqr = list(seq[:])\n",
    "    del seqr[pos]\n",
    "    seqr = \"\".join(seqr)\n",
    "    if verbose > 0:\n",
    "        print(seqr)\n",
    "    return seqr\n",
    "\n",
    "\n",
    "def print_table(table):\n",
    "    \"\"\" Display the score or path matrix as a table.\n",
    "    Parameters:\n",
    "    ----------\n",
    "    table: matrix of floats or int\n",
    "        score or path matrix\n",
    "    Return:\n",
    "    ------\n",
    "    none\n",
    "    \"\"\"\n",
    "    plt.figure()\n",
    "    tb = plt.table(cellText=table.astype(int), loc=(0,0), cellLoc='center')\n",
    "    tc = tb.properties()['child_artists']\n",
    "    for cell in tc:\n",
    "        cell.set_height(1/table.shape[0])\n",
    "        cell.set_width(1/table.shape[1])\n",
    "    ax = plt.gca()\n",
    "    ax.set_xticks([])\n",
    "    ax.set_yticks([])\n",
    "    plt.show()\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "\n",
    "    alphabet = [\"A\", \"C\", \"G\", \"T\"]\n",
    "\n",
    "    print(\"TEST ADDGAP FUNCTION\")\n",
    "    seq = choice(alphabet, 7)\n",
    "    seq = ''.join(seq)\n",
    "    gaplist = [1, 4, 7, 0, 0]\n",
    "    newseq = addgap(seq, gaplist)\n",
    "    print(\"The initial sequence equals: {} and the gaplist equals: {}. \"\n",
    "          \"The final sequence equals: {}.\".format(seq, gaplist, newseq))\n",
    "\n",
    "    print(\"TEST COST FUNCTION\")\n",
    "    for i in range(5):\n",
    "        l1 = choice(alphabet)\n",
    "        l2 = choice(alphabet)\n",
    "        print(\"The cost of ({}, {}) equals {}.\".format(l1, l2, cost(l1, l2)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# LOCAL ALIGNMENT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----------- local alignment -----------\n",
      "< case 1 >\n",
      "score: 17\n",
      "T-CACACTG\n",
      "TGCACAC-G\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "class local_alignment:\n",
    "    def __init__(self, match_award, mismatch_penalty, gap_penalty):\n",
    "        self.match_award = match_award\n",
    "        self.mismatch_penalty = mismatch_penalty\n",
    "        self.gap_penalty = gap_penalty\n",
    "        self.order_stack = []\n",
    "        self.index_stack = []\n",
    "        \n",
    "    def check_match(self, data, i, j):\n",
    "        if data.index[i] == data.columns[j]: return self.match_award\n",
    "        else: return self.mismatch_penalty\n",
    "    \n",
    "    def make_matrix(self, first_seq, second_seq):\n",
    "        first_seq2 = [\"-\"]\n",
    "        second_seq2 = [\"-\"]\n",
    "        first_seq2.extend(list(first_seq))\n",
    "        second_seq2.extend(list(second_seq))\n",
    "        \n",
    "        data = [np.zeros(len(first_seq2), dtype='int32') for i in second_seq2]\n",
    "        data = pd.DataFrame(data, columns=first_seq2, index=second_seq2)\n",
    "        \n",
    "        # save the route to trace back\n",
    "        flag = []\n",
    "        \n",
    "        # fill out the matrix\n",
    "        for i in range(1, data.shape[0]):\n",
    "            temp = []\n",
    "            for j in range(1, data.shape[1]):\n",
    "                value_score = [data.iloc[i,j-1]+self.gap_penalty,\n",
    "                               data.iloc[i-1,j-1]+self.check_match(data, i, j),\n",
    "                               data.iloc[i-1,j]+self.gap_penalty,\n",
    "                               0]\n",
    "                data.iloc[i,j] = max(value_score)\n",
    "                temp.append([data.iloc[i,j] == item for item in value_score])\n",
    "            flag.append(temp)\n",
    "        return data, flag\n",
    "    \n",
    "    def extract_max(self, data):\n",
    "        temp = np.array(data.max())\n",
    "        col = np.where(temp == temp.max())[0][0]\n",
    "        temp = data.iloc[:,col]\n",
    "        row = np.where(temp == temp.max())[0][0]\n",
    "        return data.max().max(), row, col\n",
    "        \n",
    "    def extract_position(self, i, j, position):\n",
    "        if position == 0: return [i,j-1]\n",
    "        elif position == 1: return [i-1,j-1]\n",
    "        elif position == 2: return [i-1,j]\n",
    "        else: return None\n",
    "    \n",
    "    def stack_route(self, i, j, data, flag, point):\n",
    "        term = flag[i][j]\n",
    "        cnt = flag[i][j].count(True)\n",
    "        \n",
    "        # stack order\n",
    "        for k in range(cnt): self.order_stack.append(point)\n",
    "        \n",
    "        # stack index\n",
    "        if cnt == 1:\n",
    "            pos = term.index(True)\n",
    "            self.index_stack.append(self.extract_position(i, j, pos))\n",
    "        elif cnt == 2:\n",
    "            term = np.array(term)\n",
    "            term = np.where(term == True)[0]\n",
    "            self.index_stack.append(self.extract_position(i, j, term[0]))\n",
    "            self.index_stack.append(self.extract_position(i, j, term[1]))\n",
    "        elif cnt == 3:\n",
    "            for k in range(3): self.index_stack.append(self.extract_position(i, j, k))\n",
    "        else:\n",
    "            print(\"error: 'cnt' is out of range (1,3)\")\n",
    "        return None\n",
    "    \n",
    "    def backtracking(self, data, flag, row, col):\n",
    "        i = row - 1\n",
    "        j = col - 1\n",
    "        point = 0\n",
    "        key = 0\n",
    "        seq_dic = {str(key):[[i,j]]}\n",
    "        \n",
    "        # make index list\n",
    "        while i >= 0 or j >= 0:\n",
    "            self.stack_route(i, j, data, flag, point)\n",
    "            self.order_stack.pop()\n",
    "            term = self.index_stack.pop()\n",
    "            i, j = term[0], term[1]\n",
    "            if data.iloc[i+1,j+1] == 0: break\n",
    "            seq_dic[str(key)].append(term)\n",
    "            point += 1\n",
    "            if len(self.order_stack) != 0 and i < 0 or j < 0:\n",
    "                point = self.order_stack.pop()\n",
    "                term = self.index_stack.pop()\n",
    "                i, j = term[0], term[1]\n",
    "                if data.iloc[i+1,j+1] == 0: break\n",
    "                key += 1\n",
    "                seq_dic[str(key)] = seq_dic[str(key-1)][:point+1]\n",
    "                seq_dic[str(key)].append(term)\n",
    "        \n",
    "        # convert index list to name list\n",
    "        key = 0\n",
    "        for value in seq_dic.values():\n",
    "            row, col = [i for i in zip(*value)]\n",
    "            row = [row[i] if row[i] != row[i+1] else -1 for i in range(len(row)-1)] + [row[-1]]\n",
    "            col = [col[i] if col[i] != col[i+1] else -1 for i in range(len(col)-1)] + [col[-1]]\n",
    "            \n",
    "            row = [data.index[i+1] for i in row]\n",
    "            col = [data.columns[i+1] for i in col]\n",
    "            row = \"\".join(row)[::-1]\n",
    "            col = \"\".join(col)[::-1]\n",
    "            seq_dic[str(key)] = [col, row]\n",
    "            key += 1\n",
    "        return seq_dic\n",
    "    \n",
    "if __name__ == \"__main__\":\n",
    "    # input data\n",
    "    print(\"----------- local alignment -----------\")\n",
    "    #first_seq = input(\"Please enter the first nucleotide sequence: \")\n",
    "    #second_seq = input(\"Please enter the second nucleotide sequence: \")\n",
    "\n",
    "    first_seq = \"AACTCACACTGTT\"\n",
    "    second_seq = \"TGCACACG\"\n",
    "\n",
    "    # assign award and penalties\n",
    "    match_award = 3\n",
    "    mismatch_penalty = -1\n",
    "    gap_penalty = -2\n",
    "    \n",
    "    # local alignment\n",
    "    process = local_alignment(match_award, mismatch_penalty, gap_penalty)\n",
    "    data, flag = process.make_matrix(first_seq, second_seq)\n",
    "    score, row, col = process.extract_max(data)\n",
    "    seq_dic = process.backtracking(data, flag, row, col)\n",
    "    for key in seq_dic.keys():\n",
    "        print(\"< case\",int(key)+1,\">\")\n",
    "        print(\"score:\", score)\n",
    "        print(seq_dic[key][0])\n",
    "        print(seq_dic[key][1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>-</th>\n",
       "      <th>A</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>T</th>\n",
       "      <th>C</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>T</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "      <th>T</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>-</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>10</td>\n",
       "      <td>8</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>13</td>\n",
       "      <td>11</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>16</td>\n",
       "      <td>14</td>\n",
       "      <td>12</td>\n",
       "      <td>10</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>9</td>\n",
       "      <td>10</td>\n",
       "      <td>14</td>\n",
       "      <td>15</td>\n",
       "      <td>17</td>\n",
       "      <td>15</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   -  A  A  C  T  C  A   C   A   C   T   G   T   T\n",
       "-  0  0  0  0  0  0  0   0   0   0   0   0   0   0\n",
       "T  0  0  0  0  3  1  0   0   0   0   3   1   3   3\n",
       "G  0  0  0  0  1  2  0   0   0   0   1   6   4   2\n",
       "C  0  0  0  3  1  4  2   3   1   3   1   4   5   3\n",
       "A  0  3  3  1  2  2  7   5   6   4   2   2   3   4\n",
       "C  0  1  2  6  4  5  5  10   8   9   7   5   3   2\n",
       "A  0  3  4  4  5  3  8   8  13  11   9   7   5   3\n",
       "C  0  1  2  7  5  8  6  11  11  16  14  12  10   8\n",
       "G  0  0  0  5  6  6  7   9  10  14  15  17  15  13"
      ]
     },
     "execution_count": 168,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#  Global Alignment Algorithm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----------- global alignment -----------\n",
      "< case 1 >\n",
      "score: 14\n",
      "AGTCCT-A\n",
      "AG-CCTTA\n",
      "< case 2 >\n",
      "score: 14\n",
      "AGTCC-TA\n",
      "AG-CCTTA\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "class global_alignment:\n",
    "    def __init__(self, match_award, mismatch_penalty, gap_penalty):\n",
    "        self.match_award = match_award\n",
    "        self.mismatch_penalty = mismatch_penalty\n",
    "        self.gap_penalty = gap_penalty\n",
    "        self.order_stack = []\n",
    "        self.index_stack = []\n",
    "        \n",
    "    def check_match(self, data, i, j):\n",
    "        if data.index[i] == data.columns[j]: return self.match_award\n",
    "        else: return self.mismatch_penalty\n",
    "    \n",
    "    def make_matrix(self, first_seq, second_seq):\n",
    "        first_seq2 = [\"-\"]\n",
    "        second_seq2 = [\"-\"]\n",
    "        first_seq2.extend(list(first_seq))\n",
    "        second_seq2.extend(list(second_seq))\n",
    "        \n",
    "        data = [np.zeros(len(first_seq2), dtype='int32') for i in second_seq2]\n",
    "        data = pd.DataFrame(data, columns=first_seq2, index=second_seq2)\n",
    "        \n",
    "        # fill 1st row and 1st column with gap_penalty\n",
    "        data.iloc[0,:] = [self.gap_penalty*i for i in range(data.shape[1])]\n",
    "        data.iloc[:,0] = [self.gap_penalty*i for i in range(data.shape[0])]\n",
    "        \n",
    "        # save the route to trace back\n",
    "        flag = []\n",
    "        \n",
    "        # fill out the matrix\n",
    "        for i in range(1, data.shape[0]):\n",
    "            temp = []\n",
    "            for j in range(1, data.shape[1]):\n",
    "                value_score = [data.iloc[i,j-1]+self.gap_penalty,\n",
    "                               data.iloc[i-1,j-1]+self.check_match(data, i, j),\n",
    "                               data.iloc[i-1,j]+self.gap_penalty]\n",
    "                data.iloc[i,j] = max(value_score)\n",
    "                temp.append([data.iloc[i,j] == item for item in value_score])\n",
    "            flag.append(temp)\n",
    "        return data, flag\n",
    "    \n",
    "    def extract_position(self, i, j, position):\n",
    "        if position == 0: return [i,j-1]\n",
    "        elif position == 1: return [i-1,j-1]\n",
    "        else: return [i-1,j]\n",
    "    \n",
    "    def stack_route(self, i, j, data, flag, point):\n",
    "        term = flag[i][j]\n",
    "        cnt = flag[i][j].count(True)\n",
    "        \n",
    "        # stack order\n",
    "        for k in range(cnt): self.order_stack.append(point)\n",
    "        \n",
    "        # stack index\n",
    "        if cnt == 1:\n",
    "            pos = term.index(True)\n",
    "            self.index_stack.append(self.extract_position(i, j, pos))\n",
    "        elif cnt == 2:\n",
    "            term = np.array(term)\n",
    "            term = np.where(term == True)[0]\n",
    "            self.index_stack.append(self.extract_position(i, j, term[0]))\n",
    "            self.index_stack.append(self.extract_position(i, j, term[1]))\n",
    "        elif cnt == 3:\n",
    "            for k in range(3): self.index_stack.append(self.extract_position(i, j, k))\n",
    "        else:\n",
    "            print(\"error: 'cnt' is out of range (1,3)\")\n",
    "        return None\n",
    "    \n",
    "    def backtracking(self, data, flag):\n",
    "        i = len(flag) - 1\n",
    "        j = len(flag[0]) - 1\n",
    "        maximum = max(i, j)\n",
    "        point = 0\n",
    "        key = 0\n",
    "        seq_dic = {str(key):[[i,j]]}\n",
    "        \n",
    "        # make index list\n",
    "        while point < maximum or i >= 0 or j >= 0:\n",
    "            self.stack_route(i, j, data, flag, point)\n",
    "            self.order_stack.pop()\n",
    "            term = self.index_stack.pop()\n",
    "            i, j = term[0], term[1]\n",
    "            if i < 0 or j < 0: break\n",
    "            seq_dic[str(key)].append(term)\n",
    "            point += 1\n",
    "            if point > maximum and len(self.order_stack) != 0:\n",
    "                point = self.order_stack.pop()\n",
    "                term = self.index_stack.pop()\n",
    "                i, j = term[0], term[1]\n",
    "                key += 1\n",
    "                seq_dic[str(key)] = seq_dic[str(key-1)][:point+1]\n",
    "                seq_dic[str(key)].append(term)\n",
    "        \n",
    "        # convert index list to name list\n",
    "        key = 0\n",
    "        for value in seq_dic.values():\n",
    "            row, col = [i for i in zip(*value)]\n",
    "            row = [row[i] if row[i] != row[i+1] else -1 for i in range(len(row)-1)] + [row[-1]]\n",
    "            col = [col[i] if col[i] != col[i+1] else -1 for i in range(len(col)-1)] + [col[-1]]\n",
    "            \n",
    "            row = [data.index[i+1] for i in row]\n",
    "            col = [data.columns[i+1] for i in col]\n",
    "            row = \"\".join(row)[::-1]\n",
    "            col = \"\".join(col)[::-1]\n",
    "            seq_dic[str(key)] = [col, row]\n",
    "            key += 1\n",
    "        return seq_dic\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    # input data\n",
    "    print(\"----------- global alignment -----------\")\n",
    "    #first_seq = input(\"Please enter the first nucleotide sequence: \")\n",
    "    #second_seq = input(\"Please enter the second nucleotide sequence: \")\n",
    "    \n",
    "    first_seq = \"AGTCCTA\"\n",
    "    second_seq = \"AGCCTTA\"\n",
    "    \n",
    "    # assign award and penalties\n",
    "    match_award = 3\n",
    "    mismatch_penalty = -1\n",
    "    gap_penalty = -2\n",
    "    \n",
    "    # global alignment\n",
    "    process = global_alignment(match_award, mismatch_penalty, gap_penalty)\n",
    "    data, flag = process.make_matrix(first_seq, second_seq)\n",
    "    score = data.iloc[-1,-1]\n",
    "    seq_dic = process.backtracking(data, flag)\n",
    "    for key in seq_dic.keys():\n",
    "        print(\"< case\",int(key)+1,\">\")\n",
    "        print(\"score:\", score)\n",
    "        print(seq_dic[key][0])\n",
    "        print(seq_dic[key][1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>-</th>\n",
       "      <th>A</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "      <th>C</th>\n",
       "      <th>C</th>\n",
       "      <th>T</th>\n",
       "      <th>A</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>-</th>\n",
       "      <td>0</td>\n",
       "      <td>-2</td>\n",
       "      <td>-4</td>\n",
       "      <td>-6</td>\n",
       "      <td>-8</td>\n",
       "      <td>-10</td>\n",
       "      <td>-12</td>\n",
       "      <td>-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>-2</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-3</td>\n",
       "      <td>-5</td>\n",
       "      <td>-7</td>\n",
       "      <td>-9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>-4</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>-2</td>\n",
       "      <td>-4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>-6</td>\n",
       "      <td>-1</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>-8</td>\n",
       "      <td>-3</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>10</td>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T</th>\n",
       "      <td>-10</td>\n",
       "      <td>-5</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>13</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T</th>\n",
       "      <td>-12</td>\n",
       "      <td>-7</td>\n",
       "      <td>-2</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>11</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>-14</td>\n",
       "      <td>-9</td>\n",
       "      <td>-4</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    -  A  G  T  C   C   T   A\n",
       "-   0 -2 -4 -6 -8 -10 -12 -14\n",
       "A  -2  3  1 -1 -3  -5  -7  -9\n",
       "G  -4  1  6  4  2   0  -2  -4\n",
       "C  -6 -1  4  5  7   5   3   1\n",
       "C  -8 -3  2  3  8  10   8   6\n",
       "T -10 -5  0  5  6   8  13  11\n",
       "T -12 -7 -2  3  4   6  11  12\n",
       "A -14 -9 -4  1  2   4   9  14"
      ]
     },
     "execution_count": 171,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Needleman and Wunsch Algorithm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "___A_TGAG_AT\n",
      "EXAMPLE 0 : 2a2\n",
      "\n",
      "(array([[  0.,  -3.,  -6.,  -9., -12., -15., -18., -21., -24.],\n",
      "       [ -3.,   2.,  -1.,  -4.,  -7., -10., -13., -16., -19.],\n",
      "       [ -6.,  -1.,   1.,  -2.,  -5.,  -8., -11., -14., -14.],\n",
      "       [ -9.,  -4.,   1.,   3.,   0.,  -3.,  -6.,  -9., -12.],\n",
      "       [-12.,  -7.,  -2.,   0.,   5.,   2.,  -1.,  -4.,  -7.],\n",
      "       [-15., -10.,  -5.,   0.,   2.,   7.,   4.,   1.,  -2.],\n",
      "       [-18., -13.,  -8.,  -3.,   2.,   4.,   9.,   6.,   3.],\n",
      "       [-21., -16., -11.,  -6.,  -1.,   1.,   6.,   8.,   8.]]), array([[0., 1., 1., 1., 1., 1., 1., 1., 1.],\n",
      "       [3., 2., 1., 1., 2., 1., 2., 1., 1.],\n",
      "       [3., 3., 2., 2., 2., 2., 2., 2., 2.],\n",
      "       [3., 3., 2., 2., 1., 2., 1., 2., 1.],\n",
      "       [3., 2., 3., 2., 2., 1., 2., 1., 1.],\n",
      "       [3., 3., 2., 2., 3., 2., 1., 2., 1.],\n",
      "       [3., 2., 3., 3., 2., 3., 2., 1., 1.],\n",
      "       [3., 3., 3., 3., 3., 2., 3., 2., 2.]]), ['ATGAGA_T', 'AGGAGAGT'], [])\n",
      "EXAMPLE 1 : COURSE PAGE 16\n",
      "\n",
      "pairs = [(0, 1), (0, 2), (0, 3), (0, 4), (1, 2), (1, 3), (1, 4), (2, 3), (2, 4), (3, 4)]\n",
      "scores2a2 = [8.0, -2.0, 3.0, 1.0, 1.0, 11.0, -2.0, 1.0, 1.0, -2.0]\n",
      "global scores = [10.0, 18.0, 1.0, 13.0, -2.0]\n",
      "pivot sequence =  AGGAGAGT\n",
      "initial seqs =\n",
      "ATGAGAT\n",
      "AGGAGAGT\n",
      "GGAGG\n",
      "AGGGAGT\n",
      "AGAAC\n",
      "res aligned to pivot =\n",
      "\n",
      "ATGAGA_T\n",
      "AGGAGAGT\n",
      "_GGAG_G_\n",
      "AGG_GAGT\n",
      "A_GA_A_C\n",
      "EXAMPLE 2 : TEST ON SYNTHESIZED DATA\n",
      "\n",
      "TATGGCA\n",
      "TATCGCA\n",
      "TTGCA\n",
      "TTATGGGCA\n",
      "TATGTGCAAC\n",
      "pairs = [(0, 1), (0, 2), (0, 3), (0, 4), (1, 2), (1, 3), (1, 4), (2, 3), (2, 4), (3, 4)]\n",
      "scores2a2 = [11.0, 4.0, 8.0, 5.0, 4.0, 5.0, 2.0, -2.0, -5.0, 4.0]\n",
      "global scores = [28.0, 22.0, 1.0, 15.0, 6.0]\n",
      "pivot sequence =  TATGGCA\n",
      "initial seqs =\n",
      "TATGGCA\n",
      "TATCGCA\n",
      "TTGCA\n",
      "TTATGGGCA\n",
      "TATGTGCAAC\n",
      "res aligned to pivot =\n",
      "\n",
      "_TAT_GGC_A_\n",
      "_TAT_CGC_A_\n",
      "_T_T__GC_A_\n",
      "TTATGGGC_A_\n",
      "_TATGTGCAAC\n"
     ]
    }
   ],
   "source": [
    "# -*- coding: utf-8 -*-\n",
    "\n",
    "import numpy as np\n",
    "import itertools\n",
    "import utils\n",
    "\n",
    "from utils import cost, addgap\n",
    "\n",
    "\n",
    "def align_nw_2by2(seq1, seq2):\n",
    "    \"\"\" Perform 2 by 2 sequence alignment with the Needleman and Wunsch algorithm.\n",
    "    Parameters:\n",
    "    ----------\n",
    "    seq1: string of characters of len=n1\n",
    "        sequence of nucleotides\n",
    "    seq2: string of characters of len=n2\n",
    "        sequence of nucleotides\n",
    "    Return:\n",
    "    ------\n",
    "    scores: matrix of floats of dim (n1+1, n2+1)\n",
    "        \n",
    "    paths: matrix of floats of dim (n1+1, n2+1)\n",
    "        \n",
    "    aligned: list of 2 strings\n",
    "        contains the aligned seq1 and seq2 respectively \n",
    "    L: list of int\n",
    "        list of the positions of the inserted gaps\n",
    "    \"\"\"\n",
    "    n1, n2 = len(seq1), len(seq2)\n",
    "    # initialization: path matrix, score matrix, aligned sequences list, gaps list\n",
    "    paths = np.zeros((n1 + 1, n2 + 1))\n",
    "    scores = np.zeros((n1 + 1, n2 + 1))\n",
    "    aligned = [\"\", \"\"]\n",
    "    L = []\n",
    "    for i in range(1, n1 + 1):  # browsing seq1 indexes\n",
    "        scores[i, 0] = scores[i - 1, 0] - 3\n",
    "        paths[i, 0] = 3\n",
    "        for j in range(1, n2 + 1):  # browsing seq2 indexes\n",
    "            scores[0, j] = scores[0, j - 1] - 3\n",
    "            paths[0, j] = 1\n",
    "            c1 = scores[i - 1, j - 1] + cost(seq1[i - 1], seq2[j - 1])\n",
    "            c2 = scores[i - 1, j] - 3\n",
    "            c3 = scores[i, j - 1] - 3\n",
    "            scores[i, j] = max(c1, c2, c3)\n",
    "            if scores[i, j] == c1:\n",
    "                paths[i, j] = 2\n",
    "            elif scores[i, j] == c2:\n",
    "                paths[i, j] = 3\n",
    "            elif scores[i, j] == c3:\n",
    "                paths[i, j] = 1\n",
    "\n",
    "    while i != 0 or j != 0:\n",
    "        if paths[i, j] == 1:\n",
    "            aligned[0] += '_'\n",
    "            aligned[1] += seq2[j - 1]\n",
    "            j = j - 1\n",
    "        elif paths[i, j] == 2:\n",
    "            aligned[0] += seq1[i - 1]\n",
    "            aligned[1] += seq2[j - 1]\n",
    "            j = j - 1\n",
    "            i = i - 1\n",
    "        elif paths[i, j] == 3:\n",
    "            aligned[0] += seq1[i - 1]\n",
    "            aligned[1] += '_'\n",
    "            L.append(j)  # save gaps introduced by alignment\n",
    "            i = i - 1\n",
    "    aligned[0] = aligned[0][::-1]\n",
    "    aligned[1] = aligned[1][::-1]\n",
    "    return scores, paths, aligned, L\n",
    "\n",
    "\n",
    "def align_star_multiple(seqs, verbose=2):\n",
    "    \"\"\" Perform 2 by 2 sequence alignment with the Needleman and Wunsch algorithm.\n",
    "    Parameters:\n",
    "    ----------\n",
    "    seqs: list of strings of characters\n",
    "        list of sequences of nucleotides to align\n",
    "    verbose: int (default=1)\n",
    "        level of verbosity\n",
    "    Return:\n",
    "    ------\n",
    "    res: \n",
    "    \"\"\"\n",
    "    # enumerate all possible combinations of 2 sequences\n",
    "    pairs = list(itertools.combinations(list(np.arange(len(seqs))), 2))\n",
    "    if verbose > 1:\n",
    "        print(\"pairs =\", pairs)\n",
    "    # compute 2 by 2 scores for all combinaisons\n",
    "    scores2a2 = pairs.copy()\n",
    "    for ip, p in enumerate(pairs):\n",
    "        scores2a2[ip] = align_nw_2by2(seqs[p[0]], seqs[p[1]])[0][-1, -1]\n",
    "    if verbose > 1:\n",
    "        print(\"scores2a2 =\", scores2a2)\n",
    "    # compute global scores for each sequence as sum of 2by2 scores\n",
    "    scores = [0]*len(seqs)\n",
    "    for i in range(len(seqs)):\n",
    "        scores[i] = np.sum([sc for sci, sc in enumerate(scores2a2) if i in pairs[sci]])\n",
    "    if verbose > 1:\n",
    "        print(\"global scores =\", scores)\n",
    "    # choose pivot sequence as the one with the best (highest) global score\n",
    "    imax = scores.index(max(scores))  # index of the pivot sequence\n",
    "    pivot = seqs[imax]  # pivot sequence\n",
    "    if verbose > 0:\n",
    "        print(\"pivot sequence = \", pivot)\n",
    "    # align all w.r.t pivot\n",
    "    if verbose > 0:\n",
    "        print(\"initial seqs =\")\n",
    "        print(\"\\n\".join(str(s) for s in seqs))\n",
    "    res = seqs.copy()\n",
    "    # align all sequences all together\n",
    "    for ir in range(len(res)):\n",
    "        # align sequences to pivot\n",
    "        if res[ir] != pivot:  # the pivot sequence does not need to be aligned to itself\n",
    "            alignment = align_nw_2by2(res[ir], res[imax])\n",
    "            res[ir], res[imax] = alignment[2]  # result of the alignment\n",
    "            L = alignment[3]  # result list gap position to consider\n",
    "            if res[imax] != pivot:  # if new gaps were introduced into the pivot sequence with the 2by2 alignment\n",
    "                pivot = res[imax]  # update the pivot\n",
    "            # and update all the other sequences by inserting the same gaps (positions stored in L)\n",
    "            for i in range(ir):\n",
    "                if res[i] != pivot:\n",
    "                    res[i] = addgap(res[i], L)\n",
    "    if verbose > 0:\n",
    "        print(\"res aligned to pivot =\\n\")\n",
    "        print(\"\\n\".join(str(r) for r in res))\n",
    "    return res\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    from numpy.random import choice\n",
    "\n",
    "    ex = \"ATGAGAT\"\n",
    "    print(addgap(ex, [5, 1, 0, 0, 0]))\n",
    "\n",
    "    print(\"EXAMPLE 0 : 2a2\\n\")\n",
    "    print(align_nw_2by2(\"ATGAGAT\", \"AGGAGAGT\"))\n",
    "    print(\"EXAMPLE 1 : COURSE PAGE 16\\n\")\n",
    "    align_star_multiple([\"ATGAGAT\", \"AGGAGAGT\", \"GGAGG\", \"AGGGAGT\", \"AGAAC\"])\n",
    "    print(\"EXAMPLE 2 : TEST ON SYNTHESIZED DATA\\n\")\n",
    "    alphabet = [\"A\", \"C\", \"G\", \"T\"]\n",
    "    seq0 = choice(alphabet, 7)\n",
    "    seq0 = ''.join(seq0)\n",
    "    print(seq0)\n",
    "    # seq 1 = substitution middle\n",
    "    seq1 = seq0[:]\n",
    "    alphabet.remove(seq1[3])\n",
    "    seq1 = list(seq1)\n",
    "    seq1[3] = choice(alphabet, 1)[0]\n",
    "    seq1 = \"\".join(seq1)\n",
    "    print(seq1)\n",
    "    # seq 2 = deletion pos 5 and end\n",
    "    seq2 = list(seq0[:])\n",
    "    seq2.remove(seq2[4])\n",
    "    seq2.remove(seq2[5])\n",
    "    seq2 = \"\".join(seq2)\n",
    "    print(seq2)\n",
    "    # seq 3 = insertion T beginning and insertion middle\n",
    "    seq3 = seq0[:]\n",
    "    seq3 = \"T\" + seq3\n",
    "    seq3 = seq3[:4] + 'G' + seq3[4:]\n",
    "    print(seq3)\n",
    "    # seq 4 = double insertion AC end and insertion middle\n",
    "    seq4 = seq0[:]\n",
    "    seq4 = seq4 + \"AC\"\n",
    "    seq4 = seq4[:4] + 'T' + seq4[4:]\n",
    "    print(seq4)\n",
    "    # global alignement\n",
    "    align_star_multiple([seq0, seq1, seq2, seq3, seq4])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Smith-Waterman algorithm for local alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " Identities = 59/118 (50.0%), Gaps = 50/118 (42.4%)\n",
      "\n",
      "Query  1     G-AGCAAGTGA-TGC-GAGAGCTGCCCATCCT--------CCAACCAGCATGCCCCTAGA  60  \n",
      "             | ||:|:|  | |:| ||||:|| |  | | |        | ||||  |  |||   |||\n",
      "Sbjct  1     GGAGTACG--ACTTCAGAGATCT-C--A-C-TTGGTGTATC-AACC--C--GCC---AGA  60  \n",
      "\n",
      "Query  61    ------CA--TTG-A-CACTGCATCGGAGT-CAG-GCCAAGAT----CCGCAGGACAG  118 \n",
      "                   ||  ||| | :| || |||  |:| |:| || | |||    |:||:| | ||\n",
      "Sbjct  61    GAGAATCAAATTGGATTA-TG-ATC--AATACTGTGC-A-GATGTGGCTGCTG-A-AG  118 \n",
      "\n"
     ]
    },
    {
     "ename": "SystemExit",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "An exception has occurred, use %tb to see the full traceback.\n",
      "\u001b[0;31mSystemExit\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/subinesh/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2969: UserWarning: To exit: use 'exit', 'quit', or Ctrl-D.\n",
      "  warn(\"To exit: use 'exit', 'quit', or Ctrl-D.\", stacklevel=1)\n"
     ]
    }
   ],
   "source": [
    "'''A Python implementation of the Smith-Waterman algorithm for local alignment\n",
    "of nucleotide sequences.\n",
    "'''\n",
    "\n",
    "import argparse\n",
    "import os\n",
    "import re\n",
    "import sys\n",
    "import unittest\n",
    "\n",
    "\n",
    "# These scores are taken from Wikipedia.\n",
    "# en.wikipedia.org/wiki/Smith%E2%80%93Waterman_algorithm\n",
    "match    = 2\n",
    "mismatch = -1\n",
    "gap      = -1\n",
    "seq1     = \"GTCCAGCAGAGCAAGTGATGCGAGAGCTGCCCATCCTCCAACCAGCATGCCCCTAGACATTGACACTGCATCGGAGTCAGGCCAAGATCCGCAGGACAGT\"\n",
    "seq2     = \"GGAGTACGACTTCAGAGATCTCACTTGGTGTATCAACCCGCCAGAGAGAATCAAATTGGATTATGATCAATACTGTGCAGATGTGGCTGCTGAAGAACTC\"\n",
    "\n",
    "\n",
    "def main():\n",
    "    try:\n",
    "        parse_cmd_line()\n",
    "    except ValueError as err:\n",
    "        print('error:', err)\n",
    "        return\n",
    "\n",
    "    # The scoring matrix contains an extra row and column for the gap (-), hence\n",
    "    # the +1 here.\n",
    "    rows = len(seq1) + 1\n",
    "    cols = len(seq2) + 1\n",
    "\n",
    "    # Initialize the scoring matrix.\n",
    "    score_matrix, start_pos = create_score_matrix(rows, cols)\n",
    "\n",
    "    # Traceback. Find the optimal path through the scoring matrix. This path\n",
    "    # corresponds to the optimal local sequence alignment.\n",
    "    seq1_aligned, seq2_aligned = traceback(score_matrix, start_pos)\n",
    "    assert len(seq1_aligned) == len(seq2_aligned), 'aligned strings are not the same size'\n",
    "\n",
    "    # Pretty print the results. The printing follows the format of BLAST results\n",
    "    # as closely as possible.\n",
    "    alignment_str, idents, gaps, mismatches = alignment_string(seq1_aligned, seq2_aligned)\n",
    "    alength = len(seq1_aligned)\n",
    "    print()\n",
    "    print(' Identities = {0}/{1} ({2:.1%}), Gaps = {3}/{4} ({5:.1%})'.format(idents,\n",
    "          alength, idents / alength, gaps, alength, gaps / alength))\n",
    "    print()\n",
    "    for i in range(0, alength, 60):\n",
    "        seq1_slice = seq1_aligned[i:i+60]\n",
    "        print('Query  {0:<4}  {1}  {2:<4}'.format(i + 1, seq1_slice, i + len(seq1_slice)))\n",
    "        print('             {0}'.format(alignment_str[i:i+60]))\n",
    "        seq2_slice = seq2_aligned[i:i+60]\n",
    "        print('Sbjct  {0:<4}  {1}  {2:<4}'.format(i + 1, seq2_slice, i + len(seq2_slice)))\n",
    "        print()\n",
    "\n",
    "\n",
    "def parse_cmd_line():\n",
    "    '''Parse the command line arguments.\n",
    "\n",
    "    Create a help menu, take input from the command line, and validate the\n",
    "    input by ensuring it does not contain invalid characters (i.e. characters\n",
    "    that aren't the bases A, C, G, or T).\n",
    "    '''\n",
    "\n",
    "\n",
    "\n",
    "    seq1 = \"ATAGACGACATACAGACAGCATACAGACAGCATACAGA\"\n",
    "    seq2 = \"TTTAGCATGCGCATATCAGCAATACAGACAGATACG\"\n",
    "\n",
    "\n",
    "def create_score_matrix(rows, cols):\n",
    "    '''Create a matrix of scores representing trial alignments of the two sequences.\n",
    "\n",
    "    Sequence alignment can be treated as a graph search problem. This function\n",
    "    creates a graph (2D matrix) of scores, which are based on trial alignments\n",
    "    of different base pairs. The path with the highest cummulative score is the\n",
    "    best alignment.\n",
    "    '''\n",
    "    score_matrix = [[0 for col in range(cols)] for row in range(rows)]\n",
    "\n",
    "    # Fill the scoring matrix.\n",
    "    max_score = 0\n",
    "    max_pos   = None    # The row and columbn of the highest score in matrix.\n",
    "    for i in range(1, rows):\n",
    "        for j in range(1, cols):\n",
    "            score = calc_score(score_matrix, i, j)\n",
    "            if score > max_score:\n",
    "                max_score = score\n",
    "                max_pos   = (i, j)\n",
    "\n",
    "            score_matrix[i][j] = score\n",
    "\n",
    "    assert max_pos is not None, 'the x, y position with the highest score was not found'\n",
    "\n",
    "    return score_matrix, max_pos\n",
    "\n",
    "\n",
    "def calc_score(matrix, x, y):\n",
    "    '''Calculate score for a given x, y position in the scoring matrix.\n",
    "\n",
    "    The score is based on the up, left, and upper-left neighbors.\n",
    "    '''\n",
    "    similarity = match if seq1[x - 1] == seq2[y - 1] else mismatch\n",
    "\n",
    "    diag_score = matrix[x - 1][y - 1] + similarity\n",
    "    up_score   = matrix[x - 1][y] + gap\n",
    "    left_score = matrix[x][y - 1] + gap\n",
    "\n",
    "    return max(0, diag_score, up_score, left_score)\n",
    "\n",
    "\n",
    "def traceback(score_matrix, start_pos):\n",
    "    '''Find the optimal path through the matrix.\n",
    "\n",
    "    This function traces a path from the bottom-right to the top-left corner of\n",
    "    the scoring matrix. Each move corresponds to a match, mismatch, or gap in one\n",
    "    or both of the sequences being aligned. Moves are determined by the score of\n",
    "    three adjacent squares: the upper square, the left square, and the diagonal\n",
    "    upper-left square.\n",
    "\n",
    "    WHAT EACH MOVE REPRESENTS\n",
    "        diagonal: match/mismatch\n",
    "        up:       gap in sequence 1\n",
    "        left:     gap in sequence 2\n",
    "    '''\n",
    "\n",
    "    END, DIAG, UP, LEFT = range(4)\n",
    "    aligned_seq1 = []\n",
    "    aligned_seq2 = []\n",
    "    x, y         = start_pos\n",
    "    move         = next_move(score_matrix, x, y)\n",
    "    while move != END:\n",
    "        if move == DIAG:\n",
    "            aligned_seq1.append(seq1[x - 1])\n",
    "            aligned_seq2.append(seq2[y - 1])\n",
    "            x -= 1\n",
    "            y -= 1\n",
    "        elif move == UP:\n",
    "            aligned_seq1.append(seq1[x - 1])\n",
    "            aligned_seq2.append('-')\n",
    "            x -= 1\n",
    "        else:\n",
    "            aligned_seq1.append('-')\n",
    "            aligned_seq2.append(seq2[y - 1])\n",
    "            y -= 1\n",
    "\n",
    "        move = next_move(score_matrix, x, y)\n",
    "\n",
    "    aligned_seq1.append(seq1[x - 1])\n",
    "    aligned_seq2.append(seq1[y - 1])\n",
    "\n",
    "    return ''.join(reversed(aligned_seq1)), ''.join(reversed(aligned_seq2))\n",
    "\n",
    "\n",
    "def next_move(score_matrix, x, y):\n",
    "    diag = score_matrix[x - 1][y - 1]\n",
    "    up   = score_matrix[x - 1][y]\n",
    "    left = score_matrix[x][y - 1]\n",
    "    if diag >= up and diag >= left:     # Tie goes to the DIAG move.\n",
    "        return 1 if diag != 0 else 0    # 1 signals a DIAG move. 0 signals the end.\n",
    "    elif up > diag and up >= left:      # Tie goes to UP move.\n",
    "        return 2 if up != 0 else 0      # UP move or end.\n",
    "    elif left > diag and left > up:\n",
    "        return 3 if left != 0 else 0    # LEFT move or end.\n",
    "    else:\n",
    "        # Execution should not reach here.\n",
    "        raise ValueError('invalid move during traceback')\n",
    "\n",
    "\n",
    "def alignment_string(aligned_seq1, aligned_seq2):\n",
    "    '''Construct a special string showing identities, gaps, and mismatches.\n",
    "\n",
    "    This string is printed between the two aligned sequences and shows the\n",
    "    identities (|), gaps (-), and mismatches (:). As the string is constructed,\n",
    "    it also counts number of identities, gaps, and mismatches and returns the\n",
    "    counts along with the alignment string.\n",
    "\n",
    "    AAGGATGCCTCAAATCGATCT-TTTTCTTGG-\n",
    "    ::||::::::||:|::::::: |:  :||:|   <-- alignment string\n",
    "    CTGGTACTTGCAGAGAAGGGGGTA--ATTTGG\n",
    "    '''\n",
    "    # Build the string as a list of characters to avoid costly string\n",
    "    # concatenation.\n",
    "    idents, gaps, mismatches = 0, 0, 0\n",
    "    alignment_string = []\n",
    "    for base1, base2 in zip(aligned_seq1, aligned_seq2):\n",
    "        if base1 == base2:\n",
    "            alignment_string.append('|')\n",
    "            idents += 1\n",
    "        elif '-' in (base1, base2):\n",
    "            alignment_string.append(' ')\n",
    "            gaps += 1\n",
    "        else:\n",
    "            alignment_string.append(':')\n",
    "            mismatches += 1\n",
    "\n",
    "    return ''.join(alignment_string), idents, gaps, mismatches\n",
    "\n",
    "\n",
    "def print_matrix(matrix):\n",
    "    '''Print the scoring matrix.\n",
    "\n",
    "    ex:\n",
    "    0   0   0   0   0   0\n",
    "    0   2   1   2   1   2\n",
    "    0   1   1   1   1   1\n",
    "    0   0   3   2   3   2\n",
    "    0   2   2   5   4   5\n",
    "    0   1   4   4   7   6\n",
    "    '''\n",
    "    for row in matrix:\n",
    "        for col in row:\n",
    "            print('{0:>4}'.format(col))\n",
    "        print()\n",
    "\n",
    "\n",
    "class ScoreMatrixTest(unittest.TestCase):\n",
    "    '''Compare the matrix produced by create_score_matrix() with a known matrix.'''\n",
    "    def test_matrix(self):\n",
    "        #                -   A   C   A   C   A   C   T   A\n",
    "        known_matrix = [[0,  0,  0,  0,  0,  0,  0,  0,  0],  # -\n",
    "                        [0,  2,  1,  2,  1,  2,  1,  0,  2],  # A\n",
    "                        [0,  1,  1,  1,  1,  1,  1,  0,  1],  # G\n",
    "                        [0,  0,  3,  2,  3,  2,  3,  2,  1],  # C\n",
    "                        [0,  2,  2,  5,  4,  5,  4,  3,  4],  # A\n",
    "                        [0,  1,  4,  4,  7,  6,  7,  6,  5],  # C\n",
    "                        [0,  2,  3,  6,  6,  9,  8,  7,  8],  # A\n",
    "                        [0,  1,  4,  5,  8,  8, 11, 10,  9],  # C\n",
    "                        [0,  2,  3,  6,  7, 10, 10, 10, 12]]  # A\n",
    "\n",
    "        global seq1, seq2\n",
    "        seq1 = 'AGCACACA'\n",
    "        seq2 = 'ACACACTA'\n",
    "        rows = len(seq1) + 1\n",
    "        cols = len(seq2) + 1\n",
    "\n",
    "        matrix_to_test, max_pos = create_score_matrix(rows, cols)\n",
    "        self.assertEqual(known_matrix, matrix_to_test)\n",
    "\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    sys.exit(main())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# MUTATION-DETECTION:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 93 mutation(s) for id NP_001273979.1\n",
      "Found 165 mutation(s) for id NP_001017978.1\n",
      "Found 165 mutation(s) for id sp|Q5H943.1|KKLC1_HUMAN\n",
      "Found 134 mutation(s) for id AAC35497.1\n",
      "Found 174 mutation(s) for id XP_002699503.1\n",
      "Found 161 mutation(s) for id sp|Q4R717.1|KKLC1_MACFA\n",
      "Found 138 mutation(s) for id AAN12271.1\n",
      "Found 161 mutation(s) for id NP_001270680.1\n",
      "Found 400 mutation(s) for id Sp|P63244.3|Rack1_Human\n",
      "Found 174 mutation(s) for id XP_027389897.1\n",
      "Found 169 mutation(s) for id XP_007647427.1\n",
      "Found 187 mutation(s) for id XP_014918346.1\n",
      "Found 174 mutation(s) for id XP_0026997503.1\n",
      "Found 89 mutation(s) for id XP_002699603.1\n",
      "Found 66 mutation(s) for id XP_004699503.1\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "normal_samples = SeqIO.parse(\"DNA.fasta\", \"fasta\") \n",
    "treated_samples = SeqIO.parse(\"Mutated.fasta\", \"fasta\")\n",
    "\n",
    "for normal, treated in zip(normal_samples, treated_samples): \n",
    "    if normal.id == treated.id: \n",
    "        mutations = sum(1 for n, t in zip(str(normal.seq), str(treated.seq)) if n != t) \n",
    "        print(f\"Found {mutations} mutation(s) for id {normal.id}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# OPTIONAL:--"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1.SEQUENTIAL PATTERN MINING"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "What min. support do you want to use? \n",
      "5\n",
      "\n",
      "**** Sequence with minSupport = 5 ****\n",
      "\n",
      "Sets:\n",
      "\n",
      "frozenset({''})\n",
      "frozenset({'T'})\n",
      "frozenset({'G'})\n",
      "frozenset({'C'})\n",
      "frozenset({'A'})\n",
      "frozenset({'', 'T'})\n",
      "frozenset({'', 'G'})\n",
      "frozenset({'', 'C'})\n",
      "frozenset({'', 'A'})\n",
      "frozenset({'G', 'T'})\n",
      "frozenset({'T', 'C'})\n",
      "frozenset({'A', 'T'})\n",
      "frozenset({'G', 'C'})\n",
      "frozenset({'A', 'G'})\n",
      "frozenset({'A', 'C'})\n",
      "frozenset({'', 'T', 'G'})\n",
      "frozenset({'', 'T', 'C'})\n",
      "frozenset({'', 'T', 'A'})\n",
      "frozenset({'', 'G', 'C'})\n",
      "frozenset({'', 'G', 'A'})\n",
      "frozenset({'', 'A', 'C'})\n",
      "frozenset({'G', 'T', 'C'})\n",
      "frozenset({'A', 'T', 'G'})\n",
      "frozenset({'A', 'T', 'C'})\n",
      "frozenset({'A', 'G', 'C'})\n",
      "frozenset({'', 'G', 'C', 'T'})\n",
      "frozenset({'', 'A', 'G', 'T'})\n",
      "frozenset({'', 'A', 'C', 'T'})\n",
      "frozenset({'', 'A', 'G', 'C'})\n",
      "frozenset({'A', 'G', 'C', 'T'})\n",
      "frozenset({'', 'C', 'T', 'A', 'G'})\n",
      "\n",
      "Counts:\n",
      "\n",
      "frozenset({'A'}) 81\n",
      "frozenset({'C'}) 81\n",
      "frozenset({'G'}) 81\n",
      "frozenset({'T'}) 82\n",
      "frozenset({''}) 77\n",
      "frozenset({'G', 'T'}) 81\n",
      "frozenset({'T', 'C'}) 81\n",
      "frozenset({'A', 'T'}) 81\n",
      "frozenset({'G', 'C'}) 81\n",
      "frozenset({'A', 'G'}) 81\n",
      "frozenset({'A', 'C'}) 81\n",
      "frozenset({'', 'T'}) 77\n",
      "frozenset({'', 'G'}) 77\n",
      "frozenset({'', 'C'}) 77\n",
      "frozenset({'', 'A'}) 77\n",
      "frozenset({'G', 'T', 'C'}) 81\n",
      "frozenset({'A', 'T', 'G'}) 81\n",
      "frozenset({'A', 'T', 'C'}) 81\n",
      "frozenset({'A', 'G', 'C'}) 81\n",
      "frozenset({'', 'T', 'G'}) 77\n",
      "frozenset({'', 'T', 'C'}) 77\n",
      "frozenset({'', 'T', 'A'}) 77\n",
      "frozenset({'', 'G', 'C'}) 77\n",
      "frozenset({'', 'G', 'A'}) 77\n",
      "frozenset({'', 'A', 'C'}) 77\n",
      "frozenset({'A', 'G', 'C', 'T'}) 81\n",
      "frozenset({'', 'G', 'C', 'T'}) 77\n",
      "frozenset({'', 'A', 'G', 'T'}) 77\n",
      "frozenset({'', 'A', 'C', 'T'}) 77\n",
      "frozenset({'', 'A', 'G', 'C'}) 77\n",
      "frozenset({'', 'C', 'T', 'A', 'G'}) 77\n"
     ]
    }
   ],
   "source": [
    "def createCandidateSet(data):\n",
    "    cand = []\n",
    "    for row in data:\n",
    "        for itm in row:\n",
    "            if [itm] not in cand:\n",
    "                cand.append([itm])\n",
    "    cand.sort()\n",
    "    return list(map(frozenset,cand))\n",
    "\n",
    "def scanData(data, candidateSet, minSupport):\n",
    "    subsetCount = {}\n",
    "    for curSet in data:\n",
    "        for cand in candidateSet:\n",
    "            if cand.issubset(curSet):\n",
    "                if not cand in subsetCount:\n",
    "                    subsetCount[cand] = 1\n",
    "                else:\n",
    "                    subsetCount[cand] += 1\n",
    "    n = float(len(data))\n",
    "    valid = []\n",
    "    for key in subsetCount:\n",
    "        sup = subsetCount[key]\n",
    "        if sup >= minSupport:\n",
    "            valid.insert(0,key)\n",
    "    return valid, subsetCount\n",
    "\n",
    "def genSequential(freqSets, k):\n",
    "    valid = []\n",
    "    nFreqSets = len(freqSets)\n",
    "    for i in range(nFreqSets):\n",
    "        for j in range(i+1, nFreqSets):\n",
    "            lstCands1 = list(freqSets[i])[:k-2]\n",
    "            lstCands2 = list(freqSets[j])[:k-2]\n",
    "            lstCands1.sort()\n",
    "            lstCands2.sort()\n",
    "            # if first k-2 elements are equal\n",
    "            if lstCands1 == lstCands2:\n",
    "                valid.append(freqSets[i]|freqSets[j]) # union \n",
    "    return valid\n",
    "\n",
    "def Seqtial(data, minSupport):\n",
    "    candSet = createCandidateSet(data)\n",
    "    setData = list(map(set,data))\n",
    "    lstCands, subsetCounts = scanData(setData,candSet,minSupport)\n",
    "    lstCands = [lstCands]\n",
    "    k = 2\n",
    "    while(len(lstCands[k-2]) > 0):\n",
    "        candSetX = genSequential(lstCands[k-2],k)\n",
    "        lstCandsX, subsetCountsX = scanData(setData,candSetX, minSupport)\n",
    "        subsetCounts.update(subsetCountsX)\n",
    "        lstCands.append(candSetX)\n",
    "        k += 1\n",
    "    return lstCands, subsetCounts\n",
    "\n",
    "# read in data\n",
    "data = ['ATGC','ACGT','ATCG','TTTT','AGTCG']\n",
    "dataSetFilename = 'dna.fasta'\n",
    "with open(dataSetFilename,'r') as file:\n",
    "    for line in file:\n",
    "        data.append(line.strip().split(','))\n",
    "\n",
    "\n",
    "print(\"What min. support do you want to use? \")\n",
    "minSupp = input()\n",
    "minSupp = int(minSupp)\n",
    "\n",
    "print(\"\\n**** Sequence with minSupport = {} ****\".format(minSupp))\n",
    "\n",
    "# call SPM\n",
    "sets, counts = Seqtial(data,minSupp)\n",
    "print(\"\\nSets:\\n\")\n",
    "for x in sets:\n",
    "    for y in x:\n",
    "        print(y)\n",
    "\n",
    "print(\"\\nCounts:\\n\")\n",
    "for k,v in counts.items():\n",
    "\tprint(k, v)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2.DNA TO PROTEIN TRANSLATION"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FIRST\n",
      "[SeqRecord(seq=Seq('MKTKCICELCSCGRHHCPHLPTKIYDKTEKPCLLSEYTENYPFYHSYLPRESFK...TSS', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "SECOND\n",
      "[SeqRecord(seq=Seq('*KPNAFANCAAAAAIIARICRPKFMIKPKNRAC*ANIPKTIRFIIAICRAKALN...PAA', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "THIRD\n",
      "[SeqRecord(seq=Seq('ENQMHLRTVQLRPPSLPASADQNL**NRKTVPAERIYRKLSVLS*LSAARKL*T...QQX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "FIRST\n",
      "[SeqRecord(seq=Seq('MKTKCICELCSCGRHHCPHLPTKIYDKTEKPCLLSEYTENYPFYHSYLPRESFK...TSS', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNFYLLLASSILCALIVFWKYRRFQRNTGEMSSNSTALALVRPSSSGLINSNTD...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "SECOND\n",
      "[SeqRecord(seq=Seq('*KPNAFANCAAAAAIIARICRPKFMIKPKNRAC*ANIPKTIRFIIAICRAKALN...PAA', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TFICCWRAAFCAR*LCFGNIAAFSATPAK*AATAPRWRWCARAAAA*LTATPI...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "THIRD\n",
      "[SeqRecord(seq=Seq('ENQMHLRTVQLRPPSLPASADQNL**NRKTVPAERIYRKLSVLS*LSAARKL*T...QQX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELLSAAGEQHSVRADCVLEISPLSAQHRRNEQQQHRAGAGAPEQQRPD*QQHR*...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "FIRST\n",
      "[SeqRecord(seq=Seq('MKTKCICELCSCGRHHCPHLPTKIYDKTEKPCLLSEYTENYPFYHSYLPRESFK...TSS', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNFYLLLASSILCALIVFWKYRRFQRNTGEMSSNSTALALVRPSSSGLINSNTD...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNFYLLLASSILCALIVFWKYRRFQRNTGEMSSNSTALALVRPSSSGLINSNTD...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "SECOND\n",
      "[SeqRecord(seq=Seq('*KPNAFANCAAAAAIIARICRPKFMIKPKNRAC*ANIPKTIRFIIAICRAKALN...PAA', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TFICCWRAAFCAR*LCFGNIAAFSATPAK*AATAPRWRWCARAAAA*LTATPI...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TFICCWRAAFCAR*LCFGNIAAFSATPAK*AATAPRWRWCARAAAA*LTATPI...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "THIRD\n",
      "[SeqRecord(seq=Seq('ENQMHLRTVQLRPPSLPASADQNL**NRKTVPAERIYRKLSVLS*LSAARKL*T...QQX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELLSAAGEQHSVRADCVLEISPLSAQHRRNEQQQHRAGAGAPEQQRPD*QQHR*...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELLSAAGEQHSVRADCVLEISPLSAQHRRNEQQQHRAGAGAPEQQRPD*QQHR*...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "FIRST\n",
      "[SeqRecord(seq=Seq('MKTKCICELCSCGRHHCPHLPTKIYDKTEKPCLLSEYTENYPFYHSYLPRESFK...TSS', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNFYLLLASSILCALIVFWKYRRFQRNTGEMSSNSTALALVRPSSSGLINSNTD...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNFYLLLASSILCALIVFWKYRRFQRNTGEMSSNSTALALVRPSSSGLINSNTD...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MGASGSKARGLWPFASAAGGGGSEAAGAEQALVRPRGRAVPPFVFTRRGSMFYD...YEV', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "SECOND\n",
      "[SeqRecord(seq=Seq('*KPNAFANCAAAAAIIARICRPKFMIKPKNRAC*ANIPKTIRFIIAICRAKALN...PAA', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TFICCWRAAFCAR*LCFGNIAAFSATPAK*AATAPRWRWCARAAAA*LTATPI...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TFICCWRAAFCAR*LCFGNIAAFSATPAK*AATAPRWRWCARAAAA*LTATPI...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('WARAAAKRAACGRLRARRAAAAAKRRARNRRWCARAAARCRRLCLPAAAACFMM...MKX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "THIRD\n",
      "[SeqRecord(seq=Seq('ENQMHLRTVQLRPPSLPASADQNL**NRKTVPAERIYRKLSVLS*LSAARKL*T...QQX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELLSAAGEQHSVRADCVLEISPLSAQHRRNEQQQHRAGAGAPEQQRPD*QQHR*...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELLSAAGEQHSVRADCVLEISPLSAQHRRNEQQQHRAGAGAPEQQRPD*QQHR*...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('GRERQQSARPVAVCERGGRRRQRSGGRGTGAGAPARPRGAAVCVYPPRQHVL**...*SX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "FIRST\n",
      "[SeqRecord(seq=Seq('MKTKCICELCSCGRHHCPHLPTKIYDKTEKPCLLSEYTENYPFYHSYLPRESFK...TSS', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNFYLLLASSILCALIVFWKYRRFQRNTGEMSSNSTALALVRPSSSGLINSNTD...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNFYLLLASSILCALIVFWKYRRFQRNTGEMSSNSTALALVRPSSSGLINSNTD...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MGASGSKARGLWPFASAAGGGGSEAAGAEQALVRPRGRAVPPFVFTRRGSMFYD...YEV', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MSFLFLLVSSILFAFMFVFWKTWFQRNTGEMSSNSTSLALVRSSSSTQSTKSNI...TEV', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "SECOND\n",
      "[SeqRecord(seq=Seq('*KPNAFANCAAAAAIIARICRPKFMIKPKNRAC*ANIPKTIRFIIAICRAKALN...PAA', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TFICCWRAAFCAR*LCFGNIAAFSATPAK*AATAPRWRWCARAAAA*LTATPI...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TFICCWRAAFCAR*LCFGNIAAFSATPAK*AATAPRWRWCARAAAA*LTATPI...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('WARAAAKRAACGRLRARRAAAAAKRRARNRRWCARAAARCRRLCLPAAAACFMM...MKX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*AFCFCW*AAFCLRLCLCFGKPGFSATPAK*AATAPAWRWCAAAAAPRAPKATL...PKX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "THIRD\n",
      "[SeqRecord(seq=Seq('ENQMHLRTVQLRPPSLPASADQNL**NRKTVPAERIYRKLSVLS*LSAARKL*T...QQX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELLSAAGEQHSVRADCVLEISPLSAQHRRNEQQQHRAGAGAPEQQRPD*QQHR*...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELLSAAGEQHSVRADCVLEISPLSAQHRRNEQQQHRAGAGAPEQQRPD*QQHR*...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('GRERQQSARPVAVCERGGRRRQRSGGRGTGAGAPARPRGAAVCVYPPRQHVL**...*SX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELSVSAGEQHSVCVYVCVLENLVSAQHRRNEQQQHQPGAGAQQQQHPEHQKQH*...RSX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "FIRST\n",
      "[SeqRecord(seq=Seq('MKTKCICELCSCGRHHCPHLPTKIYDKTEKPCLLSEYTENYPFYHSYLPRESFK...TSS', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNFYLLLASSILCALIVFWKYRRFQRNTGEMSSNSTALALVRPSSSGLINSNTD...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNFYLLLASSILCALIVFWKYRRFQRNTGEMSSNSTALALVRPSSSGLINSNTD...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MGASGSKARGLWPFASAAGGGGSEAAGAEQALVRPRGRAVPPFVFTRRGSMFYD...YEV', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MSFLFLLVSSILFAFMFVFWKTWFQRNTGEMSSNSTSLALVRSSSSTQSTKSNI...TEV', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNVYLLLASGILCALMTVFWKYRRFQRNTGEMSSNSTALALVRPSSTGLINSNT...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "SECOND\n",
      "[SeqRecord(seq=Seq('*KPNAFANCAAAAAIIARICRPKFMIKPKNRAC*ANIPKTIRFIIAICRAKALN...PAA', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TFICCWRAAFCAR*LCFGNIAAFSATPAK*AATAPRWRWCARAAAA*LTATPI...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TFICCWRAAFCAR*LCFGNIAAFSATPAK*AATAPRWRWCARAAAA*LTATPI...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('WARAAAKRAACGRLRARRAAAAAKRRARNRRWCARAAARCRRLCLPAAAACFMM...MKX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*AFCFCW*AAFCLRLCLCFGKPGFSATPAK*AATAPAWRWCAAAAAPRAPKATL...PKX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TCICCWRAAFCAR**PCFGNIAAFSATPAK*AATAPRWRWCARAAPA*LTATP...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "THIRD\n",
      "[SeqRecord(seq=Seq('ENQMHLRTVQLRPPSLPASADQNL**NRKTVPAERIYRKLSVLS*LSAARKL*T...QQX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELLSAAGEQHSVRADCVLEISPLSAQHRRNEQQQHRAGAGAPEQQRPD*QQHR*...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELLSAAGEQHSVRADCVLEISPLSAQHRRNEQQQHRAGAGAPEQQRPD*QQHR*...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('GRERQQSARPVAVCERGGRRRQRSGGRGTGAGAPARPRGAAVCVYPPRQHVL**...*SX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELSVSAGEQHSVCVYVCVLENLVSAQHRRNEQQQHQPGAGAQQQQHPEHQKQH*...RSX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ERVSAAGERHSVRADDRVLEISPLSAQHRRNEQQQHRAGAGAPEQHRPD*QQHR...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "FIRST\n",
      "[SeqRecord(seq=Seq('MKTKCICELCSCGRHHCPHLPTKIYDKTEKPCLLSEYTENYPFYHSYLPRESFK...TSS', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNFYLLLASSILCALIVFWKYRRFQRNTGEMSSNSTALALVRPSSSGLINSNTD...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNFYLLLASSILCALIVFWKYRRFQRNTGEMSSNSTALALVRPSSSGLINSNTD...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MGASGSKARGLWPFASAAGGGGSEAAGAEQALVRPRGRAVPPFVFTRRGSMFYD...YEV', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MSFLFLLVSSILFAFMFVFWKTWFQRNTGEMSSNSTSLALVRSSSSTQSTKSNI...TEV', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNVYLLLASGILCALMTVFWKYRRFQRNTGEMSSNSTALALVRPSSTGLINSNT...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MLLTRFEQKNGHLAQVEVDEVFGFMCHITTEVPPHDAMPGGIVLLVKFLLDMSR...GRC', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "SECOND\n",
      "[SeqRecord(seq=Seq('*KPNAFANCAAAAAIIARICRPKFMIKPKNRAC*ANIPKTIRFIIAICRAKALN...PAA', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TFICCWRAAFCAR*LCFGNIAAFSATPAK*AATAPRWRWCARAAAA*LTATPI...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TFICCWRAAFCAR*LCFGNIAAFSATPAK*AATAPRWRWCARAAAA*LTATPI...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('WARAAAKRAACGRLRARRAAAAAKRRARNRRWCARAAARCRRLCLPAAAACFMM...MKX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*AFCFCW*AAFCLRLCLCFGKPGFSATPAK*AATAPAWRWCAAAAAPRAPKATL...PKX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TCICCWRAAFCAR**PCFGNIAAFSATPAK*AATAPRWRWCARAAPA*LTATP...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('CC*PALNRKTAIWRRWKWMKCLALCAILPPKCRRMMRCRAALCCW*NFCWI*AA...AAA', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "THIRD\n",
      "[SeqRecord(seq=Seq('ENQMHLRTVQLRPPSLPASADQNL**NRKTVPAERIYRKLSVLS*LSAARKL*T...QQX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELLSAAGEQHSVRADCVLEISPLSAQHRRNEQQQHRAGAGAPEQQRPD*QQHR*...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELLSAAGEQHSVRADCVLEISPLSAQHRRNEQQQHRAGAGAPEQQRPD*QQHR*...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('GRERQQSARPVAVCERGGRRRQRSGGRGTGAGAPARPRGAAVCVYPPRQHVL**...*SX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELSVSAGEQHSVCVYVCVLENLVSAQHRRNEQQQHQPGAGAQQQQHPEHQKQH*...RSX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ERVSAAGERHSVRADDRVLEISPLSAQHRRNEQQQHRAGAGAPEQHRPD*QQHR...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('AADPL*TEKRPSGAGGSG*SVWLYVPYYHRSAAA*CDAGRHCAAGEISAGYEPQ...PLX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "FIRST\n",
      "[SeqRecord(seq=Seq('MKTKCICELCSCGRHHCPHLPTKIYDKTEKPCLLSEYTENYPFYHSYLPRESFK...TSS', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNFYLLLASSILCALIVFWKYRRFQRNTGEMSSNSTALALVRPSSSGLINSNTD...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNFYLLLASSILCALIVFWKYRRFQRNTGEMSSNSTALALVRPSSSGLINSNTD...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MGASGSKARGLWPFASAAGGGGSEAAGAEQALVRPRGRAVPPFVFTRRGSMFYD...YEV', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MSFLFLLVSSILFAFMFVFWKTWFQRNTGEMSSNSTSLALVRSSSSTQSTKSNI...TEV', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNVYLLLASGILCALMTVFWKYRRFQRNTGEMSSNSTALALVRPSSTGLINSNT...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MLLTRFEQKNGHLAQVEVDEVFGFMCHITTEVPPHDAMPGGIVLLVKFLLDMSR...GRC', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNVYLLLASGILCALMTVFWKYRRFQRNTGEMSSNSTALALVRPSSTGLINSNT...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "SECOND\n",
      "[SeqRecord(seq=Seq('*KPNAFANCAAAAAIIARICRPKFMIKPKNRAC*ANIPKTIRFIIAICRAKALN...PAA', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TFICCWRAAFCAR*LCFGNIAAFSATPAK*AATAPRWRWCARAAAA*LTATPI...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TFICCWRAAFCAR*LCFGNIAAFSATPAK*AATAPRWRWCARAAAA*LTATPI...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('WARAAAKRAACGRLRARRAAAAAKRRARNRRWCARAAARCRRLCLPAAAACFMM...MKX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*AFCFCW*AAFCLRLCLCFGKPGFSATPAK*AATAPAWRWCAAAAAPRAPKATL...PKX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TCICCWRAAFCAR**PCFGNIAAFSATPAK*AATAPRWRWCARAAPA*LTATP...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('CC*PALNRKTAIWRRWKWMKCLALCAILPPKCRRMMRCRAALCCW*NFCWI*AA...AAA', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TCICCWRAAFCAR**PCFGNIAAFSATPAK*AATAPRWRWCARAAPA*LTATP...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "THIRD\n",
      "[SeqRecord(seq=Seq('ENQMHLRTVQLRPPSLPASADQNL**NRKTVPAERIYRKLSVLS*LSAARKL*T...QQX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELLSAAGEQHSVRADCVLEISPLSAQHRRNEQQQHRAGAGAPEQQRPD*QQHR*...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELLSAAGEQHSVRADCVLEISPLSAQHRRNEQQQHRAGAGAPEQQRPD*QQHR*...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('GRERQQSARPVAVCERGGRRRQRSGGRGTGAGAPARPRGAAVCVYPPRQHVL**...*SX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELSVSAGEQHSVCVYVCVLENLVSAQHRRNEQQQHQPGAGAQQQQHPEHQKQH*...RSX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ERVSAAGERHSVRADDRVLEISPLSAQHRRNEQQQHRAGAGAPEQHRPD*QQHR...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('AADPL*TEKRPSGAGGSG*SVWLYVPYYHRSAAA*CDAGRHCAAGEISAGYEPQ...PLX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ERVSAAGERHSVRADDRVLEISPLSAQHRRNEQQQHRAGAGAPEQHRPD*QQHR...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "FIRST\n",
      "[SeqRecord(seq=Seq('MKTKCICELCSCGRHHCPHLPTKIYDKTEKPCLLSEYTENYPFYHSYLPRESFK...TSS', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNFYLLLASSILCALIVFWKYRRFQRNTGEMSSNSTALALVRPSSSGLINSNTD...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNFYLLLASSILCALIVFWKYRRFQRNTGEMSSNSTALALVRPSSSGLINSNTD...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MGASGSKARGLWPFASAAGGGGSEAAGAEQALVRPRGRAVPPFVFTRRGSMFYD...YEV', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MSFLFLLVSSILFAFMFVFWKTWFQRNTGEMSSNSTSLALVRSSSSTQSTKSNI...TEV', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNVYLLLASGILCALMTVFWKYRRFQRNTGEMSSNSTALALVRPSSTGLINSNT...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MLLTRFEQKNGHLAQVEVDEVFGFMCHITTEVPPHDAMPGGIVLLVKFLLDMSR...GRC', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNVYLLLASGILCALMTVFWKYRRFQRNTGEMSSNSTALALVRPSSTGLINSNT...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MTEQMTLRGTLKGHNGWVTQIATTPQFPDMILSASRDKTIIMWKLTRDETNYGI...GTR', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "SECOND\n",
      "[SeqRecord(seq=Seq('*KPNAFANCAAAAAIIARICRPKFMIKPKNRAC*ANIPKTIRFIIAICRAKALN...PAA', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TFICCWRAAFCAR*LCFGNIAAFSATPAK*AATAPRWRWCARAAAA*LTATPI...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TFICCWRAAFCAR*LCFGNIAAFSATPAK*AATAPRWRWCARAAAA*LTATPI...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('WARAAAKRAACGRLRARRAAAAAKRRARNRRWCARAAARCRRLCLPAAAACFMM...MKX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*AFCFCW*AAFCLRLCLCFGKPGFSATPAK*AATAPAWRWCAAAAAPRAPKATL...PKX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TCICCWRAAFCAR**PCFGNIAAFSATPAK*AATAPRWRWCARAAPA*LTATP...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('CC*PALNRKTAIWRRWKWMKCLALCAILPPKCRRMMRCRAALCCW*NFCWI*AA...AAA', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TCICCWRAAFCAR**PCFGNIAAFSATPAK*AATAPRWRWCARAAPA*LTATP...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*PNR*PCAAP*KAITAG*PRLRPPRSFRI*F*ARAAIKPLLCGN*PAMKPTMAF...APA', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "THIRD\n",
      "[SeqRecord(seq=Seq('ENQMHLRTVQLRPPSLPASADQNL**NRKTVPAERIYRKLSVLS*LSAARKL*T...QQX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELLSAAGEQHSVRADCVLEISPLSAQHRRNEQQQHRAGAGAPEQQRPD*QQHR*...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELLSAAGEQHSVRADCVLEISPLSAQHRRNEQQQHRAGAGAPEQQRPD*QQHR*...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('GRERQQSARPVAVCERGGRRRQRSGGRGTGAGAPARPRGAAVCVYPPRQHVL**...*SX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELSVSAGEQHSVCVYVCVLENLVSAQHRRNEQQQHQPGAGAQQQQHPEHQKQH*...RSX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ERVSAAGERHSVRADDRVLEISPLSAQHRRNEQQQHRAGAGAPEQHRPD*QQHR...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('AADPL*TEKRPSGAGGSG*SVWLYVPYYHRSAAA*CDAGRHCAAGEISAGYEPQ...PLX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ERVSAAGERHSVRADDRVLEISPLSAQHRRNEQQQHRAGAGAPEQHRPD*QQHR...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('DRTDDPARHPERP*RLGDPDCDHPAVSGYDSEREPR*NHYYVETDPR*NQLWHS...HPX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "FIRST\n",
      "[SeqRecord(seq=Seq('MKTKCICELCSCGRHHCPHLPTKIYDKTEKPCLLSEYTENYPFYHSYLPRESFK...TSS', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNFYLLLASSILCALIVFWKYRRFQRNTGEMSSNSTALALVRPSSSGLINSNTD...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNFYLLLASSILCALIVFWKYRRFQRNTGEMSSNSTALALVRPSSSGLINSNTD...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MGASGSKARGLWPFASAAGGGGSEAAGAEQALVRPRGRAVPPFVFTRRGSMFYD...YEV', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MSFLFLLVSSILFAFMFVFWKTWFQRNTGEMSSNSTSLALVRSSSSTQSTKSNI...TEV', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNVYLLLASGILCALMTVFWKYRRFQRNTGEMSSNSTALALVRPSSTGLINSNT...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MLLTRFEQKNGHLAQVEVDEVFGFMCHITTEVPPHDAMPGGIVLLVKFLLDMSR...GRC', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNVYLLLASGILCALMTVFWKYRRFQRNTGEMSSNSTALALVRPSSTGLINSNT...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MTEQMTLRGTLKGHNGWVTQIATTPQFPDMILSASRDKTIIMWKLTRDETNYGI...GTR', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MSFLFLLVSSILFAFMFVFWKTWFQRNTGEMSSNSTSLALVRSSSSTQSTKSNI...TEV', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "SECOND\n",
      "[SeqRecord(seq=Seq('*KPNAFANCAAAAAIIARICRPKFMIKPKNRAC*ANIPKTIRFIIAICRAKALN...PAA', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TFICCWRAAFCAR*LCFGNIAAFSATPAK*AATAPRWRWCARAAAA*LTATPI...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TFICCWRAAFCAR*LCFGNIAAFSATPAK*AATAPRWRWCARAAAA*LTATPI...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('WARAAAKRAACGRLRARRAAAAAKRRARNRRWCARAAARCRRLCLPAAAACFMM...MKX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*AFCFCW*AAFCLRLCLCFGKPGFSATPAK*AATAPAWRWCAAAAAPRAPKATL...PKX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TCICCWRAAFCAR**PCFGNIAAFSATPAK*AATAPRWRWCARAAPA*LTATP...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('CC*PALNRKTAIWRRWKWMKCLALCAILPPKCRRMMRCRAALCCW*NFCWI*AA...AAA', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TCICCWRAAFCAR**PCFGNIAAFSATPAK*AATAPRWRWCARAAPA*LTATP...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*PNR*PCAAP*KAITAG*PRLRPPRSFRI*F*ARAAIKPLLCGN*PAMKPTMAF...APA', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*AFCFCW*AAFCLRLCLCFGKPGFSATPAK*AATAPAWRWCAAAAAPRAPKATL...PKX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "THIRD\n",
      "[SeqRecord(seq=Seq('ENQMHLRTVQLRPPSLPASADQNL**NRKTVPAERIYRKLSVLS*LSAARKL*T...QQX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELLSAAGEQHSVRADCVLEISPLSAQHRRNEQQQHRAGAGAPEQQRPD*QQHR*...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELLSAAGEQHSVRADCVLEISPLSAQHRRNEQQQHRAGAGAPEQQRPD*QQHR*...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('GRERQQSARPVAVCERGGRRRQRSGGRGTGAGAPARPRGAAVCVYPPRQHVL**...*SX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELSVSAGEQHSVCVYVCVLENLVSAQHRRNEQQQHQPGAGAQQQQHPEHQKQH*...RSX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ERVSAAGERHSVRADDRVLEISPLSAQHRRNEQQQHRAGAGAPEQHRPD*QQHR...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('AADPL*TEKRPSGAGGSG*SVWLYVPYYHRSAAA*CDAGRHCAAGEISAGYEPQ...PLX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ERVSAAGERHSVRADDRVLEISPLSAQHRRNEQQQHRAGAGAPEQHRPD*QQHR...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('DRTDDPARHPERP*RLGDPDCDHPAVSGYDSEREPR*NHYYVETDPR*NQLWHS...HPX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELSVSAGEQHSVCVYVCVLENLVSAQHRRNEQQQHQPGAGAQQQQHPEHQKQH*...RSX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "FIRST\n",
      "[SeqRecord(seq=Seq('MKTKCICELCSCGRHHCPHLPTKIYDKTEKPCLLSEYTENYPFYHSYLPRESFK...TSS', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNFYLLLASSILCALIVFWKYRRFQRNTGEMSSNSTALALVRPSSSGLINSNTD...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNFYLLLASSILCALIVFWKYRRFQRNTGEMSSNSTALALVRPSSSGLINSNTD...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MGASGSKARGLWPFASAAGGGGSEAAGAEQALVRPRGRAVPPFVFTRRGSMFYD...YEV', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MSFLFLLVSSILFAFMFVFWKTWFQRNTGEMSSNSTSLALVRSSSSTQSTKSNI...TEV', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNVYLLLASGILCALMTVFWKYRRFQRNTGEMSSNSTALALVRPSSTGLINSNT...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MLLTRFEQKNGHLAQVEVDEVFGFMCHITTEVPPHDAMPGGIVLLVKFLLDMSR...GRC', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNVYLLLASGILCALMTVFWKYRRFQRNTGEMSSNSTALALVRPSSTGLINSNT...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MTEQMTLRGTLKGHNGWVTQIATTPQFPDMILSASRDKTIIMWKLTRDETNYGI...GTR', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MSFLFLLVSSILFAFMFVFWKTWFQRNTGEMSSNSTSLALVRSSSSTQSTKSNI...TEV', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNILFILVGSIIFASIFSIFKVVFQIPDHERPSNPTSPTLLRANSFWSYRNTGL...SNQ', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "SECOND\n",
      "[SeqRecord(seq=Seq('*KPNAFANCAAAAAIIARICRPKFMIKPKNRAC*ANIPKTIRFIIAICRAKALN...PAA', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TFICCWRAAFCAR*LCFGNIAAFSATPAK*AATAPRWRWCARAAAA*LTATPI...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TFICCWRAAFCAR*LCFGNIAAFSATPAK*AATAPRWRWCARAAAA*LTATPI...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('WARAAAKRAACGRLRARRAAAAAKRRARNRRWCARAAARCRRLCLPAAAACFMM...MKX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*AFCFCW*AAFCLRLCLCFGKPGFSATPAK*AATAPAWRWCAAAAAPRAPKATL...PKX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TCICCWRAAFCAR**PCFGNIAAFSATPAK*AATAPRWRWCARAAPA*LTATP...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('CC*PALNRKTAIWRRWKWMKCLALCAILPPKCRRMMRCRAALCCW*NFCWI*AA...AAA', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TCICCWRAAFCAR**PCFGNIAAFSATPAK*AATAPRWRWCARAAPA*LTATP...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*PNR*PCAAP*KAITAG*PRLRPPRSFRI*F*ARAAIKPLLCGN*PAMKPTMAF...APA', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*AFCFCW*AAFCLRLCLCFGKPGFSATPAK*AATAPAWRWCAAAAAPRAPKATL...PKX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TFCLFWWAALFLRAFLAFLKWCFRFRIMNARATRPARPCCARTAFGAIATPA*...ATX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "THIRD\n",
      "[SeqRecord(seq=Seq('ENQMHLRTVQLRPPSLPASADQNL**NRKTVPAERIYRKLSVLS*LSAARKL*T...QQX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELLSAAGEQHSVRADCVLEISPLSAQHRRNEQQQHRAGAGAPEQQRPD*QQHR*...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELLSAAGEQHSVRADCVLEISPLSAQHRRNEQQQHRAGAGAPEQQRPD*QQHR*...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('GRERQQSARPVAVCERGGRRRQRSGGRGTGAGAPARPRGAAVCVYPPRQHVL**...*SX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELSVSAGEQHSVCVYVCVLENLVSAQHRRNEQQQHQPGAGAQQQQHPEHQKQH*...RSX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ERVSAAGERHSVRADDRVLEISPLSAQHRRNEQQQHRAGAGAPEQHRPD*QQHR...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('AADPL*TEKRPSGAGGSG*SVWLYVPYYHRSAAA*CDAGRHCAAGEISAGYEPQ...PLX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ERVSAAGERHSVRADDRVLEISPLSAQHRRNEQQQHRAGAGAPEQHRPD*QQHR...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('DRTDDPARHPERP*RLGDPDCDHPAVSGYDSEREPR*NHYYVETDPR*NQLWHS...HPX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELSVSAGEQHSVCVYVCVLENLVSAQHRRNEQQQHQPGAGAQQQQHPEHQKQH*...RSX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('EHSVYSGGQHYFCEHF*HF*SGVSDSGS*TPEQPDQPDPAAREQLLELSQHRPE...QPX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "FIRST\n",
      "[SeqRecord(seq=Seq('MKTKCICELCSCGRHHCPHLPTKIYDKTEKPCLLSEYTENYPFYHSYLPRESFK...TSS', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNFYLLLASSILCALIVFWKYRRFQRNTGEMSSNSTALALVRPSSSGLINSNTD...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNFYLLLASSILCALIVFWKYRRFQRNTGEMSSNSTALALVRPSSSGLINSNTD...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MGASGSKARGLWPFASAAGGGGSEAAGAEQALVRPRGRAVPPFVFTRRGSMFYD...YEV', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MSFLFLLVSSILFAFMFVFWKTWFQRNTGEMSSNSTSLALVRSSSSTQSTKSNI...TEV', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNVYLLLASGILCALMTVFWKYRRFQRNTGEMSSNSTALALVRPSSTGLINSNT...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MLLTRFEQKNGHLAQVEVDEVFGFMCHITTEVPPHDAMPGGIVLLVKFLLDMSR...GRC', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNVYLLLASGILCALMTVFWKYRRFQRNTGEMSSNSTALALVRPSSTGLINSNT...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MTEQMTLRGTLKGHNGWVTQIATTPQFPDMILSASRDKTIIMWKLTRDETNYGI...GTR', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MSFLFLLVSSILFAFMFVFWKTWFQRNTGEMSSNSTSLALVRSSSSTQSTKSNI...TEV', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNILFILVGSIIFASIFSIFKVVFQIPDHERPSNPTSPTLLRANSFWSYRNTGL...SNQ', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MSMLLLLLSFIVFTLLLVVWKSRFQSSVGEMSSNSTSLALVRATSSTGSTKSNT...GNH', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "SECOND\n",
      "[SeqRecord(seq=Seq('*KPNAFANCAAAAAIIARICRPKFMIKPKNRAC*ANIPKTIRFIIAICRAKALN...PAA', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TFICCWRAAFCAR*LCFGNIAAFSATPAK*AATAPRWRWCARAAAA*LTATPI...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TFICCWRAAFCAR*LCFGNIAAFSATPAK*AATAPRWRWCARAAAA*LTATPI...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('WARAAAKRAACGRLRARRAAAAAKRRARNRRWCARAAARCRRLCLPAAAACFMM...MKX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*AFCFCW*AAFCLRLCLCFGKPGFSATPAK*AATAPAWRWCAAAAAPRAPKATL...PKX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TCICCWRAAFCAR**PCFGNIAAFSATPAK*AATAPRWRWCARAAPA*LTATP...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('CC*PALNRKTAIWRRWKWMKCLALCAILPPKCRRMMRCRAALCCW*NFCWI*AA...AAA', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TCICCWRAAFCAR**PCFGNIAAFSATPAK*AATAPRWRWCARAAPA*LTATP...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*PNR*PCAAP*KAITAG*PRLRPPRSFRI*F*ARAAIKPLLCGN*PAMKPTMAF...APA', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*AFCFCW*AAFCLRLCLCFGKPGFSATPAK*AATAPAWRWCAAAAAPRAPKATL...PKX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TFCLFWWAALFLRAFLAFLKWCFRFRIMNARATRPARPCCARTAFGAIATPA*...ATX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*ACCCCC*ALLCLPCCWWCGKAAFRAAWAK*AATAPAWRWCARPAAPAAPKATP...ATX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "THIRD\n",
      "[SeqRecord(seq=Seq('ENQMHLRTVQLRPPSLPASADQNL**NRKTVPAERIYRKLSVLS*LSAARKL*T...QQX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELLSAAGEQHSVRADCVLEISPLSAQHRRNEQQQHRAGAGAPEQQRPD*QQHR*...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELLSAAGEQHSVRADCVLEISPLSAQHRRNEQQQHRAGAGAPEQQRPD*QQHR*...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('GRERQQSARPVAVCERGGRRRQRSGGRGTGAGAPARPRGAAVCVYPPRQHVL**...*SX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELSVSAGEQHSVCVYVCVLENLVSAQHRRNEQQQHQPGAGAQQQQHPEHQKQH*...RSX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ERVSAAGERHSVRADDRVLEISPLSAQHRRNEQQQHRAGAGAPEQHRPD*QQHR...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('AADPL*TEKRPSGAGGSG*SVWLYVPYYHRSAAA*CDAGRHCAAGEISAGYEPQ...PLX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ERVSAAGERHSVRADDRVLEISPLSAQHRRNEQQQHRAGAGAPEQHRPD*QQHR...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('DRTDDPARHPERP*RLGDPDCDHPAVSGYDSEREPR*NHYYVETDPR*NQLWHS...HPX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELSVSAGEQHSVCVYVCVLENLVSAQHRRNEQQQHQPGAGAQQQQHPEHQKQH*...RSX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('EHSVYSGGQHYFCEHF*HF*SGVSDSGS*TPEQPDQPDPAAREQLLELSQHRPE...QPX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('EHAAAAAELYCVYPAAGGVEKPLSEQRGRNEQQQHQPGAGARDQQHRQHQKQHR...QPX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "FIRST\n",
      "[SeqRecord(seq=Seq('MKTKCICELCSCGRHHCPHLPTKIYDKTEKPCLLSEYTENYPFYHSYLPRESFK...TSS', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNFYLLLASSILCALIVFWKYRRFQRNTGEMSSNSTALALVRPSSSGLINSNTD...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNFYLLLASSILCALIVFWKYRRFQRNTGEMSSNSTALALVRPSSSGLINSNTD...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MGASGSKARGLWPFASAAGGGGSEAAGAEQALVRPRGRAVPPFVFTRRGSMFYD...YEV', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MSFLFLLVSSILFAFMFVFWKTWFQRNTGEMSSNSTSLALVRSSSSTQSTKSNI...TEV', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNVYLLLASGILCALMTVFWKYRRFQRNTGEMSSNSTALALVRPSSTGLINSNT...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MLLTRFEQKNGHLAQVEVDEVFGFMCHITTEVPPHDAMPGGIVLLVKFLLDMSR...GRC', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNVYLLLASGILCALMTVFWKYRRFQRNTGEMSSNSTALALVRPSSTGLINSNT...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MTEQMTLRGTLKGHNGWVTQIATTPQFPDMILSASRDKTIIMWKLTRDETNYGI...GTR', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MSFLFLLVSSILFAFMFVFWKTWFQRNTGEMSSNSTSLALVRSSSSTQSTKSNI...TEV', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNILFILVGSIIFASIFSIFKVVFQIPDHERPSNPTSPTLLRANSFWSYRNTGL...SNQ', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MSMLLLLLSFIVFTLLLVVWKSRFQSSVGEMSSNSTSLALVRATSSTGSTKSNT...GNH', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MSFLFLLVSSILFAFMFVFWKTWFQRNTGEMSSNSTSLALVRSSSSTQSTKSNI...TEV', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "SECOND\n",
      "[SeqRecord(seq=Seq('*KPNAFANCAAAAAIIARICRPKFMIKPKNRAC*ANIPKTIRFIIAICRAKALN...PAA', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TFICCWRAAFCAR*LCFGNIAAFSATPAK*AATAPRWRWCARAAAA*LTATPI...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TFICCWRAAFCAR*LCFGNIAAFSATPAK*AATAPRWRWCARAAAA*LTATPI...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('WARAAAKRAACGRLRARRAAAAAKRRARNRRWCARAAARCRRLCLPAAAACFMM...MKX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*AFCFCW*AAFCLRLCLCFGKPGFSATPAK*AATAPAWRWCAAAAAPRAPKATL...PKX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TCICCWRAAFCAR**PCFGNIAAFSATPAK*AATAPRWRWCARAAPA*LTATP...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('CC*PALNRKTAIWRRWKWMKCLALCAILPPKCRRMMRCRAALCCW*NFCWI*AA...AAA', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TCICCWRAAFCAR**PCFGNIAAFSATPAK*AATAPRWRWCARAAPA*LTATP...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*PNR*PCAAP*KAITAG*PRLRPPRSFRI*F*ARAAIKPLLCGN*PAMKPTMAF...APA', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*AFCFCW*AAFCLRLCLCFGKPGFSATPAK*AATAPAWRWCAAAAAPRAPKATL...PKX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TFCLFWWAALFLRAFLAFLKWCFRFRIMNARATRPARPCCARTAFGAIATPA*...ATX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*ACCCCC*ALLCLPCCWWCGKAAFRAAWAK*AATAPAWRWCARPAAPAAPKATP...ATX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*AFCFCW*AAFCLRLCLCFGKPGFSATPAK*AATAPAWRWCAAAAAPRAPKATL...PKX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "THIRD\n",
      "[SeqRecord(seq=Seq('ENQMHLRTVQLRPPSLPASADQNL**NRKTVPAERIYRKLSVLS*LSAARKL*T...QQX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELLSAAGEQHSVRADCVLEISPLSAQHRRNEQQQHRAGAGAPEQQRPD*QQHR*...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELLSAAGEQHSVRADCVLEISPLSAQHRRNEQQQHRAGAGAPEQQRPD*QQHR*...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('GRERQQSARPVAVCERGGRRRQRSGGRGTGAGAPARPRGAAVCVYPPRQHVL**...*SX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELSVSAGEQHSVCVYVCVLENLVSAQHRRNEQQQHQPGAGAQQQQHPEHQKQH*...RSX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ERVSAAGERHSVRADDRVLEISPLSAQHRRNEQQQHRAGAGAPEQHRPD*QQHR...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('AADPL*TEKRPSGAGGSG*SVWLYVPYYHRSAAA*CDAGRHCAAGEISAGYEPQ...PLX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ERVSAAGERHSVRADDRVLEISPLSAQHRRNEQQQHRAGAGAPEQHRPD*QQHR...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('DRTDDPARHPERP*RLGDPDCDHPAVSGYDSEREPR*NHYYVETDPR*NQLWHS...HPX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELSVSAGEQHSVCVYVCVLENLVSAQHRRNEQQQHQPGAGAQQQQHPEHQKQH*...RSX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('EHSVYSGGQHYFCEHF*HF*SGVSDSGS*TPEQPDQPDPAAREQLLELSQHRPE...QPX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('EHAAAAAELYCVYPAAGGVEKPLSEQRGRNEQQQHQPGAGARDQQHRQHQKQHR...QPX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELSVSAGEQHSVCVYVCVLENLVSAQHRRNEQQQHQPGAGAQQQQHPEHQKQH*...RSX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "FIRST\n",
      "[SeqRecord(seq=Seq('MKTKCICELCSCGRHHCPHLPTKIYDKTEKPCLLSEYTENYPFYHSYLPRESFK...TSS', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNFYLLLASSILCALIVFWKYRRFQRNTGEMSSNSTALALVRPSSSGLINSNTD...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNFYLLLASSILCALIVFWKYRRFQRNTGEMSSNSTALALVRPSSSGLINSNTD...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MGASGSKARGLWPFASAAGGGGSEAAGAEQALVRPRGRAVPPFVFTRRGSMFYD...YEV', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MSFLFLLVSSILFAFMFVFWKTWFQRNTGEMSSNSTSLALVRSSSSTQSTKSNI...TEV', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNVYLLLASGILCALMTVFWKYRRFQRNTGEMSSNSTALALVRPSSTGLINSNT...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MLLTRFEQKNGHLAQVEVDEVFGFMCHITTEVPPHDAMPGGIVLLVKFLLDMSR...GRC', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNVYLLLASGILCALMTVFWKYRRFQRNTGEMSSNSTALALVRPSSTGLINSNT...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MTEQMTLRGTLKGHNGWVTQIATTPQFPDMILSASRDKTIIMWKLTRDETNYGI...GTR', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MSFLFLLVSSILFAFMFVFWKTWFQRNTGEMSSNSTSLALVRSSSSTQSTKSNI...TEV', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNILFILVGSIIFASIFSIFKVVFQIPDHERPSNPTSPTLLRANSFWSYRNTGL...SNQ', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MSMLLLLLSFIVFTLLLVVWKSRFQSSVGEMSSNSTSLALVRATSSTGSTKSNT...GNH', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MSFLFLLVSSILFAFMFVFWKTWFQRNTGEMSSNSTSLALVRSSSSTQSTKSNI...TEV', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MSFLFLLVSSILFAFMFVFWKTWFQRNTGEMSSNSTSSHYRACVVVVGHKLSLS...RVK', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "SECOND\n",
      "[SeqRecord(seq=Seq('*KPNAFANCAAAAAIIARICRPKFMIKPKNRAC*ANIPKTIRFIIAICRAKALN...PAA', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TFICCWRAAFCAR*LCFGNIAAFSATPAK*AATAPRWRWCARAAAA*LTATPI...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TFICCWRAAFCAR*LCFGNIAAFSATPAK*AATAPRWRWCARAAAA*LTATPI...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('WARAAAKRAACGRLRARRAAAAAKRRARNRRWCARAAARCRRLCLPAAAACFMM...MKX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*AFCFCW*AAFCLRLCLCFGKPGFSATPAK*AATAPAWRWCAAAAAPRAPKATL...PKX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TCICCWRAAFCAR**PCFGNIAAFSATPAK*AATAPRWRWCARAAPA*LTATP...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('CC*PALNRKTAIWRRWKWMKCLALCAILPPKCRRMMRCRAALCCW*NFCWI*AA...AAA', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TCICCWRAAFCAR**PCFGNIAAFSATPAK*AATAPRWRWCARAAPA*LTATP...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*PNR*PCAAP*KAITAG*PRLRPPRSFRI*F*ARAAIKPLLCGN*PAMKPTMAF...APA', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*AFCFCW*AAFCLRLCLCFGKPGFSATPAK*AATAPAWRWCAAAAAPRAPKATL...PKX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TFCLFWWAALFLRAFLAFLKWCFRFRIMNARATRPARPCCARTAFGAIATPA*...ATX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*ACCCCC*ALLCLPCCWWCGKAAFRAAWAK*AATAPAWRWCARPAAPAAPKATP...ATX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*AFCFCW*AAFCLRLCLCFGKPGFSATPAK*AATAPAWRWCAAAAAPRAPKATL...PKX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*AFCFCW*AAFCLRLCLCFGKPGFSATPAK*AATAPAAIIARAWWWWAIN*A*A...A*X', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "THIRD\n",
      "[SeqRecord(seq=Seq('ENQMHLRTVQLRPPSLPASADQNL**NRKTVPAERIYRKLSVLS*LSAARKL*T...QQX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELLSAAGEQHSVRADCVLEISPLSAQHRRNEQQQHRAGAGAPEQQRPD*QQHR*...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELLSAAGEQHSVRADCVLEISPLSAQHRRNEQQQHRAGAGAPEQQRPD*QQHR*...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('GRERQQSARPVAVCERGGRRRQRSGGRGTGAGAPARPRGAAVCVYPPRQHVL**...*SX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELSVSAGEQHSVCVYVCVLENLVSAQHRRNEQQQHQPGAGAQQQQHPEHQKQH*...RSX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ERVSAAGERHSVRADDRVLEISPLSAQHRRNEQQQHRAGAGAPEQHRPD*QQHR...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('AADPL*TEKRPSGAGGSG*SVWLYVPYYHRSAAA*CDAGRHCAAGEISAGYEPQ...PLX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ERVSAAGERHSVRADDRVLEISPLSAQHRRNEQQQHRAGAGAPEQHRPD*QQHR...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('DRTDDPARHPERP*RLGDPDCDHPAVSGYDSEREPR*NHYYVETDPR*NQLWHS...HPX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELSVSAGEQHSVCVYVCVLENLVSAQHRRNEQQQHQPGAGAQQQQHPEHQKQH*...RSX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('EHSVYSGGQHYFCEHF*HF*SGVSDSGS*TPEQPDQPDPAAREQLLELSQHRPE...QPX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('EHAAAAAELYCVYPAAGGVEKPLSEQRGRNEQQQHQPGAGARDQQHRQHQKQHR...QPX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELSVSAGEQHSVCVYVCVLENLVSAQHRRNEQQQHQPGAGAQQQQHPEHQKQH*...RSX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELSVSAGEQHSVCVYVCVLENLVSAQHRRNEQQQHQQPLSRVRGGGGP*TEPEQ...REX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "FIRST\n",
      "[SeqRecord(seq=Seq('MKTKCICELCSCGRHHCPHLPTKIYDKTEKPCLLSEYTENYPFYHSYLPRESFK...TSS', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNFYLLLASSILCALIVFWKYRRFQRNTGEMSSNSTALALVRPSSSGLINSNTD...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNFYLLLASSILCALIVFWKYRRFQRNTGEMSSNSTALALVRPSSSGLINSNTD...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MGASGSKARGLWPFASAAGGGGSEAAGAEQALVRPRGRAVPPFVFTRRGSMFYD...YEV', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MSFLFLLVSSILFAFMFVFWKTWFQRNTGEMSSNSTSLALVRSSSSTQSTKSNI...TEV', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNVYLLLASGILCALMTVFWKYRRFQRNTGEMSSNSTALALVRPSSTGLINSNT...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MLLTRFEQKNGHLAQVEVDEVFGFMCHITTEVPPHDAMPGGIVLLVKFLLDMSR...GRC', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNVYLLLASGILCALMTVFWKYRRFQRNTGEMSSNSTALALVRPSSTGLINSNT...KST', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MTEQMTLRGTLKGHNGWVTQIATTPQFPDMILSASRDKTIIMWKLTRDETNYGI...GTR', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MSFLFLLVSSILFAFMFVFWKTWFQRNTGEMSSNSTSLALVRSSSSTQSTKSNI...TEV', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MNILFILVGSIIFASIFSIFKVVFQIPDHERPSNPTSPTLLRANSFWSYRNTGL...SNQ', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MSMLLLLLSFIVFTLLLVVWKSRFQSSVGEMSSNSTSLALVRATSSTGSTKSNT...GNH', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MSFLFLLVSSILFAFMFVFWKTWFQRNTGEMSSNSTSLALVRSSSSTQSTKSNI...TEV', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MSFLFLLVSSILFAFMFVFWKTWFQRNTGEMSSNSTSSHYRACVVVVGHKLSLS...RVK', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('MSFLFLLVSSILFAFMFVFWKTWFQRNTGEMSSNSDKPTEVNGALVNRKS', ExtendedIUPACProtein()), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "SECOND\n",
      "[SeqRecord(seq=Seq('*KPNAFANCAAAAAIIARICRPKFMIKPKNRAC*ANIPKTIRFIIAICRAKALN...PAA', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TFICCWRAAFCAR*LCFGNIAAFSATPAK*AATAPRWRWCARAAAA*LTATPI...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TFICCWRAAFCAR*LCFGNIAAFSATPAK*AATAPRWRWCARAAAA*LTATPI...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('WARAAAKRAACGRLRARRAAAAAKRRARNRRWCARAAARCRRLCLPAAAACFMM...MKX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*AFCFCW*AAFCLRLCLCFGKPGFSATPAK*AATAPAWRWCAAAAAPRAPKATL...PKX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TCICCWRAAFCAR**PCFGNIAAFSATPAK*AATAPRWRWCARAAPA*LTATP...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('CC*PALNRKTAIWRRWKWMKCLALCAILPPKCRRMMRCRAALCCW*NFCWI*AA...AAA', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TCICCWRAAFCAR**PCFGNIAAFSATPAK*AATAPRWRWCARAAPA*LTATP...KAP', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*PNR*PCAAP*KAITAG*PRLRPPRSFRI*F*ARAAIKPLLCGN*PAMKPTMAF...APA', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*AFCFCW*AAFCLRLCLCFGKPGFSATPAK*AATAPAWRWCAAAAAPRAPKATL...PKX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*TFCLFWWAALFLRAFLAFLKWCFRFRIMNARATRPARPCCARTAFGAIATPA*...ATX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*ACCCCC*ALLCLPCCWWCGKAAFRAAWAK*AATAPAWRWCARPAAPAAPKATP...ATX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*AFCFCW*AAFCLRLCLCFGKPGFSATPAK*AATAPAWRWCAAAAAPRAPKATL...PKX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*AFCFCW*AAFCLRLCLCFGKPGFSATPAK*AATAPAAIIARAWWWWAIN*A*A...A*X', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('*AFCFCW*AAFCLRLCLCFGKPGFSATPAK*AATAINRPK*TARW*TAKA', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n",
      "THIRD\n",
      "[SeqRecord(seq=Seq('ENQMHLRTVQLRPPSLPASADQNL**NRKTVPAERIYRKLSVLS*LSAARKL*T...QQX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELLSAAGEQHSVRADCVLEISPLSAQHRRNEQQQHRAGAGAPEQQRPD*QQHR*...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELLSAAGEQHSVRADCVLEISPLSAQHRRNEQQQHRAGAGAPEQQRPD*QQHR*...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('GRERQQSARPVAVCERGGRRRQRSGGRGTGAGAPARPRGAAVCVYPPRQHVL**...*SX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELSVSAGEQHSVCVYVCVLENLVSAQHRRNEQQQHQPGAGAQQQQHPEHQKQH*...RSX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ERVSAAGERHSVRADDRVLEISPLSAQHRRNEQQQHRAGAGAPEQHRPD*QQHR...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('AADPL*TEKRPSGAGGSG*SVWLYVPYYHRSAAA*CDAGRHCAAGEISAGYEPQ...PLX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ERVSAAGERHSVRADDRVLEISPLSAQHRRNEQQQHRAGAGAPEQHRPD*QQHR...KHX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('DRTDDPARHPERP*RLGDPDCDHPAVSGYDSEREPR*NHYYVETDPR*NQLWHS...HPX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELSVSAGEQHSVCVYVCVLENLVSAQHRRNEQQQHQPGAGAQQQQHPEHQKQH*...RSX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('EHSVYSGGQHYFCEHF*HF*SGVSDSGS*TPEQPDQPDPAAREQLLELSQHRPE...QPX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('EHAAAAAELYCVYPAAGGVEKPLSEQRGRNEQQQHQPGAGARDQQHRQHQKQHR...QPX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELSVSAGEQHSVCVYVCVLENLVSAQHRRNEQQQHQPGAGAQQQQHPEHQKQH*...RSX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELSVSAGEQHSVCVYVCVLENLVSAQHRRNEQQQHQQPLSRVRGGGGP*TEPEQ...REX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[]), SeqRecord(seq=Seq('ELSVSAGEQHSVCVYVCVLENLVSAQHRRNEQQQR*TDRSERRAGEPQKX', HasStopCodon(ExtendedIUPACProtein(), '*')), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 178,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Translates an DNA sequence from a fasta file to an amino acid sequence.\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "\n",
    "def pad_seq(sequence):\n",
    "    \"\"\" Pad sequence to multiple of 3 with N \"\"\"\n",
    "\n",
    "    remainder = len(sequence) % 3\n",
    "\n",
    "    return sequence if remainder == 0 else sequence + Seq('N' * (3 - remainder))\n",
    "\n",
    "seq_records = SeqIO.parse('DNA.fasta', 'fasta')\n",
    "\n",
    "\n",
    "amino_acids1 = []\n",
    "amino_acids2 = []\n",
    "amino_acids3 = []\n",
    "\n",
    "for record in seq_records:\n",
    "\n",
    "    # starting from nucleotide 1\n",
    "    amino_acids1.append(pad_seq(record).translate())\n",
    "    print(\"FIRST\")\n",
    "    print(amino_acids1)\n",
    "    # ...\n",
    "\n",
    "    # starting from nucleotide 2\n",
    "    record2 = record[1:]\n",
    "    amino_acids2.append(pad_seq(record2).translate())\n",
    "    print(\"SECOND\")\n",
    "    print(amino_acids2)\n",
    "    # ...\n",
    "\n",
    "    # starting from nucleotide 3\n",
    "    record3 = record[2:]\n",
    "    amino_acids3.append(pad_seq(record3).translate())\n",
    "    print(\"THIRD\")\n",
    "    print(amino_acids3)\n",
    "    # ...\n",
    "    \n",
    "SeqIO.write(seq_records, \"AAseq.fasta\", \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
